// @aws-sdk/client-codecommit/dist-cjs/index.js
'use strict';

var client$1 = require('@aws-sdk/core/client');
var core = require('@smithy/core');
var client = require('@smithy/core/client');
var config = require('@smithy/core/config');
var endpoints = require('@smithy/core/endpoints');
var protocols = require('@smithy/core/protocols');
var retry = require('@smithy/core/retry');
var schema = require('@smithy/core/schema');
var httpAuthSchemeProvider = require('./auth/httpAuthSchemeProvider');
var runtimeConfig = require('./runtimeConfig');
var schemas_0 = require('./schemas/schemas_0');
var errors = require('./models/errors');
var CodeCommitServiceException = require('./models/CodeCommitServiceException');

const resolveClientEndpointParameters = (options) => {
    return Object.assign(options, {
        useDualstackEndpoint: options.useDualstackEndpoint ?? false,
        useFipsEndpoint: options.useFipsEndpoint ?? false,
        defaultSigningName: "codecommit",
    });
};
const commonParams = {
    UseFIPS: { type: "builtInParams", name: "useFipsEndpoint" },
    Endpoint: { type: "builtInParams", name: "endpoint" },
    Region: { type: "builtInParams", name: "region" },
    UseDualStack: { type: "builtInParams", name: "useDualstackEndpoint" },
};

const getHttpAuthExtensionConfiguration = (runtimeConfig) => {
    const _httpAuthSchemes = runtimeConfig.httpAuthSchemes;
    let _httpAuthSchemeProvider = runtimeConfig.httpAuthSchemeProvider;
    let _credentials = runtimeConfig.credentials;
    return {
        setHttpAuthScheme(httpAuthScheme) {
            const index = _httpAuthSchemes.findIndex((scheme) => scheme.schemeId === httpAuthScheme.schemeId);
            if (index === -1) {
                _httpAuthSchemes.push(httpAuthScheme);
            }
            else {
                _httpAuthSchemes.splice(index, 1, httpAuthScheme);
            }
        },
        httpAuthSchemes() {
            return _httpAuthSchemes;
        },
        setHttpAuthSchemeProvider(httpAuthSchemeProvider) {
            _httpAuthSchemeProvider = httpAuthSchemeProvider;
        },
        httpAuthSchemeProvider() {
            return _httpAuthSchemeProvider;
        },
        setCredentials(credentials) {
            _credentials = credentials;
        },
        credentials() {
            return _credentials;
        },
    };
};
const resolveHttpAuthRuntimeConfig = (config) => {
    return {
        httpAuthSchemes: config.httpAuthSchemes(),
        httpAuthSchemeProvider: config.httpAuthSchemeProvider(),
        credentials: config.credentials(),
    };
};

const resolveRuntimeExtensions = (runtimeConfig, extensions) => {
    const extensionConfiguration = Object.assign(client$1.getAwsRegionExtensionConfiguration(runtimeConfig), client.getDefaultExtensionConfiguration(runtimeConfig), protocols.getHttpHandlerExtensionConfiguration(runtimeConfig), getHttpAuthExtensionConfiguration(runtimeConfig));
    extensions.forEach((extension) => extension.configure(extensionConfiguration));
    return Object.assign(runtimeConfig, client$1.resolveAwsRegionExtensionConfiguration(extensionConfiguration), client.resolveDefaultRuntimeConfig(extensionConfiguration), protocols.resolveHttpHandlerRuntimeConfig(extensionConfiguration), resolveHttpAuthRuntimeConfig(extensionConfiguration));
};

class CodeCommitClient extends client.Client {
    config;
    constructor(...[configuration]) {
        const _config_0 = runtimeConfig.getRuntimeConfig(configuration || {});
        super(_config_0);
        this.initConfig = _config_0;
        const _config_1 = resolveClientEndpointParameters(_config_0);
        const _config_2 = client$1.resolveUserAgentConfig(_config_1);
        const _config_3 = retry.resolveRetryConfig(_config_2);
        const _config_4 = config.resolveRegionConfig(_config_3);
        const _config_5 = client$1.resolveHostHeaderConfig(_config_4);
        const _config_6 = endpoints.resolveEndpointConfig(_config_5);
        const _config_7 = httpAuthSchemeProvider.resolveHttpAuthSchemeConfig(_config_6);
        const _config_8 = resolveRuntimeExtensions(_config_7, configuration?.extensions || []);
        this.config = _config_8;
        this.middlewareStack.use(schema.getSchemaSerdePlugin(this.config));
        this.middlewareStack.use(client$1.getUserAgentPlugin(this.config));
        this.middlewareStack.use(retry.getRetryPlugin(this.config));
        this.middlewareStack.use(protocols.getContentLengthPlugin(this.config));
        this.middlewareStack.use(client$1.getHostHeaderPlugin(this.config));
        this.middlewareStack.use(client$1.getLoggerPlugin(this.config));
        this.middlewareStack.use(client$1.getRecursionDetectionPlugin(this.config));
        this.middlewareStack.use(core.getHttpAuthSchemeEndpointRuleSetPlugin(this.config, {
            httpAuthSchemeParametersProvider: httpAuthSchemeProvider.defaultCodeCommitHttpAuthSchemeParametersProvider,
            identityProviderConfigProvider: async (config) => new core.DefaultIdentityProviderConfig({
                "aws.auth#sigv4": config.credentials,
            }),
        }));
        this.middlewareStack.use(core.getHttpSigningPlugin(this.config));
    }
    destroy() {
        super.destroy();
    }
}

class AssociateApprovalRuleTemplateWithRepositoryCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "AssociateApprovalRuleTemplateWithRepository", {})
    .n("CodeCommitClient", "AssociateApprovalRuleTemplateWithRepositoryCommand")
    .sc(schemas_0.AssociateApprovalRuleTemplateWithRepository$)
    .build() {
}

class BatchAssociateApprovalRuleTemplateWithRepositoriesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "BatchAssociateApprovalRuleTemplateWithRepositories", {})
    .n("CodeCommitClient", "BatchAssociateApprovalRuleTemplateWithRepositoriesCommand")
    .sc(schemas_0.BatchAssociateApprovalRuleTemplateWithRepositories$)
    .build() {
}

class BatchDescribeMergeConflictsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "BatchDescribeMergeConflicts", {})
    .n("CodeCommitClient", "BatchDescribeMergeConflictsCommand")
    .sc(schemas_0.BatchDescribeMergeConflicts$)
    .build() {
}

class BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "BatchDisassociateApprovalRuleTemplateFromRepositories", {})
    .n("CodeCommitClient", "BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand")
    .sc(schemas_0.BatchDisassociateApprovalRuleTemplateFromRepositories$)
    .build() {
}

class BatchGetCommitsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "BatchGetCommits", {})
    .n("CodeCommitClient", "BatchGetCommitsCommand")
    .sc(schemas_0.BatchGetCommits$)
    .build() {
}

class BatchGetRepositoriesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "BatchGetRepositories", {})
    .n("CodeCommitClient", "BatchGetRepositoriesCommand")
    .sc(schemas_0.BatchGetRepositories$)
    .build() {
}

class CreateApprovalRuleTemplateCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "CreateApprovalRuleTemplate", {})
    .n("CodeCommitClient", "CreateApprovalRuleTemplateCommand")
    .sc(schemas_0.CreateApprovalRuleTemplate$)
    .build() {
}

class CreateBranchCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "CreateBranch", {})
    .n("CodeCommitClient", "CreateBranchCommand")
    .sc(schemas_0.CreateBranch$)
    .build() {
}

class CreateCommitCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "CreateCommit", {})
    .n("CodeCommitClient", "CreateCommitCommand")
    .sc(schemas_0.CreateCommit$)
    .build() {
}

class CreatePullRequestApprovalRuleCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "CreatePullRequestApprovalRule", {})
    .n("CodeCommitClient", "CreatePullRequestApprovalRuleCommand")
    .sc(schemas_0.CreatePullRequestApprovalRule$)
    .build() {
}

class CreatePullRequestCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "CreatePullRequest", {})
    .n("CodeCommitClient", "CreatePullRequestCommand")
    .sc(schemas_0.CreatePullRequest$)
    .build() {
}

class CreateRepositoryCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "CreateRepository", {})
    .n("CodeCommitClient", "CreateRepositoryCommand")
    .sc(schemas_0.CreateRepository$)
    .build() {
}

class CreateUnreferencedMergeCommitCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "CreateUnreferencedMergeCommit", {})
    .n("CodeCommitClient", "CreateUnreferencedMergeCommitCommand")
    .sc(schemas_0.CreateUnreferencedMergeCommit$)
    .build() {
}

class DeleteApprovalRuleTemplateCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DeleteApprovalRuleTemplate", {})
    .n("CodeCommitClient", "DeleteApprovalRuleTemplateCommand")
    .sc(schemas_0.DeleteApprovalRuleTemplate$)
    .build() {
}

class DeleteBranchCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DeleteBranch", {})
    .n("CodeCommitClient", "DeleteBranchCommand")
    .sc(schemas_0.DeleteBranch$)
    .build() {
}

class DeleteCommentContentCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DeleteCommentContent", {})
    .n("CodeCommitClient", "DeleteCommentContentCommand")
    .sc(schemas_0.DeleteCommentContent$)
    .build() {
}

class DeleteFileCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DeleteFile", {})
    .n("CodeCommitClient", "DeleteFileCommand")
    .sc(schemas_0.DeleteFile$)
    .build() {
}

class DeletePullRequestApprovalRuleCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DeletePullRequestApprovalRule", {})
    .n("CodeCommitClient", "DeletePullRequestApprovalRuleCommand")
    .sc(schemas_0.DeletePullRequestApprovalRule$)
    .build() {
}

class DeleteRepositoryCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DeleteRepository", {})
    .n("CodeCommitClient", "DeleteRepositoryCommand")
    .sc(schemas_0.DeleteRepository$)
    .build() {
}

class DescribeMergeConflictsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DescribeMergeConflicts", {})
    .n("CodeCommitClient", "DescribeMergeConflictsCommand")
    .sc(schemas_0.DescribeMergeConflicts$)
    .build() {
}

class DescribePullRequestEventsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DescribePullRequestEvents", {})
    .n("CodeCommitClient", "DescribePullRequestEventsCommand")
    .sc(schemas_0.DescribePullRequestEvents$)
    .build() {
}

class DisassociateApprovalRuleTemplateFromRepositoryCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "DisassociateApprovalRuleTemplateFromRepository", {})
    .n("CodeCommitClient", "DisassociateApprovalRuleTemplateFromRepositoryCommand")
    .sc(schemas_0.DisassociateApprovalRuleTemplateFromRepository$)
    .build() {
}

class EvaluatePullRequestApprovalRulesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "EvaluatePullRequestApprovalRules", {})
    .n("CodeCommitClient", "EvaluatePullRequestApprovalRulesCommand")
    .sc(schemas_0.EvaluatePullRequestApprovalRules$)
    .build() {
}

class GetApprovalRuleTemplateCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetApprovalRuleTemplate", {})
    .n("CodeCommitClient", "GetApprovalRuleTemplateCommand")
    .sc(schemas_0.GetApprovalRuleTemplate$)
    .build() {
}

class GetBlobCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetBlob", {})
    .n("CodeCommitClient", "GetBlobCommand")
    .sc(schemas_0.GetBlob$)
    .build() {
}

class GetBranchCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetBranch", {})
    .n("CodeCommitClient", "GetBranchCommand")
    .sc(schemas_0.GetBranch$)
    .build() {
}

class GetCommentCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetComment", {})
    .n("CodeCommitClient", "GetCommentCommand")
    .sc(schemas_0.GetComment$)
    .build() {
}

class GetCommentReactionsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetCommentReactions", {})
    .n("CodeCommitClient", "GetCommentReactionsCommand")
    .sc(schemas_0.GetCommentReactions$)
    .build() {
}

class GetCommentsForComparedCommitCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetCommentsForComparedCommit", {})
    .n("CodeCommitClient", "GetCommentsForComparedCommitCommand")
    .sc(schemas_0.GetCommentsForComparedCommit$)
    .build() {
}

class GetCommentsForPullRequestCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetCommentsForPullRequest", {})
    .n("CodeCommitClient", "GetCommentsForPullRequestCommand")
    .sc(schemas_0.GetCommentsForPullRequest$)
    .build() {
}

class GetCommitCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetCommit", {})
    .n("CodeCommitClient", "GetCommitCommand")
    .sc(schemas_0.GetCommit$)
    .build() {
}

class GetDifferencesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetDifferences", {})
    .n("CodeCommitClient", "GetDifferencesCommand")
    .sc(schemas_0.GetDifferences$)
    .build() {
}

class GetFileCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetFile", {})
    .n("CodeCommitClient", "GetFileCommand")
    .sc(schemas_0.GetFile$)
    .build() {
}

class GetFolderCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetFolder", {})
    .n("CodeCommitClient", "GetFolderCommand")
    .sc(schemas_0.GetFolder$)
    .build() {
}

class GetMergeCommitCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetMergeCommit", {})
    .n("CodeCommitClient", "GetMergeCommitCommand")
    .sc(schemas_0.GetMergeCommit$)
    .build() {
}

class GetMergeConflictsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetMergeConflicts", {})
    .n("CodeCommitClient", "GetMergeConflictsCommand")
    .sc(schemas_0.GetMergeConflicts$)
    .build() {
}

class GetMergeOptionsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetMergeOptions", {})
    .n("CodeCommitClient", "GetMergeOptionsCommand")
    .sc(schemas_0.GetMergeOptions$)
    .build() {
}

class GetPullRequestApprovalStatesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetPullRequestApprovalStates", {})
    .n("CodeCommitClient", "GetPullRequestApprovalStatesCommand")
    .sc(schemas_0.GetPullRequestApprovalStates$)
    .build() {
}

class GetPullRequestCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetPullRequest", {})
    .n("CodeCommitClient", "GetPullRequestCommand")
    .sc(schemas_0.GetPullRequest$)
    .build() {
}

class GetPullRequestOverrideStateCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetPullRequestOverrideState", {})
    .n("CodeCommitClient", "GetPullRequestOverrideStateCommand")
    .sc(schemas_0.GetPullRequestOverrideState$)
    .build() {
}

class GetRepositoryCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetRepository", {})
    .n("CodeCommitClient", "GetRepositoryCommand")
    .sc(schemas_0.GetRepository$)
    .build() {
}

class GetRepositoryTriggersCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "GetRepositoryTriggers", {})
    .n("CodeCommitClient", "GetRepositoryTriggersCommand")
    .sc(schemas_0.GetRepositoryTriggers$)
    .build() {
}

class ListApprovalRuleTemplatesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListApprovalRuleTemplates", {})
    .n("CodeCommitClient", "ListApprovalRuleTemplatesCommand")
    .sc(schemas_0.ListApprovalRuleTemplates$)
    .build() {
}

class ListAssociatedApprovalRuleTemplatesForRepositoryCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListAssociatedApprovalRuleTemplatesForRepository", {})
    .n("CodeCommitClient", "ListAssociatedApprovalRuleTemplatesForRepositoryCommand")
    .sc(schemas_0.ListAssociatedApprovalRuleTemplatesForRepository$)
    .build() {
}

class ListBranchesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListBranches", {})
    .n("CodeCommitClient", "ListBranchesCommand")
    .sc(schemas_0.ListBranches$)
    .build() {
}

class ListFileCommitHistoryCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListFileCommitHistory", {})
    .n("CodeCommitClient", "ListFileCommitHistoryCommand")
    .sc(schemas_0.ListFileCommitHistory$)
    .build() {
}

class ListPullRequestsCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListPullRequests", {})
    .n("CodeCommitClient", "ListPullRequestsCommand")
    .sc(schemas_0.ListPullRequests$)
    .build() {
}

class ListRepositoriesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListRepositories", {})
    .n("CodeCommitClient", "ListRepositoriesCommand")
    .sc(schemas_0.ListRepositories$)
    .build() {
}

class ListRepositoriesForApprovalRuleTemplateCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListRepositoriesForApprovalRuleTemplate", {})
    .n("CodeCommitClient", "ListRepositoriesForApprovalRuleTemplateCommand")
    .sc(schemas_0.ListRepositoriesForApprovalRuleTemplate$)
    .build() {
}

class ListTagsForResourceCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "ListTagsForResource", {})
    .n("CodeCommitClient", "ListTagsForResourceCommand")
    .sc(schemas_0.ListTagsForResource$)
    .build() {
}

class MergeBranchesByFastForwardCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "MergeBranchesByFastForward", {})
    .n("CodeCommitClient", "MergeBranchesByFastForwardCommand")
    .sc(schemas_0.MergeBranchesByFastForward$)
    .build() {
}

class MergeBranchesBySquashCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "MergeBranchesBySquash", {})
    .n("CodeCommitClient", "MergeBranchesBySquashCommand")
    .sc(schemas_0.MergeBranchesBySquash$)
    .build() {
}

class MergeBranchesByThreeWayCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "MergeBranchesByThreeWay", {})
    .n("CodeCommitClient", "MergeBranchesByThreeWayCommand")
    .sc(schemas_0.MergeBranchesByThreeWay$)
    .build() {
}

class MergePullRequestByFastForwardCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "MergePullRequestByFastForward", {})
    .n("CodeCommitClient", "MergePullRequestByFastForwardCommand")
    .sc(schemas_0.MergePullRequestByFastForward$)
    .build() {
}

class MergePullRequestBySquashCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "MergePullRequestBySquash", {})
    .n("CodeCommitClient", "MergePullRequestBySquashCommand")
    .sc(schemas_0.MergePullRequestBySquash$)
    .build() {
}

class MergePullRequestByThreeWayCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "MergePullRequestByThreeWay", {})
    .n("CodeCommitClient", "MergePullRequestByThreeWayCommand")
    .sc(schemas_0.MergePullRequestByThreeWay$)
    .build() {
}

class OverridePullRequestApprovalRulesCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "OverridePullRequestApprovalRules", {})
    .n("CodeCommitClient", "OverridePullRequestApprovalRulesCommand")
    .sc(schemas_0.OverridePullRequestApprovalRules$)
    .build() {
}

class PostCommentForComparedCommitCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "PostCommentForComparedCommit", {})
    .n("CodeCommitClient", "PostCommentForComparedCommitCommand")
    .sc(schemas_0.PostCommentForComparedCommit$)
    .build() {
}

class PostCommentForPullRequestCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "PostCommentForPullRequest", {})
    .n("CodeCommitClient", "PostCommentForPullRequestCommand")
    .sc(schemas_0.PostCommentForPullRequest$)
    .build() {
}

class PostCommentReplyCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "PostCommentReply", {})
    .n("CodeCommitClient", "PostCommentReplyCommand")
    .sc(schemas_0.PostCommentReply$)
    .build() {
}

class PutCommentReactionCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "PutCommentReaction", {})
    .n("CodeCommitClient", "PutCommentReactionCommand")
    .sc(schemas_0.PutCommentReaction$)
    .build() {
}

class PutFileCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "PutFile", {})
    .n("CodeCommitClient", "PutFileCommand")
    .sc(schemas_0.PutFile$)
    .build() {
}

class PutRepositoryTriggersCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "PutRepositoryTriggers", {})
    .n("CodeCommitClient", "PutRepositoryTriggersCommand")
    .sc(schemas_0.PutRepositoryTriggers$)
    .build() {
}

class TagResourceCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "TagResource", {})
    .n("CodeCommitClient", "TagResourceCommand")
    .sc(schemas_0.TagResource$)
    .build() {
}

class TestRepositoryTriggersCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "TestRepositoryTriggers", {})
    .n("CodeCommitClient", "TestRepositoryTriggersCommand")
    .sc(schemas_0.TestRepositoryTriggers$)
    .build() {
}

class UntagResourceCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UntagResource", {})
    .n("CodeCommitClient", "UntagResourceCommand")
    .sc(schemas_0.UntagResource$)
    .build() {
}

class UpdateApprovalRuleTemplateContentCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateApprovalRuleTemplateContent", {})
    .n("CodeCommitClient", "UpdateApprovalRuleTemplateContentCommand")
    .sc(schemas_0.UpdateApprovalRuleTemplateContent$)
    .build() {
}

class UpdateApprovalRuleTemplateDescriptionCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateApprovalRuleTemplateDescription", {})
    .n("CodeCommitClient", "UpdateApprovalRuleTemplateDescriptionCommand")
    .sc(schemas_0.UpdateApprovalRuleTemplateDescription$)
    .build() {
}

class UpdateApprovalRuleTemplateNameCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateApprovalRuleTemplateName", {})
    .n("CodeCommitClient", "UpdateApprovalRuleTemplateNameCommand")
    .sc(schemas_0.UpdateApprovalRuleTemplateName$)
    .build() {
}

class UpdateCommentCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateComment", {})
    .n("CodeCommitClient", "UpdateCommentCommand")
    .sc(schemas_0.UpdateComment$)
    .build() {
}

class UpdateDefaultBranchCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateDefaultBranch", {})
    .n("CodeCommitClient", "UpdateDefaultBranchCommand")
    .sc(schemas_0.UpdateDefaultBranch$)
    .build() {
}

class UpdatePullRequestApprovalRuleContentCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdatePullRequestApprovalRuleContent", {})
    .n("CodeCommitClient", "UpdatePullRequestApprovalRuleContentCommand")
    .sc(schemas_0.UpdatePullRequestApprovalRuleContent$)
    .build() {
}

class UpdatePullRequestApprovalStateCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdatePullRequestApprovalState", {})
    .n("CodeCommitClient", "UpdatePullRequestApprovalStateCommand")
    .sc(schemas_0.UpdatePullRequestApprovalState$)
    .build() {
}

class UpdatePullRequestDescriptionCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdatePullRequestDescription", {})
    .n("CodeCommitClient", "UpdatePullRequestDescriptionCommand")
    .sc(schemas_0.UpdatePullRequestDescription$)
    .build() {
}

class UpdatePullRequestStatusCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdatePullRequestStatus", {})
    .n("CodeCommitClient", "UpdatePullRequestStatusCommand")
    .sc(schemas_0.UpdatePullRequestStatus$)
    .build() {
}

class UpdatePullRequestTitleCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdatePullRequestTitle", {})
    .n("CodeCommitClient", "UpdatePullRequestTitleCommand")
    .sc(schemas_0.UpdatePullRequestTitle$)
    .build() {
}

class UpdateRepositoryDescriptionCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateRepositoryDescription", {})
    .n("CodeCommitClient", "UpdateRepositoryDescriptionCommand")
    .sc(schemas_0.UpdateRepositoryDescription$)
    .build() {
}

class UpdateRepositoryEncryptionKeyCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateRepositoryEncryptionKey", {})
    .n("CodeCommitClient", "UpdateRepositoryEncryptionKeyCommand")
    .sc(schemas_0.UpdateRepositoryEncryptionKey$)
    .build() {
}

class UpdateRepositoryNameCommand extends client.Command
    .classBuilder()
    .ep(commonParams)
    .m(function (Command, cs, config, o) {
    return [endpoints.getEndpointPlugin(config, Command.getEndpointParameterInstructions())];
})
    .s("CodeCommit_20150413", "UpdateRepositoryName", {})
    .n("CodeCommitClient", "UpdateRepositoryNameCommand")
    .sc(schemas_0.UpdateRepositoryName$)
    .build() {
}

const paginateDescribeMergeConflicts = core.createPaginator(CodeCommitClient, DescribeMergeConflictsCommand, "nextToken", "nextToken", "maxMergeHunks");

const paginateDescribePullRequestEvents = core.createPaginator(CodeCommitClient, DescribePullRequestEventsCommand, "nextToken", "nextToken", "maxResults");

const paginateGetCommentReactions = core.createPaginator(CodeCommitClient, GetCommentReactionsCommand, "nextToken", "nextToken", "maxResults");

const paginateGetCommentsForComparedCommit = core.createPaginator(CodeCommitClient, GetCommentsForComparedCommitCommand, "nextToken", "nextToken", "maxResults");

const paginateGetCommentsForPullRequest = core.createPaginator(CodeCommitClient, GetCommentsForPullRequestCommand, "nextToken", "nextToken", "maxResults");

const paginateGetDifferences = core.createPaginator(CodeCommitClient, GetDifferencesCommand, "NextToken", "NextToken", "MaxResults");

const paginateGetMergeConflicts = core.createPaginator(CodeCommitClient, GetMergeConflictsCommand, "nextToken", "nextToken", "maxConflictFiles");

const paginateListApprovalRuleTemplates = core.createPaginator(CodeCommitClient, ListApprovalRuleTemplatesCommand, "nextToken", "nextToken", "maxResults");

const paginateListAssociatedApprovalRuleTemplatesForRepository = core.createPaginator(CodeCommitClient, ListAssociatedApprovalRuleTemplatesForRepositoryCommand, "nextToken", "nextToken", "maxResults");

const paginateListBranches = core.createPaginator(CodeCommitClient, ListBranchesCommand, "nextToken", "nextToken", "");

const paginateListFileCommitHistory = core.createPaginator(CodeCommitClient, ListFileCommitHistoryCommand, "nextToken", "nextToken", "maxResults");

const paginateListPullRequests = core.createPaginator(CodeCommitClient, ListPullRequestsCommand, "nextToken", "nextToken", "maxResults");

const paginateListRepositoriesForApprovalRuleTemplate = core.createPaginator(CodeCommitClient, ListRepositoriesForApprovalRuleTemplateCommand, "nextToken", "nextToken", "maxResults");

const paginateListRepositories = core.createPaginator(CodeCommitClient, ListRepositoriesCommand, "nextToken", "nextToken", "");

const commands = {
    AssociateApprovalRuleTemplateWithRepositoryCommand,
    BatchAssociateApprovalRuleTemplateWithRepositoriesCommand,
    BatchDescribeMergeConflictsCommand,
    BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand,
    BatchGetCommitsCommand,
    BatchGetRepositoriesCommand,
    CreateApprovalRuleTemplateCommand,
    CreateBranchCommand,
    CreateCommitCommand,
    CreatePullRequestCommand,
    CreatePullRequestApprovalRuleCommand,
    CreateRepositoryCommand,
    CreateUnreferencedMergeCommitCommand,
    DeleteApprovalRuleTemplateCommand,
    DeleteBranchCommand,
    DeleteCommentContentCommand,
    DeleteFileCommand,
    DeletePullRequestApprovalRuleCommand,
    DeleteRepositoryCommand,
    DescribeMergeConflictsCommand,
    DescribePullRequestEventsCommand,
    DisassociateApprovalRuleTemplateFromRepositoryCommand,
    EvaluatePullRequestApprovalRulesCommand,
    GetApprovalRuleTemplateCommand,
    GetBlobCommand,
    GetBranchCommand,
    GetCommentCommand,
    GetCommentReactionsCommand,
    GetCommentsForComparedCommitCommand,
    GetCommentsForPullRequestCommand,
    GetCommitCommand,
    GetDifferencesCommand,
    GetFileCommand,
    GetFolderCommand,
    GetMergeCommitCommand,
    GetMergeConflictsCommand,
    GetMergeOptionsCommand,
    GetPullRequestCommand,
    GetPullRequestApprovalStatesCommand,
    GetPullRequestOverrideStateCommand,
    GetRepositoryCommand,
    GetRepositoryTriggersCommand,
    ListApprovalRuleTemplatesCommand,
    ListAssociatedApprovalRuleTemplatesForRepositoryCommand,
    ListBranchesCommand,
    ListFileCommitHistoryCommand,
    ListPullRequestsCommand,
    ListRepositoriesCommand,
    ListRepositoriesForApprovalRuleTemplateCommand,
    ListTagsForResourceCommand,
    MergeBranchesByFastForwardCommand,
    MergeBranchesBySquashCommand,
    MergeBranchesByThreeWayCommand,
    MergePullRequestByFastForwardCommand,
    MergePullRequestBySquashCommand,
    MergePullRequestByThreeWayCommand,
    OverridePullRequestApprovalRulesCommand,
    PostCommentForComparedCommitCommand,
    PostCommentForPullRequestCommand,
    PostCommentReplyCommand,
    PutCommentReactionCommand,
    PutFileCommand,
    PutRepositoryTriggersCommand,
    TagResourceCommand,
    TestRepositoryTriggersCommand,
    UntagResourceCommand,
    UpdateApprovalRuleTemplateContentCommand,
    UpdateApprovalRuleTemplateDescriptionCommand,
    UpdateApprovalRuleTemplateNameCommand,
    UpdateCommentCommand,
    UpdateDefaultBranchCommand,
    UpdatePullRequestApprovalRuleContentCommand,
    UpdatePullRequestApprovalStateCommand,
    UpdatePullRequestDescriptionCommand,
    UpdatePullRequestStatusCommand,
    UpdatePullRequestTitleCommand,
    UpdateRepositoryDescriptionCommand,
    UpdateRepositoryEncryptionKeyCommand,
    UpdateRepositoryNameCommand,
};
const paginators = {
    paginateDescribeMergeConflicts,
    paginateDescribePullRequestEvents,
    paginateGetCommentReactions,
    paginateGetCommentsForComparedCommit,
    paginateGetCommentsForPullRequest,
    paginateGetDifferences,
    paginateGetMergeConflicts,
    paginateListApprovalRuleTemplates,
    paginateListAssociatedApprovalRuleTemplatesForRepository,
    paginateListBranches,
    paginateListFileCommitHistory,
    paginateListPullRequests,
    paginateListRepositories,
    paginateListRepositoriesForApprovalRuleTemplate,
};
class CodeCommit extends CodeCommitClient {
}
client.createAggregatedClient(commands, CodeCommit, { paginators });

const ApprovalState = {
    APPROVE: "APPROVE",
    REVOKE: "REVOKE",
};
const OverrideStatus = {
    OVERRIDE: "OVERRIDE",
    REVOKE: "REVOKE",
};
const ConflictDetailLevelTypeEnum = {
    FILE_LEVEL: "FILE_LEVEL",
    LINE_LEVEL: "LINE_LEVEL",
};
const ConflictResolutionStrategyTypeEnum = {
    ACCEPT_DESTINATION: "ACCEPT_DESTINATION",
    ACCEPT_SOURCE: "ACCEPT_SOURCE",
    AUTOMERGE: "AUTOMERGE",
    NONE: "NONE",
};
const MergeOptionTypeEnum = {
    FAST_FORWARD_MERGE: "FAST_FORWARD_MERGE",
    SQUASH_MERGE: "SQUASH_MERGE",
    THREE_WAY_MERGE: "THREE_WAY_MERGE",
};
const FileModeTypeEnum = {
    EXECUTABLE: "EXECUTABLE",
    NORMAL: "NORMAL",
    SYMLINK: "SYMLINK",
};
const ChangeTypeEnum = {
    ADDED: "A",
    DELETED: "D",
    MODIFIED: "M",
};
const ObjectTypeEnum = {
    DIRECTORY: "DIRECTORY",
    FILE: "FILE",
    GIT_LINK: "GIT_LINK",
    SYMBOLIC_LINK: "SYMBOLIC_LINK",
};
const BatchGetRepositoriesErrorCodeEnum = {
    ENCRYPTION_INTEGRITY_CHECKS_FAILED_EXCEPTION: "EncryptionIntegrityChecksFailedException",
    ENCRYPTION_KEY_ACCESS_DENIED_EXCEPTION: "EncryptionKeyAccessDeniedException",
    ENCRYPTION_KEY_DISABLED_EXCEPTION: "EncryptionKeyDisabledException",
    ENCRYPTION_KEY_NOT_FOUND_EXCEPTION: "EncryptionKeyNotFoundException",
    ENCRYPTION_KEY_UNAVAILABLE_EXCEPTION: "EncryptionKeyUnavailableException",
    REPOSITORY_DOES_NOT_EXIST_EXCEPTION: "RepositoryDoesNotExistException",
};
const PullRequestStatusEnum = {
    CLOSED: "CLOSED",
    OPEN: "OPEN",
};
const ReplacementTypeEnum = {
    KEEP_BASE: "KEEP_BASE",
    KEEP_DESTINATION: "KEEP_DESTINATION",
    KEEP_SOURCE: "KEEP_SOURCE",
    USE_NEW_CONTENT: "USE_NEW_CONTENT",
};
const PullRequestEventType = {
    PULL_REQUEST_APPROVAL_RULE_CREATED: "PULL_REQUEST_APPROVAL_RULE_CREATED",
    PULL_REQUEST_APPROVAL_RULE_DELETED: "PULL_REQUEST_APPROVAL_RULE_DELETED",
    PULL_REQUEST_APPROVAL_RULE_OVERRIDDEN: "PULL_REQUEST_APPROVAL_RULE_OVERRIDDEN",
    PULL_REQUEST_APPROVAL_RULE_UPDATED: "PULL_REQUEST_APPROVAL_RULE_UPDATED",
    PULL_REQUEST_APPROVAL_STATE_CHANGED: "PULL_REQUEST_APPROVAL_STATE_CHANGED",
    PULL_REQUEST_CREATED: "PULL_REQUEST_CREATED",
    PULL_REQUEST_MERGE_STATE_CHANGED: "PULL_REQUEST_MERGE_STATE_CHANGED",
    PULL_REQUEST_SOURCE_REFERENCE_UPDATED: "PULL_REQUEST_SOURCE_REFERENCE_UPDATED",
    PULL_REQUEST_STATUS_CHANGED: "PULL_REQUEST_STATUS_CHANGED",
};
const RelativeFileVersionEnum = {
    AFTER: "AFTER",
    BEFORE: "BEFORE",
};
const RepositoryTriggerEventEnum = {
    ALL: "all",
    CREATE_REFERENCE: "createReference",
    DELETE_REFERENCE: "deleteReference",
    UPDATE_REFERENCE: "updateReference",
};
const OrderEnum = {
    ASCENDING: "ascending",
    DESCENDING: "descending",
};
const SortByEnum = {
    MODIFIED_DATE: "lastModifiedDate",
    REPOSITORY_NAME: "repositoryName",
};

exports.$Command = client.Command;
exports.__Client = client.Client;
exports.CodeCommitServiceException = CodeCommitServiceException.CodeCommitServiceException;
exports.ApprovalState = ApprovalState;
exports.AssociateApprovalRuleTemplateWithRepositoryCommand = AssociateApprovalRuleTemplateWithRepositoryCommand;
exports.BatchAssociateApprovalRuleTemplateWithRepositoriesCommand = BatchAssociateApprovalRuleTemplateWithRepositoriesCommand;
exports.BatchDescribeMergeConflictsCommand = BatchDescribeMergeConflictsCommand;
exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand = BatchDisassociateApprovalRuleTemplateFromRepositoriesCommand;
exports.BatchGetCommitsCommand = BatchGetCommitsCommand;
exports.BatchGetRepositoriesCommand = BatchGetRepositoriesCommand;
exports.BatchGetRepositoriesErrorCodeEnum = BatchGetRepositoriesErrorCodeEnum;
exports.ChangeTypeEnum = ChangeTypeEnum;
exports.CodeCommit = CodeCommit;
exports.CodeCommitClient = CodeCommitClient;
exports.ConflictDetailLevelTypeEnum = ConflictDetailLevelTypeEnum;
exports.ConflictResolutionStrategyTypeEnum = ConflictResolutionStrategyTypeEnum;
exports.CreateApprovalRuleTemplateCommand = CreateApprovalRuleTemplateCommand;
exports.CreateBranchCommand = CreateBranchCommand;
exports.CreateCommitCommand = CreateCommitCommand;
exports.CreatePullRequestApprovalRuleCommand = CreatePullRequestApprovalRuleCommand;
exports.CreatePullRequestCommand = CreatePullRequestCommand;
exports.CreateRepositoryCommand = CreateRepositoryCommand;
exports.CreateUnreferencedMergeCommitCommand = CreateUnreferencedMergeCommitCommand;
exports.DeleteApprovalRuleTemplateCommand = DeleteApprovalRuleTemplateCommand;
exports.DeleteBranchCommand = DeleteBranchCommand;
exports.DeleteCommentContentCommand = DeleteCommentContentCommand;
exports.DeleteFileCommand = DeleteFileCommand;
exports.DeletePullRequestApprovalRuleCommand = DeletePullRequestApprovalRuleCommand;
exports.DeleteRepositoryCommand = DeleteRepositoryCommand;
exports.DescribeMergeConflictsCommand = DescribeMergeConflictsCommand;
exports.DescribePullRequestEventsCommand = DescribePullRequestEventsCommand;
exports.DisassociateApprovalRuleTemplateFromRepositoryCommand = DisassociateApprovalRuleTemplateFromRepositoryCommand;
exports.EvaluatePullRequestApprovalRulesCommand = EvaluatePullRequestApprovalRulesCommand;
exports.FileModeTypeEnum = FileModeTypeEnum;
exports.GetApprovalRuleTemplateCommand = GetApprovalRuleTemplateCommand;
exports.GetBlobCommand = GetBlobCommand;
exports.GetBranchCommand = GetBranchCommand;
exports.GetCommentCommand = GetCommentCommand;
exports.GetCommentReactionsCommand = GetCommentReactionsCommand;
exports.GetCommentsForComparedCommitCommand = GetCommentsForComparedCommitCommand;
exports.GetCommentsForPullRequestCommand = GetCommentsForPullRequestCommand;
exports.GetCommitCommand = GetCommitCommand;
exports.GetDifferencesCommand = GetDifferencesCommand;
exports.GetFileCommand = GetFileCommand;
exports.GetFolderCommand = GetFolderCommand;
exports.GetMergeCommitCommand = GetMergeCommitCommand;
exports.GetMergeConflictsCommand = GetMergeConflictsCommand;
exports.GetMergeOptionsCommand = GetMergeOptionsCommand;
exports.GetPullRequestApprovalStatesCommand = GetPullRequestApprovalStatesCommand;
exports.GetPullRequestCommand = GetPullRequestCommand;
exports.GetPullRequestOverrideStateCommand = GetPullRequestOverrideStateCommand;
exports.GetRepositoryCommand = GetRepositoryCommand;
exports.GetRepositoryTriggersCommand = GetRepositoryTriggersCommand;
exports.ListApprovalRuleTemplatesCommand = ListApprovalRuleTemplatesCommand;
exports.ListAssociatedApprovalRuleTemplatesForRepositoryCommand = ListAssociatedApprovalRuleTemplatesForRepositoryCommand;
exports.ListBranchesCommand = ListBranchesCommand;
exports.ListFileCommitHistoryCommand = ListFileCommitHistoryCommand;
exports.ListPullRequestsCommand = ListPullRequestsCommand;
exports.ListRepositoriesCommand = ListRepositoriesCommand;
exports.ListRepositoriesForApprovalRuleTemplateCommand = ListRepositoriesForApprovalRuleTemplateCommand;
exports.ListTagsForResourceCommand = ListTagsForResourceCommand;
exports.MergeBranchesByFastForwardCommand = MergeBranchesByFastForwardCommand;
exports.MergeBranchesBySquashCommand = MergeBranchesBySquashCommand;
exports.MergeBranchesByThreeWayCommand = MergeBranchesByThreeWayCommand;
exports.MergeOptionTypeEnum = MergeOptionTypeEnum;
exports.MergePullRequestByFastForwardCommand = MergePullRequestByFastForwardCommand;
exports.MergePullRequestBySquashCommand = MergePullRequestBySquashCommand;
exports.MergePullRequestByThreeWayCommand = MergePullRequestByThreeWayCommand;
exports.ObjectTypeEnum = ObjectTypeEnum;
exports.OrderEnum = OrderEnum;
exports.OverridePullRequestApprovalRulesCommand = OverridePullRequestApprovalRulesCommand;
exports.OverrideStatus = OverrideStatus;
exports.PostCommentForComparedCommitCommand = PostCommentForComparedCommitCommand;
exports.PostCommentForPullRequestCommand = PostCommentForPullRequestCommand;
exports.PostCommentReplyCommand = PostCommentReplyCommand;
exports.PullRequestEventType = PullRequestEventType;
exports.PullRequestStatusEnum = PullRequestStatusEnum;
exports.PutCommentReactionCommand = PutCommentReactionCommand;
exports.PutFileCommand = PutFileCommand;
exports.PutRepositoryTriggersCommand = PutRepositoryTriggersCommand;
exports.RelativeFileVersionEnum = RelativeFileVersionEnum;
exports.ReplacementTypeEnum = ReplacementTypeEnum;
exports.RepositoryTriggerEventEnum = RepositoryTriggerEventEnum;
exports.SortByEnum = SortByEnum;
exports.TagResourceCommand = TagResourceCommand;
exports.TestRepositoryTriggersCommand = TestRepositoryTriggersCommand;
exports.UntagResourceCommand = UntagResourceCommand;
exports.UpdateApprovalRuleTemplateContentCommand = UpdateApprovalRuleTemplateContentCommand;
exports.UpdateApprovalRuleTemplateDescriptionCommand = UpdateApprovalRuleTemplateDescriptionCommand;
exports.UpdateApprovalRuleTemplateNameCommand = UpdateApprovalRuleTemplateNameCommand;
exports.UpdateCommentCommand = UpdateCommentCommand;
exports.UpdateDefaultBranchCommand = UpdateDefaultBranchCommand;
exports.UpdatePullRequestApprovalRuleContentCommand = UpdatePullRequestApprovalRuleContentCommand;
exports.UpdatePullRequestApprovalStateCommand = UpdatePullRequestApprovalStateCommand;
exports.UpdatePullRequestDescriptionCommand = UpdatePullRequestDescriptionCommand;
exports.UpdatePullRequestStatusCommand = UpdatePullRequestStatusCommand;
exports.UpdatePullRequestTitleCommand = UpdatePullRequestTitleCommand;
exports.UpdateRepositoryDescriptionCommand = UpdateRepositoryDescriptionCommand;
exports.UpdateRepositoryEncryptionKeyCommand = UpdateRepositoryEncryptionKeyCommand;
exports.UpdateRepositoryNameCommand = UpdateRepositoryNameCommand;
exports.paginateDescribeMergeConflicts = paginateDescribeMergeConflicts;
exports.paginateDescribePullRequestEvents = paginateDescribePullRequestEvents;
exports.paginateGetCommentReactions = paginateGetCommentReactions;
exports.paginateGetCommentsForComparedCommit = paginateGetCommentsForComparedCommit;
exports.paginateGetCommentsForPullRequest = paginateGetCommentsForPullRequest;
exports.paginateGetDifferences = paginateGetDifferences;
exports.paginateGetMergeConflicts = paginateGetMergeConflicts;
exports.paginateListApprovalRuleTemplates = paginateListApprovalRuleTemplates;
exports.paginateListAssociatedApprovalRuleTemplatesForRepository = paginateListAssociatedApprovalRuleTemplatesForRepository;
exports.paginateListBranches = paginateListBranches;
exports.paginateListFileCommitHistory = paginateListFileCommitHistory;
exports.paginateListPullRequests = paginateListPullRequests;
exports.paginateListRepositories = paginateListRepositories;
exports.paginateListRepositoriesForApprovalRuleTemplate = paginateListRepositoriesForApprovalRuleTemplate;
Object.prototype.hasOwnProperty.call(schemas_0, '__proto__') &&
    !Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
    Object.defineProperty(exports, '__proto__', {
        enumerable: true,
        value: schemas_0['__proto__']
    });

Object.keys(schemas_0).forEach(function (k) {
    if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = schemas_0[k];
});
Object.prototype.hasOwnProperty.call(errors, '__proto__') &&
    !Object.prototype.hasOwnProperty.call(exports, '__proto__') &&
    Object.defineProperty(exports, '__proto__', {
        enumerable: true,
        value: errors['__proto__']
    });

Object.keys(errors).forEach(function (k) {
    if (k !== 'default' && !Object.prototype.hasOwnProperty.call(exports, k)) exports[k] = errors[k];
});
