// @aws-sdk/client-eks/dist-cjs/models/errors.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InvalidStateException = exports.ResourcePropagationDelayException = exports.NotFoundException = exports.BadRequestException = exports.UnsupportedAvailabilityZoneException = exports.ServiceUnavailableException = exports.ResourceLimitExceededException = exports.ThrottlingException = exports.ResourceInUseException = exports.ClientException = exports.ServerException = exports.ResourceNotFoundException = exports.InvalidRequestException = exports.InvalidParameterException = exports.AccessDeniedException = void 0;
const EKSServiceException_1 = require("./EKSServiceException");
class AccessDeniedException extends EKSServiceException_1.EKSServiceException {
    name = "AccessDeniedException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "AccessDeniedException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, AccessDeniedException.prototype);
    }
}
exports.AccessDeniedException = AccessDeniedException;
class InvalidParameterException extends EKSServiceException_1.EKSServiceException {
    name = "InvalidParameterException";
    $fault = "client";
    clusterName;
    nodegroupName;
    fargateProfileName;
    addonName;
    subscriptionId;
    constructor(opts) {
        super({
            name: "InvalidParameterException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidParameterException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.fargateProfileName = opts.fargateProfileName;
        this.addonName = opts.addonName;
        this.subscriptionId = opts.subscriptionId;
    }
}
exports.InvalidParameterException = InvalidParameterException;
class InvalidRequestException extends EKSServiceException_1.EKSServiceException {
    name = "InvalidRequestException";
    $fault = "client";
    clusterName;
    nodegroupName;
    addonName;
    subscriptionId;
    constructor(opts) {
        super({
            name: "InvalidRequestException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidRequestException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.addonName = opts.addonName;
        this.subscriptionId = opts.subscriptionId;
    }
}
exports.InvalidRequestException = InvalidRequestException;
class ResourceNotFoundException extends EKSServiceException_1.EKSServiceException {
    name = "ResourceNotFoundException";
    $fault = "client";
    clusterName;
    nodegroupName;
    fargateProfileName;
    addonName;
    subscriptionId;
    constructor(opts) {
        super({
            name: "ResourceNotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceNotFoundException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.fargateProfileName = opts.fargateProfileName;
        this.addonName = opts.addonName;
        this.subscriptionId = opts.subscriptionId;
    }
}
exports.ResourceNotFoundException = ResourceNotFoundException;
class ServerException extends EKSServiceException_1.EKSServiceException {
    name = "ServerException";
    $fault = "server";
    clusterName;
    nodegroupName;
    addonName;
    subscriptionId;
    constructor(opts) {
        super({
            name: "ServerException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, ServerException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.addonName = opts.addonName;
        this.subscriptionId = opts.subscriptionId;
    }
}
exports.ServerException = ServerException;
class ClientException extends EKSServiceException_1.EKSServiceException {
    name = "ClientException";
    $fault = "client";
    clusterName;
    nodegroupName;
    addonName;
    subscriptionId;
    constructor(opts) {
        super({
            name: "ClientException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ClientException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.addonName = opts.addonName;
        this.subscriptionId = opts.subscriptionId;
    }
}
exports.ClientException = ClientException;
class ResourceInUseException extends EKSServiceException_1.EKSServiceException {
    name = "ResourceInUseException";
    $fault = "client";
    clusterName;
    nodegroupName;
    addonName;
    constructor(opts) {
        super({
            name: "ResourceInUseException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceInUseException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.addonName = opts.addonName;
    }
}
exports.ResourceInUseException = ResourceInUseException;
class ThrottlingException extends EKSServiceException_1.EKSServiceException {
    name = "ThrottlingException";
    $fault = "client";
    clusterName;
    constructor(opts) {
        super({
            name: "ThrottlingException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ThrottlingException.prototype);
        this.clusterName = opts.clusterName;
    }
}
exports.ThrottlingException = ThrottlingException;
class ResourceLimitExceededException extends EKSServiceException_1.EKSServiceException {
    name = "ResourceLimitExceededException";
    $fault = "client";
    clusterName;
    nodegroupName;
    subscriptionId;
    constructor(opts) {
        super({
            name: "ResourceLimitExceededException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourceLimitExceededException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.subscriptionId = opts.subscriptionId;
    }
}
exports.ResourceLimitExceededException = ResourceLimitExceededException;
class ServiceUnavailableException extends EKSServiceException_1.EKSServiceException {
    name = "ServiceUnavailableException";
    $fault = "server";
    constructor(opts) {
        super({
            name: "ServiceUnavailableException",
            $fault: "server",
            ...opts,
        });
        Object.setPrototypeOf(this, ServiceUnavailableException.prototype);
    }
}
exports.ServiceUnavailableException = ServiceUnavailableException;
class UnsupportedAvailabilityZoneException extends EKSServiceException_1.EKSServiceException {
    name = "UnsupportedAvailabilityZoneException";
    $fault = "client";
    clusterName;
    nodegroupName;
    validZones;
    constructor(opts) {
        super({
            name: "UnsupportedAvailabilityZoneException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, UnsupportedAvailabilityZoneException.prototype);
        this.clusterName = opts.clusterName;
        this.nodegroupName = opts.nodegroupName;
        this.validZones = opts.validZones;
    }
}
exports.UnsupportedAvailabilityZoneException = UnsupportedAvailabilityZoneException;
class BadRequestException extends EKSServiceException_1.EKSServiceException {
    name = "BadRequestException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "BadRequestException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, BadRequestException.prototype);
    }
}
exports.BadRequestException = BadRequestException;
class NotFoundException extends EKSServiceException_1.EKSServiceException {
    name = "NotFoundException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "NotFoundException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, NotFoundException.prototype);
    }
}
exports.NotFoundException = NotFoundException;
class ResourcePropagationDelayException extends EKSServiceException_1.EKSServiceException {
    name = "ResourcePropagationDelayException";
    $fault = "client";
    constructor(opts) {
        super({
            name: "ResourcePropagationDelayException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, ResourcePropagationDelayException.prototype);
    }
}
exports.ResourcePropagationDelayException = ResourcePropagationDelayException;
class InvalidStateException extends EKSServiceException_1.EKSServiceException {
    name = "InvalidStateException";
    $fault = "client";
    clusterName;
    constructor(opts) {
        super({
            name: "InvalidStateException",
            $fault: "client",
            ...opts,
        });
        Object.setPrototypeOf(this, InvalidStateException.prototype);
        this.clusterName = opts.clusterName;
    }
}
exports.InvalidStateException = InvalidStateException;
