// renovate/dist/modules/manager/mise/artifacts.js
import "../../../constants/error-messages.js";
import { logger } from "../../../logger/index.js";
import { find } from "../../../util/host-rules.js";
import { readLocalFile, writeLocalFile } from "../../../util/fs/index.js";
import { exec } from "../../../util/exec/index.js";
import { getRepoStatus } from "../../../util/git/index.js";
import { findGithubToken } from "../../../util/check-token.js";
import { getConfigType, getLockFileName } from "./lockfile.js";
import { isNonEmptyStringAndNotWhitespace, isString } from "@sindresorhus/is";
import { quote } from "shlex";
//#region lib/modules/manager/mise/artifacts.ts
/**
* Updates mise lock files when dependencies are updated.
* Runs `mise lock` for lock file maintenance or `mise lock <tools>` for targeted updates.
*/
async function updateArtifacts({ packageFileName, updatedDeps, newPackageFileContent, config }) {
	const lockFileName = getLockFileName(packageFileName);
	if (!await readLocalFile(lockFileName, "utf8")) {
		logger.debug({ lockFileName }, "No mise lock file found");
		return null;
	}
	await writeLocalFile(packageFileName, newPackageFileContent);
	const { isLocal, env } = getConfigType(packageFileName);
	const localFlag = isLocal ? " --local" : "";
	let cmd;
	if (config.isLockFileMaintenance) cmd = `mise lock${localFlag}`;
	else {
		const tools = updatedDeps.map(({ depName }) => depName).filter(isNonEmptyStringAndNotWhitespace).map(quote).join(" ");
		cmd = tools ? `mise lock${localFlag} ${tools}` : `mise lock${localFlag}`;
	}
	const extraEnv = {};
	if (env) extraEnv.MISE_ENV = env;
	const token = findGithubToken(find({
		hostType: "github",
		url: "https://api.github.com/"
	}));
	if (token) extraEnv.GITHUB_TOKEN = token;
	const execOptions = {
		cwdFile: packageFileName,
		extraEnv,
		toolConstraints: [{
			toolName: "mise",
			constraint: config.constraints?.mise
		}],
		docker: {}
	};
	try {
		await exec(cmd, execOptions);
		if (!(await getRepoStatus()).modified.includes(lockFileName)) return null;
		logger.debug({ lockFileName }, "Returning updated mise lock file");
		return [{ file: {
			type: "addition",
			path: lockFileName,
			contents: await readLocalFile(lockFileName)
		} }];
	} catch (err) {
		// istanbul ignore if: not worth testing
		if (err.message === "temporary-error") throw err;
		const errorOutput = [
			err.stdout,
			err.stderr,
			err.message
		].filter(isString).join("\n");
		logger.warn({ err }, `Error updating ${lockFileName}`);
		return [{ artifactError: {
			fileName: lockFileName,
			stderr: errorOutput
		} }];
	}
}
//#endregion
export { updateArtifacts };
