// renovate/dist/modules/datasource/npm/schema.js
import { z } from "zod/v4";
//#region lib/modules/datasource/npm/schema.ts
const Repository = z.union([z.string(), z.object({
	url: z.string().optional(),
	directory: z.string().optional()
})]);
const Attestations = z.object({ url: z.string().optional() });
const Distribution = z.object({ attestations: Attestations.optional() });
const Version = z.object({
	repository: Repository.optional(),
	homepage: z.string().optional().catch(void 0),
	deprecated: z.union([z.string(), z.boolean()]).optional(),
	gitHead: z.string().optional(),
	dependencies: z.record(z.string(), z.string()).optional(),
	devDependencies: z.record(z.string(), z.string()).optional(),
	engines: z.object({ node: z.string().optional() }).optional().catch(void 0),
	dist: Distribution.optional()
});
const CachedPackument = z.object({
	versions: z.record(z.string(), Version).optional(),
	repository: Repository.optional(),
	homepage: z.string().optional(),
	time: z.record(z.string(), z.string()).optional(),
	"dist-tags": z.record(z.string(), z.string()).optional()
});
//#endregion
export { CachedPackument };
