// @aws-sdk/client-codecommit/dist-cjs/schemas/schemas_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FileContentSizeLimitExceededException$ = exports.FileContentRequiredException$ = exports.FileContentAndSourceFileSpecifiedException$ = exports.EncryptionKeyUnavailableException$ = exports.EncryptionKeyRequiredException$ = exports.EncryptionKeyNotFoundException$ = exports.EncryptionKeyInvalidUsageException$ = exports.EncryptionKeyInvalidIdException$ = exports.EncryptionKeyDisabledException$ = exports.EncryptionKeyAccessDeniedException$ = exports.EncryptionIntegrityChecksFailedException$ = exports.DirectoryNameConflictsWithFileNameException$ = exports.DefaultBranchCannotBeDeletedException$ = exports.ConcurrentReferenceUpdateException$ = exports.CommitRequiredException$ = exports.CommitMessageLengthExceededException$ = exports.CommitIdsListRequiredException$ = exports.CommitIdsLimitExceededException$ = exports.CommitIdRequiredException$ = exports.CommitIdDoesNotExistException$ = exports.CommitDoesNotExistException$ = exports.CommentNotCreatedByCallerException$ = exports.CommentIdRequiredException$ = exports.CommentDoesNotExistException$ = exports.CommentDeletedException$ = exports.CommentContentSizeLimitExceededException$ = exports.CommentContentRequiredException$ = exports.ClientRequestTokenRequiredException$ = exports.CannotModifyApprovalRuleFromTemplateException$ = exports.CannotDeleteApprovalRuleFromTemplateException$ = exports.BranchNameRequiredException$ = exports.BranchNameIsTagNameException$ = exports.BranchNameExistsException$ = exports.BranchDoesNotExistException$ = exports.BlobIdRequiredException$ = exports.BlobIdDoesNotExistException$ = exports.BeforeCommitIdAndAfterCommitIdAreSameException$ = exports.AuthorDoesNotExistException$ = exports.ApprovalStateRequiredException$ = exports.ApprovalRuleTemplateNameRequiredException$ = exports.ApprovalRuleTemplateNameAlreadyExistsException$ = exports.ApprovalRuleTemplateInUseException$ = exports.ApprovalRuleTemplateDoesNotExistException$ = exports.ApprovalRuleTemplateContentRequiredException$ = exports.ApprovalRuleNameRequiredException$ = exports.ApprovalRuleNameAlreadyExistsException$ = exports.ApprovalRuleDoesNotExistException$ = exports.ApprovalRuleContentRequiredException$ = exports.ActorDoesNotExistException$ = exports.CodeCommitServiceException$ = void 0;
exports.InvalidRelativeFileVersionEnumException$ = exports.InvalidReferenceNameException$ = exports.InvalidReactionValueException$ = exports.InvalidReactionUserArnException$ = exports.InvalidPullRequestStatusUpdateException$ = exports.InvalidPullRequestStatusException$ = exports.InvalidPullRequestIdException$ = exports.InvalidPullRequestEventTypeException$ = exports.InvalidPathException$ = exports.InvalidParentCommitIdException$ = exports.InvalidOverrideStatusException$ = exports.InvalidOrderException$ = exports.InvalidMergeOptionException$ = exports.InvalidMaxResultsException$ = exports.InvalidMaxMergeHunksException$ = exports.InvalidMaxConflictFilesException$ = exports.InvalidFilePositionException$ = exports.InvalidFileModeException$ = exports.InvalidFileLocationException$ = exports.InvalidEmailException$ = exports.InvalidDestinationCommitSpecifierException$ = exports.InvalidDescriptionException$ = exports.InvalidDeletionParameterException$ = exports.InvalidContinuationTokenException$ = exports.InvalidConflictResolutionStrategyException$ = exports.InvalidConflictResolutionException$ = exports.InvalidConflictDetailLevelException$ = exports.InvalidCommitIdException$ = exports.InvalidCommitException$ = exports.InvalidCommentIdException$ = exports.InvalidClientRequestTokenException$ = exports.InvalidBranchNameException$ = exports.InvalidBlobIdException$ = exports.InvalidAuthorArnException$ = exports.InvalidApprovalStateException$ = exports.InvalidApprovalRuleTemplateNameException$ = exports.InvalidApprovalRuleTemplateDescriptionException$ = exports.InvalidApprovalRuleTemplateContentException$ = exports.InvalidApprovalRuleNameException$ = exports.InvalidApprovalRuleContentException$ = exports.InvalidActorArnException$ = exports.IdempotencyParameterMismatchException$ = exports.FolderDoesNotExistException$ = exports.FolderContentSizeLimitExceededException$ = exports.FileTooLargeException$ = exports.FilePathConflictsWithSubmodulePathException$ = exports.FileNameConflictsWithDirectoryNameException$ = exports.FileModeRequiredException$ = exports.FileEntryRequiredException$ = exports.FileDoesNotExistException$ = void 0;
exports.PullRequestApprovalRulesNotSatisfiedException$ = exports.PullRequestAlreadyClosedException$ = exports.PathRequiredException$ = exports.PathDoesNotExistException$ = exports.ParentCommitIdRequiredException$ = exports.ParentCommitIdOutdatedException$ = exports.ParentCommitDoesNotExistException$ = exports.OverrideStatusRequiredException$ = exports.OverrideAlreadySetException$ = exports.OperationNotAllowedException$ = exports.NumberOfRuleTemplatesExceededException$ = exports.NumberOfRulesExceededException$ = exports.NoChangeException$ = exports.NameLengthExceededException$ = exports.MultipleRepositoriesInPullRequestException$ = exports.MultipleConflictResolutionEntriesException$ = exports.MergeOptionRequiredException$ = exports.MaximumRuleTemplatesAssociatedWithRepositoryException$ = exports.MaximumRepositoryTriggersExceededException$ = exports.MaximumRepositoryNamesExceededException$ = exports.MaximumOpenPullRequestsExceededException$ = exports.MaximumNumberOfApprovalsExceededException$ = exports.MaximumItemsToCompareExceededException$ = exports.MaximumFileEntriesExceededException$ = exports.MaximumFileContentToLoadExceededException$ = exports.MaximumConflictResolutionEntriesExceededException$ = exports.MaximumBranchesExceededException$ = exports.ManualMergeRequiredException$ = exports.InvalidTitleException$ = exports.InvalidTargetsException$ = exports.InvalidTargetException$ = exports.InvalidTargetBranchException$ = exports.InvalidTagsMapException$ = exports.InvalidTagKeysListException$ = exports.InvalidSystemTagUsageException$ = exports.InvalidSourceCommitSpecifierException$ = exports.InvalidSortByException$ = exports.InvalidRuleContentSha256Exception$ = exports.InvalidRevisionIdException$ = exports.InvalidResourceArnException$ = exports.InvalidRepositoryTriggerRegionException$ = exports.InvalidRepositoryTriggerNameException$ = exports.InvalidRepositoryTriggerEventsException$ = exports.InvalidRepositoryTriggerDestinationArnException$ = exports.InvalidRepositoryTriggerCustomDataException$ = exports.InvalidRepositoryTriggerBranchNameException$ = exports.InvalidRepositoryNameException$ = exports.InvalidRepositoryDescriptionException$ = exports.InvalidReplacementTypeException$ = exports.InvalidReplacementContentException$ = void 0;
exports.BatchAssociateApprovalRuleTemplateWithRepositoriesInput$ = exports.BatchAssociateApprovalRuleTemplateWithRepositoriesError$ = exports.AssociateApprovalRuleTemplateWithRepositoryInput$ = exports.ApprovalStateChangedEventMetadata$ = exports.ApprovalRuleTemplate$ = exports.ApprovalRuleOverriddenEventMetadata$ = exports.ApprovalRuleEventMetadata$ = exports.ApprovalRule$ = exports.Approval$ = exports.errorTypeRegistries = exports.TooManyTagsException$ = exports.TitleRequiredException$ = exports.TipsDivergenceExceededException$ = exports.TipOfSourceReferenceIsDifferentException$ = exports.TargetsRequiredException$ = exports.TargetRequiredException$ = exports.TagsMapRequiredException$ = exports.TagPolicyException$ = exports.TagKeysListRequiredException$ = exports.SourceFileOrContentRequiredException$ = exports.SourceAndDestinationAreSameException$ = exports.SamePathRequestException$ = exports.SameFileContentException$ = exports.RevisionNotCurrentException$ = exports.RevisionIdRequiredException$ = exports.RestrictedSourceFileException$ = exports.ResourceArnRequiredException$ = exports.RepositoryTriggersListRequiredException$ = exports.RepositoryTriggerNameRequiredException$ = exports.RepositoryTriggerEventsListRequiredException$ = exports.RepositoryTriggerDestinationArnRequiredException$ = exports.RepositoryTriggerBranchNameListRequiredException$ = exports.RepositoryNotAssociatedWithPullRequestException$ = exports.RepositoryNamesRequiredException$ = exports.RepositoryNameRequiredException$ = exports.RepositoryNameExistsException$ = exports.RepositoryLimitExceededException$ = exports.RepositoryDoesNotExistException$ = exports.ReplacementTypeRequiredException$ = exports.ReplacementContentRequiredException$ = exports.ReferenceTypeNotSupportedException$ = exports.ReferenceNameRequiredException$ = exports.ReferenceDoesNotExistException$ = exports.ReactionValueRequiredException$ = exports.ReactionLimitExceededException$ = exports.PutFileEntryConflictException$ = exports.PullRequestStatusRequiredException$ = exports.PullRequestIdRequiredException$ = exports.PullRequestDoesNotExistException$ = exports.PullRequestCannotBeApprovedByAuthorException$ = void 0;
exports.DescribeMergeConflictsOutput$ = exports.DescribeMergeConflictsInput$ = exports.DeleteRepositoryOutput$ = exports.DeleteRepositoryInput$ = exports.DeletePullRequestApprovalRuleOutput$ = exports.DeletePullRequestApprovalRuleInput$ = exports.DeleteFileOutput$ = exports.DeleteFileInput$ = exports.DeleteFileEntry$ = exports.DeleteCommentContentOutput$ = exports.DeleteCommentContentInput$ = exports.DeleteBranchOutput$ = exports.DeleteBranchInput$ = exports.DeleteApprovalRuleTemplateOutput$ = exports.DeleteApprovalRuleTemplateInput$ = exports.CreateUnreferencedMergeCommitOutput$ = exports.CreateUnreferencedMergeCommitInput$ = exports.CreateRepositoryOutput$ = exports.CreateRepositoryInput$ = exports.CreatePullRequestOutput$ = exports.CreatePullRequestInput$ = exports.CreatePullRequestApprovalRuleOutput$ = exports.CreatePullRequestApprovalRuleInput$ = exports.CreateCommitOutput$ = exports.CreateCommitInput$ = exports.CreateBranchInput$ = exports.CreateApprovalRuleTemplateOutput$ = exports.CreateApprovalRuleTemplateInput$ = exports.ConflictResolution$ = exports.ConflictMetadata$ = exports.Conflict$ = exports.Commit$ = exports.CommentsForPullRequest$ = exports.CommentsForComparedCommit$ = exports.Comment$ = exports.BranchInfo$ = exports.BlobMetadata$ = exports.BatchGetRepositoriesOutput$ = exports.BatchGetRepositoriesInput$ = exports.BatchGetRepositoriesError$ = exports.BatchGetCommitsOutput$ = exports.BatchGetCommitsInput$ = exports.BatchGetCommitsError$ = exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesOutput$ = exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesInput$ = exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesError$ = exports.BatchDescribeMergeConflictsOutput$ = exports.BatchDescribeMergeConflictsInput$ = exports.BatchDescribeMergeConflictsError$ = exports.BatchAssociateApprovalRuleTemplateWithRepositoriesOutput$ = void 0;
exports.GetRepositoryTriggersInput$ = exports.GetRepositoryOutput$ = exports.GetRepositoryInput$ = exports.GetPullRequestOverrideStateOutput$ = exports.GetPullRequestOverrideStateInput$ = exports.GetPullRequestOutput$ = exports.GetPullRequestInput$ = exports.GetPullRequestApprovalStatesOutput$ = exports.GetPullRequestApprovalStatesInput$ = exports.GetMergeOptionsOutput$ = exports.GetMergeOptionsInput$ = exports.GetMergeConflictsOutput$ = exports.GetMergeConflictsInput$ = exports.GetMergeCommitOutput$ = exports.GetMergeCommitInput$ = exports.GetFolderOutput$ = exports.GetFolderInput$ = exports.GetFileOutput$ = exports.GetFileInput$ = exports.GetDifferencesOutput$ = exports.GetDifferencesInput$ = exports.GetCommitOutput$ = exports.GetCommitInput$ = exports.GetCommentsForPullRequestOutput$ = exports.GetCommentsForPullRequestInput$ = exports.GetCommentsForComparedCommitOutput$ = exports.GetCommentsForComparedCommitInput$ = exports.GetCommentReactionsOutput$ = exports.GetCommentReactionsInput$ = exports.GetCommentOutput$ = exports.GetCommentInput$ = exports.GetBranchOutput$ = exports.GetBranchInput$ = exports.GetBlobOutput$ = exports.GetBlobInput$ = exports.GetApprovalRuleTemplateOutput$ = exports.GetApprovalRuleTemplateInput$ = exports.Folder$ = exports.FileVersion$ = exports.FileSizes$ = exports.FileModes$ = exports.FileMetadata$ = exports.File$ = exports.Evaluation$ = exports.EvaluatePullRequestApprovalRulesOutput$ = exports.EvaluatePullRequestApprovalRulesInput$ = exports.DisassociateApprovalRuleTemplateFromRepositoryInput$ = exports.Difference$ = exports.DescribePullRequestEventsOutput$ = exports.DescribePullRequestEventsInput$ = void 0;
exports.PullRequestStatusChangedEventMetadata$ = exports.PullRequestSourceReferenceUpdatedEventMetadata$ = exports.PullRequestMergedStateChangedEventMetadata$ = exports.PullRequestEvent$ = exports.PullRequestCreatedEventMetadata$ = exports.PullRequest$ = exports.PostCommentReplyOutput$ = exports.PostCommentReplyInput$ = exports.PostCommentForPullRequestOutput$ = exports.PostCommentForPullRequestInput$ = exports.PostCommentForComparedCommitOutput$ = exports.PostCommentForComparedCommitInput$ = exports.OverridePullRequestApprovalRulesInput$ = exports.OriginApprovalRuleTemplate$ = exports.ObjectTypes$ = exports.MergePullRequestByThreeWayOutput$ = exports.MergePullRequestByThreeWayInput$ = exports.MergePullRequestBySquashOutput$ = exports.MergePullRequestBySquashInput$ = exports.MergePullRequestByFastForwardOutput$ = exports.MergePullRequestByFastForwardInput$ = exports.MergeOperations$ = exports.MergeMetadata$ = exports.MergeHunkDetail$ = exports.MergeHunk$ = exports.MergeBranchesByThreeWayOutput$ = exports.MergeBranchesByThreeWayInput$ = exports.MergeBranchesBySquashOutput$ = exports.MergeBranchesBySquashInput$ = exports.MergeBranchesByFastForwardOutput$ = exports.MergeBranchesByFastForwardInput$ = exports.Location$ = exports.ListTagsForResourceOutput$ = exports.ListTagsForResourceInput$ = exports.ListRepositoriesOutput$ = exports.ListRepositoriesInput$ = exports.ListRepositoriesForApprovalRuleTemplateOutput$ = exports.ListRepositoriesForApprovalRuleTemplateInput$ = exports.ListPullRequestsOutput$ = exports.ListPullRequestsInput$ = exports.ListFileCommitHistoryResponse$ = exports.ListFileCommitHistoryRequest$ = exports.ListBranchesOutput$ = exports.ListBranchesInput$ = exports.ListAssociatedApprovalRuleTemplatesForRepositoryOutput$ = exports.ListAssociatedApprovalRuleTemplatesForRepositoryInput$ = exports.ListApprovalRuleTemplatesOutput$ = exports.ListApprovalRuleTemplatesInput$ = exports.IsBinaryFile$ = exports.GetRepositoryTriggersOutput$ = void 0;
exports.BatchDisassociateApprovalRuleTemplateFromRepositories$ = exports.BatchDescribeMergeConflicts$ = exports.BatchAssociateApprovalRuleTemplateWithRepositories$ = exports.AssociateApprovalRuleTemplateWithRepository$ = exports.UserInfo$ = exports.UpdateRepositoryNameInput$ = exports.UpdateRepositoryEncryptionKeyOutput$ = exports.UpdateRepositoryEncryptionKeyInput$ = exports.UpdateRepositoryDescriptionInput$ = exports.UpdatePullRequestTitleOutput$ = exports.UpdatePullRequestTitleInput$ = exports.UpdatePullRequestStatusOutput$ = exports.UpdatePullRequestStatusInput$ = exports.UpdatePullRequestDescriptionOutput$ = exports.UpdatePullRequestDescriptionInput$ = exports.UpdatePullRequestApprovalStateInput$ = exports.UpdatePullRequestApprovalRuleContentOutput$ = exports.UpdatePullRequestApprovalRuleContentInput$ = exports.UpdateDefaultBranchInput$ = exports.UpdateCommentOutput$ = exports.UpdateCommentInput$ = exports.UpdateApprovalRuleTemplateNameOutput$ = exports.UpdateApprovalRuleTemplateNameInput$ = exports.UpdateApprovalRuleTemplateDescriptionOutput$ = exports.UpdateApprovalRuleTemplateDescriptionInput$ = exports.UpdateApprovalRuleTemplateContentOutput$ = exports.UpdateApprovalRuleTemplateContentInput$ = exports.UntagResourceInput$ = exports.TestRepositoryTriggersOutput$ = exports.TestRepositoryTriggersInput$ = exports.Target$ = exports.TagResourceInput$ = exports.SymbolicLink$ = exports.SubModule$ = exports.SourceFileSpecifier$ = exports.SetFileModeEntry$ = exports.RepositoryTriggerExecutionFailure$ = exports.RepositoryTrigger$ = exports.RepositoryNameIdPair$ = exports.RepositoryMetadata$ = exports.ReplaceContentEntry$ = exports.ReactionValueFormats$ = exports.ReactionForComment$ = exports.PutRepositoryTriggersOutput$ = exports.PutRepositoryTriggersInput$ = exports.PutFileOutput$ = exports.PutFileInput$ = exports.PutFileEntry$ = exports.PutCommentReactionInput$ = exports.PullRequestTarget$ = void 0;
exports.MergePullRequestByFastForward$ = exports.MergeBranchesByThreeWay$ = exports.MergeBranchesBySquash$ = exports.MergeBranchesByFastForward$ = exports.ListTagsForResource$ = exports.ListRepositoriesForApprovalRuleTemplate$ = exports.ListRepositories$ = exports.ListPullRequests$ = exports.ListFileCommitHistory$ = exports.ListBranches$ = exports.ListAssociatedApprovalRuleTemplatesForRepository$ = exports.ListApprovalRuleTemplates$ = exports.GetRepositoryTriggers$ = exports.GetRepository$ = exports.GetPullRequestOverrideState$ = exports.GetPullRequestApprovalStates$ = exports.GetPullRequest$ = exports.GetMergeOptions$ = exports.GetMergeConflicts$ = exports.GetMergeCommit$ = exports.GetFolder$ = exports.GetFile$ = exports.GetDifferences$ = exports.GetCommit$ = exports.GetCommentsForPullRequest$ = exports.GetCommentsForComparedCommit$ = exports.GetCommentReactions$ = exports.GetComment$ = exports.GetBranch$ = exports.GetBlob$ = exports.GetApprovalRuleTemplate$ = exports.EvaluatePullRequestApprovalRules$ = exports.DisassociateApprovalRuleTemplateFromRepository$ = exports.DescribePullRequestEvents$ = exports.DescribeMergeConflicts$ = exports.DeleteRepository$ = exports.DeletePullRequestApprovalRule$ = exports.DeleteFile$ = exports.DeleteCommentContent$ = exports.DeleteBranch$ = exports.DeleteApprovalRuleTemplate$ = exports.CreateUnreferencedMergeCommit$ = exports.CreateRepository$ = exports.CreatePullRequestApprovalRule$ = exports.CreatePullRequest$ = exports.CreateCommit$ = exports.CreateBranch$ = exports.CreateApprovalRuleTemplate$ = exports.BatchGetRepositories$ = exports.BatchGetCommits$ = void 0;
exports.UpdateRepositoryName$ = exports.UpdateRepositoryEncryptionKey$ = exports.UpdateRepositoryDescription$ = exports.UpdatePullRequestTitle$ = exports.UpdatePullRequestStatus$ = exports.UpdatePullRequestDescription$ = exports.UpdatePullRequestApprovalState$ = exports.UpdatePullRequestApprovalRuleContent$ = exports.UpdateDefaultBranch$ = exports.UpdateComment$ = exports.UpdateApprovalRuleTemplateName$ = exports.UpdateApprovalRuleTemplateDescription$ = exports.UpdateApprovalRuleTemplateContent$ = exports.UntagResource$ = exports.TestRepositoryTriggers$ = exports.TagResource$ = exports.PutRepositoryTriggers$ = exports.PutFile$ = exports.PutCommentReaction$ = exports.PostCommentReply$ = exports.PostCommentForPullRequest$ = exports.PostCommentForComparedCommit$ = exports.OverridePullRequestApprovalRules$ = exports.MergePullRequestByThreeWay$ = exports.MergePullRequestBySquash$ = void 0;
const _A = "Approval";
const _AARTWR = "AssociateApprovalRuleTemplateWithRepository";
const _AARTWRI = "AssociateApprovalRuleTemplateWithRepositoryInput";
const _ADNEE = "ActorDoesNotExistException";
const _ADNEEu = "AuthorDoesNotExistException";
const _AL = "ApprovalList";
const _AR = "ApprovalRule";
const _ARCRE = "ApprovalRuleContentRequiredException";
const _ARDNEE = "ApprovalRuleDoesNotExistException";
const _AREM = "ApprovalRuleEventMetadata";
const _ARL = "ApprovalRulesList";
const _ARNAEE = "ApprovalRuleNameAlreadyExistsException";
const _ARNRE = "ApprovalRuleNameRequiredException";
const _AROEM = "ApprovalRuleOverriddenEventMetadata";
const _ART = "ApprovalRuleTemplate";
const _ARTCRE = "ApprovalRuleTemplateContentRequiredException";
const _ARTDNEE = "ApprovalRuleTemplateDoesNotExistException";
const _ARTIUE = "ApprovalRuleTemplateInUseException";
const _ARTNAEE = "ApprovalRuleTemplateNameAlreadyExistsException";
const _ARTNRE = "ApprovalRuleTemplateNameRequiredException";
const _ASCEM = "ApprovalStateChangedEventMetadata";
const _ASRE = "ApprovalStateRequiredException";
const _Ar = "Arn";
const _BAARTWR = "BatchAssociateApprovalRuleTemplateWithRepositories";
const _BAARTWRE = "BatchAssociateApprovalRuleTemplateWithRepositoriesError";
const _BAARTWREL = "BatchAssociateApprovalRuleTemplateWithRepositoriesErrorsList";
const _BAARTWRI = "BatchAssociateApprovalRuleTemplateWithRepositoriesInput";
const _BAARTWRO = "BatchAssociateApprovalRuleTemplateWithRepositoriesOutput";
const _BCIAACIASE = "BeforeCommitIdAndAfterCommitIdAreSameException";
const _BDARTFR = "BatchDisassociateApprovalRuleTemplateFromRepositories";
const _BDARTFRE = "BatchDisassociateApprovalRuleTemplateFromRepositoriesError";
const _BDARTFREL = "BatchDisassociateApprovalRuleTemplateFromRepositoriesErrorsList";
const _BDARTFRI = "BatchDisassociateApprovalRuleTemplateFromRepositoriesInput";
const _BDARTFRO = "BatchDisassociateApprovalRuleTemplateFromRepositoriesOutput";
const _BDMC = "BatchDescribeMergeConflicts";
const _BDMCE = "BatchDescribeMergeConflictsError";
const _BDMCEa = "BatchDescribeMergeConflictsErrors";
const _BDMCI = "BatchDescribeMergeConflictsInput";
const _BDMCO = "BatchDescribeMergeConflictsOutput";
const _BDNEE = "BranchDoesNotExistException";
const _BGC = "BatchGetCommits";
const _BGCE = "BatchGetCommitsError";
const _BGCEL = "BatchGetCommitsErrorsList";
const _BGCI = "BatchGetCommitsInput";
const _BGCO = "BatchGetCommitsOutput";
const _BGR = "BatchGetRepositories";
const _BGRE = "BatchGetRepositoriesError";
const _BGREL = "BatchGetRepositoriesErrorsList";
const _BGRI = "BatchGetRepositoriesInput";
const _BGRO = "BatchGetRepositoriesOutput";
const _BI = "BranchInfo";
const _BIDNEE = "BlobIdDoesNotExistException";
const _BIRE = "BlobIdRequiredException";
const _BM = "BlobMetadata";
const _BNEE = "BranchNameExistsException";
const _BNITNE = "BranchNameIsTagNameException";
const _BNRE = "BranchNameRequiredException";
const _C = "Comment";
const _CART = "CreateApprovalRuleTemplate";
const _CARTI = "CreateApprovalRuleTemplateInput";
const _CARTO = "CreateApprovalRuleTemplateOutput";
const _CB = "CreateBranch";
const _CBI = "CreateBranchInput";
const _CC = "CreateCommit";
const _CCI = "CreateCommitInput";
const _CCO = "CreateCommitOutput";
const _CCRE = "CommentContentRequiredException";
const _CCSLEE = "CommentContentSizeLimitExceededException";
const _CDARFTE = "CannotDeleteApprovalRuleFromTemplateException";
const _CDE = "CommentDeletedException";
const _CDNEE = "CommentDoesNotExistException";
const _CDNEEo = "CommitDoesNotExistException";
const _CFCC = "CommentsForComparedCommit";
const _CFCCD = "CommentsForComparedCommitData";
const _CFPR = "CommentsForPullRequest";
const _CFPRD = "CommentsForPullRequestData";
const _CIDNEE = "CommitIdDoesNotExistException";
const _CILEE = "CommitIdsLimitExceededException";
const _CILRE = "CommitIdsListRequiredException";
const _CIRE = "CommentIdRequiredException";
const _CIREo = "CommitIdRequiredException";
const _CM = "ConflictMetadata";
const _CMARFTE = "CannotModifyApprovalRuleFromTemplateException";
const _CML = "ConflictMetadataList";
const _CMLEE = "CommitMessageLengthExceededException";
const _CNCBCE = "CommentNotCreatedByCallerException";
const _COL = "CommitObjectsList";
const _CPR = "CreatePullRequest";
const _CPRAR = "CreatePullRequestApprovalRule";
const _CPRARI = "CreatePullRequestApprovalRuleInput";
const _CPRARO = "CreatePullRequestApprovalRuleOutput";
const _CPRI = "CreatePullRequestInput";
const _CPRO = "CreatePullRequestOutput";
const _CR = "ConflictResolution";
const _CRE = "CommitRequiredException";
const _CRI = "CreateRepositoryInput";
const _CRO = "CreateRepositoryOutput";
const _CRTRE = "ClientRequestTokenRequiredException";
const _CRUE = "ConcurrentReferenceUpdateException";
const _CRr = "CreateRepository";
const _CUMC = "CreateUnreferencedMergeCommit";
const _CUMCI = "CreateUnreferencedMergeCommitInput";
const _CUMCO = "CreateUnreferencedMergeCommitOutput";
const _Co = "Commit";
const _Com = "Comments";
const _Con = "Conflict";
const _Conf = "Conflicts";
const _D = "Difference";
const _DART = "DeleteApprovalRuleTemplate";
const _DARTFR = "DisassociateApprovalRuleTemplateFromRepository";
const _DARTFRI = "DisassociateApprovalRuleTemplateFromRepositoryInput";
const _DARTI = "DeleteApprovalRuleTemplateInput";
const _DARTO = "DeleteApprovalRuleTemplateOutput";
const _DB = "DeleteBranch";
const _DBCBDE = "DefaultBranchCannotBeDeletedException";
const _DBI = "DeleteBranchInput";
const _DBO = "DeleteBranchOutput";
const _DCC = "DeleteCommentContent";
const _DCCI = "DeleteCommentContentInput";
const _DCCO = "DeleteCommentContentOutput";
const _DF = "DeleteFile";
const _DFE = "DeleteFileEntry";
const _DFEe = "DeleteFileEntries";
const _DFI = "DeleteFileInput";
const _DFO = "DeleteFileOutput";
const _DL = "DifferenceList";
const _DMC = "DescribeMergeConflicts";
const _DMCI = "DescribeMergeConflictsInput";
const _DMCO = "DescribeMergeConflictsOutput";
const _DNCWFNE = "DirectoryNameConflictsWithFileNameException";
const _DPRAR = "DeletePullRequestApprovalRule";
const _DPRARI = "DeletePullRequestApprovalRuleInput";
const _DPRARO = "DeletePullRequestApprovalRuleOutput";
const _DPRE = "DescribePullRequestEvents";
const _DPREI = "DescribePullRequestEventsInput";
const _DPREO = "DescribePullRequestEventsOutput";
const _DR = "DeleteRepository";
const _DRI = "DeleteRepositoryInput";
const _DRO = "DeleteRepositoryOutput";
const _E = "Evaluation";
const _EICFE = "EncryptionIntegrityChecksFailedException";
const _EKADE = "EncryptionKeyAccessDeniedException";
const _EKDE = "EncryptionKeyDisabledException";
const _EKIIE = "EncryptionKeyInvalidIdException";
const _EKIUE = "EncryptionKeyInvalidUsageException";
const _EKNFE = "EncryptionKeyNotFoundException";
const _EKRE = "EncryptionKeyRequiredException";
const _EKUE = "EncryptionKeyUnavailableException";
const _EPRAR = "EvaluatePullRequestApprovalRules";
const _EPRARI = "EvaluatePullRequestApprovalRulesInput";
const _EPRARO = "EvaluatePullRequestApprovalRulesOutput";
const _F = "File";
const _FCASFSE = "FileContentAndSourceFileSpecifiedException";
const _FCRE = "FileContentRequiredException";
const _FCSLEE = "FileContentSizeLimitExceededException";
const _FCSLEEo = "FolderContentSizeLimitExceededException";
const _FDNEE = "FileDoesNotExistException";
const _FDNEEo = "FolderDoesNotExistException";
const _FERE = "FileEntryRequiredException";
const _FL = "FileList";
const _FLo = "FolderList";
const _FM = "FileMetadata";
const _FMRE = "FileModeRequiredException";
const _FMi = "FileModes";
const _FMil = "FilesMetadata";
const _FNCWDNE = "FileNameConflictsWithDirectoryNameException";
const _FPCWSPE = "FilePathConflictsWithSubmodulePathException";
const _FS = "FileSizes";
const _FTLE = "FileTooLargeException";
const _FV = "FileVersion";
const _Fo = "Folder";
const _GART = "GetApprovalRuleTemplate";
const _GARTI = "GetApprovalRuleTemplateInput";
const _GARTO = "GetApprovalRuleTemplateOutput";
const _GB = "GetBlob";
const _GBI = "GetBlobInput";
const _GBIe = "GetBranchInput";
const _GBO = "GetBlobOutput";
const _GBOe = "GetBranchOutput";
const _GBe = "GetBranch";
const _GC = "GetComment";
const _GCFCC = "GetCommentsForComparedCommit";
const _GCFCCI = "GetCommentsForComparedCommitInput";
const _GCFCCO = "GetCommentsForComparedCommitOutput";
const _GCFPR = "GetCommentsForPullRequest";
const _GCFPRI = "GetCommentsForPullRequestInput";
const _GCFPRO = "GetCommentsForPullRequestOutput";
const _GCI = "GetCommentInput";
const _GCIe = "GetCommitInput";
const _GCO = "GetCommentOutput";
const _GCOe = "GetCommitOutput";
const _GCR = "GetCommentReactions";
const _GCRI = "GetCommentReactionsInput";
const _GCRO = "GetCommentReactionsOutput";
const _GCe = "GetCommit";
const _GD = "GetDifferences";
const _GDI = "GetDifferencesInput";
const _GDO = "GetDifferencesOutput";
const _GF = "GetFile";
const _GFI = "GetFileInput";
const _GFIe = "GetFolderInput";
const _GFO = "GetFileOutput";
const _GFOe = "GetFolderOutput";
const _GFe = "GetFolder";
const _GMC = "GetMergeCommit";
const _GMCI = "GetMergeCommitInput";
const _GMCIe = "GetMergeConflictsInput";
const _GMCO = "GetMergeCommitOutput";
const _GMCOe = "GetMergeConflictsOutput";
const _GMCe = "GetMergeConflicts";
const _GMO = "GetMergeOptions";
const _GMOI = "GetMergeOptionsInput";
const _GMOO = "GetMergeOptionsOutput";
const _GPR = "GetPullRequest";
const _GPRAS = "GetPullRequestApprovalStates";
const _GPRASI = "GetPullRequestApprovalStatesInput";
const _GPRASO = "GetPullRequestApprovalStatesOutput";
const _GPRI = "GetPullRequestInput";
const _GPRO = "GetPullRequestOutput";
const _GPROS = "GetPullRequestOverrideState";
const _GPROSI = "GetPullRequestOverrideStateInput";
const _GPROSO = "GetPullRequestOverrideStateOutput";
const _GR = "GetRepository";
const _GRI = "GetRepositoryInput";
const _GRO = "GetRepositoryOutput";
const _GRT = "GetRepositoryTriggers";
const _GRTI = "GetRepositoryTriggersInput";
const _GRTO = "GetRepositoryTriggersOutput";
const _IAAE = "InvalidActorArnException";
const _IAAEn = "InvalidAuthorArnException";
const _IARCE = "InvalidApprovalRuleContentException";
const _IARNE = "InvalidApprovalRuleNameException";
const _IARTCE = "InvalidApprovalRuleTemplateContentException";
const _IARTDE = "InvalidApprovalRuleTemplateDescriptionException";
const _IARTNE = "InvalidApprovalRuleTemplateNameException";
const _IASE = "InvalidApprovalStateException";
const _IBF = "IsBinaryFile";
const _IBIE = "InvalidBlobIdException";
const _IBNE = "InvalidBranchNameException";
const _ICDLE = "InvalidConflictDetailLevelException";
const _ICE = "InvalidCommitException";
const _ICIE = "InvalidCommentIdException";
const _ICIEn = "InvalidCommitIdException";
const _ICRE = "InvalidConflictResolutionException";
const _ICRSE = "InvalidConflictResolutionStrategyException";
const _ICRTE = "InvalidClientRequestTokenException";
const _ICTE = "InvalidContinuationTokenException";
const _IDCSE = "InvalidDestinationCommitSpecifierException";
const _IDE = "InvalidDescriptionException";
const _IDPE = "InvalidDeletionParameterException";
const _IEE = "InvalidEmailException";
const _IFLE = "InvalidFileLocationException";
const _IFME = "InvalidFileModeException";
const _IFPE = "InvalidFilePositionException";
const _IMCFE = "InvalidMaxConflictFilesException";
const _IMMHE = "InvalidMaxMergeHunksException";
const _IMOE = "InvalidMergeOptionException";
const _IMRE = "InvalidMaxResultsException";
const _IOE = "InvalidOrderException";
const _IOSE = "InvalidOverrideStatusException";
const _IPCIE = "InvalidParentCommitIdException";
const _IPE = "InvalidPathException";
const _IPME = "IdempotencyParameterMismatchException";
const _IPRETE = "InvalidPullRequestEventTypeException";
const _IPRIE = "InvalidPullRequestIdException";
const _IPRSE = "InvalidPullRequestStatusException";
const _IPRSUE = "InvalidPullRequestStatusUpdateException";
const _IRAE = "InvalidResourceArnException";
const _IRCE = "InvalidReplacementContentException";
const _IRCSE = "InvalidRuleContentSha256Exception";
const _IRDE = "InvalidRepositoryDescriptionException";
const _IRFVEE = "InvalidRelativeFileVersionEnumException";
const _IRIE = "InvalidRevisionIdException";
const _IRNE = "InvalidReferenceNameException";
const _IRNEn = "InvalidRepositoryNameException";
const _IRTBNE = "InvalidRepositoryTriggerBranchNameException";
const _IRTCDE = "InvalidRepositoryTriggerCustomDataException";
const _IRTDAE = "InvalidRepositoryTriggerDestinationArnException";
const _IRTE = "InvalidReplacementTypeException";
const _IRTEE = "InvalidRepositoryTriggerEventsException";
const _IRTNE = "InvalidRepositoryTriggerNameException";
const _IRTRE = "InvalidRepositoryTriggerRegionException";
const _IRUAE = "InvalidReactionUserArnException";
const _IRVE = "InvalidReactionValueException";
const _ISBE = "InvalidSortByException";
const _ISCSE = "InvalidSourceCommitSpecifierException";
const _ISTUE = "InvalidSystemTagUsageException";
const _ITBE = "InvalidTargetBranchException";
const _ITE = "InvalidTargetException";
const _ITEn = "InvalidTargetsException";
const _ITEnv = "InvalidTitleException";
const _ITKLE = "InvalidTagKeysListException";
const _ITME = "InvalidTagsMapException";
const _L = "Location";
const _LAARTFR = "ListAssociatedApprovalRuleTemplatesForRepository";
const _LAARTFRI = "ListAssociatedApprovalRuleTemplatesForRepositoryInput";
const _LAARTFRO = "ListAssociatedApprovalRuleTemplatesForRepositoryOutput";
const _LART = "ListApprovalRuleTemplates";
const _LARTI = "ListApprovalRuleTemplatesInput";
const _LARTO = "ListApprovalRuleTemplatesOutput";
const _LB = "ListBranches";
const _LBI = "ListBranchesInput";
const _LBO = "ListBranchesOutput";
const _LFCH = "ListFileCommitHistory";
const _LFCHR = "ListFileCommitHistoryRequest";
const _LFCHRi = "ListFileCommitHistoryResponse";
const _LPR = "ListPullRequests";
const _LPRI = "ListPullRequestsInput";
const _LPRO = "ListPullRequestsOutput";
const _LR = "ListRepositories";
const _LRFART = "ListRepositoriesForApprovalRuleTemplate";
const _LRFARTI = "ListRepositoriesForApprovalRuleTemplateInput";
const _LRFARTO = "ListRepositoriesForApprovalRuleTemplateOutput";
const _LRI = "ListRepositoriesInput";
const _LRO = "ListRepositoriesOutput";
const _LTFR = "ListTagsForResource";
const _LTFRI = "ListTagsForResourceInput";
const _LTFRO = "ListTagsForResourceOutput";
const _MBBFF = "MergeBranchesByFastForward";
const _MBBFFI = "MergeBranchesByFastForwardInput";
const _MBBFFO = "MergeBranchesByFastForwardOutput";
const _MBBS = "MergeBranchesBySquash";
const _MBBSI = "MergeBranchesBySquashInput";
const _MBBSO = "MergeBranchesBySquashOutput";
const _MBBTW = "MergeBranchesByThreeWay";
const _MBBTWI = "MergeBranchesByThreeWayInput";
const _MBBTWO = "MergeBranchesByThreeWayOutput";
const _MBEE = "MaximumBranchesExceededException";
const _MCREE = "MultipleConflictResolutionEntriesException";
const _MCREEE = "MaximumConflictResolutionEntriesExceededException";
const _MFCTLEE = "MaximumFileContentToLoadExceededException";
const _MFEEE = "MaximumFileEntriesExceededException";
const _MH = "MergeHunk";
const _MHD = "MergeHunkDetail";
const _MHe = "MergeHunks";
const _MITCEE = "MaximumItemsToCompareExceededException";
const _MM = "MergeMetadata";
const _MMRE = "ManualMergeRequiredException";
const _MNOAEE = "MaximumNumberOfApprovalsExceededException";
const _MO = "MergeOperations";
const _MOPREE = "MaximumOpenPullRequestsExceededException";
const _MORE = "MergeOptionRequiredException";
const _MPRBFF = "MergePullRequestByFastForward";
const _MPRBFFI = "MergePullRequestByFastForwardInput";
const _MPRBFFO = "MergePullRequestByFastForwardOutput";
const _MPRBS = "MergePullRequestBySquash";
const _MPRBSI = "MergePullRequestBySquashInput";
const _MPRBSO = "MergePullRequestBySquashOutput";
const _MPRBTW = "MergePullRequestByThreeWay";
const _MPRBTWI = "MergePullRequestByThreeWayInput";
const _MPRBTWO = "MergePullRequestByThreeWayOutput";
const _MR = "MaxResults";
const _MRIPRE = "MultipleRepositoriesInPullRequestException";
const _MRNEE = "MaximumRepositoryNamesExceededException";
const _MRTAWRE = "MaximumRuleTemplatesAssociatedWithRepositoryException";
const _MRTEE = "MaximumRepositoryTriggersExceededException";
const _NCE = "NoChangeException";
const _NLEE = "NameLengthExceededException";
const _NOREE = "NumberOfRulesExceededException";
const _NORTEE = "NumberOfRuleTemplatesExceededException";
const _NT = "NextToken";
const _OART = "OriginApprovalRuleTemplate";
const _OASE = "OverrideAlreadySetException";
const _ONAE = "OperationNotAllowedException";
const _OPRAR = "OverridePullRequestApprovalRules";
const _OPRARI = "OverridePullRequestApprovalRulesInput";
const _OSRE = "OverrideStatusRequiredException";
const _OT = "ObjectTypes";
const _PCDNEE = "ParentCommitDoesNotExistException";
const _PCFCC = "PostCommentForComparedCommit";
const _PCFCCI = "PostCommentForComparedCommitInput";
const _PCFCCO = "PostCommentForComparedCommitOutput";
const _PCFPR = "PostCommentForPullRequest";
const _PCFPRI = "PostCommentForPullRequestInput";
const _PCFPRO = "PostCommentForPullRequestOutput";
const _PCIOE = "ParentCommitIdOutdatedException";
const _PCIRE = "ParentCommitIdRequiredException";
const _PCR = "PostCommentReply";
const _PCRI = "PostCommentReplyInput";
const _PCRIu = "PutCommentReactionInput";
const _PCRO = "PostCommentReplyOutput";
const _PCRu = "PutCommentReaction";
const _PDNEE = "PathDoesNotExistException";
const _PF = "PutFile";
const _PFE = "PutFileEntry";
const _PFECE = "PutFileEntryConflictException";
const _PFEu = "PutFileEntries";
const _PFI = "PutFileInput";
const _PFO = "PutFileOutput";
const _PR = "PullRequest";
const _PRACE = "PullRequestAlreadyClosedException";
const _PRARNSE = "PullRequestApprovalRulesNotSatisfiedException";
const _PRCBABAE = "PullRequestCannotBeApprovedByAuthorException";
const _PRCEM = "PullRequestCreatedEventMetadata";
const _PRDNEE = "PullRequestDoesNotExistException";
const _PRE = "PathRequiredException";
const _PREL = "PullRequestEventList";
const _PREu = "PullRequestEvent";
const _PRIRE = "PullRequestIdRequiredException";
const _PRMSCEM = "PullRequestMergedStateChangedEventMetadata";
const _PRSCEM = "PullRequestStatusChangedEventMetadata";
const _PRSRE = "PullRequestStatusRequiredException";
const _PRSRUEM = "PullRequestSourceReferenceUpdatedEventMetadata";
const _PRT = "PullRequestTarget";
const _PRTI = "PutRepositoryTriggersInput";
const _PRTL = "PullRequestTargetList";
const _PRTO = "PutRepositoryTriggersOutput";
const _PRTu = "PutRepositoryTriggers";
const _RARE = "ResourceArnRequiredException";
const _RCE = "ReplaceContentEntry";
const _RCEe = "ReplaceContentEntries";
const _RCRE = "ReplacementContentRequiredException";
const _RD = "RevisionDag";
const _RDNEE = "ReferenceDoesNotExistException";
const _RDNEEe = "RepositoryDoesNotExistException";
const _RFC = "ReactionForComment";
const _RFCL = "ReactionsForCommentList";
const _RIRE = "RevisionIdRequiredException";
const _RLEE = "ReactionLimitExceededException";
const _RLEEe = "RepositoryLimitExceededException";
const _RM = "RepositoryMetadata";
const _RML = "RepositoryMetadataList";
const _RNAWPRE = "RepositoryNotAssociatedWithPullRequestException";
const _RNCE = "RevisionNotCurrentException";
const _RNEE = "RepositoryNameExistsException";
const _RNIP = "RepositoryNameIdPair";
const _RNIPL = "RepositoryNameIdPairList";
const _RNRE = "ReferenceNameRequiredException";
const _RNREe = "RepositoryNameRequiredException";
const _RNREep = "RepositoryNamesRequiredException";
const _RSFE = "RestrictedSourceFileException";
const _RT = "RepositoryTrigger";
const _RTBNLRE = "RepositoryTriggerBranchNameListRequiredException";
const _RTDARE = "RepositoryTriggerDestinationArnRequiredException";
const _RTEF = "RepositoryTriggerExecutionFailure";
const _RTEFL = "RepositoryTriggerExecutionFailureList";
const _RTELRE = "RepositoryTriggerEventsListRequiredException";
const _RTL = "RepositoryTriggersList";
const _RTLRE = "RepositoryTriggersListRequiredException";
const _RTNRE = "RepositoryTriggerNameRequiredException";
const _RTNSE = "ReferenceTypeNotSupportedException";
const _RTRE = "ReplacementTypeRequiredException";
const _RVF = "ReactionValueFormats";
const _RVRE = "ReactionValueRequiredException";
const _SADASE = "SourceAndDestinationAreSameException";
const _SFCE = "SameFileContentException";
const _SFME = "SetFileModeEntry";
const _SFMEe = "SetFileModeEntries";
const _SFOCRE = "SourceFileOrContentRequiredException";
const _SFS = "SourceFileSpecifier";
const _SL = "SymbolicLink";
const _SLL = "SymbolicLinkList";
const _SM = "SubModule";
const _SML = "SubModuleList";
const _SPRE = "SamePathRequestException";
const _T = "Target";
const _TDEE = "TipsDivergenceExceededException";
const _TKLRE = "TagKeysListRequiredException";
const _TL = "TargetList";
const _TMRE = "TagsMapRequiredException";
const _TMTE = "TooManyTagsException";
const _TOSRIDE = "TipOfSourceReferenceIsDifferentException";
const _TPE = "TagPolicyException";
const _TR = "TagResource";
const _TRE = "TargetRequiredException";
const _TREa = "TargetsRequiredException";
const _TREi = "TitleRequiredException";
const _TRI = "TagResourceInput";
const _TRT = "TestRepositoryTriggers";
const _TRTI = "TestRepositoryTriggersInput";
const _TRTO = "TestRepositoryTriggersOutput";
const _UARTC = "UpdateApprovalRuleTemplateContent";
const _UARTCI = "UpdateApprovalRuleTemplateContentInput";
const _UARTCO = "UpdateApprovalRuleTemplateContentOutput";
const _UARTD = "UpdateApprovalRuleTemplateDescription";
const _UARTDI = "UpdateApprovalRuleTemplateDescriptionInput";
const _UARTDO = "UpdateApprovalRuleTemplateDescriptionOutput";
const _UARTN = "UpdateApprovalRuleTemplateName";
const _UARTNI = "UpdateApprovalRuleTemplateNameInput";
const _UARTNO = "UpdateApprovalRuleTemplateNameOutput";
const _UC = "UpdateComment";
const _UCI = "UpdateCommentInput";
const _UCO = "UpdateCommentOutput";
const _UDB = "UpdateDefaultBranch";
const _UDBI = "UpdateDefaultBranchInput";
const _UI = "UserInfo";
const _UPRARC = "UpdatePullRequestApprovalRuleContent";
const _UPRARCI = "UpdatePullRequestApprovalRuleContentInput";
const _UPRARCO = "UpdatePullRequestApprovalRuleContentOutput";
const _UPRAS = "UpdatePullRequestApprovalState";
const _UPRASI = "UpdatePullRequestApprovalStateInput";
const _UPRD = "UpdatePullRequestDescription";
const _UPRDI = "UpdatePullRequestDescriptionInput";
const _UPRDO = "UpdatePullRequestDescriptionOutput";
const _UPRS = "UpdatePullRequestStatus";
const _UPRSI = "UpdatePullRequestStatusInput";
const _UPRSO = "UpdatePullRequestStatusOutput";
const _UPRT = "UpdatePullRequestTitle";
const _UPRTI = "UpdatePullRequestTitleInput";
const _UPRTO = "UpdatePullRequestTitleOutput";
const _UR = "UntagResource";
const _URD = "UpdateRepositoryDescription";
const _URDI = "UpdateRepositoryDescriptionInput";
const _UREK = "UpdateRepositoryEncryptionKey";
const _UREKI = "UpdateRepositoryEncryptionKeyInput";
const _UREKO = "UpdateRepositoryEncryptionKeyOutput";
const _URI = "UntagResourceInput";
const _URN = "UpdateRepositoryName";
const _URNI = "UpdateRepositoryNameInput";
const _a = "author";
const _aA = "authorArn";
const _aAc = "actorArn";
const _aB = "afterBlob";
const _aBI = "afterBlobId";
const _aCI = "afterCommitId";
const _aCS = "afterCommitSpecifier";
const _aD = "additionalData";
const _aI = "accountId";
const _aN = "authorName";
const _aP = "absolutePath";
const _aPf = "afterPath";
const _aR = "approvalRule";
const _aRC = "approvalRuleContent";
const _aREM = "approvalRuleEventMetadata";
const _aRI = "approvalRuleId";
const _aRN = "approvalRuleName";
const _aRNS = "approvalRulesNotSatisfied";
const _aRNs = "associatedRepositoryNames";
const _aROEM = "approvalRuleOverriddenEventMetadata";
const _aRS = "approvalRulesSatisfied";
const _aRT = "approvalRuleTemplate";
const _aRTC = "approvalRuleTemplateContent";
const _aRTD = "approvalRuleTemplateDescription";
const _aRTI = "approvalRuleTemplateId";
const _aRTN = "approvalRuleTemplateName";
const _aRTNp = "approvalRuleTemplateNames";
const _aRp = "approvalRules";
const _aS = "approvalState";
const _aSCEM = "approvalStateChangedEventMetadata";
const _aSp = "approvalStatus";
const _ap = "approved";
const _app = "approvals";
const _b = "base";
const _bB = "beforeBlob";
const _bBI = "beforeBlobId";
const _bCI = "baseCommitId";
const _bCIe = "beforeCommitId";
const _bCS = "beforeCommitSpecifier";
const _bI = "blobId";
const _bN = "branchName";
const _bP = "beforePath";
const _br = "branch";
const _bra = "branches";
const _c = "client";
const _cC = "contentConflict";
const _cD = "creationDate";
const _cDL = "conflictDetailLevel";
const _cDu = "customData";
const _cFCCD = "commentsForComparedCommitData";
const _cFPRD = "commentsForPullRequestData";
const _cI = "commitId";
const _cIo = "commitIds";
const _cIom = "commentId";
const _cIon = "configurationId";
const _cM = "conflictMetadata";
const _cML = "conflictMetadataList";
const _cMo = "commitMessage";
const _cR = "callerReactions";
const _cRS = "conflictResolutionStrategy";
const _cRT = "clientRequestToken";
const _cRo = "conflictResolution";
const _cS = "commitSpecifier";
const _cT = "changeType";
const _cUH = "cloneUrlHttp";
const _cUS = "cloneUrlSsh";
const _co = "conflicts";
const _com = "commits";
const _comm = "comments";
const _comme = "comment";
const _commi = "committer";
const _commit = "commit";
const _con = "content";
const _d = "deleted";
const _dA = "destinationArn";
const _dB = "deletedBranch";
const _dBN = "defaultBranchName";
const _dBe = "defaultBranch";
const _dC = "destinationCommit";
const _dCI = "destinationCommitId";
const _dCS = "destinationCommitSpecifier";
const _dF = "deleteFiles";
const _dR = "destinationReference";
const _dRN = "disassociatedRepositoryNames";
const _da = "date";
const _de = "description";
const _des = "destination";
const _di = "differences";
const _e = "error";
const _eC = "errorCode";
const _eD = "eventDate";
const _eL = "endLine";
const _eM = "errorMessage";
const _eN = "exceptionName";
const _eRCS = "existingRuleContentSha256";
const _em = "email";
const _emo = "emoji";
const _er = "errors";
const _ev = "evaluation";
const _eve = "events";
const _f = "files";
const _fA = "filesAdded";
const _fC = "fileContent";
const _fD = "filesDeleted";
const _fE = "failedExecutions";
const _fM = "fileModes";
const _fMC = "fileModeConflict";
const _fMa = "failureMessage";
const _fMi = "fileMode";
const _fP = "filePath";
const _fPi = "filePaths";
const _fPil = "filePosition";
const _fPo = "folderPath";
const _fS = "fileSizes";
const _fSi = "fileSize";
const _fU = "filesUpdated";
const _hC = "hunkContent";
const _iBF = "isBinaryFile";
const _iC = "isConflict";
const _iM = "isMerged";
const _iMs = "isMove";
const _iRT = "inReplyTo";
const _kEF = "keepEmptyFolders";
const _kKI = "kmsKeyId";
const _l = "location";
const _lAD = "lastActivityDate";
const _lMD = "lastModifiedDate";
const _lMU = "lastModifiedUser";
const _m = "message";
const _mB = "mergedBy";
const _mBe = "mergeBase";
const _mCF = "maxConflictFiles";
const _mCI = "mergedCommitId";
const _mCIe = "mergeCommitId";
const _mH = "mergeHunks";
const _mM = "mergeMetadata";
const _mMH = "maxMergeHunks";
const _mO = "mergeOption";
const _mOe = "mergeOperations";
const _mOer = "mergeOptions";
const _mR = "maxResults";
const _me = "mergeable";
const _mo = "mode";
const _n = "name";
const _nARTN = "newApprovalRuleTemplateName";
const _nN = "newName";
const _nOC = "numberOfConflicts";
const _nRC = "newRuleContent";
const _nT = "nextToken";
const _o = "overridden";
const _oART = "originApprovalRuleTemplate";
const _oARTN = "oldApprovalRuleTemplateName";
const _oKKI = "originalKmsKeyId";
const _oN = "oldName";
const _oS = "overrideStatus";
const _oT = "objectTypes";
const _oTC = "objectTypeConflict";
const _or = "order";
const _ov = "overrider";
const _p = "path";
const _pCI = "parentCommitId";
const _pF = "putFiles";
const _pR = "pullRequest";
const _pRCEM = "pullRequestCreatedEventMetadata";
const _pRE = "pullRequestEvents";
const _pRET = "pullRequestEventType";
const _pRI = "pullRequestId";
const _pRIu = "pullRequestIds";
const _pRMSCEM = "pullRequestMergedStateChangedEventMetadata";
const _pRS = "pullRequestStatus";
const _pRSCEM = "pullRequestStatusChangedEventMetadata";
const _pRSRUEM = "pullRequestSourceReferenceUpdatedEventMetadata";
const _pRT = "pullRequestTargets";
const _pa = "parents";
const _r = "repositories";
const _rA = "resourceArn";
const _rC = "reactionCounts";
const _rCS = "ruleContentSha256";
const _rCe = "replaceContents";
const _rCev = "revisionChildren";
const _rD = "repositoryDescription";
const _rDe = "revisionDag";
const _rFC = "reactionsForComment";
const _rFDUC = "reactionsFromDeletedUsersCount";
const _rFV = "relativeFileVersion";
const _rI = "revisionId";
const _rIe = "repositoryId";
const _rM = "repositoryMetadata";
const _rN = "repositoryName";
const _rNF = "repositoriesNotFound";
const _rNe = "repositoryNames";
const _rP = "relativePath";
const _rT = "replacementType";
const _rU = "reactionUsers";
const _rUA = "reactionUserArn";
const _rV = "reactionValue";
const _re = "reaction";
const _s = "smithy.ts.sdk.synthetic.com.amazonaws.codecommit";
const _sB = "sortBy";
const _sC = "sourceCommit";
const _sCI = "sourceCommitId";
const _sCS = "sourceCommitSpecifier";
const _sCh = "shortCode";
const _sE = "successfulExecutions";
const _sF = "subFolders";
const _sFM = "setFileModes";
const _sFo = "sourceFile";
const _sL = "symbolicLinks";
const _sLt = "startLine";
const _sM = "subModules";
const _sR = "sourceReference";
const _se = "server";
const _so = "source";
const _t = "title";
const _tB = "targetBranch";
const _tI = "treeId";
const _tK = "tagKeys";
const _ta = "targets";
const _tag = "tags";
const _tr = "triggers";
const _tri = "trigger";
const _u = "unicode";
const _uA = "userArn";
const n0 = "com.amazonaws.codecommit";
const schema_1 = require("@smithy/core/schema");
const CodeCommitServiceException_1 = require("../models/CodeCommitServiceException");
const errors_1 = require("../models/errors");
const _s_registry = schema_1.TypeRegistry.for(_s);
exports.CodeCommitServiceException$ = [-3, _s, "CodeCommitServiceException", 0, [], []];
_s_registry.registerError(exports.CodeCommitServiceException$, CodeCommitServiceException_1.CodeCommitServiceException);
const n0_registry = schema_1.TypeRegistry.for(n0);
exports.ActorDoesNotExistException$ = [-3, n0, _ADNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ActorDoesNotExistException$, errors_1.ActorDoesNotExistException);
exports.ApprovalRuleContentRequiredException$ = [-3, n0, _ARCRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleContentRequiredException$, errors_1.ApprovalRuleContentRequiredException);
exports.ApprovalRuleDoesNotExistException$ = [-3, n0, _ARDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleDoesNotExistException$, errors_1.ApprovalRuleDoesNotExistException);
exports.ApprovalRuleNameAlreadyExistsException$ = [-3, n0, _ARNAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleNameAlreadyExistsException$, errors_1.ApprovalRuleNameAlreadyExistsException);
exports.ApprovalRuleNameRequiredException$ = [-3, n0, _ARNRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleNameRequiredException$, errors_1.ApprovalRuleNameRequiredException);
exports.ApprovalRuleTemplateContentRequiredException$ = [-3, n0, _ARTCRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleTemplateContentRequiredException$, errors_1.ApprovalRuleTemplateContentRequiredException);
exports.ApprovalRuleTemplateDoesNotExistException$ = [-3, n0, _ARTDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleTemplateDoesNotExistException$, errors_1.ApprovalRuleTemplateDoesNotExistException);
exports.ApprovalRuleTemplateInUseException$ = [-3, n0, _ARTIUE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleTemplateInUseException$, errors_1.ApprovalRuleTemplateInUseException);
exports.ApprovalRuleTemplateNameAlreadyExistsException$ = [-3, n0, _ARTNAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleTemplateNameAlreadyExistsException$, errors_1.ApprovalRuleTemplateNameAlreadyExistsException);
exports.ApprovalRuleTemplateNameRequiredException$ = [-3, n0, _ARTNRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalRuleTemplateNameRequiredException$, errors_1.ApprovalRuleTemplateNameRequiredException);
exports.ApprovalStateRequiredException$ = [-3, n0, _ASRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ApprovalStateRequiredException$, errors_1.ApprovalStateRequiredException);
exports.AuthorDoesNotExistException$ = [-3, n0, _ADNEEu,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.AuthorDoesNotExistException$, errors_1.AuthorDoesNotExistException);
exports.BeforeCommitIdAndAfterCommitIdAreSameException$ = [-3, n0, _BCIAACIASE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.BeforeCommitIdAndAfterCommitIdAreSameException$, errors_1.BeforeCommitIdAndAfterCommitIdAreSameException);
exports.BlobIdDoesNotExistException$ = [-3, n0, _BIDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.BlobIdDoesNotExistException$, errors_1.BlobIdDoesNotExistException);
exports.BlobIdRequiredException$ = [-3, n0, _BIRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.BlobIdRequiredException$, errors_1.BlobIdRequiredException);
exports.BranchDoesNotExistException$ = [-3, n0, _BDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.BranchDoesNotExistException$, errors_1.BranchDoesNotExistException);
exports.BranchNameExistsException$ = [-3, n0, _BNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.BranchNameExistsException$, errors_1.BranchNameExistsException);
exports.BranchNameIsTagNameException$ = [-3, n0, _BNITNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.BranchNameIsTagNameException$, errors_1.BranchNameIsTagNameException);
exports.BranchNameRequiredException$ = [-3, n0, _BNRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.BranchNameRequiredException$, errors_1.BranchNameRequiredException);
exports.CannotDeleteApprovalRuleFromTemplateException$ = [-3, n0, _CDARFTE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CannotDeleteApprovalRuleFromTemplateException$, errors_1.CannotDeleteApprovalRuleFromTemplateException);
exports.CannotModifyApprovalRuleFromTemplateException$ = [-3, n0, _CMARFTE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CannotModifyApprovalRuleFromTemplateException$, errors_1.CannotModifyApprovalRuleFromTemplateException);
exports.ClientRequestTokenRequiredException$ = [-3, n0, _CRTRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ClientRequestTokenRequiredException$, errors_1.ClientRequestTokenRequiredException);
exports.CommentContentRequiredException$ = [-3, n0, _CCRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommentContentRequiredException$, errors_1.CommentContentRequiredException);
exports.CommentContentSizeLimitExceededException$ = [-3, n0, _CCSLEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommentContentSizeLimitExceededException$, errors_1.CommentContentSizeLimitExceededException);
exports.CommentDeletedException$ = [-3, n0, _CDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommentDeletedException$, errors_1.CommentDeletedException);
exports.CommentDoesNotExistException$ = [-3, n0, _CDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommentDoesNotExistException$, errors_1.CommentDoesNotExistException);
exports.CommentIdRequiredException$ = [-3, n0, _CIRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommentIdRequiredException$, errors_1.CommentIdRequiredException);
exports.CommentNotCreatedByCallerException$ = [-3, n0, _CNCBCE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommentNotCreatedByCallerException$, errors_1.CommentNotCreatedByCallerException);
exports.CommitDoesNotExistException$ = [-3, n0, _CDNEEo,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommitDoesNotExistException$, errors_1.CommitDoesNotExistException);
exports.CommitIdDoesNotExistException$ = [-3, n0, _CIDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommitIdDoesNotExistException$, errors_1.CommitIdDoesNotExistException);
exports.CommitIdRequiredException$ = [-3, n0, _CIREo,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommitIdRequiredException$, errors_1.CommitIdRequiredException);
exports.CommitIdsLimitExceededException$ = [-3, n0, _CILEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommitIdsLimitExceededException$, errors_1.CommitIdsLimitExceededException);
exports.CommitIdsListRequiredException$ = [-3, n0, _CILRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommitIdsListRequiredException$, errors_1.CommitIdsListRequiredException);
exports.CommitMessageLengthExceededException$ = [-3, n0, _CMLEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommitMessageLengthExceededException$, errors_1.CommitMessageLengthExceededException);
exports.CommitRequiredException$ = [-3, n0, _CRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.CommitRequiredException$, errors_1.CommitRequiredException);
exports.ConcurrentReferenceUpdateException$ = [-3, n0, _CRUE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ConcurrentReferenceUpdateException$, errors_1.ConcurrentReferenceUpdateException);
exports.DefaultBranchCannotBeDeletedException$ = [-3, n0, _DBCBDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.DefaultBranchCannotBeDeletedException$, errors_1.DefaultBranchCannotBeDeletedException);
exports.DirectoryNameConflictsWithFileNameException$ = [-3, n0, _DNCWFNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.DirectoryNameConflictsWithFileNameException$, errors_1.DirectoryNameConflictsWithFileNameException);
exports.EncryptionIntegrityChecksFailedException$ = [-3, n0, _EICFE,
    { [_e]: _se },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionIntegrityChecksFailedException$, errors_1.EncryptionIntegrityChecksFailedException);
exports.EncryptionKeyAccessDeniedException$ = [-3, n0, _EKADE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionKeyAccessDeniedException$, errors_1.EncryptionKeyAccessDeniedException);
exports.EncryptionKeyDisabledException$ = [-3, n0, _EKDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionKeyDisabledException$, errors_1.EncryptionKeyDisabledException);
exports.EncryptionKeyInvalidIdException$ = [-3, n0, _EKIIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionKeyInvalidIdException$, errors_1.EncryptionKeyInvalidIdException);
exports.EncryptionKeyInvalidUsageException$ = [-3, n0, _EKIUE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionKeyInvalidUsageException$, errors_1.EncryptionKeyInvalidUsageException);
exports.EncryptionKeyNotFoundException$ = [-3, n0, _EKNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionKeyNotFoundException$, errors_1.EncryptionKeyNotFoundException);
exports.EncryptionKeyRequiredException$ = [-3, n0, _EKRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionKeyRequiredException$, errors_1.EncryptionKeyRequiredException);
exports.EncryptionKeyUnavailableException$ = [-3, n0, _EKUE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.EncryptionKeyUnavailableException$, errors_1.EncryptionKeyUnavailableException);
exports.FileContentAndSourceFileSpecifiedException$ = [-3, n0, _FCASFSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileContentAndSourceFileSpecifiedException$, errors_1.FileContentAndSourceFileSpecifiedException);
exports.FileContentRequiredException$ = [-3, n0, _FCRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileContentRequiredException$, errors_1.FileContentRequiredException);
exports.FileContentSizeLimitExceededException$ = [-3, n0, _FCSLEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileContentSizeLimitExceededException$, errors_1.FileContentSizeLimitExceededException);
exports.FileDoesNotExistException$ = [-3, n0, _FDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileDoesNotExistException$, errors_1.FileDoesNotExistException);
exports.FileEntryRequiredException$ = [-3, n0, _FERE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileEntryRequiredException$, errors_1.FileEntryRequiredException);
exports.FileModeRequiredException$ = [-3, n0, _FMRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileModeRequiredException$, errors_1.FileModeRequiredException);
exports.FileNameConflictsWithDirectoryNameException$ = [-3, n0, _FNCWDNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileNameConflictsWithDirectoryNameException$, errors_1.FileNameConflictsWithDirectoryNameException);
exports.FilePathConflictsWithSubmodulePathException$ = [-3, n0, _FPCWSPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FilePathConflictsWithSubmodulePathException$, errors_1.FilePathConflictsWithSubmodulePathException);
exports.FileTooLargeException$ = [-3, n0, _FTLE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FileTooLargeException$, errors_1.FileTooLargeException);
exports.FolderContentSizeLimitExceededException$ = [-3, n0, _FCSLEEo,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FolderContentSizeLimitExceededException$, errors_1.FolderContentSizeLimitExceededException);
exports.FolderDoesNotExistException$ = [-3, n0, _FDNEEo,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.FolderDoesNotExistException$, errors_1.FolderDoesNotExistException);
exports.IdempotencyParameterMismatchException$ = [-3, n0, _IPME,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.IdempotencyParameterMismatchException$, errors_1.IdempotencyParameterMismatchException);
exports.InvalidActorArnException$ = [-3, n0, _IAAE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidActorArnException$, errors_1.InvalidActorArnException);
exports.InvalidApprovalRuleContentException$ = [-3, n0, _IARCE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidApprovalRuleContentException$, errors_1.InvalidApprovalRuleContentException);
exports.InvalidApprovalRuleNameException$ = [-3, n0, _IARNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidApprovalRuleNameException$, errors_1.InvalidApprovalRuleNameException);
exports.InvalidApprovalRuleTemplateContentException$ = [-3, n0, _IARTCE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidApprovalRuleTemplateContentException$, errors_1.InvalidApprovalRuleTemplateContentException);
exports.InvalidApprovalRuleTemplateDescriptionException$ = [-3, n0, _IARTDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidApprovalRuleTemplateDescriptionException$, errors_1.InvalidApprovalRuleTemplateDescriptionException);
exports.InvalidApprovalRuleTemplateNameException$ = [-3, n0, _IARTNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidApprovalRuleTemplateNameException$, errors_1.InvalidApprovalRuleTemplateNameException);
exports.InvalidApprovalStateException$ = [-3, n0, _IASE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidApprovalStateException$, errors_1.InvalidApprovalStateException);
exports.InvalidAuthorArnException$ = [-3, n0, _IAAEn,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidAuthorArnException$, errors_1.InvalidAuthorArnException);
exports.InvalidBlobIdException$ = [-3, n0, _IBIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidBlobIdException$, errors_1.InvalidBlobIdException);
exports.InvalidBranchNameException$ = [-3, n0, _IBNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidBranchNameException$, errors_1.InvalidBranchNameException);
exports.InvalidClientRequestTokenException$ = [-3, n0, _ICRTE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidClientRequestTokenException$, errors_1.InvalidClientRequestTokenException);
exports.InvalidCommentIdException$ = [-3, n0, _ICIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidCommentIdException$, errors_1.InvalidCommentIdException);
exports.InvalidCommitException$ = [-3, n0, _ICE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidCommitException$, errors_1.InvalidCommitException);
exports.InvalidCommitIdException$ = [-3, n0, _ICIEn,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidCommitIdException$, errors_1.InvalidCommitIdException);
exports.InvalidConflictDetailLevelException$ = [-3, n0, _ICDLE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidConflictDetailLevelException$, errors_1.InvalidConflictDetailLevelException);
exports.InvalidConflictResolutionException$ = [-3, n0, _ICRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidConflictResolutionException$, errors_1.InvalidConflictResolutionException);
exports.InvalidConflictResolutionStrategyException$ = [-3, n0, _ICRSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidConflictResolutionStrategyException$, errors_1.InvalidConflictResolutionStrategyException);
exports.InvalidContinuationTokenException$ = [-3, n0, _ICTE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidContinuationTokenException$, errors_1.InvalidContinuationTokenException);
exports.InvalidDeletionParameterException$ = [-3, n0, _IDPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidDeletionParameterException$, errors_1.InvalidDeletionParameterException);
exports.InvalidDescriptionException$ = [-3, n0, _IDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidDescriptionException$, errors_1.InvalidDescriptionException);
exports.InvalidDestinationCommitSpecifierException$ = [-3, n0, _IDCSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidDestinationCommitSpecifierException$, errors_1.InvalidDestinationCommitSpecifierException);
exports.InvalidEmailException$ = [-3, n0, _IEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidEmailException$, errors_1.InvalidEmailException);
exports.InvalidFileLocationException$ = [-3, n0, _IFLE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidFileLocationException$, errors_1.InvalidFileLocationException);
exports.InvalidFileModeException$ = [-3, n0, _IFME,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidFileModeException$, errors_1.InvalidFileModeException);
exports.InvalidFilePositionException$ = [-3, n0, _IFPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidFilePositionException$, errors_1.InvalidFilePositionException);
exports.InvalidMaxConflictFilesException$ = [-3, n0, _IMCFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidMaxConflictFilesException$, errors_1.InvalidMaxConflictFilesException);
exports.InvalidMaxMergeHunksException$ = [-3, n0, _IMMHE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidMaxMergeHunksException$, errors_1.InvalidMaxMergeHunksException);
exports.InvalidMaxResultsException$ = [-3, n0, _IMRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidMaxResultsException$, errors_1.InvalidMaxResultsException);
exports.InvalidMergeOptionException$ = [-3, n0, _IMOE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidMergeOptionException$, errors_1.InvalidMergeOptionException);
exports.InvalidOrderException$ = [-3, n0, _IOE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidOrderException$, errors_1.InvalidOrderException);
exports.InvalidOverrideStatusException$ = [-3, n0, _IOSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidOverrideStatusException$, errors_1.InvalidOverrideStatusException);
exports.InvalidParentCommitIdException$ = [-3, n0, _IPCIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidParentCommitIdException$, errors_1.InvalidParentCommitIdException);
exports.InvalidPathException$ = [-3, n0, _IPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidPathException$, errors_1.InvalidPathException);
exports.InvalidPullRequestEventTypeException$ = [-3, n0, _IPRETE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidPullRequestEventTypeException$, errors_1.InvalidPullRequestEventTypeException);
exports.InvalidPullRequestIdException$ = [-3, n0, _IPRIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidPullRequestIdException$, errors_1.InvalidPullRequestIdException);
exports.InvalidPullRequestStatusException$ = [-3, n0, _IPRSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidPullRequestStatusException$, errors_1.InvalidPullRequestStatusException);
exports.InvalidPullRequestStatusUpdateException$ = [-3, n0, _IPRSUE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidPullRequestStatusUpdateException$, errors_1.InvalidPullRequestStatusUpdateException);
exports.InvalidReactionUserArnException$ = [-3, n0, _IRUAE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidReactionUserArnException$, errors_1.InvalidReactionUserArnException);
exports.InvalidReactionValueException$ = [-3, n0, _IRVE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidReactionValueException$, errors_1.InvalidReactionValueException);
exports.InvalidReferenceNameException$ = [-3, n0, _IRNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidReferenceNameException$, errors_1.InvalidReferenceNameException);
exports.InvalidRelativeFileVersionEnumException$ = [-3, n0, _IRFVEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRelativeFileVersionEnumException$, errors_1.InvalidRelativeFileVersionEnumException);
exports.InvalidReplacementContentException$ = [-3, n0, _IRCE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidReplacementContentException$, errors_1.InvalidReplacementContentException);
exports.InvalidReplacementTypeException$ = [-3, n0, _IRTE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidReplacementTypeException$, errors_1.InvalidReplacementTypeException);
exports.InvalidRepositoryDescriptionException$ = [-3, n0, _IRDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryDescriptionException$, errors_1.InvalidRepositoryDescriptionException);
exports.InvalidRepositoryNameException$ = [-3, n0, _IRNEn,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryNameException$, errors_1.InvalidRepositoryNameException);
exports.InvalidRepositoryTriggerBranchNameException$ = [-3, n0, _IRTBNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryTriggerBranchNameException$, errors_1.InvalidRepositoryTriggerBranchNameException);
exports.InvalidRepositoryTriggerCustomDataException$ = [-3, n0, _IRTCDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryTriggerCustomDataException$, errors_1.InvalidRepositoryTriggerCustomDataException);
exports.InvalidRepositoryTriggerDestinationArnException$ = [-3, n0, _IRTDAE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryTriggerDestinationArnException$, errors_1.InvalidRepositoryTriggerDestinationArnException);
exports.InvalidRepositoryTriggerEventsException$ = [-3, n0, _IRTEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryTriggerEventsException$, errors_1.InvalidRepositoryTriggerEventsException);
exports.InvalidRepositoryTriggerNameException$ = [-3, n0, _IRTNE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryTriggerNameException$, errors_1.InvalidRepositoryTriggerNameException);
exports.InvalidRepositoryTriggerRegionException$ = [-3, n0, _IRTRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRepositoryTriggerRegionException$, errors_1.InvalidRepositoryTriggerRegionException);
exports.InvalidResourceArnException$ = [-3, n0, _IRAE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidResourceArnException$, errors_1.InvalidResourceArnException);
exports.InvalidRevisionIdException$ = [-3, n0, _IRIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRevisionIdException$, errors_1.InvalidRevisionIdException);
exports.InvalidRuleContentSha256Exception$ = [-3, n0, _IRCSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidRuleContentSha256Exception$, errors_1.InvalidRuleContentSha256Exception);
exports.InvalidSortByException$ = [-3, n0, _ISBE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidSortByException$, errors_1.InvalidSortByException);
exports.InvalidSourceCommitSpecifierException$ = [-3, n0, _ISCSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidSourceCommitSpecifierException$, errors_1.InvalidSourceCommitSpecifierException);
exports.InvalidSystemTagUsageException$ = [-3, n0, _ISTUE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidSystemTagUsageException$, errors_1.InvalidSystemTagUsageException);
exports.InvalidTagKeysListException$ = [-3, n0, _ITKLE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidTagKeysListException$, errors_1.InvalidTagKeysListException);
exports.InvalidTagsMapException$ = [-3, n0, _ITME,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidTagsMapException$, errors_1.InvalidTagsMapException);
exports.InvalidTargetBranchException$ = [-3, n0, _ITBE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidTargetBranchException$, errors_1.InvalidTargetBranchException);
exports.InvalidTargetException$ = [-3, n0, _ITE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidTargetException$, errors_1.InvalidTargetException);
exports.InvalidTargetsException$ = [-3, n0, _ITEn,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidTargetsException$, errors_1.InvalidTargetsException);
exports.InvalidTitleException$ = [-3, n0, _ITEnv,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.InvalidTitleException$, errors_1.InvalidTitleException);
exports.ManualMergeRequiredException$ = [-3, n0, _MMRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ManualMergeRequiredException$, errors_1.ManualMergeRequiredException);
exports.MaximumBranchesExceededException$ = [-3, n0, _MBEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumBranchesExceededException$, errors_1.MaximumBranchesExceededException);
exports.MaximumConflictResolutionEntriesExceededException$ = [-3, n0, _MCREEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumConflictResolutionEntriesExceededException$, errors_1.MaximumConflictResolutionEntriesExceededException);
exports.MaximumFileContentToLoadExceededException$ = [-3, n0, _MFCTLEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumFileContentToLoadExceededException$, errors_1.MaximumFileContentToLoadExceededException);
exports.MaximumFileEntriesExceededException$ = [-3, n0, _MFEEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumFileEntriesExceededException$, errors_1.MaximumFileEntriesExceededException);
exports.MaximumItemsToCompareExceededException$ = [-3, n0, _MITCEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumItemsToCompareExceededException$, errors_1.MaximumItemsToCompareExceededException);
exports.MaximumNumberOfApprovalsExceededException$ = [-3, n0, _MNOAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumNumberOfApprovalsExceededException$, errors_1.MaximumNumberOfApprovalsExceededException);
exports.MaximumOpenPullRequestsExceededException$ = [-3, n0, _MOPREE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumOpenPullRequestsExceededException$, errors_1.MaximumOpenPullRequestsExceededException);
exports.MaximumRepositoryNamesExceededException$ = [-3, n0, _MRNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumRepositoryNamesExceededException$, errors_1.MaximumRepositoryNamesExceededException);
exports.MaximumRepositoryTriggersExceededException$ = [-3, n0, _MRTEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumRepositoryTriggersExceededException$, errors_1.MaximumRepositoryTriggersExceededException);
exports.MaximumRuleTemplatesAssociatedWithRepositoryException$ = [-3, n0, _MRTAWRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MaximumRuleTemplatesAssociatedWithRepositoryException$, errors_1.MaximumRuleTemplatesAssociatedWithRepositoryException);
exports.MergeOptionRequiredException$ = [-3, n0, _MORE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MergeOptionRequiredException$, errors_1.MergeOptionRequiredException);
exports.MultipleConflictResolutionEntriesException$ = [-3, n0, _MCREE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MultipleConflictResolutionEntriesException$, errors_1.MultipleConflictResolutionEntriesException);
exports.MultipleRepositoriesInPullRequestException$ = [-3, n0, _MRIPRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.MultipleRepositoriesInPullRequestException$, errors_1.MultipleRepositoriesInPullRequestException);
exports.NameLengthExceededException$ = [-3, n0, _NLEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.NameLengthExceededException$, errors_1.NameLengthExceededException);
exports.NoChangeException$ = [-3, n0, _NCE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.NoChangeException$, errors_1.NoChangeException);
exports.NumberOfRulesExceededException$ = [-3, n0, _NOREE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.NumberOfRulesExceededException$, errors_1.NumberOfRulesExceededException);
exports.NumberOfRuleTemplatesExceededException$ = [-3, n0, _NORTEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.NumberOfRuleTemplatesExceededException$, errors_1.NumberOfRuleTemplatesExceededException);
exports.OperationNotAllowedException$ = [-3, n0, _ONAE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.OperationNotAllowedException$, errors_1.OperationNotAllowedException);
exports.OverrideAlreadySetException$ = [-3, n0, _OASE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.OverrideAlreadySetException$, errors_1.OverrideAlreadySetException);
exports.OverrideStatusRequiredException$ = [-3, n0, _OSRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.OverrideStatusRequiredException$, errors_1.OverrideStatusRequiredException);
exports.ParentCommitDoesNotExistException$ = [-3, n0, _PCDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ParentCommitDoesNotExistException$, errors_1.ParentCommitDoesNotExistException);
exports.ParentCommitIdOutdatedException$ = [-3, n0, _PCIOE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ParentCommitIdOutdatedException$, errors_1.ParentCommitIdOutdatedException);
exports.ParentCommitIdRequiredException$ = [-3, n0, _PCIRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ParentCommitIdRequiredException$, errors_1.ParentCommitIdRequiredException);
exports.PathDoesNotExistException$ = [-3, n0, _PDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PathDoesNotExistException$, errors_1.PathDoesNotExistException);
exports.PathRequiredException$ = [-3, n0, _PRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PathRequiredException$, errors_1.PathRequiredException);
exports.PullRequestAlreadyClosedException$ = [-3, n0, _PRACE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PullRequestAlreadyClosedException$, errors_1.PullRequestAlreadyClosedException);
exports.PullRequestApprovalRulesNotSatisfiedException$ = [-3, n0, _PRARNSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PullRequestApprovalRulesNotSatisfiedException$, errors_1.PullRequestApprovalRulesNotSatisfiedException);
exports.PullRequestCannotBeApprovedByAuthorException$ = [-3, n0, _PRCBABAE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PullRequestCannotBeApprovedByAuthorException$, errors_1.PullRequestCannotBeApprovedByAuthorException);
exports.PullRequestDoesNotExistException$ = [-3, n0, _PRDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PullRequestDoesNotExistException$, errors_1.PullRequestDoesNotExistException);
exports.PullRequestIdRequiredException$ = [-3, n0, _PRIRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PullRequestIdRequiredException$, errors_1.PullRequestIdRequiredException);
exports.PullRequestStatusRequiredException$ = [-3, n0, _PRSRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PullRequestStatusRequiredException$, errors_1.PullRequestStatusRequiredException);
exports.PutFileEntryConflictException$ = [-3, n0, _PFECE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.PutFileEntryConflictException$, errors_1.PutFileEntryConflictException);
exports.ReactionLimitExceededException$ = [-3, n0, _RLEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ReactionLimitExceededException$, errors_1.ReactionLimitExceededException);
exports.ReactionValueRequiredException$ = [-3, n0, _RVRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ReactionValueRequiredException$, errors_1.ReactionValueRequiredException);
exports.ReferenceDoesNotExistException$ = [-3, n0, _RDNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ReferenceDoesNotExistException$, errors_1.ReferenceDoesNotExistException);
exports.ReferenceNameRequiredException$ = [-3, n0, _RNRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ReferenceNameRequiredException$, errors_1.ReferenceNameRequiredException);
exports.ReferenceTypeNotSupportedException$ = [-3, n0, _RTNSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ReferenceTypeNotSupportedException$, errors_1.ReferenceTypeNotSupportedException);
exports.ReplacementContentRequiredException$ = [-3, n0, _RCRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ReplacementContentRequiredException$, errors_1.ReplacementContentRequiredException);
exports.ReplacementTypeRequiredException$ = [-3, n0, _RTRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ReplacementTypeRequiredException$, errors_1.ReplacementTypeRequiredException);
exports.RepositoryDoesNotExistException$ = [-3, n0, _RDNEEe,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryDoesNotExistException$, errors_1.RepositoryDoesNotExistException);
exports.RepositoryLimitExceededException$ = [-3, n0, _RLEEe,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryLimitExceededException$, errors_1.RepositoryLimitExceededException);
exports.RepositoryNameExistsException$ = [-3, n0, _RNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryNameExistsException$, errors_1.RepositoryNameExistsException);
exports.RepositoryNameRequiredException$ = [-3, n0, _RNREe,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryNameRequiredException$, errors_1.RepositoryNameRequiredException);
exports.RepositoryNamesRequiredException$ = [-3, n0, _RNREep,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryNamesRequiredException$, errors_1.RepositoryNamesRequiredException);
exports.RepositoryNotAssociatedWithPullRequestException$ = [-3, n0, _RNAWPRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryNotAssociatedWithPullRequestException$, errors_1.RepositoryNotAssociatedWithPullRequestException);
exports.RepositoryTriggerBranchNameListRequiredException$ = [-3, n0, _RTBNLRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryTriggerBranchNameListRequiredException$, errors_1.RepositoryTriggerBranchNameListRequiredException);
exports.RepositoryTriggerDestinationArnRequiredException$ = [-3, n0, _RTDARE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryTriggerDestinationArnRequiredException$, errors_1.RepositoryTriggerDestinationArnRequiredException);
exports.RepositoryTriggerEventsListRequiredException$ = [-3, n0, _RTELRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryTriggerEventsListRequiredException$, errors_1.RepositoryTriggerEventsListRequiredException);
exports.RepositoryTriggerNameRequiredException$ = [-3, n0, _RTNRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryTriggerNameRequiredException$, errors_1.RepositoryTriggerNameRequiredException);
exports.RepositoryTriggersListRequiredException$ = [-3, n0, _RTLRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RepositoryTriggersListRequiredException$, errors_1.RepositoryTriggersListRequiredException);
exports.ResourceArnRequiredException$ = [-3, n0, _RARE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.ResourceArnRequiredException$, errors_1.ResourceArnRequiredException);
exports.RestrictedSourceFileException$ = [-3, n0, _RSFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RestrictedSourceFileException$, errors_1.RestrictedSourceFileException);
exports.RevisionIdRequiredException$ = [-3, n0, _RIRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RevisionIdRequiredException$, errors_1.RevisionIdRequiredException);
exports.RevisionNotCurrentException$ = [-3, n0, _RNCE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.RevisionNotCurrentException$, errors_1.RevisionNotCurrentException);
exports.SameFileContentException$ = [-3, n0, _SFCE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.SameFileContentException$, errors_1.SameFileContentException);
exports.SamePathRequestException$ = [-3, n0, _SPRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.SamePathRequestException$, errors_1.SamePathRequestException);
exports.SourceAndDestinationAreSameException$ = [-3, n0, _SADASE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.SourceAndDestinationAreSameException$, errors_1.SourceAndDestinationAreSameException);
exports.SourceFileOrContentRequiredException$ = [-3, n0, _SFOCRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.SourceFileOrContentRequiredException$, errors_1.SourceFileOrContentRequiredException);
exports.TagKeysListRequiredException$ = [-3, n0, _TKLRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TagKeysListRequiredException$, errors_1.TagKeysListRequiredException);
exports.TagPolicyException$ = [-3, n0, _TPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TagPolicyException$, errors_1.TagPolicyException);
exports.TagsMapRequiredException$ = [-3, n0, _TMRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TagsMapRequiredException$, errors_1.TagsMapRequiredException);
exports.TargetRequiredException$ = [-3, n0, _TRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TargetRequiredException$, errors_1.TargetRequiredException);
exports.TargetsRequiredException$ = [-3, n0, _TREa,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TargetsRequiredException$, errors_1.TargetsRequiredException);
exports.TipOfSourceReferenceIsDifferentException$ = [-3, n0, _TOSRIDE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TipOfSourceReferenceIsDifferentException$, errors_1.TipOfSourceReferenceIsDifferentException);
exports.TipsDivergenceExceededException$ = [-3, n0, _TDEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TipsDivergenceExceededException$, errors_1.TipsDivergenceExceededException);
exports.TitleRequiredException$ = [-3, n0, _TREi,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TitleRequiredException$, errors_1.TitleRequiredException);
exports.TooManyTagsException$ = [-3, n0, _TMTE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(exports.TooManyTagsException$, errors_1.TooManyTagsException);
exports.errorTypeRegistries = [
    _s_registry,
    n0_registry,
];
exports.Approval$ = [3, n0, _A,
    0,
    [_uA, _aS],
    [0, 0]
];
exports.ApprovalRule$ = [3, n0, _AR,
    0,
    [_aRI, _aRN, _aRC, _rCS, _lMD, _cD, _lMU, _oART],
    [0, 0, 0, 0, 4, 4, 0, () => exports.OriginApprovalRuleTemplate$]
];
exports.ApprovalRuleEventMetadata$ = [3, n0, _AREM,
    0,
    [_aRN, _aRI, _aRC],
    [0, 0, 0]
];
exports.ApprovalRuleOverriddenEventMetadata$ = [3, n0, _AROEM,
    0,
    [_rI, _oS],
    [0, 0]
];
exports.ApprovalRuleTemplate$ = [3, n0, _ART,
    0,
    [_aRTI, _aRTN, _aRTD, _aRTC, _rCS, _lMD, _cD, _lMU],
    [0, 0, 0, 0, 0, 4, 4, 0]
];
exports.ApprovalStateChangedEventMetadata$ = [3, n0, _ASCEM,
    0,
    [_rI, _aSp],
    [0, 0]
];
exports.AssociateApprovalRuleTemplateWithRepositoryInput$ = [3, n0, _AARTWRI,
    0,
    [_aRTN, _rN],
    [0, 0], 2
];
exports.BatchAssociateApprovalRuleTemplateWithRepositoriesError$ = [3, n0, _BAARTWRE,
    0,
    [_rN, _eC, _eM],
    [0, 0, 0]
];
exports.BatchAssociateApprovalRuleTemplateWithRepositoriesInput$ = [3, n0, _BAARTWRI,
    0,
    [_aRTN, _rNe],
    [0, 64 | 0], 2
];
exports.BatchAssociateApprovalRuleTemplateWithRepositoriesOutput$ = [3, n0, _BAARTWRO,
    0,
    [_aRNs, _er],
    [64 | 0, () => BatchAssociateApprovalRuleTemplateWithRepositoriesErrorsList], 2
];
exports.BatchDescribeMergeConflictsError$ = [3, n0, _BDMCE,
    0,
    [_fP, _eN, _m],
    [0, 0, 0], 3
];
exports.BatchDescribeMergeConflictsInput$ = [3, n0, _BDMCI,
    0,
    [_rN, _dCS, _sCS, _mO, _mMH, _mCF, _fPi, _cDL, _cRS, _nT],
    [0, 0, 0, 0, 1, 1, 64 | 0, 0, 0, 0], 4
];
exports.BatchDescribeMergeConflictsOutput$ = [3, n0, _BDMCO,
    0,
    [_co, _dCI, _sCI, _nT, _er, _bCI],
    [() => Conflicts, 0, 0, 0, () => BatchDescribeMergeConflictsErrors, 0], 3
];
exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesError$ = [3, n0, _BDARTFRE,
    0,
    [_rN, _eC, _eM],
    [0, 0, 0]
];
exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesInput$ = [3, n0, _BDARTFRI,
    0,
    [_aRTN, _rNe],
    [0, 64 | 0], 2
];
exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesOutput$ = [3, n0, _BDARTFRO,
    0,
    [_dRN, _er],
    [64 | 0, () => BatchDisassociateApprovalRuleTemplateFromRepositoriesErrorsList], 2
];
exports.BatchGetCommitsError$ = [3, n0, _BGCE,
    0,
    [_cI, _eC, _eM],
    [0, 0, 0]
];
exports.BatchGetCommitsInput$ = [3, n0, _BGCI,
    0,
    [_cIo, _rN],
    [64 | 0, 0], 2
];
exports.BatchGetCommitsOutput$ = [3, n0, _BGCO,
    0,
    [_com, _er],
    [() => CommitObjectsList, () => BatchGetCommitsErrorsList]
];
exports.BatchGetRepositoriesError$ = [3, n0, _BGRE,
    0,
    [_rIe, _rN, _eC, _eM],
    [0, 0, 0, 0]
];
exports.BatchGetRepositoriesInput$ = [3, n0, _BGRI,
    0,
    [_rNe],
    [64 | 0], 1
];
exports.BatchGetRepositoriesOutput$ = [3, n0, _BGRO,
    0,
    [_r, _rNF, _er],
    [() => RepositoryMetadataList, 64 | 0, () => BatchGetRepositoriesErrorsList]
];
exports.BlobMetadata$ = [3, n0, _BM,
    0,
    [_bI, _p, _mo],
    [0, 0, 0]
];
exports.BranchInfo$ = [3, n0, _BI,
    0,
    [_bN, _cI],
    [0, 0]
];
exports.Comment$ = [3, n0, _C,
    0,
    [_cIom, _con, _iRT, _cD, _lMD, _aA, _d, _cRT, _cR, _rC],
    [0, 0, 0, 4, 4, 0, 2, 0, 64 | 0, 128 | 1]
];
exports.CommentsForComparedCommit$ = [3, n0, _CFCC,
    0,
    [_rN, _bCIe, _aCI, _bBI, _aBI, _l, _comm],
    [0, 0, 0, 0, 0, () => exports.Location$, () => Comments]
];
exports.CommentsForPullRequest$ = [3, n0, _CFPR,
    0,
    [_pRI, _rN, _bCIe, _aCI, _bBI, _aBI, _l, _comm],
    [0, 0, 0, 0, 0, 0, () => exports.Location$, () => Comments]
];
exports.Commit$ = [3, n0, _Co,
    0,
    [_cI, _tI, _pa, _m, _a, _commi, _aD],
    [0, 0, 64 | 0, 0, () => exports.UserInfo$, () => exports.UserInfo$, 0]
];
exports.Conflict$ = [3, n0, _Con,
    0,
    [_cM, _mH],
    [() => exports.ConflictMetadata$, () => MergeHunks]
];
exports.ConflictMetadata$ = [3, n0, _CM,
    0,
    [_fP, _fS, _fM, _oT, _nOC, _iBF, _cC, _fMC, _oTC, _mOe],
    [0, () => exports.FileSizes$, () => exports.FileModes$, () => exports.ObjectTypes$, 1, () => exports.IsBinaryFile$, 2, 2, 2, () => exports.MergeOperations$]
];
exports.ConflictResolution$ = [3, n0, _CR,
    0,
    [_rCe, _dF, _sFM],
    [() => ReplaceContentEntries, () => DeleteFileEntries, () => SetFileModeEntries]
];
exports.CreateApprovalRuleTemplateInput$ = [3, n0, _CARTI,
    0,
    [_aRTN, _aRTC, _aRTD],
    [0, 0, 0], 2
];
exports.CreateApprovalRuleTemplateOutput$ = [3, n0, _CARTO,
    0,
    [_aRT],
    [() => exports.ApprovalRuleTemplate$], 1
];
exports.CreateBranchInput$ = [3, n0, _CBI,
    0,
    [_rN, _bN, _cI],
    [0, 0, 0], 3
];
exports.CreateCommitInput$ = [3, n0, _CCI,
    0,
    [_rN, _bN, _pCI, _aN, _em, _cMo, _kEF, _pF, _dF, _sFM],
    [0, 0, 0, 0, 0, 0, 2, () => PutFileEntries, () => DeleteFileEntries, () => SetFileModeEntries], 2
];
exports.CreateCommitOutput$ = [3, n0, _CCO,
    0,
    [_cI, _tI, _fA, _fU, _fD],
    [0, 0, () => FilesMetadata, () => FilesMetadata, () => FilesMetadata]
];
exports.CreatePullRequestApprovalRuleInput$ = [3, n0, _CPRARI,
    0,
    [_pRI, _aRN, _aRC],
    [0, 0, 0], 3
];
exports.CreatePullRequestApprovalRuleOutput$ = [3, n0, _CPRARO,
    0,
    [_aR],
    [() => exports.ApprovalRule$], 1
];
exports.CreatePullRequestInput$ = [3, n0, _CPRI,
    0,
    [_t, _ta, _de, _cRT],
    [0, () => TargetList, 0, [0, 4]], 2
];
exports.CreatePullRequestOutput$ = [3, n0, _CPRO,
    0,
    [_pR],
    [() => exports.PullRequest$], 1
];
exports.CreateRepositoryInput$ = [3, n0, _CRI,
    0,
    [_rN, _rD, _tag, _kKI],
    [0, 0, 128 | 0, 0], 1
];
exports.CreateRepositoryOutput$ = [3, n0, _CRO,
    0,
    [_rM],
    [() => exports.RepositoryMetadata$]
];
exports.CreateUnreferencedMergeCommitInput$ = [3, n0, _CUMCI,
    0,
    [_rN, _sCS, _dCS, _mO, _cDL, _cRS, _aN, _em, _cMo, _kEF, _cRo],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 2, () => exports.ConflictResolution$], 4
];
exports.CreateUnreferencedMergeCommitOutput$ = [3, n0, _CUMCO,
    0,
    [_cI, _tI],
    [0, 0]
];
exports.DeleteApprovalRuleTemplateInput$ = [3, n0, _DARTI,
    0,
    [_aRTN],
    [0], 1
];
exports.DeleteApprovalRuleTemplateOutput$ = [3, n0, _DARTO,
    0,
    [_aRTI],
    [0], 1
];
exports.DeleteBranchInput$ = [3, n0, _DBI,
    0,
    [_rN, _bN],
    [0, 0], 2
];
exports.DeleteBranchOutput$ = [3, n0, _DBO,
    0,
    [_dB],
    [() => exports.BranchInfo$]
];
exports.DeleteCommentContentInput$ = [3, n0, _DCCI,
    0,
    [_cIom],
    [0], 1
];
exports.DeleteCommentContentOutput$ = [3, n0, _DCCO,
    0,
    [_comme],
    [() => exports.Comment$]
];
exports.DeleteFileEntry$ = [3, n0, _DFE,
    0,
    [_fP],
    [0], 1
];
exports.DeleteFileInput$ = [3, n0, _DFI,
    0,
    [_rN, _bN, _fP, _pCI, _kEF, _cMo, _n, _em],
    [0, 0, 0, 0, 2, 0, 0, 0], 4
];
exports.DeleteFileOutput$ = [3, n0, _DFO,
    0,
    [_cI, _bI, _tI, _fP],
    [0, 0, 0, 0], 4
];
exports.DeletePullRequestApprovalRuleInput$ = [3, n0, _DPRARI,
    0,
    [_pRI, _aRN],
    [0, 0], 2
];
exports.DeletePullRequestApprovalRuleOutput$ = [3, n0, _DPRARO,
    0,
    [_aRI],
    [0], 1
];
exports.DeleteRepositoryInput$ = [3, n0, _DRI,
    0,
    [_rN],
    [0], 1
];
exports.DeleteRepositoryOutput$ = [3, n0, _DRO,
    0,
    [_rIe],
    [0]
];
exports.DescribeMergeConflictsInput$ = [3, n0, _DMCI,
    0,
    [_rN, _dCS, _sCS, _mO, _fP, _mMH, _cDL, _cRS, _nT],
    [0, 0, 0, 0, 0, 1, 0, 0, 0], 5
];
exports.DescribeMergeConflictsOutput$ = [3, n0, _DMCO,
    0,
    [_cM, _mH, _dCI, _sCI, _nT, _bCI],
    [() => exports.ConflictMetadata$, () => MergeHunks, 0, 0, 0, 0], 4
];
exports.DescribePullRequestEventsInput$ = [3, n0, _DPREI,
    0,
    [_pRI, _pRET, _aAc, _nT, _mR],
    [0, 0, 0, 0, 1], 1
];
exports.DescribePullRequestEventsOutput$ = [3, n0, _DPREO,
    0,
    [_pRE, _nT],
    [() => PullRequestEventList, 0], 1
];
exports.Difference$ = [3, n0, _D,
    0,
    [_bB, _aB, _cT],
    [() => exports.BlobMetadata$, () => exports.BlobMetadata$, 0]
];
exports.DisassociateApprovalRuleTemplateFromRepositoryInput$ = [3, n0, _DARTFRI,
    0,
    [_aRTN, _rN],
    [0, 0], 2
];
exports.EvaluatePullRequestApprovalRulesInput$ = [3, n0, _EPRARI,
    0,
    [_pRI, _rI],
    [0, 0], 2
];
exports.EvaluatePullRequestApprovalRulesOutput$ = [3, n0, _EPRARO,
    0,
    [_ev],
    [() => exports.Evaluation$], 1
];
exports.Evaluation$ = [3, n0, _E,
    0,
    [_ap, _o, _aRS, _aRNS],
    [2, 2, 64 | 0, 64 | 0]
];
exports.File$ = [3, n0, _F,
    0,
    [_bI, _aP, _rP, _fMi],
    [0, 0, 0, 0]
];
exports.FileMetadata$ = [3, n0, _FM,
    0,
    [_aP, _bI, _fMi],
    [0, 0, 0]
];
exports.FileModes$ = [3, n0, _FMi,
    0,
    [_so, _des, _b],
    [0, 0, 0]
];
exports.FileSizes$ = [3, n0, _FS,
    0,
    [_so, _des, _b],
    [1, 1, 1]
];
exports.FileVersion$ = [3, n0, _FV,
    0,
    [_commit, _bI, _p, _rCev],
    [() => exports.Commit$, 0, 0, 64 | 0]
];
exports.Folder$ = [3, n0, _Fo,
    0,
    [_tI, _aP, _rP],
    [0, 0, 0]
];
exports.GetApprovalRuleTemplateInput$ = [3, n0, _GARTI,
    0,
    [_aRTN],
    [0], 1
];
exports.GetApprovalRuleTemplateOutput$ = [3, n0, _GARTO,
    0,
    [_aRT],
    [() => exports.ApprovalRuleTemplate$], 1
];
exports.GetBlobInput$ = [3, n0, _GBI,
    0,
    [_rN, _bI],
    [0, 0], 2
];
exports.GetBlobOutput$ = [3, n0, _GBO,
    0,
    [_con],
    [21], 1
];
exports.GetBranchInput$ = [3, n0, _GBIe,
    0,
    [_rN, _bN],
    [0, 0]
];
exports.GetBranchOutput$ = [3, n0, _GBOe,
    0,
    [_br],
    [() => exports.BranchInfo$]
];
exports.GetCommentInput$ = [3, n0, _GCI,
    0,
    [_cIom],
    [0], 1
];
exports.GetCommentOutput$ = [3, n0, _GCO,
    0,
    [_comme],
    [() => exports.Comment$]
];
exports.GetCommentReactionsInput$ = [3, n0, _GCRI,
    0,
    [_cIom, _rUA, _nT, _mR],
    [0, 0, 0, 1], 1
];
exports.GetCommentReactionsOutput$ = [3, n0, _GCRO,
    0,
    [_rFC, _nT],
    [() => ReactionsForCommentList, 0], 1
];
exports.GetCommentsForComparedCommitInput$ = [3, n0, _GCFCCI,
    0,
    [_rN, _aCI, _bCIe, _nT, _mR],
    [0, 0, 0, 0, 1], 2
];
exports.GetCommentsForComparedCommitOutput$ = [3, n0, _GCFCCO,
    0,
    [_cFCCD, _nT],
    [() => CommentsForComparedCommitData, 0]
];
exports.GetCommentsForPullRequestInput$ = [3, n0, _GCFPRI,
    0,
    [_pRI, _rN, _bCIe, _aCI, _nT, _mR],
    [0, 0, 0, 0, 0, 1], 1
];
exports.GetCommentsForPullRequestOutput$ = [3, n0, _GCFPRO,
    0,
    [_cFPRD, _nT],
    [() => CommentsForPullRequestData, 0]
];
exports.GetCommitInput$ = [3, n0, _GCIe,
    0,
    [_rN, _cI],
    [0, 0], 2
];
exports.GetCommitOutput$ = [3, n0, _GCOe,
    0,
    [_commit],
    [() => exports.Commit$], 1
];
exports.GetDifferencesInput$ = [3, n0, _GDI,
    0,
    [_rN, _aCS, _bCS, _bP, _aPf, _MR, _NT],
    [0, 0, 0, 0, 0, 1, 0], 2
];
exports.GetDifferencesOutput$ = [3, n0, _GDO,
    0,
    [_di, _NT],
    [() => DifferenceList, 0]
];
exports.GetFileInput$ = [3, n0, _GFI,
    0,
    [_rN, _fP, _cS],
    [0, 0, 0], 2
];
exports.GetFileOutput$ = [3, n0, _GFO,
    0,
    [_cI, _bI, _fP, _fMi, _fSi, _fC],
    [0, 0, 0, 0, 1, 21], 6
];
exports.GetFolderInput$ = [3, n0, _GFIe,
    0,
    [_rN, _fPo, _cS],
    [0, 0, 0], 2
];
exports.GetFolderOutput$ = [3, n0, _GFOe,
    0,
    [_cI, _fPo, _tI, _sF, _f, _sL, _sM],
    [0, 0, 0, () => FolderList, () => FileList, () => SymbolicLinkList, () => SubModuleList], 2
];
exports.GetMergeCommitInput$ = [3, n0, _GMCI,
    0,
    [_rN, _sCS, _dCS, _cDL, _cRS],
    [0, 0, 0, 0, 0], 3
];
exports.GetMergeCommitOutput$ = [3, n0, _GMCO,
    0,
    [_sCI, _dCI, _bCI, _mCI],
    [0, 0, 0, 0]
];
exports.GetMergeConflictsInput$ = [3, n0, _GMCIe,
    0,
    [_rN, _dCS, _sCS, _mO, _cDL, _mCF, _cRS, _nT],
    [0, 0, 0, 0, 0, 1, 0, 0], 4
];
exports.GetMergeConflictsOutput$ = [3, n0, _GMCOe,
    0,
    [_me, _dCI, _sCI, _cML, _bCI, _nT],
    [2, 0, 0, () => ConflictMetadataList, 0, 0], 4
];
exports.GetMergeOptionsInput$ = [3, n0, _GMOI,
    0,
    [_rN, _sCS, _dCS, _cDL, _cRS],
    [0, 0, 0, 0, 0], 3
];
exports.GetMergeOptionsOutput$ = [3, n0, _GMOO,
    0,
    [_mOer, _sCI, _dCI, _bCI],
    [64 | 0, 0, 0, 0], 4
];
exports.GetPullRequestApprovalStatesInput$ = [3, n0, _GPRASI,
    0,
    [_pRI, _rI],
    [0, 0], 2
];
exports.GetPullRequestApprovalStatesOutput$ = [3, n0, _GPRASO,
    0,
    [_app],
    [() => ApprovalList]
];
exports.GetPullRequestInput$ = [3, n0, _GPRI,
    0,
    [_pRI],
    [0], 1
];
exports.GetPullRequestOutput$ = [3, n0, _GPRO,
    0,
    [_pR],
    [() => exports.PullRequest$], 1
];
exports.GetPullRequestOverrideStateInput$ = [3, n0, _GPROSI,
    0,
    [_pRI, _rI],
    [0, 0], 2
];
exports.GetPullRequestOverrideStateOutput$ = [3, n0, _GPROSO,
    0,
    [_o, _ov],
    [2, 0]
];
exports.GetRepositoryInput$ = [3, n0, _GRI,
    0,
    [_rN],
    [0], 1
];
exports.GetRepositoryOutput$ = [3, n0, _GRO,
    0,
    [_rM],
    [() => exports.RepositoryMetadata$]
];
exports.GetRepositoryTriggersInput$ = [3, n0, _GRTI,
    0,
    [_rN],
    [0], 1
];
exports.GetRepositoryTriggersOutput$ = [3, n0, _GRTO,
    0,
    [_cIon, _tr],
    [0, () => RepositoryTriggersList]
];
exports.IsBinaryFile$ = [3, n0, _IBF,
    0,
    [_so, _des, _b],
    [2, 2, 2]
];
exports.ListApprovalRuleTemplatesInput$ = [3, n0, _LARTI,
    0,
    [_nT, _mR],
    [0, 1]
];
exports.ListApprovalRuleTemplatesOutput$ = [3, n0, _LARTO,
    0,
    [_aRTNp, _nT],
    [64 | 0, 0]
];
exports.ListAssociatedApprovalRuleTemplatesForRepositoryInput$ = [3, n0, _LAARTFRI,
    0,
    [_rN, _nT, _mR],
    [0, 0, 1], 1
];
exports.ListAssociatedApprovalRuleTemplatesForRepositoryOutput$ = [3, n0, _LAARTFRO,
    0,
    [_aRTNp, _nT],
    [64 | 0, 0]
];
exports.ListBranchesInput$ = [3, n0, _LBI,
    0,
    [_rN, _nT],
    [0, 0], 1
];
exports.ListBranchesOutput$ = [3, n0, _LBO,
    0,
    [_bra, _nT],
    [64 | 0, 0]
];
exports.ListFileCommitHistoryRequest$ = [3, n0, _LFCHR,
    0,
    [_rN, _fP, _cS, _mR, _nT],
    [0, 0, 0, 1, 0], 2
];
exports.ListFileCommitHistoryResponse$ = [3, n0, _LFCHRi,
    0,
    [_rDe, _nT],
    [() => RevisionDag, 0], 1
];
exports.ListPullRequestsInput$ = [3, n0, _LPRI,
    0,
    [_rN, _aA, _pRS, _nT, _mR],
    [0, 0, 0, 0, 1], 1
];
exports.ListPullRequestsOutput$ = [3, n0, _LPRO,
    0,
    [_pRIu, _nT],
    [64 | 0, 0], 1
];
exports.ListRepositoriesForApprovalRuleTemplateInput$ = [3, n0, _LRFARTI,
    0,
    [_aRTN, _nT, _mR],
    [0, 0, 1], 1
];
exports.ListRepositoriesForApprovalRuleTemplateOutput$ = [3, n0, _LRFARTO,
    0,
    [_rNe, _nT],
    [64 | 0, 0]
];
exports.ListRepositoriesInput$ = [3, n0, _LRI,
    0,
    [_nT, _sB, _or],
    [0, 0, 0]
];
exports.ListRepositoriesOutput$ = [3, n0, _LRO,
    0,
    [_r, _nT],
    [() => RepositoryNameIdPairList, 0]
];
exports.ListTagsForResourceInput$ = [3, n0, _LTFRI,
    0,
    [_rA, _nT],
    [0, 0], 1
];
exports.ListTagsForResourceOutput$ = [3, n0, _LTFRO,
    0,
    [_tag, _nT],
    [128 | 0, 0]
];
exports.Location$ = [3, n0, _L,
    0,
    [_fP, _fPil, _rFV],
    [0, 1, 0]
];
exports.MergeBranchesByFastForwardInput$ = [3, n0, _MBBFFI,
    0,
    [_rN, _sCS, _dCS, _tB],
    [0, 0, 0, 0], 3
];
exports.MergeBranchesByFastForwardOutput$ = [3, n0, _MBBFFO,
    0,
    [_cI, _tI],
    [0, 0]
];
exports.MergeBranchesBySquashInput$ = [3, n0, _MBBSI,
    0,
    [_rN, _sCS, _dCS, _tB, _cDL, _cRS, _aN, _em, _cMo, _kEF, _cRo],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 2, () => exports.ConflictResolution$], 3
];
exports.MergeBranchesBySquashOutput$ = [3, n0, _MBBSO,
    0,
    [_cI, _tI],
    [0, 0]
];
exports.MergeBranchesByThreeWayInput$ = [3, n0, _MBBTWI,
    0,
    [_rN, _sCS, _dCS, _tB, _cDL, _cRS, _aN, _em, _cMo, _kEF, _cRo],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 2, () => exports.ConflictResolution$], 3
];
exports.MergeBranchesByThreeWayOutput$ = [3, n0, _MBBTWO,
    0,
    [_cI, _tI],
    [0, 0]
];
exports.MergeHunk$ = [3, n0, _MH,
    0,
    [_iC, _so, _des, _b],
    [2, () => exports.MergeHunkDetail$, () => exports.MergeHunkDetail$, () => exports.MergeHunkDetail$]
];
exports.MergeHunkDetail$ = [3, n0, _MHD,
    0,
    [_sLt, _eL, _hC],
    [1, 1, 0]
];
exports.MergeMetadata$ = [3, n0, _MM,
    0,
    [_iM, _mB, _mCIe, _mO],
    [2, 0, 0, 0]
];
exports.MergeOperations$ = [3, n0, _MO,
    0,
    [_so, _des],
    [0, 0]
];
exports.MergePullRequestByFastForwardInput$ = [3, n0, _MPRBFFI,
    0,
    [_pRI, _rN, _sCI],
    [0, 0, 0], 2
];
exports.MergePullRequestByFastForwardOutput$ = [3, n0, _MPRBFFO,
    0,
    [_pR],
    [() => exports.PullRequest$]
];
exports.MergePullRequestBySquashInput$ = [3, n0, _MPRBSI,
    0,
    [_pRI, _rN, _sCI, _cDL, _cRS, _cMo, _aN, _em, _kEF, _cRo],
    [0, 0, 0, 0, 0, 0, 0, 0, 2, () => exports.ConflictResolution$], 2
];
exports.MergePullRequestBySquashOutput$ = [3, n0, _MPRBSO,
    0,
    [_pR],
    [() => exports.PullRequest$]
];
exports.MergePullRequestByThreeWayInput$ = [3, n0, _MPRBTWI,
    0,
    [_pRI, _rN, _sCI, _cDL, _cRS, _cMo, _aN, _em, _kEF, _cRo],
    [0, 0, 0, 0, 0, 0, 0, 0, 2, () => exports.ConflictResolution$], 2
];
exports.MergePullRequestByThreeWayOutput$ = [3, n0, _MPRBTWO,
    0,
    [_pR],
    [() => exports.PullRequest$]
];
exports.ObjectTypes$ = [3, n0, _OT,
    0,
    [_so, _des, _b],
    [0, 0, 0]
];
exports.OriginApprovalRuleTemplate$ = [3, n0, _OART,
    0,
    [_aRTI, _aRTN],
    [0, 0]
];
exports.OverridePullRequestApprovalRulesInput$ = [3, n0, _OPRARI,
    0,
    [_pRI, _rI, _oS],
    [0, 0, 0], 3
];
exports.PostCommentForComparedCommitInput$ = [3, n0, _PCFCCI,
    0,
    [_rN, _aCI, _con, _bCIe, _l, _cRT],
    [0, 0, 0, 0, () => exports.Location$, [0, 4]], 3
];
exports.PostCommentForComparedCommitOutput$ = [3, n0, _PCFCCO,
    0,
    [_rN, _bCIe, _aCI, _bBI, _aBI, _l, _comme],
    [0, 0, 0, 0, 0, () => exports.Location$, () => exports.Comment$]
];
exports.PostCommentForPullRequestInput$ = [3, n0, _PCFPRI,
    0,
    [_pRI, _rN, _bCIe, _aCI, _con, _l, _cRT],
    [0, 0, 0, 0, 0, () => exports.Location$, [0, 4]], 5
];
exports.PostCommentForPullRequestOutput$ = [3, n0, _PCFPRO,
    0,
    [_rN, _pRI, _bCIe, _aCI, _bBI, _aBI, _l, _comme],
    [0, 0, 0, 0, 0, 0, () => exports.Location$, () => exports.Comment$]
];
exports.PostCommentReplyInput$ = [3, n0, _PCRI,
    0,
    [_iRT, _con, _cRT],
    [0, 0, [0, 4]], 2
];
exports.PostCommentReplyOutput$ = [3, n0, _PCRO,
    0,
    [_comme],
    [() => exports.Comment$]
];
exports.PullRequest$ = [3, n0, _PR,
    0,
    [_pRI, _t, _de, _lAD, _cD, _pRS, _aA, _pRT, _cRT, _rI, _aRp],
    [0, 0, 0, 4, 4, 0, 0, () => PullRequestTargetList, 0, 0, () => ApprovalRulesList]
];
exports.PullRequestCreatedEventMetadata$ = [3, n0, _PRCEM,
    0,
    [_rN, _sCI, _dCI, _mBe],
    [0, 0, 0, 0]
];
exports.PullRequestEvent$ = [3, n0, _PREu,
    0,
    [_pRI, _eD, _pRET, _aAc, _pRCEM, _pRSCEM, _pRSRUEM, _pRMSCEM, _aREM, _aSCEM, _aROEM],
    [0, 4, 0, 0, () => exports.PullRequestCreatedEventMetadata$, () => exports.PullRequestStatusChangedEventMetadata$, () => exports.PullRequestSourceReferenceUpdatedEventMetadata$, () => exports.PullRequestMergedStateChangedEventMetadata$, () => exports.ApprovalRuleEventMetadata$, () => exports.ApprovalStateChangedEventMetadata$, () => exports.ApprovalRuleOverriddenEventMetadata$]
];
exports.PullRequestMergedStateChangedEventMetadata$ = [3, n0, _PRMSCEM,
    0,
    [_rN, _dR, _mM],
    [0, 0, () => exports.MergeMetadata$]
];
exports.PullRequestSourceReferenceUpdatedEventMetadata$ = [3, n0, _PRSRUEM,
    0,
    [_rN, _bCIe, _aCI, _mBe],
    [0, 0, 0, 0]
];
exports.PullRequestStatusChangedEventMetadata$ = [3, n0, _PRSCEM,
    0,
    [_pRS],
    [0]
];
exports.PullRequestTarget$ = [3, n0, _PRT,
    0,
    [_rN, _sR, _dR, _dC, _sC, _mBe, _mM],
    [0, 0, 0, 0, 0, 0, () => exports.MergeMetadata$]
];
exports.PutCommentReactionInput$ = [3, n0, _PCRIu,
    0,
    [_cIom, _rV],
    [0, 0], 2
];
exports.PutFileEntry$ = [3, n0, _PFE,
    0,
    [_fP, _fMi, _fC, _sFo],
    [0, 0, 21, () => exports.SourceFileSpecifier$], 1
];
exports.PutFileInput$ = [3, n0, _PFI,
    0,
    [_rN, _bN, _fC, _fP, _fMi, _pCI, _cMo, _n, _em],
    [0, 0, 21, 0, 0, 0, 0, 0, 0], 4
];
exports.PutFileOutput$ = [3, n0, _PFO,
    0,
    [_cI, _bI, _tI],
    [0, 0, 0], 3
];
exports.PutRepositoryTriggersInput$ = [3, n0, _PRTI,
    0,
    [_rN, _tr],
    [0, () => RepositoryTriggersList], 2
];
exports.PutRepositoryTriggersOutput$ = [3, n0, _PRTO,
    0,
    [_cIon],
    [0]
];
exports.ReactionForComment$ = [3, n0, _RFC,
    0,
    [_re, _rU, _rFDUC],
    [() => exports.ReactionValueFormats$, 64 | 0, 1]
];
exports.ReactionValueFormats$ = [3, n0, _RVF,
    0,
    [_emo, _sCh, _u],
    [0, 0, 0]
];
exports.ReplaceContentEntry$ = [3, n0, _RCE,
    0,
    [_fP, _rT, _con, _fMi],
    [0, 0, 21, 0], 2
];
exports.RepositoryMetadata$ = [3, n0, _RM,
    0,
    [_aI, _rIe, _rN, _rD, _dBe, _lMD, _cD, _cUH, _cUS, _Ar, _kKI],
    [0, 0, 0, 0, 0, 4, 4, 0, 0, 0, 0]
];
exports.RepositoryNameIdPair$ = [3, n0, _RNIP,
    0,
    [_rN, _rIe],
    [0, 0]
];
exports.RepositoryTrigger$ = [3, n0, _RT,
    0,
    [_n, _dA, _eve, _cDu, _bra],
    [0, 0, 64 | 0, 0, 64 | 0], 3
];
exports.RepositoryTriggerExecutionFailure$ = [3, n0, _RTEF,
    0,
    [_tri, _fMa],
    [0, 0]
];
exports.SetFileModeEntry$ = [3, n0, _SFME,
    0,
    [_fP, _fMi],
    [0, 0], 2
];
exports.SourceFileSpecifier$ = [3, n0, _SFS,
    0,
    [_fP, _iMs],
    [0, 2], 1
];
exports.SubModule$ = [3, n0, _SM,
    0,
    [_cI, _aP, _rP],
    [0, 0, 0]
];
exports.SymbolicLink$ = [3, n0, _SL,
    0,
    [_bI, _aP, _rP, _fMi],
    [0, 0, 0, 0]
];
exports.TagResourceInput$ = [3, n0, _TRI,
    0,
    [_rA, _tag],
    [0, 128 | 0], 2
];
exports.Target$ = [3, n0, _T,
    0,
    [_rN, _sR, _dR],
    [0, 0, 0], 2
];
exports.TestRepositoryTriggersInput$ = [3, n0, _TRTI,
    0,
    [_rN, _tr],
    [0, () => RepositoryTriggersList], 2
];
exports.TestRepositoryTriggersOutput$ = [3, n0, _TRTO,
    0,
    [_sE, _fE],
    [64 | 0, () => RepositoryTriggerExecutionFailureList]
];
exports.UntagResourceInput$ = [3, n0, _URI,
    0,
    [_rA, _tK],
    [0, 64 | 0], 2
];
exports.UpdateApprovalRuleTemplateContentInput$ = [3, n0, _UARTCI,
    0,
    [_aRTN, _nRC, _eRCS],
    [0, 0, 0], 2
];
exports.UpdateApprovalRuleTemplateContentOutput$ = [3, n0, _UARTCO,
    0,
    [_aRT],
    [() => exports.ApprovalRuleTemplate$], 1
];
exports.UpdateApprovalRuleTemplateDescriptionInput$ = [3, n0, _UARTDI,
    0,
    [_aRTN, _aRTD],
    [0, 0], 2
];
exports.UpdateApprovalRuleTemplateDescriptionOutput$ = [3, n0, _UARTDO,
    0,
    [_aRT],
    [() => exports.ApprovalRuleTemplate$], 1
];
exports.UpdateApprovalRuleTemplateNameInput$ = [3, n0, _UARTNI,
    0,
    [_oARTN, _nARTN],
    [0, 0], 2
];
exports.UpdateApprovalRuleTemplateNameOutput$ = [3, n0, _UARTNO,
    0,
    [_aRT],
    [() => exports.ApprovalRuleTemplate$], 1
];
exports.UpdateCommentInput$ = [3, n0, _UCI,
    0,
    [_cIom, _con],
    [0, 0], 2
];
exports.UpdateCommentOutput$ = [3, n0, _UCO,
    0,
    [_comme],
    [() => exports.Comment$]
];
exports.UpdateDefaultBranchInput$ = [3, n0, _UDBI,
    0,
    [_rN, _dBN],
    [0, 0], 2
];
exports.UpdatePullRequestApprovalRuleContentInput$ = [3, n0, _UPRARCI,
    0,
    [_pRI, _aRN, _nRC, _eRCS],
    [0, 0, 0, 0], 3
];
exports.UpdatePullRequestApprovalRuleContentOutput$ = [3, n0, _UPRARCO,
    0,
    [_aR],
    [() => exports.ApprovalRule$], 1
];
exports.UpdatePullRequestApprovalStateInput$ = [3, n0, _UPRASI,
    0,
    [_pRI, _rI, _aS],
    [0, 0, 0], 3
];
exports.UpdatePullRequestDescriptionInput$ = [3, n0, _UPRDI,
    0,
    [_pRI, _de],
    [0, 0], 2
];
exports.UpdatePullRequestDescriptionOutput$ = [3, n0, _UPRDO,
    0,
    [_pR],
    [() => exports.PullRequest$], 1
];
exports.UpdatePullRequestStatusInput$ = [3, n0, _UPRSI,
    0,
    [_pRI, _pRS],
    [0, 0], 2
];
exports.UpdatePullRequestStatusOutput$ = [3, n0, _UPRSO,
    0,
    [_pR],
    [() => exports.PullRequest$], 1
];
exports.UpdatePullRequestTitleInput$ = [3, n0, _UPRTI,
    0,
    [_pRI, _t],
    [0, 0], 2
];
exports.UpdatePullRequestTitleOutput$ = [3, n0, _UPRTO,
    0,
    [_pR],
    [() => exports.PullRequest$], 1
];
exports.UpdateRepositoryDescriptionInput$ = [3, n0, _URDI,
    0,
    [_rN, _rD],
    [0, 0], 1
];
exports.UpdateRepositoryEncryptionKeyInput$ = [3, n0, _UREKI,
    0,
    [_rN, _kKI],
    [0, 0], 2
];
exports.UpdateRepositoryEncryptionKeyOutput$ = [3, n0, _UREKO,
    0,
    [_rIe, _kKI, _oKKI],
    [0, 0, 0]
];
exports.UpdateRepositoryNameInput$ = [3, n0, _URNI,
    0,
    [_oN, _nN],
    [0, 0], 2
];
exports.UserInfo$ = [3, n0, _UI,
    0,
    [_n, _em, _da],
    [0, 0, 0]
];
var __Unit = "unit";
var ApprovalList = [1, n0, _AL,
    0, () => exports.Approval$
];
var ApprovalRulesList = [1, n0, _ARL,
    0, () => exports.ApprovalRule$
];
var ApprovalRulesNotSatisfiedList = 64 | 0;
var ApprovalRulesSatisfiedList = 64 | 0;
var ApprovalRuleTemplateNameList = 64 | 0;
var BatchAssociateApprovalRuleTemplateWithRepositoriesErrorsList = [1, n0, _BAARTWREL,
    0, () => exports.BatchAssociateApprovalRuleTemplateWithRepositoriesError$
];
var BatchDescribeMergeConflictsErrors = [1, n0, _BDMCEa,
    0, () => exports.BatchDescribeMergeConflictsError$
];
var BatchDisassociateApprovalRuleTemplateFromRepositoriesErrorsList = [1, n0, _BDARTFREL,
    0, () => exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesError$
];
var BatchGetCommitsErrorsList = [1, n0, _BGCEL,
    0, () => exports.BatchGetCommitsError$
];
var BatchGetRepositoriesErrorsList = [1, n0, _BGREL,
    0, () => exports.BatchGetRepositoriesError$
];
var BranchNameList = 64 | 0;
var CallerReactions = 64 | 0;
var Comments = [1, n0, _Com,
    0, () => exports.Comment$
];
var CommentsForComparedCommitData = [1, n0, _CFCCD,
    0, () => exports.CommentsForComparedCommit$
];
var CommentsForPullRequestData = [1, n0, _CFPRD,
    0, () => exports.CommentsForPullRequest$
];
var CommitIdsInputList = 64 | 0;
var CommitObjectsList = [1, n0, _COL,
    0, () => exports.Commit$
];
var ConflictMetadataList = [1, n0, _CML,
    0, () => exports.ConflictMetadata$
];
var Conflicts = [1, n0, _Conf,
    0, () => exports.Conflict$
];
var DeleteFileEntries = [1, n0, _DFEe,
    0, () => exports.DeleteFileEntry$
];
var DifferenceList = [1, n0, _DL,
    0, () => exports.Difference$
];
var FileList = [1, n0, _FL,
    0, () => exports.File$
];
var FilePaths = 64 | 0;
var FilesMetadata = [1, n0, _FMil,
    0, () => exports.FileMetadata$
];
var FolderList = [1, n0, _FLo,
    0, () => exports.Folder$
];
var MergeHunks = [1, n0, _MHe,
    0, () => exports.MergeHunk$
];
var MergeOptions = 64 | 0;
var ParentList = 64 | 0;
var PullRequestEventList = [1, n0, _PREL,
    0, () => exports.PullRequestEvent$
];
var PullRequestIdList = 64 | 0;
var PullRequestTargetList = [1, n0, _PRTL,
    0, () => exports.PullRequestTarget$
];
var PutFileEntries = [1, n0, _PFEu,
    0, () => exports.PutFileEntry$
];
var ReactionsForCommentList = [1, n0, _RFCL,
    0, () => exports.ReactionForComment$
];
var ReactionUsersList = 64 | 0;
var ReplaceContentEntries = [1, n0, _RCEe,
    0, () => exports.ReplaceContentEntry$
];
var RepositoryMetadataList = [1, n0, _RML,
    0, () => exports.RepositoryMetadata$
];
var RepositoryNameIdPairList = [1, n0, _RNIPL,
    0, () => exports.RepositoryNameIdPair$
];
var RepositoryNameList = 64 | 0;
var RepositoryNotFoundList = 64 | 0;
var RepositoryTriggerEventList = 64 | 0;
var RepositoryTriggerExecutionFailureList = [1, n0, _RTEFL,
    0, () => exports.RepositoryTriggerExecutionFailure$
];
var RepositoryTriggerNameList = 64 | 0;
var RepositoryTriggersList = [1, n0, _RTL,
    0, () => exports.RepositoryTrigger$
];
var RevisionChildren = 64 | 0;
var RevisionDag = [1, n0, _RD,
    0, () => exports.FileVersion$
];
var SetFileModeEntries = [1, n0, _SFMEe,
    0, () => exports.SetFileModeEntry$
];
var SubModuleList = [1, n0, _SML,
    0, () => exports.SubModule$
];
var SymbolicLinkList = [1, n0, _SLL,
    0, () => exports.SymbolicLink$
];
var TagKeysList = 64 | 0;
var TargetList = [1, n0, _TL,
    0, () => exports.Target$
];
var ReactionCountsMap = 128 | 1;
var TagsMap = 128 | 0;
exports.AssociateApprovalRuleTemplateWithRepository$ = [9, n0, _AARTWR,
    0, () => exports.AssociateApprovalRuleTemplateWithRepositoryInput$, () => __Unit
];
exports.BatchAssociateApprovalRuleTemplateWithRepositories$ = [9, n0, _BAARTWR,
    0, () => exports.BatchAssociateApprovalRuleTemplateWithRepositoriesInput$, () => exports.BatchAssociateApprovalRuleTemplateWithRepositoriesOutput$
];
exports.BatchDescribeMergeConflicts$ = [9, n0, _BDMC,
    0, () => exports.BatchDescribeMergeConflictsInput$, () => exports.BatchDescribeMergeConflictsOutput$
];
exports.BatchDisassociateApprovalRuleTemplateFromRepositories$ = [9, n0, _BDARTFR,
    0, () => exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesInput$, () => exports.BatchDisassociateApprovalRuleTemplateFromRepositoriesOutput$
];
exports.BatchGetCommits$ = [9, n0, _BGC,
    0, () => exports.BatchGetCommitsInput$, () => exports.BatchGetCommitsOutput$
];
exports.BatchGetRepositories$ = [9, n0, _BGR,
    0, () => exports.BatchGetRepositoriesInput$, () => exports.BatchGetRepositoriesOutput$
];
exports.CreateApprovalRuleTemplate$ = [9, n0, _CART,
    0, () => exports.CreateApprovalRuleTemplateInput$, () => exports.CreateApprovalRuleTemplateOutput$
];
exports.CreateBranch$ = [9, n0, _CB,
    0, () => exports.CreateBranchInput$, () => __Unit
];
exports.CreateCommit$ = [9, n0, _CC,
    0, () => exports.CreateCommitInput$, () => exports.CreateCommitOutput$
];
exports.CreatePullRequest$ = [9, n0, _CPR,
    0, () => exports.CreatePullRequestInput$, () => exports.CreatePullRequestOutput$
];
exports.CreatePullRequestApprovalRule$ = [9, n0, _CPRAR,
    0, () => exports.CreatePullRequestApprovalRuleInput$, () => exports.CreatePullRequestApprovalRuleOutput$
];
exports.CreateRepository$ = [9, n0, _CRr,
    0, () => exports.CreateRepositoryInput$, () => exports.CreateRepositoryOutput$
];
exports.CreateUnreferencedMergeCommit$ = [9, n0, _CUMC,
    0, () => exports.CreateUnreferencedMergeCommitInput$, () => exports.CreateUnreferencedMergeCommitOutput$
];
exports.DeleteApprovalRuleTemplate$ = [9, n0, _DART,
    0, () => exports.DeleteApprovalRuleTemplateInput$, () => exports.DeleteApprovalRuleTemplateOutput$
];
exports.DeleteBranch$ = [9, n0, _DB,
    0, () => exports.DeleteBranchInput$, () => exports.DeleteBranchOutput$
];
exports.DeleteCommentContent$ = [9, n0, _DCC,
    0, () => exports.DeleteCommentContentInput$, () => exports.DeleteCommentContentOutput$
];
exports.DeleteFile$ = [9, n0, _DF,
    0, () => exports.DeleteFileInput$, () => exports.DeleteFileOutput$
];
exports.DeletePullRequestApprovalRule$ = [9, n0, _DPRAR,
    0, () => exports.DeletePullRequestApprovalRuleInput$, () => exports.DeletePullRequestApprovalRuleOutput$
];
exports.DeleteRepository$ = [9, n0, _DR,
    0, () => exports.DeleteRepositoryInput$, () => exports.DeleteRepositoryOutput$
];
exports.DescribeMergeConflicts$ = [9, n0, _DMC,
    0, () => exports.DescribeMergeConflictsInput$, () => exports.DescribeMergeConflictsOutput$
];
exports.DescribePullRequestEvents$ = [9, n0, _DPRE,
    0, () => exports.DescribePullRequestEventsInput$, () => exports.DescribePullRequestEventsOutput$
];
exports.DisassociateApprovalRuleTemplateFromRepository$ = [9, n0, _DARTFR,
    0, () => exports.DisassociateApprovalRuleTemplateFromRepositoryInput$, () => __Unit
];
exports.EvaluatePullRequestApprovalRules$ = [9, n0, _EPRAR,
    0, () => exports.EvaluatePullRequestApprovalRulesInput$, () => exports.EvaluatePullRequestApprovalRulesOutput$
];
exports.GetApprovalRuleTemplate$ = [9, n0, _GART,
    0, () => exports.GetApprovalRuleTemplateInput$, () => exports.GetApprovalRuleTemplateOutput$
];
exports.GetBlob$ = [9, n0, _GB,
    0, () => exports.GetBlobInput$, () => exports.GetBlobOutput$
];
exports.GetBranch$ = [9, n0, _GBe,
    0, () => exports.GetBranchInput$, () => exports.GetBranchOutput$
];
exports.GetComment$ = [9, n0, _GC,
    0, () => exports.GetCommentInput$, () => exports.GetCommentOutput$
];
exports.GetCommentReactions$ = [9, n0, _GCR,
    0, () => exports.GetCommentReactionsInput$, () => exports.GetCommentReactionsOutput$
];
exports.GetCommentsForComparedCommit$ = [9, n0, _GCFCC,
    0, () => exports.GetCommentsForComparedCommitInput$, () => exports.GetCommentsForComparedCommitOutput$
];
exports.GetCommentsForPullRequest$ = [9, n0, _GCFPR,
    0, () => exports.GetCommentsForPullRequestInput$, () => exports.GetCommentsForPullRequestOutput$
];
exports.GetCommit$ = [9, n0, _GCe,
    0, () => exports.GetCommitInput$, () => exports.GetCommitOutput$
];
exports.GetDifferences$ = [9, n0, _GD,
    0, () => exports.GetDifferencesInput$, () => exports.GetDifferencesOutput$
];
exports.GetFile$ = [9, n0, _GF,
    0, () => exports.GetFileInput$, () => exports.GetFileOutput$
];
exports.GetFolder$ = [9, n0, _GFe,
    0, () => exports.GetFolderInput$, () => exports.GetFolderOutput$
];
exports.GetMergeCommit$ = [9, n0, _GMC,
    0, () => exports.GetMergeCommitInput$, () => exports.GetMergeCommitOutput$
];
exports.GetMergeConflicts$ = [9, n0, _GMCe,
    0, () => exports.GetMergeConflictsInput$, () => exports.GetMergeConflictsOutput$
];
exports.GetMergeOptions$ = [9, n0, _GMO,
    0, () => exports.GetMergeOptionsInput$, () => exports.GetMergeOptionsOutput$
];
exports.GetPullRequest$ = [9, n0, _GPR,
    0, () => exports.GetPullRequestInput$, () => exports.GetPullRequestOutput$
];
exports.GetPullRequestApprovalStates$ = [9, n0, _GPRAS,
    0, () => exports.GetPullRequestApprovalStatesInput$, () => exports.GetPullRequestApprovalStatesOutput$
];
exports.GetPullRequestOverrideState$ = [9, n0, _GPROS,
    0, () => exports.GetPullRequestOverrideStateInput$, () => exports.GetPullRequestOverrideStateOutput$
];
exports.GetRepository$ = [9, n0, _GR,
    0, () => exports.GetRepositoryInput$, () => exports.GetRepositoryOutput$
];
exports.GetRepositoryTriggers$ = [9, n0, _GRT,
    0, () => exports.GetRepositoryTriggersInput$, () => exports.GetRepositoryTriggersOutput$
];
exports.ListApprovalRuleTemplates$ = [9, n0, _LART,
    0, () => exports.ListApprovalRuleTemplatesInput$, () => exports.ListApprovalRuleTemplatesOutput$
];
exports.ListAssociatedApprovalRuleTemplatesForRepository$ = [9, n0, _LAARTFR,
    0, () => exports.ListAssociatedApprovalRuleTemplatesForRepositoryInput$, () => exports.ListAssociatedApprovalRuleTemplatesForRepositoryOutput$
];
exports.ListBranches$ = [9, n0, _LB,
    0, () => exports.ListBranchesInput$, () => exports.ListBranchesOutput$
];
exports.ListFileCommitHistory$ = [9, n0, _LFCH,
    0, () => exports.ListFileCommitHistoryRequest$, () => exports.ListFileCommitHistoryResponse$
];
exports.ListPullRequests$ = [9, n0, _LPR,
    0, () => exports.ListPullRequestsInput$, () => exports.ListPullRequestsOutput$
];
exports.ListRepositories$ = [9, n0, _LR,
    0, () => exports.ListRepositoriesInput$, () => exports.ListRepositoriesOutput$
];
exports.ListRepositoriesForApprovalRuleTemplate$ = [9, n0, _LRFART,
    0, () => exports.ListRepositoriesForApprovalRuleTemplateInput$, () => exports.ListRepositoriesForApprovalRuleTemplateOutput$
];
exports.ListTagsForResource$ = [9, n0, _LTFR,
    0, () => exports.ListTagsForResourceInput$, () => exports.ListTagsForResourceOutput$
];
exports.MergeBranchesByFastForward$ = [9, n0, _MBBFF,
    0, () => exports.MergeBranchesByFastForwardInput$, () => exports.MergeBranchesByFastForwardOutput$
];
exports.MergeBranchesBySquash$ = [9, n0, _MBBS,
    0, () => exports.MergeBranchesBySquashInput$, () => exports.MergeBranchesBySquashOutput$
];
exports.MergeBranchesByThreeWay$ = [9, n0, _MBBTW,
    0, () => exports.MergeBranchesByThreeWayInput$, () => exports.MergeBranchesByThreeWayOutput$
];
exports.MergePullRequestByFastForward$ = [9, n0, _MPRBFF,
    0, () => exports.MergePullRequestByFastForwardInput$, () => exports.MergePullRequestByFastForwardOutput$
];
exports.MergePullRequestBySquash$ = [9, n0, _MPRBS,
    0, () => exports.MergePullRequestBySquashInput$, () => exports.MergePullRequestBySquashOutput$
];
exports.MergePullRequestByThreeWay$ = [9, n0, _MPRBTW,
    0, () => exports.MergePullRequestByThreeWayInput$, () => exports.MergePullRequestByThreeWayOutput$
];
exports.OverridePullRequestApprovalRules$ = [9, n0, _OPRAR,
    0, () => exports.OverridePullRequestApprovalRulesInput$, () => __Unit
];
exports.PostCommentForComparedCommit$ = [9, n0, _PCFCC,
    2, () => exports.PostCommentForComparedCommitInput$, () => exports.PostCommentForComparedCommitOutput$
];
exports.PostCommentForPullRequest$ = [9, n0, _PCFPR,
    2, () => exports.PostCommentForPullRequestInput$, () => exports.PostCommentForPullRequestOutput$
];
exports.PostCommentReply$ = [9, n0, _PCR,
    2, () => exports.PostCommentReplyInput$, () => exports.PostCommentReplyOutput$
];
exports.PutCommentReaction$ = [9, n0, _PCRu,
    0, () => exports.PutCommentReactionInput$, () => __Unit
];
exports.PutFile$ = [9, n0, _PF,
    0, () => exports.PutFileInput$, () => exports.PutFileOutput$
];
exports.PutRepositoryTriggers$ = [9, n0, _PRTu,
    0, () => exports.PutRepositoryTriggersInput$, () => exports.PutRepositoryTriggersOutput$
];
exports.TagResource$ = [9, n0, _TR,
    0, () => exports.TagResourceInput$, () => __Unit
];
exports.TestRepositoryTriggers$ = [9, n0, _TRT,
    0, () => exports.TestRepositoryTriggersInput$, () => exports.TestRepositoryTriggersOutput$
];
exports.UntagResource$ = [9, n0, _UR,
    0, () => exports.UntagResourceInput$, () => __Unit
];
exports.UpdateApprovalRuleTemplateContent$ = [9, n0, _UARTC,
    0, () => exports.UpdateApprovalRuleTemplateContentInput$, () => exports.UpdateApprovalRuleTemplateContentOutput$
];
exports.UpdateApprovalRuleTemplateDescription$ = [9, n0, _UARTD,
    0, () => exports.UpdateApprovalRuleTemplateDescriptionInput$, () => exports.UpdateApprovalRuleTemplateDescriptionOutput$
];
exports.UpdateApprovalRuleTemplateName$ = [9, n0, _UARTN,
    0, () => exports.UpdateApprovalRuleTemplateNameInput$, () => exports.UpdateApprovalRuleTemplateNameOutput$
];
exports.UpdateComment$ = [9, n0, _UC,
    0, () => exports.UpdateCommentInput$, () => exports.UpdateCommentOutput$
];
exports.UpdateDefaultBranch$ = [9, n0, _UDB,
    0, () => exports.UpdateDefaultBranchInput$, () => __Unit
];
exports.UpdatePullRequestApprovalRuleContent$ = [9, n0, _UPRARC,
    0, () => exports.UpdatePullRequestApprovalRuleContentInput$, () => exports.UpdatePullRequestApprovalRuleContentOutput$
];
exports.UpdatePullRequestApprovalState$ = [9, n0, _UPRAS,
    0, () => exports.UpdatePullRequestApprovalStateInput$, () => __Unit
];
exports.UpdatePullRequestDescription$ = [9, n0, _UPRD,
    0, () => exports.UpdatePullRequestDescriptionInput$, () => exports.UpdatePullRequestDescriptionOutput$
];
exports.UpdatePullRequestStatus$ = [9, n0, _UPRS,
    0, () => exports.UpdatePullRequestStatusInput$, () => exports.UpdatePullRequestStatusOutput$
];
exports.UpdatePullRequestTitle$ = [9, n0, _UPRT,
    0, () => exports.UpdatePullRequestTitleInput$, () => exports.UpdatePullRequestTitleOutput$
];
exports.UpdateRepositoryDescription$ = [9, n0, _URD,
    0, () => exports.UpdateRepositoryDescriptionInput$, () => __Unit
];
exports.UpdateRepositoryEncryptionKey$ = [9, n0, _UREK,
    0, () => exports.UpdateRepositoryEncryptionKeyInput$, () => exports.UpdateRepositoryEncryptionKeyOutput$
];
exports.UpdateRepositoryName$ = [9, n0, _URN,
    0, () => exports.UpdateRepositoryNameInput$, () => __Unit
];
