// jsonc-morph/esm/lib/rs_lib.internal.js
// @generated file from wasmbuild -- do not edit
// @ts-nocheck: generated
// deno-lint-ignore-file
// deno-fmt-ignore-file
let wasm;
export function __wbg_set_wasm(val) {
    wasm = val;
}
function addToExternrefTable0(obj) {
    const idx = wasm.__externref_table_alloc();
    wasm.__wbindgen_externrefs.set(idx, obj);
    return idx;
}
function debugString(val) {
    // primitive types
    const type = typeof val;
    if (type == "number" || type == "boolean" || val == null) {
        return `${val}`;
    }
    if (type == "string") {
        return `"${val}"`;
    }
    if (type == "symbol") {
        const description = val.description;
        if (description == null) {
            return "Symbol";
        }
        else {
            return `Symbol(${description})`;
        }
    }
    if (type == "function") {
        const name = val.name;
        if (typeof name == "string" && name.length > 0) {
            return `Function(${name})`;
        }
        else {
            return "Function";
        }
    }
    // objects
    if (Array.isArray(val)) {
        const length = val.length;
        let debug = "[";
        if (length > 0) {
            debug += debugString(val[0]);
        }
        for (let i = 1; i < length; i++) {
            debug += ", " + debugString(val[i]);
        }
        debug += "]";
        return debug;
    }
    // Test for built-in
    const builtInMatches = /\[object ([^\]]+)\]/.exec(toString.call(val));
    let className;
    if (builtInMatches && builtInMatches.length > 1) {
        className = builtInMatches[1];
    }
    else {
        // Failed to match the standard '[object ClassName]'
        return toString.call(val);
    }
    if (className == "Object") {
        // we're a user defined class or Object
        // JSON.stringify avoids problems with cycles, and is generally much
        // easier than looping through ownProperties of `val`.
        try {
            return "Object(" + JSON.stringify(val) + ")";
        }
        catch (_) {
            return "Object";
        }
    }
    // errors
    if (val instanceof Error) {
        return `${val.name}: ${val.message}\n${val.stack}`;
    }
    // TODO we could test for more things here, like `Set`s and `Map`s.
    return className;
}
function getArrayJsValueFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    const mem = getDataViewMemory0();
    const result = [];
    for (let i = ptr; i < ptr + 4 * len; i += 4) {
        result.push(wasm.__wbindgen_externrefs.get(mem.getUint32(i, true)));
    }
    wasm.__externref_drop_slice(ptr, len);
    return result;
}
function getArrayU8FromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return getUint8ArrayMemory0().subarray(ptr / 1, ptr / 1 + len);
}
let cachedDataViewMemory0 = null;
function getDataViewMemory0() {
    if (cachedDataViewMemory0 === null ||
        cachedDataViewMemory0.buffer.detached === true ||
        (cachedDataViewMemory0.buffer.detached === undefined &&
            cachedDataViewMemory0.buffer !== wasm.memory.buffer)) {
        cachedDataViewMemory0 = new DataView(wasm.memory.buffer);
    }
    return cachedDataViewMemory0;
}
function getStringFromWasm0(ptr, len) {
    ptr = ptr >>> 0;
    return decodeText(ptr, len);
}
let cachedUint8ArrayMemory0 = null;
function getUint8ArrayMemory0() {
    if (cachedUint8ArrayMemory0 === null || cachedUint8ArrayMemory0.byteLength === 0) {
        cachedUint8ArrayMemory0 = new Uint8Array(wasm.memory.buffer);
    }
    return cachedUint8ArrayMemory0;
}
function handleError(f, args) {
    try {
        return f.apply(this, args);
    }
    catch (e) {
        const idx = addToExternrefTable0(e);
        wasm.__wbindgen_exn_store(idx);
    }
}
function isLikeNone(x) {
    return x === undefined || x === null;
}
function passStringToWasm0(arg, malloc, realloc) {
    if (realloc === undefined) {
        const buf = cachedTextEncoder.encode(arg);
        const ptr = malloc(buf.length, 1) >>> 0;
        getUint8ArrayMemory0().subarray(ptr, ptr + buf.length).set(buf);
        WASM_VECTOR_LEN = buf.length;
        return ptr;
    }
    let len = arg.length;
    let ptr = malloc(len, 1) >>> 0;
    const mem = getUint8ArrayMemory0();
    let offset = 0;
    for (; offset < len; offset++) {
        const code = arg.charCodeAt(offset);
        if (code > 0x7F)
            break;
        mem[ptr + offset] = code;
    }
    if (offset !== len) {
        if (offset !== 0) {
            arg = arg.slice(offset);
        }
        ptr = realloc(ptr, len, len = offset + arg.length * 3, 1) >>> 0;
        const view = getUint8ArrayMemory0().subarray(ptr + offset, ptr + len);
        const ret = cachedTextEncoder.encodeInto(arg, view);
        offset += ret.written;
        ptr = realloc(ptr, len, offset, 1) >>> 0;
    }
    WASM_VECTOR_LEN = offset;
    return ptr;
}
function takeFromExternrefTable0(idx) {
    const value = wasm.__wbindgen_externrefs.get(idx);
    wasm.__externref_table_dealloc(idx);
    return value;
}
let cachedTextDecoder = new TextDecoder("utf-8", {
    ignoreBOM: true,
    fatal: true,
});
cachedTextDecoder.decode();
const MAX_SAFARI_DECODE_BYTES = 2146435072;
let numBytesDecoded = 0;
function decodeText(ptr, len) {
    numBytesDecoded += len;
    if (numBytesDecoded >= MAX_SAFARI_DECODE_BYTES) {
        cachedTextDecoder = new TextDecoder("utf-8", {
            ignoreBOM: true,
            fatal: true,
        });
        cachedTextDecoder.decode();
        numBytesDecoded = len;
    }
    return cachedTextDecoder.decode(getUint8ArrayMemory0().subarray(ptr, ptr + len));
}
const cachedTextEncoder = new TextEncoder();
if (!("encodeInto" in cachedTextEncoder)) {
    cachedTextEncoder.encodeInto = function (arg, view) {
        const buf = cachedTextEncoder.encode(arg);
        view.set(buf);
        return {
            read: arg.length,
            written: buf.length,
        };
    };
}
let WASM_VECTOR_LEN = 0;
const BooleanLitFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_booleanlit_free(ptr >>> 0, 1));
const JsonArrayFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_jsonarray_free(ptr >>> 0, 1));
const JsonObjectFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_jsonobject_free(ptr >>> 0, 1));
const NodeFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_node_free(ptr >>> 0, 1));
const NullKeywordFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_nullkeyword_free(ptr >>> 0, 1));
const NumberLitFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_numberlit_free(ptr >>> 0, 1));
const ObjectPropFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_objectprop_free(ptr >>> 0, 1));
const ObjectPropNameFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_objectpropname_free(ptr >>> 0, 1));
const RootNodeFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_rootnode_free(ptr >>> 0, 1));
const StringLitFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_stringlit_free(ptr >>> 0, 1));
const WordLitFinalization = (typeof FinalizationRegistry === "undefined")
    ? { register: () => { }, unregister: () => { } }
    : new FinalizationRegistry((ptr) => wasm.__wbg_wordlit_free(ptr >>> 0, 1));
/**
 * Represents a boolean literal node in the CST.
 * Provides methods for manipulating boolean values.
 */
export class BooleanLit {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(BooleanLit.prototype);
        obj.__wbg_ptr = ptr;
        BooleanLitFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        BooleanLitFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_booleanlit_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.booleanlit_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.booleanlit_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.booleanlit_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this boolean literal with a new value.
     * @param replacement - The new value to replace this boolean with
     * @returns The new node that replaced this one, or undefined if this was the root value
     * @param {any} replacement
     * @returns {Node | undefined}
     */
    replaceWith(replacement) {
        const ret = wasm.booleanlit_replaceWith(this.__wbg_ptr, replacement);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.booleanlit_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.booleanlit_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.booleanlit_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.booleanlit_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the boolean value (true or false).
     * @returns The boolean value
     * @returns {boolean}
     */
    value() {
        const ret = wasm.booleanlit_value(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.booleanlit_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this node from its parent.
     * After calling this method, the node is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.booleanlit_remove(ptr);
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.booleanlit_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.booleanlit_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
    /**
     * Sets the boolean value.
     * @param value - The new boolean value (true or false)
     * @param {boolean} value
     */
    setValue(value) {
        wasm.booleanlit_setValue(this.__wbg_ptr, value);
    }
}
if (Symbol.dispose) {
    BooleanLit.prototype[Symbol.dispose] = BooleanLit.prototype.free;
}
/**
 * Represents a JSON array node in the CST.
 * Provides methods for manipulating array elements.
 */
export class JsonArray {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(JsonArray.prototype);
        obj.__wbg_ptr = ptr;
        JsonArrayFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        JsonArrayFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_jsonarray_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.jsonarray_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.jsonarray_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.jsonarray_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this array with a new value.
     * @param value - The new value to replace this array with
     * @returns The new node that replaced this one, or undefined if this was the root value
     * @param {any} value
     * @returns {Node | undefined}
     */
    replaceWith(value) {
        const ret = wasm.jsonarray_replaceWith(this.__wbg_ptr, value);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.jsonarray_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the child node at the specified index.
     * @param index - The child index
     * @returns The child node, or undefined if index is out of bounds
     * @param {number} index
     * @returns {Node | undefined}
     */
    childAtIndex(index) {
        const ret = wasm.jsonarray_childAtIndex(this.__wbg_ptr, index);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Ensures the array is formatted with each element on its own line.
     */
    ensureMultiline() {
        wasm.jsonarray_ensureMultiline(this.__wbg_ptr);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.jsonarray_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.jsonarray_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Configures whether trailing commas should be used in this array.
     * When enabled, trailing commas are added for multiline formatting.
     * @param enabled - Whether to enable trailing commas
     * @param {boolean} enabled
     */
    setTrailingCommas(enabled) {
        wasm.jsonarray_setTrailingCommas(this.__wbg_ptr, enabled);
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.jsonarray_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns child nodes excluding whitespace, comments, and punctuation.
     * @returns Array of significant child nodes
     * @returns {Node[]}
     */
    childrenExcludeTriviaAndTokens() {
        const ret = wasm.jsonarray_childrenExcludeTriviaAndTokens(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Appends a new element to the end of the array.
     * @param value - The value to append
     * @returns The newly created element node
     * @param {any} value
     * @returns {Node}
     */
    append(value) {
        const ret = wasm.jsonarray_append(this.__wbg_ptr, value);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return Node.__wrap(ret[0]);
    }
    /**
     * Inserts a new element at the specified index.
     * @param index - The position to insert at
     * @param value - The value to insert
     * @returns The newly created element node
     * @param {number} index
     * @param {any} value
     * @returns {Node}
     */
    insert(index, value) {
        const ret = wasm.jsonarray_insert(this.__wbg_ptr, index, value);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return Node.__wrap(ret[0]);
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.jsonarray_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this array from its parent.
     * After calling this method, the array is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.jsonarray_remove(ptr);
    }
    /**
     * Returns all child nodes including whitespace and punctuation.
     * @returns Array of all child nodes
     * @returns {Node[]}
     */
    children() {
        const ret = wasm.jsonarray_children(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns all element nodes in the array.
     * @returns Array of element nodes
     * @returns {Node[]}
     */
    elements() {
        const ret = wasm.jsonarray_elements(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.jsonarray_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.jsonarray_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose) {
    JsonArray.prototype[Symbol.dispose] = JsonArray.prototype.free;
}
/**
 * Represents a JSON object node in the CST.
 * Provides methods for manipulating object properties.
 */
export class JsonObject {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(JsonObject.prototype);
        obj.__wbg_ptr = ptr;
        JsonObjectFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        JsonObjectFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_jsonobject_free(ptr, 0);
    }
    /**
     * Returns all properties in the object.
     * @returns Array of object properties
     * @returns {ObjectProp[]}
     */
    properties() {
        const ret = wasm.jsonobject_properties(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.jsonobject_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.jsonobject_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Gets a property value if it's an array.
     * @param name - The property name to look up
     * @returns The array value, or undefined if property doesn't exist or is not an array
     * @param {string} name
     * @returns {JsonArray | undefined}
     */
    getIfArray(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfArray(this.__wbg_ptr, ptr0, len0);
        return ret === 0 ? undefined : JsonArray.__wrap(ret);
    }
    /**
     * Gets a property by name, throwing if not found.
     * @param key - The property name to look up
     * @returns The property
     * @throws If the property is not found
     * @param {string} key
     * @returns {ObjectProp}
     */
    getOrThrow(key) {
        const ptr0 = passStringToWasm0(key, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getOrThrow(this.__wbg_ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ObjectProp.__wrap(ret[0]);
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.jsonobject_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this object with a new value.
     * @param replacement - The new value to replace this object with
     * @returns The new node that replaced this one, or undefined if this was the root value
     * @param {any} replacement
     * @returns {Node | undefined}
     */
    replaceWith(replacement) {
        const ret = wasm.jsonobject_replaceWith(this.__wbg_ptr, replacement);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Gets a property value if it's an object.
     * @param name - The property name to look up
     * @returns The object value, or undefined if property doesn't exist or is not an object
     * @param {string} name
     * @returns {JsonObject | undefined}
     */
    getIfObject(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfObject(this.__wbg_ptr, ptr0, len0);
        return ret === 0 ? undefined : JsonObject.__wrap(ret);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.jsonobject_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the child node at the specified index.
     * @param index - The child index
     * @returns The child node, or undefined if index is out of bounds
     * @param {number} index
     * @returns {Node | undefined}
     */
    childAtIndex(index) {
        const ret = wasm.jsonobject_childAtIndex(this.__wbg_ptr, index);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Ensures the object is formatted with each property on its own line.
     */
    ensureMultiline() {
        wasm.jsonobject_ensureMultiline(this.__wbg_ptr);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.jsonobject_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.jsonobject_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Configures whether trailing commas should be used in this object.
     * When enabled, trailing commas are added for multiline formatting.
     * @param enabled - Whether to enable trailing commas
     * @param {boolean} enabled
     */
    setTrailingCommas(enabled) {
        wasm.jsonobject_setTrailingCommas(this.__wbg_ptr, enabled);
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.jsonobject_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Gets a property value as an array, creating or replacing the value with an empty array if needed.
     * Unlike getIfArrayOrCreate, this always returns an array by replacing non-array values.
     * @param name - The property name to get
     * @returns The array value (always succeeds)
     * @param {string} name
     * @returns {JsonArray}
     */
    getIfArrayOrForce(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfArrayOrForce(this.__wbg_ptr, ptr0, len0);
        return JsonArray.__wrap(ret);
    }
    /**
     * Gets a property value as an array, throwing if not found or wrong type.
     * @param name - The property name to look up
     * @returns The array value
     * @throws If the property doesn't exist or is not an array
     * @param {string} name
     * @returns {JsonArray}
     */
    getIfArrayOrThrow(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfArrayOrThrow(this.__wbg_ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonArray.__wrap(ret[0]);
    }
    /**
     * Gets a property value as an array, creating an empty array if the property doesn't exist.
     * Returns undefined if the property exists but has a non-array value.
     * @param name - The property name to get
     * @returns The array value, or undefined if property has a non-array value
     * @param {string} name
     * @returns {JsonArray | undefined}
     */
    getIfArrayOrCreate(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfArrayOrCreate(this.__wbg_ptr, ptr0, len0);
        return ret === 0 ? undefined : JsonArray.__wrap(ret);
    }
    /**
     * Gets a property value as an object, creating or replacing the value with an empty object if needed.
     * Unlike getIfObjectOrCreate, this always returns an object by replacing non-object values.
     * @param name - The property name to get
     * @returns The object value (always succeeds)
     * @param {string} name
     * @returns {JsonObject}
     */
    getIfObjectOrForce(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfObjectOrForce(this.__wbg_ptr, ptr0, len0);
        return JsonObject.__wrap(ret);
    }
    /**
     * Gets a property value as an object, throwing if not found or wrong type.
     * @param name - The property name to look up
     * @returns The object value
     * @throws If the property doesn't exist or is not an object
     * @param {string} name
     * @returns {JsonObject}
     */
    getIfObjectOrThrow(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfObjectOrThrow(this.__wbg_ptr, ptr0, len0);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonObject.__wrap(ret[0]);
    }
    /**
     * Gets a property value as an object, creating an empty object if the property doesn't exist.
     * Returns undefined if the property exists but has a non-object value.
     * @param name - The property name to get
     * @returns The object value, or undefined if property has a non-object value
     * @param {string} name
     * @returns {JsonObject | undefined}
     */
    getIfObjectOrCreate(name) {
        const ptr0 = passStringToWasm0(name, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_getIfObjectOrCreate(this.__wbg_ptr, ptr0, len0);
        return ret === 0 ? undefined : JsonObject.__wrap(ret);
    }
    /**
     * Returns child nodes excluding whitespace, comments, and punctuation.
     * @returns Array of significant child nodes
     * @returns {Node[]}
     */
    childrenExcludeTriviaAndTokens() {
        const ret = wasm.jsonobject_childrenExcludeTriviaAndTokens(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Gets a property by name.
     * @param key - The property name to look up
     * @returns The property, or undefined if not found
     * @param {string} key
     * @returns {ObjectProp | undefined}
     */
    get(key) {
        const ptr0 = passStringToWasm0(key, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_get(this.__wbg_ptr, ptr0, len0);
        return ret === 0 ? undefined : ObjectProp.__wrap(ret);
    }
    /**
     * Appends a new property to the object.
     * @param key - The name of the property to add
     * @param value - The value to set for the property
     * @returns The newly created property
     * @param {string} key
     * @param {any} value
     * @returns {ObjectProp}
     */
    append(key, value) {
        const ptr0 = passStringToWasm0(key, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_append(this.__wbg_ptr, ptr0, len0, value);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ObjectProp.__wrap(ret[0]);
    }
    /**
     * Inserts a new property at the specified index.
     * @param index - The position to insert the property at
     * @param key - The name of the property to add
     * @param value - The value to set for the property
     * @returns The newly created property
     * @param {number} index
     * @param {string} key
     * @param {any} value
     * @returns {ObjectProp}
     */
    insert(index, key, value) {
        const ptr0 = passStringToWasm0(key, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.jsonobject_insert(this.__wbg_ptr, index, ptr0, len0, value);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ObjectProp.__wrap(ret[0]);
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.jsonobject_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this object from its parent.
     * After calling this method, the object is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.jsonobject_remove(ptr);
    }
    /**
     * Returns all child nodes including whitespace and punctuation.
     * @returns Array of all child nodes
     * @returns {Node[]}
     */
    children() {
        const ret = wasm.jsonobject_children(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.jsonobject_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.jsonobject_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose) {
    JsonObject.prototype[Symbol.dispose] = JsonObject.prototype.free;
}
/**
 * Represents a generic node in the CST.
 * Can be a container node (object, array, property) or a leaf node (string, number, boolean, null).
 */
export class Node {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(Node.prototype);
        obj.__wbg_ptr = ptr;
        NodeFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NodeFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_node_free(ptr, 0);
    }
    /**
     * Returns the boolean value if this node is a boolean literal.
     * @returns The boolean value, or undefined if this node is not a boolean
     * @returns {boolean | undefined}
     */
    asBoolean() {
        const ret = wasm.node_asBoolean(this.__wbg_ptr);
        return ret === 0xFFFFFF ? undefined : ret !== 0;
    }
    /**
     * Returns true if this node is a boolean literal.
     * @returns true if this node is a boolean
     * @returns {boolean}
     */
    isBoolean() {
        const ret = wasm.node_isBoolean(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns true if this node is a comment.
     * @returns true if this node is a comment
     * @returns {boolean}
     */
    isComment() {
        const ret = wasm.node_isComment(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns true if this node is a newline character.
     * @returns true if this node is a newline
     * @returns {boolean}
     */
    isNewline() {
        const ret = wasm.node_isNewline(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the character if this node is a single-character token.
     * @returns The token character, or undefined if not a token
     * @returns {string | undefined}
     */
    tokenChar() {
        const ret = wasm.node_tokenChar(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns this node as a WordLit if it is one.
     * @returns The WordLit, or undefined if this node is not a word literal
     * @returns {WordLit | undefined}
     */
    asWordLit() {
        const ret = wasm.node_asWordLit(this.__wbg_ptr);
        return ret === 0 ? undefined : WordLit.__wrap(ret);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.node_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.node_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Converts this node to the root node if it is one.
     * @returns The root node, or undefined if this is not a root node
     * @returns {RootNode | undefined}
     */
    asRootNode() {
        const ret = wasm.node_asRootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
    /**
     * Returns true if this node is a container (object, array, or property).
     * @returns true if this is a container node
     * @returns {boolean}
     */
    isContainer() {
        const ret = wasm.node_isContainer(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.node_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns the raw string representation of a number literal.
     * Returns a string to preserve the exact formatting (e.g., "1.0" vs "1", "1e10" vs "10000000000").
     * @returns The number as a string, or undefined if this node is not a number
     * @returns {string | undefined}
     */
    numberValue() {
        const ret = wasm.node_numberValue(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns this node as a NumberLit if it is one.
     * @returns The NumberLit, or undefined if this node is not a number literal
     * @returns {NumberLit | undefined}
     */
    asNumberLit() {
        const ret = wasm.node_asNumberLit(this.__wbg_ptr);
        return ret === 0 ? undefined : NumberLit.__wrap(ret);
    }
    /**
     * Returns this node as a StringLit if it is one.
     * @returns The StringLit, or undefined if this node is not a string literal
     * @returns {StringLit | undefined}
     */
    asStringLit() {
        const ret = wasm.node_asStringLit(this.__wbg_ptr);
        return ret === 0 ? undefined : StringLit.__wrap(ret);
    }
    /**
     * Returns the element index if this node is an array element.
     * @returns The element index, or undefined if not an array element
     * @returns {number | undefined}
     */
    elementIndex() {
        const ret = wasm.node_elementIndex(this.__wbg_ptr);
        return ret === 0x100000001 ? undefined : ret;
    }
    /**
     * Returns true if this node is whitespace.
     * @returns true if this node is whitespace
     * @returns {boolean}
     */
    isWhitespace() {
        const ret = wasm.node_isWhitespace(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.node_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns this node as a BooleanLit if it is one.
     * @returns The BooleanLit, or undefined if this node is not a boolean literal
     * @returns {BooleanLit | undefined}
     */
    asBooleanLit() {
        const ret = wasm.node_asBooleanLit(this.__wbg_ptr);
        return ret === 0 ? undefined : BooleanLit.__wrap(ret);
    }
    /**
     * Returns the child node at the specified index.
     * @param index - The child index
     * @returns The child node, or undefined if index is out of bounds
     * @param {number} index
     * @returns {Node | undefined}
     */
    childAtIndex(index) {
        const ret = wasm.node_childAtIndex(this.__wbg_ptr, index);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns this node as a NullKeyword if it is one.
     * @returns The NullKeyword, or undefined if this node is not a null keyword
     * @returns {NullKeyword | undefined}
     */
    asNullKeyword() {
        const ret = wasm.node_asNullKeyword(this.__wbg_ptr);
        return ret === 0 ? undefined : NullKeyword.__wrap(ret);
    }
    /**
     * Returns the parent node, throwing if this is the root.
     * @returns The parent node
     * @throws If this node has no parent
     * @returns {Node}
     */
    parentOrThrow() {
        const ret = wasm.node_parentOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return Node.__wrap(ret[0]);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.node_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Converts this node to an array, throwing if it's not an array.
     * @returns The array
     * @throws If this node is not an array
     * @returns {JsonArray}
     */
    asArrayOrThrow() {
        const ret = wasm.node_asArrayOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonArray.__wrap(ret[0]);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.node_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Converts this node to an object, throwing if it's not an object.
     * @returns The object
     * @throws If this node is not an object
     * @returns {JsonObject}
     */
    asObjectOrThrow() {
        const ret = wasm.node_asObjectOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonObject.__wrap(ret[0]);
    }
    /**
     * Returns the decoded string value, throwing if not a string.
     * @returns The string value
     * @throws If this node is not a string
     * @returns {string}
     */
    asStringOrThrow() {
        let deferred2_0;
        let deferred2_1;
        try {
            const ret = wasm.node_asStringOrThrow(this.__wbg_ptr);
            var ptr1 = ret[0];
            var len1 = ret[1];
            if (ret[3]) {
                ptr1 = 0;
                len1 = 0;
                throw takeFromExternrefTable0(ret[2]);
            }
            deferred2_0 = ptr1;
            deferred2_1 = len1;
            return getStringFromWasm0(ptr1, len1);
        }
        finally {
            wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
        }
    }
    /**
     * Returns the root node, throwing if detached.
     * @returns The root node
     * @throws If this node is detached from the CST
     * @returns {RootNode}
     */
    rootNodeOrThrow() {
        const ret = wasm.node_rootNodeOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return RootNode.__wrap(ret[0]);
    }
    /**
     * Returns the boolean value, throwing if not a boolean.
     * @returns The boolean value
     * @throws If this node is not a boolean
     * @returns {boolean}
     */
    asBooleanOrThrow() {
        const ret = wasm.node_asBooleanOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] !== 0;
    }
    /**
     * Returns this node as a WordLit, throwing if it's not a word literal.
     * @returns The WordLit
     * @throws If this node is not a word literal
     * @returns {WordLit}
     */
    asWordLitOrThrow() {
        const ret = wasm.node_asWordLitOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return WordLit.__wrap(ret[0]);
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.node_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Converts this node to the root node, throwing if it's not a root node.
     * @returns The root node
     * @throws If this node is not a root node
     * @returns {RootNode}
     */
    asRootNodeOrThrow() {
        const ret = wasm.node_asRootNodeOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return RootNode.__wrap(ret[0]);
    }
    /**
     * Returns the raw string representation of a number literal, throwing if not a number.
     * Returns a string to preserve the exact formatting (e.g., "1.0" vs "1", "1e10" vs "10000000000").
     * @returns The number as a string
     * @throws If this node is not a number
     * @returns {string}
     */
    numberValueOrThrow() {
        let deferred2_0;
        let deferred2_1;
        try {
            const ret = wasm.node_numberValueOrThrow(this.__wbg_ptr);
            var ptr1 = ret[0];
            var len1 = ret[1];
            if (ret[3]) {
                ptr1 = 0;
                len1 = 0;
                throw takeFromExternrefTable0(ret[2]);
            }
            deferred2_0 = ptr1;
            deferred2_1 = len1;
            return getStringFromWasm0(ptr1, len1);
        }
        finally {
            wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
        }
    }
    /**
     * Returns this node as a NumberLit, throwing if it's not a number literal.
     * @returns The NumberLit
     * @throws If this node is not a number literal
     * @returns {NumberLit}
     */
    asNumberLitOrThrow() {
        const ret = wasm.node_asNumberLitOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return NumberLit.__wrap(ret[0]);
    }
    /**
     * Returns this node as a StringLit, throwing if it's not a string literal.
     * @returns The StringLit
     * @throws If this node is not a string literal
     * @returns {StringLit}
     */
    asStringLitOrThrow() {
        const ret = wasm.node_asStringLitOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return StringLit.__wrap(ret[0]);
    }
    /**
     * Returns this node as a BooleanLit, throwing if it's not a boolean literal.
     * @returns The BooleanLit
     * @throws If this node is not a boolean literal
     * @returns {BooleanLit}
     */
    asBooleanLitOrThrow() {
        const ret = wasm.node_asBooleanLitOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return BooleanLit.__wrap(ret[0]);
    }
    /**
     * Returns this node as a NullKeyword, throwing if it's not a null keyword.
     * @returns The NullKeyword
     * @throws If this node is not a null keyword
     * @returns {NullKeyword}
     */
    asNullKeywordOrThrow() {
        const ret = wasm.node_asNullKeywordOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return NullKeyword.__wrap(ret[0]);
    }
    /**
     * Returns child nodes excluding whitespace, comments, and punctuation.
     * @returns Array of significant child nodes
     * @returns {Node[]}
     */
    childrenExcludeTriviaAndTokens() {
        const ret = wasm.node_childrenExcludeTriviaAndTokens(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.node_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this node from its parent.
     * Works for any node kind, including whitespace, comments, tokens, containers, and leaf values.
     * After calling this method, the node is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.node_remove(ptr);
    }
    /**
     * Returns true if this node is a leaf value (string, number, boolean, null).
     * @returns true if this is a leaf node
     * @returns {boolean}
     */
    isLeaf() {
        const ret = wasm.node_isLeaf(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns true if this node is a null keyword.
     * @returns true if this node represents null
     * @returns {boolean}
     */
    isNull() {
        const ret = wasm.node_isNull(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Converts this node to an array if it is one.
     * @returns The array, or undefined if this node is not an array
     * @returns {JsonArray | undefined}
     */
    asArray() {
        const ret = wasm.node_asArray(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonArray.__wrap(ret);
    }
    /**
     * Returns all child nodes including whitespace and punctuation.
     * @returns Array of all child nodes
     * @returns {Node[]}
     */
    children() {
        const ret = wasm.node_children(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns true if this node is a comma token.
     * @returns true if this node is a comma
     * @returns {boolean}
     */
    isComma() {
        const ret = wasm.node_isComma(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns true if this node is a punctuation token (bracket, brace, colon, comma).
     * @returns true if this node is a token
     * @returns {boolean}
     */
    isToken() {
        const ret = wasm.node_isToken(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Converts this CST node to a plain JavaScript value.
     * This recursively converts objects, arrays, and primitives to their JavaScript equivalents.
     * Comments and formatting information are discarded.
     * @returns The plain JavaScript value (object, array, string, number, boolean, or null)
     * @throws If the node contains invalid values that cannot be converted
     * @returns {any}
     */
    toValue() {
        const ret = wasm.node_toValue(this.__wbg_ptr);
        return ret;
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.node_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Converts this node to an object if it is one.
     * @returns The object, or undefined if this node is not an object
     * @returns {JsonObject | undefined}
     */
    asObject() {
        const ret = wasm.node_asObject(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonObject.__wrap(ret);
    }
    /**
     * Returns the decoded string value if this node is a string literal.
     * @returns The string value, or undefined if this node is not a string
     * @returns {string | undefined}
     */
    asString() {
        const ret = wasm.node_asString(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns true if this node is a number literal.
     * @returns true if this node is a number
     * @returns {boolean}
     */
    isNumber() {
        const ret = wasm.node_isNumber(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns true if this node is a string literal.
     * @returns true if this node is a string
     * @returns {boolean}
     */
    isString() {
        const ret = wasm.node_isString(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns true if this node is trivia (whitespace or comments).
     * @returns true if this node is trivia
     * @returns {boolean}
     */
    isTrivia() {
        const ret = wasm.node_isTrivia(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.node_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose)
    Node.prototype[Symbol.dispose] = Node.prototype.free;
/**
 * Represents a null keyword node in the CST.
 */
export class NullKeyword {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NullKeyword.prototype);
        obj.__wbg_ptr = ptr;
        NullKeywordFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NullKeywordFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_nullkeyword_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.nullkeyword_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.nullkeyword_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.nullkeyword_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this null keyword with a new value.
     * @param replacement - The new value to replace this null with
     * @returns The new node that replaced this one, or undefined if this was the root value
     * @param {any} replacement
     * @returns {Node | undefined}
     */
    replaceWith(replacement) {
        const ret = wasm.nullkeyword_replaceWith(this.__wbg_ptr, replacement);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.nullkeyword_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.nullkeyword_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.nullkeyword_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.nullkeyword_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.nullkeyword_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this node from its parent.
     * After calling this method, the node is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.nullkeyword_remove(ptr);
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.nullkeyword_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.nullkeyword_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose) {
    NullKeyword.prototype[Symbol.dispose] = NullKeyword.prototype.free;
}
/**
 * Represents a number literal node in the CST.
 * Provides methods for manipulating number values.
 */
export class NumberLit {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(NumberLit.prototype);
        obj.__wbg_ptr = ptr;
        NumberLitFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        NumberLitFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_numberlit_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.numberlit_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.numberlit_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.numberlit_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this number literal with a new value.
     * @param replacement - The new value to replace this number with
     * @returns The new node that replaced this one, or undefined if this was the root value
     * @param {any} replacement
     * @returns {Node | undefined}
     */
    replaceWith(replacement) {
        const ret = wasm.numberlit_replaceWith(this.__wbg_ptr, replacement);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.numberlit_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Sets the raw number value.
     * The value should be a valid JSON number string (e.g., "42", "3.14", "1e10").
     * @param value - The raw number string to set
     * @param {string} value
     */
    setRawValue(value) {
        const ptr0 = passStringToWasm0(value, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.numberlit_setRawValue(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.numberlit_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.numberlit_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.numberlit_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the raw string representation of the number.
     * Returns a string to preserve the exact formatting (e.g., "1.0" vs "1", "1e10" vs "10000000000").
     * @returns The number as a string
     * @returns {string}
     */
    value() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.numberlit_value(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        }
        finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.numberlit_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this node from its parent.
     * After calling this method, the node is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.numberlit_remove(ptr);
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.numberlit_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.numberlit_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose) {
    NumberLit.prototype[Symbol.dispose] = NumberLit.prototype.free;
}
/**
 * Represents an object property (key-value pair) in the CST.
 * Provides methods for accessing and manipulating both the property name and its value.
 */
export class ObjectProp {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(ObjectProp.prototype);
        obj.__wbg_ptr = ptr;
        ObjectPropFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        ObjectPropFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_objectprop_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.objectprop_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.objectprop_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.objectprop_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this property with a new property.
     * This allows changing both the property name and its value.
     * @param key - The new property name
     * @param replacement - The new value for the property
     * @returns The new node that replaced this property, or undefined if this was the root value
     * @param {string} key
     * @param {any} replacement
     * @returns {Node | undefined}
     */
    replaceWith(key, replacement) {
        const ptr0 = passStringToWasm0(key, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        const ret = wasm.objectprop_replaceWith(this.__wbg_ptr, ptr0, len0, replacement);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Returns the property name, throwing if malformed.
     * @returns The property name
     * @throws If the property name is malformed
     * @returns {ObjectPropName}
     */
    nameOrThrow() {
        const ret = wasm.objectprop_nameOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ObjectPropName.__wrap(ret[0]);
    }
    /**
     * Returns the next property in the same object.
     * @returns The next property, or undefined if this is the last property
     * @returns {ObjectProp | undefined}
     */
    nextProperty() {
        const ret = wasm.objectprop_nextProperty(this.__wbg_ptr);
        return ret === 0 ? undefined : ObjectProp.__wrap(ret);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.objectprop_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the child node at the specified index.
     * @param index - The child index
     * @returns The child node, or undefined if index is out of bounds
     * @param {number} index
     * @returns {Node | undefined}
     */
    childAtIndex(index) {
        const ret = wasm.objectprop_childAtIndex(this.__wbg_ptr, index);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns the index of this property within its parent object.
     * @returns The property index
     * @returns {number}
     */
    propertyIndex() {
        const ret = wasm.objectprop_propertyIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the property value if it's an array.
     * @returns The array value, or undefined if not an array
     * @returns {JsonArray | undefined}
     */
    valueIfArray() {
        const ret = wasm.objectprop_valueIfArray(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonArray.__wrap(ret);
    }
    /**
     * Returns the property value, throwing if malformed.
     * @returns The property value
     * @throws If the property value is malformed
     * @returns {Node}
     */
    valueOrThrow() {
        const ret = wasm.objectprop_valueOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return Node.__wrap(ret[0]);
    }
    /**
     * Returns the property value if it's an object.
     * @returns The object value, or undefined if not an object
     * @returns {JsonObject | undefined}
     */
    valueIfObject() {
        const ret = wasm.objectprop_valueIfObject(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonObject.__wrap(ret);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.objectprop_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns the previous property in the same object.
     * @returns The previous property, or undefined if this is the first property
     * @returns {ObjectProp | undefined}
     */
    previousProperty() {
        const ret = wasm.objectprop_previousProperty(this.__wbg_ptr);
        return ret === 0 ? undefined : ObjectProp.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.objectprop_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.objectprop_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Gets the property value as an array, replacing the value with an empty array if needed.
     * Always returns an array by replacing non-array values.
     * @returns The array value (always succeeds)
     * @returns {JsonArray}
     */
    valueIfArrayOrForce() {
        const ret = wasm.objectprop_valueIfArrayOrForce(this.__wbg_ptr);
        return JsonArray.__wrap(ret);
    }
    /**
     * Returns the property value as an array, throwing if not an array.
     * @returns The array value
     * @throws If the property value is not an array
     * @returns {JsonArray}
     */
    valueIfArrayOrThrow() {
        const ret = wasm.objectprop_valueIfArrayOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonArray.__wrap(ret[0]);
    }
    /**
     * Gets the property value as an object, replacing the value with an empty object if needed.
     * Always returns an object by replacing non-object values.
     * @returns The object value (always succeeds)
     * @returns {JsonObject}
     */
    valueIfObjectOrForce() {
        const ret = wasm.objectprop_valueIfObjectOrForce(this.__wbg_ptr);
        return JsonObject.__wrap(ret);
    }
    /**
     * Returns the property value as an object, throwing if not an object.
     * @returns The object value
     * @throws If the property value is not an object
     * @returns {JsonObject}
     */
    valueIfObjectOrThrow() {
        const ret = wasm.objectprop_valueIfObjectOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonObject.__wrap(ret[0]);
    }
    /**
     * Returns child nodes excluding whitespace, comments, and punctuation.
     * @returns Array of significant child nodes
     * @returns {Node[]}
     */
    childrenExcludeTriviaAndTokens() {
        const ret = wasm.objectprop_childrenExcludeTriviaAndTokens(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the property name.
     * @returns The property name, or undefined if malformed
     * @returns {ObjectPropName | undefined}
     */
    name() {
        const ret = wasm.objectprop_name(this.__wbg_ptr);
        return ret === 0 ? undefined : ObjectPropName.__wrap(ret);
    }
    /**
     * Returns the property value.
     * @returns The property value, or undefined if malformed
     * @returns {Node | undefined}
     */
    value() {
        const ret = wasm.objectprop_value(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.objectprop_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this property from its parent object.
     * After calling this method, the property is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.objectprop_remove(ptr);
    }
    /**
     * Returns all child nodes including whitespace and punctuation.
     * @returns Array of all child nodes
     * @returns {Node[]}
     */
    children() {
        const ret = wasm.objectprop_children(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.objectprop_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.objectprop_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
    /**
     * Sets the value of this property.
     * @param value - The new value to set
     * @param {any} value
     */
    setValue(value) {
        const ret = wasm.objectprop_setValue(this.__wbg_ptr, value);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
}
if (Symbol.dispose) {
    ObjectProp.prototype[Symbol.dispose] = ObjectProp.prototype.free;
}
/**
 * Represents the name part of an object property in the CST.
 * Can be either a quoted string or an unquoted word literal (when allowLooseObjectPropertyNames is enabled).
 */
export class ObjectPropName {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(ObjectPropName.prototype);
        obj.__wbg_ptr = ptr;
        ObjectPropNameFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        ObjectPropNameFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_objectpropname_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.objectpropname_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.objectpropname_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.objectpropname_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns the decoded property name (unquoted and unescaped).
     * @returns The decoded property name
     * @returns {string}
     */
    decodedValue() {
        let deferred2_0;
        let deferred2_1;
        try {
            const ret = wasm.objectpropname_decodedValue(this.__wbg_ptr);
            var ptr1 = ret[0];
            var len1 = ret[1];
            if (ret[3]) {
                ptr1 = 0;
                len1 = 0;
                throw takeFromExternrefTable0(ret[2]);
            }
            deferred2_0 = ptr1;
            deferred2_1 = len1;
            return getStringFromWasm0(ptr1, len1);
        }
        finally {
            wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
        }
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.objectpropname_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.objectpropname_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.objectpropname_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.objectpropname_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.objectpropname_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose) {
    ObjectPropName.prototype[Symbol.dispose] = ObjectPropName.prototype.free;
}
/**
 * Represents the root node of a JSONC document.
 * This is the entry point for manipulating the concrete syntax tree.
 */
export class RootNode {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(RootNode.prototype);
        obj.__wbg_ptr = ptr;
        RootNodeFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        RootNodeFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_rootnode_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.rootnode_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.rootnode_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the newline kind used in the document.
     * @returns Either "\n" or "\r\n"
     * @returns {string}
     */
    newlineKind() {
        const ret = wasm.rootnode_newlineKind(this.__wbg_ptr);
        return ret;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.rootnode_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.rootnode_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the child node at the specified index.
     * @param index - The child index
     * @returns The child node, or undefined if index is out of bounds
     * @param {number} index
     * @returns {Node | undefined}
     */
    childAtIndex(index) {
        const ret = wasm.rootnode_childAtIndex(this.__wbg_ptr, index);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Clears all children from the root node, leaving an empty document.
     */
    clearChildren() {
        wasm.rootnode_clearChildren(this.__wbg_ptr);
    }
    /**
     * Returns the root value node, throwing if empty.
     * @returns The root value
     * @throws If the document is empty
     * @returns {Node}
     */
    valueOrThrow() {
        const ret = wasm.rootnode_valueOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return Node.__wrap(ret[0]);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.rootnode_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Converts the CST back to a string representation.
     * @returns The JSONC string
     * @returns {string}
     */
    toString() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.rootnode_toString(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        }
        finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns the root value as an array, replacing any existing value with an empty array if needed.
     * Unlike asArrayOrCreate, this always returns an array by replacing non-array values.
     * @returns The array (always succeeds)
     * @returns {JsonArray}
     */
    asArrayOrForce() {
        const ret = wasm.rootnode_asArrayOrForce(this.__wbg_ptr);
        return JsonArray.__wrap(ret);
    }
    /**
     * Returns the root value as an array, throwing if it's not an array.
     * @returns The array
     * @throws If the root is not an array
     * @returns {JsonArray}
     */
    asArrayOrThrow() {
        const ret = wasm.rootnode_asArrayOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonArray.__wrap(ret[0]);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.rootnode_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root value as an array, creating an empty array if the root is empty.
     * Returns undefined if the root contains a value of a different type.
     * @returns The array, or undefined if a non-array value exists
     * @returns {JsonArray | undefined}
     */
    asArrayOrCreate() {
        const ret = wasm.rootnode_asArrayOrCreate(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonArray.__wrap(ret);
    }
    /**
     * Returns the root value as an object, replacing any existing value with an empty object if needed.
     * Unlike asObjectOrCreate, this always returns an object by replacing non-object values.
     * @returns The object (always succeeds)
     * @returns {JsonObject}
     */
    asObjectOrForce() {
        const ret = wasm.rootnode_asObjectOrForce(this.__wbg_ptr);
        return JsonObject.__wrap(ret);
    }
    /**
     * Returns the root value as an object, throwing if it's not an object.
     * @returns The object
     * @throws If the root is not an object
     * @returns {JsonObject}
     */
    asObjectOrThrow() {
        const ret = wasm.rootnode_asObjectOrThrow(this.__wbg_ptr);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return JsonObject.__wrap(ret[0]);
    }
    /**
     * Returns the indentation string used for a single level.
     * @returns The single-level indentation string (e.g., "  " or "\t")
     * @returns {string | undefined}
     */
    singleIndentText() {
        const ret = wasm.rootnode_singleIndentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the root value as an object, creating an empty object if the root is empty.
     * Returns undefined if the root contains a value of a different type.
     * @returns The object, or undefined if a non-object value exists
     * @returns {JsonObject | undefined}
     */
    asObjectOrCreate() {
        const ret = wasm.rootnode_asObjectOrCreate(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonObject.__wrap(ret);
    }
    /**
     * Configures whether trailing commas should be used throughout the document.
     * When enabled, trailing commas are added for multiline formatting in objects and arrays.
     * @param enabled - Whether to enable trailing commas
     * @param {boolean} enabled
     */
    setTrailingCommas(enabled) {
        wasm.rootnode_setTrailingCommas(this.__wbg_ptr, enabled);
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.rootnode_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns child nodes excluding whitespace, comments, and punctuation.
     * @returns Array of significant child nodes
     * @returns {Node[]}
     */
    childrenExcludeTriviaAndTokens() {
        const ret = wasm.rootnode_childrenExcludeTriviaAndTokens(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root value node.
     * @returns The root value, or undefined if the document is empty
     * @returns {Node | undefined}
     */
    value() {
        const ret = wasm.rootnode_value(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.rootnode_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns the root value as an array if it is one.
     * @returns The array, or undefined if root is not an array
     * @returns {JsonArray | undefined}
     */
    asArray() {
        const ret = wasm.rootnode_asArray(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonArray.__wrap(ret);
    }
    /**
     * Returns all child nodes including whitespace and punctuation.
     * @returns Array of all child nodes
     * @returns {Node[]}
     */
    children() {
        const ret = wasm.rootnode_children(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Converts the CST to a plain JavaScript value, similar to JSON.parse.
     * This recursively converts the root value to its JavaScript equivalent.
     * Comments and formatting information are discarded.
     * @returns The plain JavaScript value (object, array, string, number, boolean, or null)
     * @throws If the document contains invalid values that cannot be converted
     * @returns {any}
     */
    toValue() {
        const ret = wasm.rootnode_toValue(this.__wbg_ptr);
        return ret;
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.rootnode_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root value as an object if it is one.
     * @returns The object, or undefined if root is not an object
     * @returns {JsonObject | undefined}
     */
    asObject() {
        const ret = wasm.rootnode_asObject(this.__wbg_ptr);
        return ret === 0 ? undefined : JsonObject.__wrap(ret);
    }
    /**
     * Sets the root value of the document.
     * Accepts any JSON value: string, number, boolean, null, array, or object.
     * @param value - The new value to set
     * @param {any} value
     */
    setValue(value) {
        const ret = wasm.rootnode_setValue(this.__wbg_ptr, value);
        if (ret[1]) {
            throw takeFromExternrefTable0(ret[0]);
        }
    }
}
if (Symbol.dispose) {
    RootNode.prototype[Symbol.dispose] = RootNode.prototype.free;
}
/**
 * Represents a string literal node in the CST.
 * Provides methods for manipulating string values and their formatting.
 */
export class StringLit {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(StringLit.prototype);
        obj.__wbg_ptr = ptr;
        StringLitFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        StringLitFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_stringlit_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.stringlit_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.stringlit_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.stringlit_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this string literal with a new value.
     * @param replacement - The new value to replace this string with
     * @returns The new node that replaced this one, or undefined if this was the root value
     * @param {any} replacement
     * @returns {Node | undefined}
     */
    replaceWith(replacement) {
        const ret = wasm.stringlit_replaceWith(this.__wbg_ptr, replacement);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Returns the decoded string value (without quotes and with escape sequences processed).
     * @returns The decoded string value
     * @returns {string}
     */
    decodedValue() {
        let deferred2_0;
        let deferred2_1;
        try {
            const ret = wasm.stringlit_decodedValue(this.__wbg_ptr);
            var ptr1 = ret[0];
            var len1 = ret[1];
            if (ret[3]) {
                ptr1 = 0;
                len1 = 0;
                throw takeFromExternrefTable0(ret[2]);
            }
            deferred2_0 = ptr1;
            deferred2_1 = len1;
            return getStringFromWasm0(ptr1, len1);
        }
        finally {
            wasm.__wbindgen_free(deferred2_0, deferred2_1, 1);
        }
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.stringlit_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Sets the raw string value (should include quotes).
     * @param value - The new raw string value
     * @param {string} value
     */
    setRawValue(value) {
        const ptr0 = passStringToWasm0(value, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.stringlit_setRawValue(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.stringlit_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.stringlit_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.stringlit_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.stringlit_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this string literal from its parent.
     * After calling this method, the node is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.stringlit_remove(ptr);
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.stringlit_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the raw string value including quotes and escape sequences.
     * @returns The raw string representation
     * @returns {string}
     */
    rawValue() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.stringlit_rawValue(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        }
        finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.stringlit_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose) {
    StringLit.prototype[Symbol.dispose] = StringLit.prototype.free;
}
/**
 * Represents an unquoted word literal node in the CST.
 * Used for unquoted property names when `allowLooseObjectPropertyNames` is enabled.
 */
export class WordLit {
    static __wrap(ptr) {
        ptr = ptr >>> 0;
        const obj = Object.create(WordLit.prototype);
        obj.__wbg_ptr = ptr;
        WordLitFinalization.register(obj, obj.__wbg_ptr, obj);
        return obj;
    }
    __destroy_into_raw() {
        const ptr = this.__wbg_ptr;
        this.__wbg_ptr = 0;
        WordLitFinalization.unregister(this);
        return ptr;
    }
    free() {
        const ptr = this.__destroy_into_raw();
        wasm.__wbg_wordlit_free(ptr, 0);
    }
    /**
     * Returns the index of this node within its parent's children.
     * @returns The child index
     * @returns {number}
     */
    childIndex() {
        const ret = wasm.wordlit_childIndex(this.__wbg_ptr);
        return ret >>> 0;
    }
    /**
     * Returns the indentation string used at this node's depth.
     * @returns The indentation string, or undefined if not applicable
     * @returns {string | undefined}
     */
    indentText() {
        const ret = wasm.wordlit_indentText(this.__wbg_ptr);
        let v1;
        if (ret[0] !== 0) {
            v1 = getStringFromWasm0(ret[0], ret[1]).slice();
            wasm.__wbindgen_free(ret[0], ret[1] * 1, 1);
        }
        return v1;
    }
    /**
     * Returns the next sibling node.
     * @returns The next sibling, or undefined if this is the last child
     * @returns {Node | undefined}
     */
    nextSibling() {
        const ret = wasm.wordlit_nextSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Replaces this word literal with a new value.
     * @param replacement - The new value to replace this word with
     * @returns The new node that replaced this one, or undefined if this was the root value
     * @param {any} replacement
     * @returns {Node | undefined}
     */
    replaceWith(replacement) {
        const ret = wasm.wordlit_replaceWith(this.__wbg_ptr, replacement);
        if (ret[2]) {
            throw takeFromExternrefTable0(ret[1]);
        }
        return ret[0] === 0 ? undefined : Node.__wrap(ret[0]);
    }
    /**
     * Returns all next sibling nodes.
     * @returns Array of next siblings
     * @returns {Node[]}
     */
    nextSiblings() {
        const ret = wasm.wordlit_nextSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Sets the raw word value.
     * The value should be a valid unquoted identifier (alphanumeric and underscores).
     * @param value - The raw word string to set
     * @param {string} value
     */
    setRawValue(value) {
        const ptr0 = passStringToWasm0(value, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
        const len0 = WASM_VECTOR_LEN;
        wasm.wordlit_setRawValue(this.__wbg_ptr, ptr0, len0);
    }
    /**
     * Returns the previous sibling node.
     * @returns The previous sibling, or undefined if this is the first child
     * @returns {Node | undefined}
     */
    previousSibling() {
        const ret = wasm.wordlit_previousSibling(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Returns all previous sibling nodes.
     * @returns Array of previous siblings
     * @returns {Node[]}
     */
    previousSiblings() {
        const ret = wasm.wordlit_previousSiblings(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns whether this node's container uses trailing commas.
     * @returns true if trailing commas are used
     * @returns {boolean}
     */
    usesTrailingCommas() {
        const ret = wasm.wordlit_usesTrailingCommas(this.__wbg_ptr);
        return ret !== 0;
    }
    /**
     * Returns the unquoted word value.
     * @returns The word literal as a string
     * @returns {string}
     */
    value() {
        let deferred1_0;
        let deferred1_1;
        try {
            const ret = wasm.wordlit_value(this.__wbg_ptr);
            deferred1_0 = ret[0];
            deferred1_1 = ret[1];
            return getStringFromWasm0(ret[0], ret[1]);
        }
        finally {
            wasm.__wbindgen_free(deferred1_0, deferred1_1, 1);
        }
    }
    /**
     * Returns the parent node in the CST.
     * @returns The parent node, or undefined if this is the root
     * @returns {Node | undefined}
     */
    parent() {
        const ret = wasm.wordlit_parent(this.__wbg_ptr);
        return ret === 0 ? undefined : Node.__wrap(ret);
    }
    /**
     * Removes this node from its parent.
     * After calling this method, the node is detached from the CST and can no longer be used.
     */
    remove() {
        const ptr = this.__destroy_into_raw();
        wasm.wordlit_remove(ptr);
    }
    /**
     * Returns all ancestor nodes from parent to root.
     * @returns Array of ancestor nodes
     * @returns {Node[]}
     */
    ancestors() {
        const ret = wasm.wordlit_ancestors(this.__wbg_ptr);
        var v1 = getArrayJsValueFromWasm0(ret[0], ret[1]).slice();
        wasm.__wbindgen_free(ret[0], ret[1] * 4, 4);
        return v1;
    }
    /**
     * Returns the root node of the document.
     * @returns The root node, or undefined if detached
     * @returns {RootNode | undefined}
     */
    rootNode() {
        const ret = wasm.wordlit_rootNode(this.__wbg_ptr);
        return ret === 0 ? undefined : RootNode.__wrap(ret);
    }
}
if (Symbol.dispose)
    WordLit.prototype[Symbol.dispose] = WordLit.prototype.free;
/**
 * Parses a JSONC (JSON with Comments) string into a concrete syntax tree.
 * @param text - The JSONC text to parse
 * @param options - Optional parsing options
 * @returns The root node of the parsed CST
 * @param {string} text
 * @param {ParseOptions | null} [options]
 * @returns {RootNode}
 */
export function parse(text, options) {
    const ptr0 = passStringToWasm0(text, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.parse(ptr0, len0, isLikeNone(options) ? 0 : addToExternrefTable0(options));
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return RootNode.__wrap(ret[0]);
}
/**
 * Parses a JSONC (JSON with Comments) string directly to a JavaScript object.
 * @param text - The JSONC text to parse
 * @param options - Optional parsing options
 * @returns The plain JavaScript value (object, array, string, number, boolean, or null)
 * @throws If the text cannot be parsed or converted
 * @param {string} text
 * @param {ParseOptions | null} [options]
 * @returns {any}
 */
export function parseToValue(text, options) {
    const ptr0 = passStringToWasm0(text, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len0 = WASM_VECTOR_LEN;
    const ret = wasm.parseToValue(ptr0, len0, isLikeNone(options) ? 0 : addToExternrefTable0(options));
    if (ret[2]) {
        throw takeFromExternrefTable0(ret[1]);
    }
    return takeFromExternrefTable0(ret[0]);
}
export function __wbg_Error_52673b7de5a0ca89(arg0, arg1) {
    const ret = Error(getStringFromWasm0(arg0, arg1));
    return ret;
}
export function __wbg_String_8f0eb39a4a4c2f66(arg0, arg1) {
    const ret = String(arg1);
    const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
}
export function __wbg___wbindgen_bigint_get_as_i64_6e32f5e6aff02e1d(arg0, arg1) {
    const v = arg1;
    const ret = typeof v === "bigint" ? v : undefined;
    getDataViewMemory0().setBigInt64(arg0 + 8 * 1, isLikeNone(ret) ? BigInt(0) : ret, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
}
export function __wbg___wbindgen_boolean_get_dea25b33882b895b(arg0) {
    const v = arg0;
    const ret = typeof v === "boolean" ? v : undefined;
    return isLikeNone(ret) ? 0xFFFFFF : ret ? 1 : 0;
}
export function __wbg___wbindgen_debug_string_adfb662ae34724b6(arg0, arg1) {
    const ret = debugString(arg1);
    const ptr1 = passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    const len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
}
export function __wbg___wbindgen_in_0d3e1e8f0c669317(arg0, arg1) {
    const ret = arg0 in arg1;
    return ret;
}
export function __wbg___wbindgen_is_bigint_0e1a2e3f55cfae27(arg0) {
    const ret = typeof arg0 === "bigint";
    return ret;
}
export function __wbg___wbindgen_is_function_8d400b8b1af978cd(arg0) {
    const ret = typeof arg0 === "function";
    return ret;
}
export function __wbg___wbindgen_is_object_ce774f3490692386(arg0) {
    const val = arg0;
    const ret = typeof val === "object" && val !== null;
    return ret;
}
export function __wbg___wbindgen_is_string_704ef9c8fc131030(arg0) {
    const ret = typeof arg0 === "string";
    return ret;
}
export function __wbg___wbindgen_jsval_eq_b6101cc9cef1fe36(arg0, arg1) {
    const ret = arg0 === arg1;
    return ret;
}
export function __wbg___wbindgen_jsval_loose_eq_766057600fdd1b0d(arg0, arg1) {
    const ret = arg0 == arg1;
    return ret;
}
export function __wbg___wbindgen_number_get_9619185a74197f95(arg0, arg1) {
    const obj = arg1;
    const ret = typeof obj === "number" ? obj : undefined;
    getDataViewMemory0().setFloat64(arg0 + 8 * 1, isLikeNone(ret) ? 0 : ret, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, !isLikeNone(ret), true);
}
export function __wbg___wbindgen_string_get_a2a31e16edf96e42(arg0, arg1) {
    const obj = arg1;
    const ret = typeof obj === "string" ? obj : undefined;
    var ptr1 = isLikeNone(ret)
        ? 0
        : passStringToWasm0(ret, wasm.__wbindgen_malloc, wasm.__wbindgen_realloc);
    var len1 = WASM_VECTOR_LEN;
    getDataViewMemory0().setInt32(arg0 + 4 * 1, len1, true);
    getDataViewMemory0().setInt32(arg0 + 4 * 0, ptr1, true);
}
export function __wbg___wbindgen_throw_dd24417ed36fc46e(arg0, arg1) {
    throw new Error(getStringFromWasm0(arg0, arg1));
}
export function __wbg_call_abb4ff46ce38be40() {
    return handleError(function (arg0, arg1) {
        const ret = arg0.call(arg1);
        return ret;
    }, arguments);
}
export function __wbg_done_62ea16af4ce34b24(arg0) {
    const ret = arg0.done;
    return ret;
}
export function __wbg_entries_83c79938054e065f(arg0) {
    const ret = Object.entries(arg0);
    return ret;
}
export function __wbg_get_6b7bd52aca3f9671(arg0, arg1) {
    const ret = arg0[arg1 >>> 0];
    return ret;
}
export function __wbg_get_af9dab7e9603ea93() {
    return handleError(function (arg0, arg1) {
        const ret = Reflect.get(arg0, arg1);
        return ret;
    }, arguments);
}
export function __wbg_instanceof_ArrayBuffer_f3320d2419cd0355(arg0) {
    let result;
    try {
        result = arg0 instanceof ArrayBuffer;
    }
    catch (_) {
        result = false;
    }
    const ret = result;
    return ret;
}
export function __wbg_instanceof_Map_084be8da74364158(arg0) {
    let result;
    try {
        result = arg0 instanceof Map;
    }
    catch (_) {
        result = false;
    }
    const ret = result;
    return ret;
}
export function __wbg_instanceof_Uint8Array_da54ccc9d3e09434(arg0) {
    let result;
    try {
        result = arg0 instanceof Uint8Array;
    }
    catch (_) {
        result = false;
    }
    const ret = result;
    return ret;
}
export function __wbg_isArray_51fd9e6422c0a395(arg0) {
    const ret = Array.isArray(arg0);
    return ret;
}
export function __wbg_isSafeInteger_ae7d3f054d55fa16(arg0) {
    const ret = Number.isSafeInteger(arg0);
    return ret;
}
export function __wbg_iterator_27b7c8b35ab3e86b() {
    const ret = Symbol.iterator;
    return ret;
}
export function __wbg_length_22ac23eaec9d8053(arg0) {
    const ret = arg0.length;
    return ret;
}
export function __wbg_length_d45040a40c570362(arg0) {
    const ret = arg0.length;
    return ret;
}
export function __wbg_new_1ba21ce319a06297() {
    const ret = new Object();
    return ret;
}
export function __wbg_new_25f239778d6112b9() {
    const ret = new Array();
    return ret;
}
export function __wbg_new_6421f6084cc5bc5a(arg0) {
    const ret = new Uint8Array(arg0);
    return ret;
}
export function __wbg_new_b546ae120718850e() {
    const ret = new Map();
    return ret;
}
export function __wbg_new_df1173567d5ff028(arg0, arg1) {
    const ret = new Error(getStringFromWasm0(arg0, arg1));
    return ret;
}
export function __wbg_next_138a17bbf04e926c(arg0) {
    const ret = arg0.next;
    return ret;
}
export function __wbg_next_3cfe5c0fe2a4cc53() {
    return handleError(function (arg0) {
        const ret = arg0.next();
        return ret;
    }, arguments);
}
export function __wbg_node_new(arg0) {
    const ret = Node.__wrap(arg0);
    return ret;
}
export function __wbg_objectprop_new(arg0) {
    const ret = ObjectProp.__wrap(arg0);
    return ret;
}
export function __wbg_prototypesetcall_dfe9b766cdc1f1fd(arg0, arg1, arg2) {
    Uint8Array.prototype.set.call(getArrayU8FromWasm0(arg0, arg1), arg2);
}
export function __wbg_set_3f1d0b984ed272ed(arg0, arg1, arg2) {
    arg0[arg1] = arg2;
}
export function __wbg_set_7df433eea03a5c14(arg0, arg1, arg2) {
    arg0[arg1 >>> 0] = arg2;
}
export function __wbg_set_efaaf145b9377369(arg0, arg1, arg2) {
    const ret = arg0.set(arg1, arg2);
    return ret;
}
export function __wbg_value_57b7b035e117f7ee(arg0) {
    const ret = arg0.value;
    return ret;
}
export function __wbindgen_cast_2241b6af4c4b2941(arg0, arg1) {
    // Cast intrinsic for `Ref(String) -> Externref`.
    const ret = getStringFromWasm0(arg0, arg1);
    return ret;
}
export function __wbindgen_cast_4625c577ab2ec9ee(arg0) {
    // Cast intrinsic for `U64 -> Externref`.
    const ret = BigInt.asUintN(64, arg0);
    return ret;
}
export function __wbindgen_cast_9ae0607507abb057(arg0) {
    // Cast intrinsic for `I64 -> Externref`.
    const ret = arg0;
    return ret;
}
export function __wbindgen_cast_d6cd19b81560fd6e(arg0) {
    // Cast intrinsic for `F64 -> Externref`.
    const ret = arg0;
    return ret;
}
export function __wbindgen_init_externref_table() {
    const table = wasm.__wbindgen_externrefs;
    const offset = table.grow(4);
    table.set(0, undefined);
    table.set(offset + 0, undefined);
    table.set(offset + 1, null);
    table.set(offset + 2, true);
    table.set(offset + 3, false);
}
