// jsonc-morph/script/mod.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WordLit = exports.StringLit = exports.RootNode = exports.parseToValue = exports.parse = exports.ObjectPropName = exports.ObjectProp = exports.NumberLit = exports.NullKeyword = exports.Node = exports.JsonObject = exports.JsonArray = exports.BooleanLit = void 0;
exports.parseStrict = parseStrict;
exports.parseToValueStrict = parseToValueStrict;
var rs_lib_js_1 = require("./lib/rs_lib.js");
Object.defineProperty(exports, "BooleanLit", { enumerable: true, get: function () { return rs_lib_js_1.BooleanLit; } });
Object.defineProperty(exports, "JsonArray", { enumerable: true, get: function () { return rs_lib_js_1.JsonArray; } });
Object.defineProperty(exports, "JsonObject", { enumerable: true, get: function () { return rs_lib_js_1.JsonObject; } });
Object.defineProperty(exports, "Node", { enumerable: true, get: function () { return rs_lib_js_1.Node; } });
Object.defineProperty(exports, "NullKeyword", { enumerable: true, get: function () { return rs_lib_js_1.NullKeyword; } });
Object.defineProperty(exports, "NumberLit", { enumerable: true, get: function () { return rs_lib_js_1.NumberLit; } });
Object.defineProperty(exports, "ObjectProp", { enumerable: true, get: function () { return rs_lib_js_1.ObjectProp; } });
Object.defineProperty(exports, "ObjectPropName", { enumerable: true, get: function () { return rs_lib_js_1.ObjectPropName; } });
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return rs_lib_js_1.parse; } });
Object.defineProperty(exports, "parseToValue", { enumerable: true, get: function () { return rs_lib_js_1.parseToValue; } });
Object.defineProperty(exports, "RootNode", { enumerable: true, get: function () { return rs_lib_js_1.RootNode; } });
Object.defineProperty(exports, "StringLit", { enumerable: true, get: function () { return rs_lib_js_1.StringLit; } });
Object.defineProperty(exports, "WordLit", { enumerable: true, get: function () { return rs_lib_js_1.WordLit; } });
const rs_lib_js_2 = require("./lib/rs_lib.js");
const STRICT_DEFAULTS = {
    allowComments: false,
    allowTrailingCommas: false,
    allowLooseObjectPropertyNames: false,
    allowMissingCommas: false,
    allowSingleQuotedStrings: false,
    allowHexadecimalNumbers: false,
    allowUnaryPlusNumbers: false,
};
/**
 * Parses a strict JSON string into a concrete syntax tree.
 * By default, all JSONC extensions are disabled (no comments, no trailing commas, etc.).
 * You can selectively enable extensions by setting options to `true`.
 * @param text - The JSON text to parse
 * @param options - Optional parsing options (all default to `false`)
 * @returns The root node of the parsed CST
 */
function parseStrict(text, options) {
    return (0, rs_lib_js_2.parse)(text, { ...STRICT_DEFAULTS, ...options });
}
/**
 * Parses a strict JSON string directly to a JavaScript value.
 * By default, all JSONC extensions are disabled (no comments, no trailing commas, etc.).
 * You can selectively enable extensions by setting options to `true`.
 * @param text - The JSON text to parse
 * @param options - Optional parsing options (all default to `false`)
 * @returns The plain JavaScript value (object, array, string, number, boolean, or null)
 * @throws If the text cannot be parsed or converted
 */
function parseToValueStrict(text, options) {
    return (0, rs_lib_js_2.parseToValue)(text, { ...STRICT_DEFAULTS, ...options });
}
