// @aws-sdk/client-ecr/dist-es/schemas/schemas_0.js
const _A = "Attribute";
const _AD = "AuthorizationData";
const _ADL = "AuthorizationDataList";
const _AECID = "AwsEcrContainerImageDetails";
const _AL = "AttributeList";
const _BBOPE = "BlockedByOrganizationPolicyException";
const _BCLA = "BatchCheckLayerAvailability";
const _BCLAR = "BatchCheckLayerAvailabilityRequest";
const _BCLARa = "BatchCheckLayerAvailabilityResponse";
const _BDI = "BatchDeleteImage";
const _BDIR = "BatchDeleteImageRequest";
const _BDIRa = "BatchDeleteImageResponse";
const _BGI = "BatchGetImage";
const _BGIR = "BatchGetImageRequest";
const _BGIRa = "BatchGetImageResponse";
const _BGRSC = "BatchGetRepositoryScanningConfiguration";
const _BGRSCR = "BatchGetRepositoryScanningConfigurationRequest";
const _BGRSCRa = "BatchGetRepositoryScanningConfigurationResponse";
const _CLU = "CompleteLayerUpload";
const _CLUR = "CompleteLayerUploadRequest";
const _CLURo = "CompleteLayerUploadResponse";
const _CPTCR = "CreatePullThroughCacheRule";
const _CPTCRR = "CreatePullThroughCacheRuleRequest";
const _CPTCRRr = "CreatePullThroughCacheRuleResponse";
const _CR = "CreateRepository";
const _CRCT = "CreateRepositoryCreationTemplate";
const _CRCTR = "CreateRepositoryCreationTemplateRequest";
const _CRCTRr = "CreateRepositoryCreationTemplateResponse";
const _CRR = "CreateRepositoryRequest";
const _CRRr = "CreateRepositoryResponse";
const _CS = "CvssScore";
const _CSA = "CvssScoreAdjustment";
const _CSAL = "CvssScoreAdjustmentList";
const _CSD = "CvssScoreDetails";
const _CSL = "CvssScoreList";
const _DI = "DescribeImages";
const _DIF = "DescribeImagesFilter";
const _DIR = "DescribeImagesRequest";
const _DIRS = "DescribeImageReplicationStatus";
const _DIRSR = "DescribeImageReplicationStatusRequest";
const _DIRSRe = "DescribeImageReplicationStatusResponse";
const _DIRe = "DescribeImagesResponse";
const _DISF = "DescribeImageScanFindings";
const _DISFR = "DescribeImageScanFindingsRequest";
const _DISFRe = "DescribeImageScanFindingsResponse";
const _DISS = "DescribeImageSigningStatus";
const _DISSR = "DescribeImageSigningStatusRequest";
const _DISSRe = "DescribeImageSigningStatusResponse";
const _DLP = "DeleteLifecyclePolicy";
const _DLPR = "DeleteLifecyclePolicyRequest";
const _DLPRe = "DeleteLifecyclePolicyResponse";
const _DPTCR = "DeletePullThroughCacheRule";
const _DPTCRR = "DeletePullThroughCacheRuleRequest";
const _DPTCRRe = "DeletePullThroughCacheRuleResponse";
const _DPTCRRes = "DescribePullThroughCacheRulesRequest";
const _DPTCRResc = "DescribePullThroughCacheRulesResponse";
const _DPTCRe = "DescribePullThroughCacheRules";
const _DPTUE = "DeregisterPullTimeUpdateExclusion";
const _DPTUER = "DeregisterPullTimeUpdateExclusionRequest";
const _DPTUERe = "DeregisterPullTimeUpdateExclusionResponse";
const _DR = "DeleteRepository";
const _DRCT = "DeleteRepositoryCreationTemplate";
const _DRCTR = "DeleteRepositoryCreationTemplateRequest";
const _DRCTRe = "DeleteRepositoryCreationTemplateResponse";
const _DRCTRes = "DescribeRepositoryCreationTemplatesRequest";
const _DRCTResc = "DescribeRepositoryCreationTemplatesResponse";
const _DRCTe = "DescribeRepositoryCreationTemplates";
const _DRP = "DeleteRegistryPolicy";
const _DRPR = "DeleteRegistryPolicyRequest";
const _DRPRe = "DeleteRegistryPolicyResponse";
const _DRPRel = "DeleteRepositoryPolicyRequest";
const _DRPRele = "DeleteRepositoryPolicyResponse";
const _DRPe = "DeleteRepositoryPolicy";
const _DRR = "DeleteRepositoryRequest";
const _DRRe = "DeleteRepositoryResponse";
const _DRRes = "DescribeRegistryRequest";
const _DRResc = "DescribeRegistryResponse";
const _DRRescr = "DescribeRepositoriesRequest";
const _DRRescri = "DescribeRepositoriesResponse";
const _DRe = "DescribeRegistry";
const _DRes = "DescribeRepositories";
const _DSC = "DeleteSigningConfiguration";
const _DSCR = "DeleteSigningConfigurationRequest";
const _DSCRe = "DeleteSigningConfigurationResponse";
const _EAEE = "ExclusionAlreadyExistsException";
const _EC = "EncryptionConfiguration";
const _ECFRCT = "EncryptionConfigurationForRepositoryCreationTemplate";
const _EISF = "EnhancedImageScanFinding";
const _EISFL = "EnhancedImageScanFindingList";
const _ENFE = "ExclusionNotFoundException";
const _EUE = "EmptyUploadException";
const _GAS = "GetAccountSetting";
const _GASR = "GetAccountSettingRequest";
const _GASRe = "GetAccountSettingResponse";
const _GAT = "GetAuthorizationToken";
const _GATR = "GetAuthorizationTokenRequest";
const _GATRe = "GetAuthorizationTokenResponse";
const _GDUFL = "GetDownloadUrlForLayer";
const _GDUFLR = "GetDownloadUrlForLayerRequest";
const _GDUFLRe = "GetDownloadUrlForLayerResponse";
const _GLP = "GetLifecyclePolicy";
const _GLPP = "GetLifecyclePolicyPreview";
const _GLPPR = "GetLifecyclePolicyPreviewRequest";
const _GLPPRe = "GetLifecyclePolicyPreviewResponse";
const _GLPR = "GetLifecyclePolicyRequest";
const _GLPRe = "GetLifecyclePolicyResponse";
const _GRP = "GetRegistryPolicy";
const _GRPR = "GetRegistryPolicyRequest";
const _GRPRe = "GetRegistryPolicyResponse";
const _GRPRet = "GetRepositoryPolicyRequest";
const _GRPRete = "GetRepositoryPolicyResponse";
const _GRPe = "GetRepositoryPolicy";
const _GRSC = "GetRegistryScanningConfiguration";
const _GRSCR = "GetRegistryScanningConfigurationRequest";
const _GRSCRe = "GetRegistryScanningConfigurationResponse";
const _GSC = "GetSigningConfiguration";
const _GSCR = "GetSigningConfigurationRequest";
const _GSCRe = "GetSigningConfigurationResponse";
const _I = "Image";
const _IAE = "ImageArchivedException";
const _IAEE = "ImageAlreadyExistsException";
const _ID = "ImageDetail";
const _IDDNME = "ImageDigestDoesNotMatchException";
const _IDL = "ImageDetailList";
const _IF = "ImageFailure";
const _IFL = "ImageFailureList";
const _II = "ImageIdentifier";
const _IIL = "ImageIdentifierList";
const _IL = "ImageList";
const _ILE = "InvalidLayerException";
const _ILPE = "InvalidLayerPartException";
const _ILU = "InitiateLayerUpload";
const _ILUR = "InitiateLayerUploadRequest";
const _ILURn = "InitiateLayerUploadResponse";
const _INFE = "ImageNotFoundException";
const _IPE = "InvalidParameterException";
const _IR = "ImageReferrer";
const _IRL = "ImageReferrerList";
const _IRS = "ImageReplicationStatus";
const _IRSL = "ImageReplicationStatusList";
const _ISC = "ImageScanningConfiguration";
const _ISCUNSE = "ImageStorageClassUpdateNotSupportedException";
const _ISF = "ImageScanFinding";
const _ISFL = "ImageScanFindingList";
const _ISFS = "ImageScanFindingsSummary";
const _ISFm = "ImageScanFindings";
const _ISS = "ImageScanStatus";
const _ISSL = "ImageSigningStatusList";
const _ISSm = "ImageSigningStatus";
const _ITAEE = "ImageTagAlreadyExistsException";
const _ITMEF = "ImageTagMutabilityExclusionFilter";
const _ITMEFm = "ImageTagMutabilityExclusionFilters";
const _ITPE = "InvalidTagParameterException";
const _K = "Key";
const _KE = "KmsException";
const _L = "Layer";
const _LAEE = "LayerAlreadyExistsException";
const _LEE = "LimitExceededException";
const _LF = "LayerFailure";
const _LFL = "LayerFailureList";
const _LI = "ListImages";
const _LIE = "LayerInaccessibleException";
const _LIF = "ListImagesFilter";
const _LIR = "ListImagesRequest";
const _LIRF = "ListImageReferrersFilter";
const _LIRR = "ListImageReferrersRequest";
const _LIRRi = "ListImageReferrersResponse";
const _LIRi = "ListImagesResponse";
const _LIRis = "ListImageReferrers";
const _LL = "LayerList";
const _LNFE = "LayersNotFoundException";
const _LPNFE = "LifecyclePolicyNotFoundException";
const _LPPF = "LifecyclePolicyPreviewFilter";
const _LPPIPE = "LifecyclePolicyPreviewInProgressException";
const _LPPNFE = "LifecyclePolicyPreviewNotFoundException";
const _LPPR = "LifecyclePolicyPreviewResult";
const _LPPRL = "LifecyclePolicyPreviewResultList";
const _LPPS = "LifecyclePolicyPreviewSummary";
const _LPRA = "LifecyclePolicyRuleAction";
const _LPTSE = "LayerPartTooSmallException";
const _LPTUE = "ListPullTimeUpdateExclusions";
const _LPTUER = "ListPullTimeUpdateExclusionsRequest";
const _LPTUERi = "ListPullTimeUpdateExclusionsResponse";
const _LTFR = "ListTagsForResource";
const _LTFRR = "ListTagsForResourceRequest";
const _LTFRRi = "ListTagsForResourceResponse";
const _PAS = "PutAccountSetting";
const _PASR = "PutAccountSettingRequest";
const _PASRu = "PutAccountSettingResponse";
const _PI = "PutImage";
const _PIR = "PutImageRequest";
const _PIRu = "PutImageResponse";
const _PISC = "PutImageScanningConfiguration";
const _PISCR = "PutImageScanningConfigurationRequest";
const _PISCRu = "PutImageScanningConfigurationResponse";
const _PITM = "PutImageTagMutability";
const _PITMR = "PutImageTagMutabilityRequest";
const _PITMRu = "PutImageTagMutabilityResponse";
const _PLP = "PutLifecyclePolicy";
const _PLPR = "PutLifecyclePolicyRequest";
const _PLPRu = "PutLifecyclePolicyResponse";
const _PRC = "PutReplicationConfiguration";
const _PRCR = "PutReplicationConfigurationRequest";
const _PRCRu = "PutReplicationConfigurationResponse";
const _PRP = "PutRegistryPolicy";
const _PRPR = "PutRegistryPolicyRequest";
const _PRPRu = "PutRegistryPolicyResponse";
const _PRSC = "PutRegistryScanningConfiguration";
const _PRSCR = "PutRegistryScanningConfigurationRequest";
const _PRSCRu = "PutRegistryScanningConfigurationResponse";
const _PSC = "PutSigningConfiguration";
const _PSCR = "PutSigningConfigurationRequest";
const _PSCRu = "PutSigningConfigurationResponse";
const _PTCR = "PullThroughCacheRule";
const _PTCRAEE = "PullThroughCacheRuleAlreadyExistsException";
const _PTCRL = "PullThroughCacheRuleList";
const _PTCRNFE = "PullThroughCacheRuleNotFoundException";
const _PVD = "PackageVulnerabilityDetails";
const _R = "Recommendation";
const _RAEE = "RepositoryAlreadyExistsException";
const _RC = "ReplicationConfiguration";
const _RCT = "RepositoryCreationTemplate";
const _RCTL = "RepositoryCreationTemplateList";
const _RD = "ReplicationDestination";
const _RDL = "ReplicationDestinationList";
const _RDe = "ResourceDetails";
const _RF = "RepositoryFilter";
const _RFL = "RepositoryFilterList";
const _RINFE = "ReferencedImagesNotFoundException";
const _RL = "RepositoryList";
const _RLe = "ResourceList";
const _RNEE = "RepositoryNotEmptyException";
const _RNFE = "RepositoryNotFoundException";
const _RPNFE = "RegistryPolicyNotFoundException";
const _RPNFEe = "RepositoryPolicyNotFoundException";
const _RPTUE = "RegisterPullTimeUpdateExclusion";
const _RPTUER = "RegisterPullTimeUpdateExclusionRequest";
const _RPTUERe = "RegisterPullTimeUpdateExclusionResponse";
const _RR = "ReplicationRule";
const _RRL = "ReplicationRuleList";
const _RSC = "RegistryScanningConfiguration";
const _RSCF = "RepositoryScanningConfigurationFailure";
const _RSCFL = "RepositoryScanningConfigurationFailureList";
const _RSCL = "RepositoryScanningConfigurationList";
const _RSCe = "RepositoryScanningConfiguration";
const _RSR = "RegistryScanningRule";
const _RSRL = "RegistryScanningRuleList";
const _Re = "Remediation";
const _Rep = "Repository";
const _Res = "Resource";
const _SC = "SigningConfiguration";
const _SCNFE = "SigningConfigurationNotFoundException";
const _SD = "ScoreDetails";
const _SE = "ServerException";
const _SI = "SubjectIdentifier";
const _SIS = "StartImageScan";
const _SISR = "StartImageScanRequest";
const _SISRt = "StartImageScanResponse";
const _SLPP = "StartLifecyclePolicyPreview";
const _SLPPR = "StartLifecyclePolicyPreviewRequest";
const _SLPPRt = "StartLifecyclePolicyPreviewResponse";
const _SNFE = "ScanNotFoundException";
const _SNFEe = "SecretNotFoundException";
const _SR = "SigningRule";
const _SRF = "ScanningRepositoryFilter";
const _SRFL = "ScanningRepositoryFilterList";
const _SRFLi = "SigningRepositoryFilterList";
const _SRFi = "SigningRepositoryFilter";
const _SRL = "SigningRuleList";
const _SRP = "SetRepositoryPolicy";
const _SRPR = "SetRepositoryPolicyRequest";
const _SRPRe = "SetRepositoryPolicyResponse";
const _T = "Tag";
const _TAEE = "TemplateAlreadyExistsException";
const _TITC = "TransitioningImageTotalCount";
const _TITCr = "TransitioningImageTotalCounts";
const _TL = "TagList";
const _TMTE = "TooManyTagsException";
const _TNFE = "TemplateNotFoundException";
const _TR = "TagResource";
const _TRR = "TagResourceRequest";
const _TRRa = "TagResourceResponse";
const _UISC = "UpdateImageStorageClass";
const _UISCR = "UpdateImageStorageClassRequest";
const _UISCRp = "UpdateImageStorageClassResponse";
const _UITE = "UnsupportedImageTypeException";
const _ULP = "UploadLayerPart";
const _ULPR = "UploadLayerPartRequest";
const _ULPRp = "UploadLayerPartResponse";
const _UNFE = "UploadNotFoundException";
const _UPTCR = "UpdatePullThroughCacheRule";
const _UPTCRR = "UpdatePullThroughCacheRuleRequest";
const _UPTCRRp = "UpdatePullThroughCacheRuleResponse";
const _UR = "UntagResource";
const _URCT = "UpdateRepositoryCreationTemplate";
const _URCTR = "UpdateRepositoryCreationTemplateRequest";
const _URCTRp = "UpdateRepositoryCreationTemplateResponse";
const _URR = "UntagResourceRequest";
const _URRn = "UntagResourceResponse";
const _UTASE = "UnableToAccessSecretException";
const _UTDSVE = "UnableToDecryptSecretValueException";
const _UTGUIE = "UnableToGetUpstreamImageException";
const _UTGULE = "UnableToGetUpstreamLayerException";
const _UTLUIRE = "UnableToListUpstreamImageReferrersException";
const _UURE = "UnsupportedUpstreamRegistryException";
const _V = "Value";
const _VE = "ValidationException";
const _VP = "VulnerablePackage";
const _VPL = "VulnerablePackagesList";
const _VPTCR = "ValidatePullThroughCacheRule";
const _VPTCRR = "ValidatePullThroughCacheRuleRequest";
const _VPTCRRa = "ValidatePullThroughCacheRuleResponse";
const _a = "architecture";
const _aAI = "awsAccountId";
const _aD = "authorizationData";
const _aECI = "awsEcrContainerImage";
const _aF = "appliedFor";
const _aMT = "acceptedMediaTypes";
const _aMTr = "artifactMediaType";
const _aRP = "appliedRulePriority";
const _aS = "artifactStatus";
const _aSF = "appliedScanFilters";
const _aT = "authorizationToken";
const _aTr = "artifactType";
const _aTrt = "artifactTypes";
const _ac = "action";
const _ad = "adjustments";
const _an = "annotations";
const _ar = "arch";
const _at = "attributes";
const _au = "author";
const _bS = "baseScore";
const _c = "client";
const _cA = "credentialArn";
const _cAr = "createdAt";
const _cRA = "customRoleArn";
const _cv = "cvss";
const _d = "description";
const _dU = "downloadUrl";
const _de = "destinations";
const _det = "details";
const _di = "digest";
const _e = "error";
const _eA = "expiresAt";
const _eAx = "exploitAvailable";
const _eC = "encryptionConfiguration";
const _eF = "enhancedFindings";
const _eITC = "expiringImageTotalCount";
const _eRP = "ecrRepositoryPrefix";
const _eRPc = "ecrRepositoryPrefixes";
const _eT = "encryptionType";
const _ep = "epoch";
const _f = "failures";
const _fA = "findingArn";
const _fAi = "fixAvailable";
const _fC = "failureCode";
const _fIV = "fixedInVersion";
const _fOA = "firstObservedAt";
const _fP = "filePath";
const _fR = "failureReason";
const _fSC = "findingSeverityCounts";
const _fT = "filterType";
const _fa = "failure";
const _fi = "filter";
const _fin = "findings";
const _fo = "force";
const _hE = "httpError";
const _i = "images";
const _iD = "imageDetails";
const _iDm = "imageDigest";
const _iH = "imageHash";
const _iI = "imageIds";
const _iIm = "imageId";
const _iM = "imageManifest";
const _iMMT = "imageManifestMediaType";
const _iPA = "imagePushedAt";
const _iS = "imageStatus";
const _iSC = "imageScanningConfiguration";
const _iSCA = "imageScanCompletedAt";
const _iSF = "imageScanFindings";
const _iSFS = "imageScanFindingsSummary";
const _iSIB = "imageSizeInBytes";
const _iSS = "imageScanStatus";
const _iT = "imageTags";
const _iTC = "imageTotalCount";
const _iTM = "imageTagMutability";
const _iTMEF = "imageTagMutabilityExclusionFilters";
const _iTm = "imageTag";
const _iUC = "inUseCount";
const _iV = "isValid";
const _id = "id";
const _im = "image";
const _k = "key";
const _kE = "kmsError";
const _kK = "kmsKey";
const _l = "layers";
const _lA = "layerAvailability";
const _lAA = "lastArchivedAt";
const _lAAa = "lastActivatedAt";
const _lBR = "lastByteReceived";
const _lD = "layerDigests";
const _lDa = "layerDigest";
const _lEA = "lastEvaluatedAt";
const _lIUA = "lastInUseAt";
const _lOA = "lastObservedAt";
const _lP = "lifecyclePolicy";
const _lPB = "layerPartBlob";
const _lPT = "lifecyclePolicyText";
const _lRPT = "lastRecordedPullTime";
const _lS = "layerSize";
const _lVBR = "lastValidByteReceived";
const _m = "message";
const _mR = "maxResults";
const _mT = "mediaType";
const _me = "metric";
const _n = "name";
const _nT = "nextToken";
const _p = "platform";
const _pA = "pushedAt";
const _pAr = "principalArn";
const _pE = "proxyEndpoint";
const _pFB = "partFirstByte";
const _pLB = "partLastByte";
const _pM = "packageManager";
const _pR = "previewResults";
const _pS = "partSize";
const _pT = "policyText";
const _pTCR = "pullThroughCacheRules";
const _pTUE = "pullTimeUpdateExclusions";
const _pVD = "packageVulnerabilityDetails";
const _pr = "prefix";
const _pre = "prefixes";
const _r = "registry";
const _rA = "resourceArn";
const _rAe = "repositoryArn";
const _rC = "replicationConfiguration";
const _rCT = "repositoryCreationTemplate";
const _rCTe = "repositoryCreationTemplates";
const _rF = "repositoryFilters";
const _rI = "registryId";
const _rIe = "registryIds";
const _rN = "repositoryName";
const _rNe = "repositoryNames";
const _rP = "repositoryPolicy";
const _rS = "replicationStatuses";
const _rSC = "registryScanningConfiguration";
const _rT = "resourceTags";
const _rU = "referenceUrls";
const _rUe = "repositoryUri";
const _rV = "relatedVulnerabilities";
const _re = "repository";
const _rea = "reason";
const _rec = "recommendation";
const _ref = "referrers";
const _reg = "region";
const _rel = "release";
const _rem = "remediation";
const _rep = "repositories";
const _res = "resources";
const _ru = "rules";
const _s = "smithy.ts.sdk.synthetic.com.amazonaws.ecr";
const _sC = "scanningConfigurations";
const _sCc = "scanningConfiguration";
const _sCi = "signingConfiguration";
const _sCt = "storageClass";
const _sD = "scoreDetails";
const _sF = "scanFrequency";
const _sI = "subjectId";
const _sLH = "sourceLayerHash";
const _sMD = "subjectManifestDigest";
const _sOP = "scanOnPush";
const _sPA = "signingProfileArn";
const _sS = "scoreSource";
const _sSi = "signingStatuses";
const _sT = "scanType";
const _sU = "sourceUrl";
const _sV = "scoringVector";
const _sc = "score";
const _se = "server";
const _sev = "severity";
const _si = "size";
const _so = "source";
const _st = "status";
const _su = "summary";
const _t = "tags";
const _tITC = "transitioningImageTotalCounts";
const _tK = "tagKeys";
const _tS = "tagStatus";
const _tSC = "targetStorageClass";
const _te = "text";
const _ti = "title";
const _ty = "type";
const _u = "uri";
const _uA = "updatedAt";
const _uI = "uploadId";
const _uR = "upstreamRegistry";
const _uRP = "upstreamRepositoryPrefix";
const _uRU = "upstreamRegistryUrl";
const _ur = "url";
const _v = "value";
const _vCA = "vendorCreatedAt";
const _vI = "vulnerabilityId";
const _vP = "vulnerablePackages";
const _vS = "vendorSeverity";
const _vSUA = "vulnerabilitySourceUpdatedAt";
const _vUA = "vendorUpdatedAt";
const _ve = "version";
const n0 = "com.amazonaws.ecr";
import { TypeRegistry } from "@smithy/core/schema";
import { ECRServiceException } from "../models/ECRServiceException";
import { BlockedByOrganizationPolicyException, EmptyUploadException, ExclusionAlreadyExistsException, ExclusionNotFoundException, ImageAlreadyExistsException, ImageArchivedException, ImageDigestDoesNotMatchException, ImageNotFoundException, ImageStorageClassUpdateNotSupportedException, ImageTagAlreadyExistsException, InvalidLayerException, InvalidLayerPartException, InvalidParameterException, InvalidTagParameterException, KmsException, LayerAlreadyExistsException, LayerInaccessibleException, LayerPartTooSmallException, LayersNotFoundException, LifecyclePolicyNotFoundException, LifecyclePolicyPreviewInProgressException, LifecyclePolicyPreviewNotFoundException, LimitExceededException, PullThroughCacheRuleAlreadyExistsException, PullThroughCacheRuleNotFoundException, ReferencedImagesNotFoundException, RegistryPolicyNotFoundException, RepositoryAlreadyExistsException, RepositoryNotEmptyException, RepositoryNotFoundException, RepositoryPolicyNotFoundException, ScanNotFoundException, SecretNotFoundException, ServerException, SigningConfigurationNotFoundException, TemplateAlreadyExistsException, TemplateNotFoundException, TooManyTagsException, UnableToAccessSecretException, UnableToDecryptSecretValueException, UnableToGetUpstreamImageException, UnableToGetUpstreamLayerException, UnableToListUpstreamImageReferrersException, UnsupportedImageTypeException, UnsupportedUpstreamRegistryException, UploadNotFoundException, ValidationException, } from "../models/errors";
const _s_registry = TypeRegistry.for(_s);
export var ECRServiceException$ = [-3, _s, "ECRServiceException", 0, [], []];
_s_registry.registerError(ECRServiceException$, ECRServiceException);
const n0_registry = TypeRegistry.for(n0);
export var BlockedByOrganizationPolicyException$ = [-3, n0, _BBOPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(BlockedByOrganizationPolicyException$, BlockedByOrganizationPolicyException);
export var EmptyUploadException$ = [-3, n0, _EUE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(EmptyUploadException$, EmptyUploadException);
export var ExclusionAlreadyExistsException$ = [-3, n0, _EAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ExclusionAlreadyExistsException$, ExclusionAlreadyExistsException);
export var ExclusionNotFoundException$ = [-3, n0, _ENFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ExclusionNotFoundException$, ExclusionNotFoundException);
export var ImageAlreadyExistsException$ = [-3, n0, _IAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ImageAlreadyExistsException$, ImageAlreadyExistsException);
export var ImageArchivedException$ = [-3, n0, _IAE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ImageArchivedException$, ImageArchivedException);
export var ImageDigestDoesNotMatchException$ = [-3, n0, _IDDNME,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ImageDigestDoesNotMatchException$, ImageDigestDoesNotMatchException);
export var ImageNotFoundException$ = [-3, n0, _INFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ImageNotFoundException$, ImageNotFoundException);
export var ImageStorageClassUpdateNotSupportedException$ = [-3, n0, _ISCUNSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ImageStorageClassUpdateNotSupportedException$, ImageStorageClassUpdateNotSupportedException);
export var ImageTagAlreadyExistsException$ = [-3, n0, _ITAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ImageTagAlreadyExistsException$, ImageTagAlreadyExistsException);
export var InvalidLayerException$ = [-3, n0, _ILE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(InvalidLayerException$, InvalidLayerException);
export var InvalidLayerPartException$ = [-3, n0, _ILPE,
    { [_e]: _c },
    [_rI, _rN, _uI, _lVBR, _m],
    [0, 0, 0, 1, 0]
];
n0_registry.registerError(InvalidLayerPartException$, InvalidLayerPartException);
export var InvalidParameterException$ = [-3, n0, _IPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(InvalidParameterException$, InvalidParameterException);
export var InvalidTagParameterException$ = [-3, n0, _ITPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(InvalidTagParameterException$, InvalidTagParameterException);
export var KmsException$ = [-3, n0, _KE,
    { [_e]: _c },
    [_m, _kE],
    [0, 0]
];
n0_registry.registerError(KmsException$, KmsException);
export var LayerAlreadyExistsException$ = [-3, n0, _LAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LayerAlreadyExistsException$, LayerAlreadyExistsException);
export var LayerInaccessibleException$ = [-3, n0, _LIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LayerInaccessibleException$, LayerInaccessibleException);
export var LayerPartTooSmallException$ = [-3, n0, _LPTSE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LayerPartTooSmallException$, LayerPartTooSmallException);
export var LayersNotFoundException$ = [-3, n0, _LNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LayersNotFoundException$, LayersNotFoundException);
export var LifecyclePolicyNotFoundException$ = [-3, n0, _LPNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LifecyclePolicyNotFoundException$, LifecyclePolicyNotFoundException);
export var LifecyclePolicyPreviewInProgressException$ = [-3, n0, _LPPIPE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LifecyclePolicyPreviewInProgressException$, LifecyclePolicyPreviewInProgressException);
export var LifecyclePolicyPreviewNotFoundException$ = [-3, n0, _LPPNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LifecyclePolicyPreviewNotFoundException$, LifecyclePolicyPreviewNotFoundException);
export var LimitExceededException$ = [-3, n0, _LEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(LimitExceededException$, LimitExceededException);
export var PullThroughCacheRuleAlreadyExistsException$ = [-3, n0, _PTCRAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(PullThroughCacheRuleAlreadyExistsException$, PullThroughCacheRuleAlreadyExistsException);
export var PullThroughCacheRuleNotFoundException$ = [-3, n0, _PTCRNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(PullThroughCacheRuleNotFoundException$, PullThroughCacheRuleNotFoundException);
export var ReferencedImagesNotFoundException$ = [-3, n0, _RINFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ReferencedImagesNotFoundException$, ReferencedImagesNotFoundException);
export var RegistryPolicyNotFoundException$ = [-3, n0, _RPNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(RegistryPolicyNotFoundException$, RegistryPolicyNotFoundException);
export var RepositoryAlreadyExistsException$ = [-3, n0, _RAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(RepositoryAlreadyExistsException$, RepositoryAlreadyExistsException);
export var RepositoryNotEmptyException$ = [-3, n0, _RNEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(RepositoryNotEmptyException$, RepositoryNotEmptyException);
export var RepositoryNotFoundException$ = [-3, n0, _RNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(RepositoryNotFoundException$, RepositoryNotFoundException);
export var RepositoryPolicyNotFoundException$ = [-3, n0, _RPNFEe,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(RepositoryPolicyNotFoundException$, RepositoryPolicyNotFoundException);
export var ScanNotFoundException$ = [-3, n0, _SNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(ScanNotFoundException$, ScanNotFoundException);
export var SecretNotFoundException$ = [-3, n0, _SNFEe,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(SecretNotFoundException$, SecretNotFoundException);
export var ServerException$ = [-3, n0, _SE,
    { [_e]: _se },
    [_m],
    [0]
];
n0_registry.registerError(ServerException$, ServerException);
export var SigningConfigurationNotFoundException$ = [-3, n0, _SCNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(SigningConfigurationNotFoundException$, SigningConfigurationNotFoundException);
export var TemplateAlreadyExistsException$ = [-3, n0, _TAEE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(TemplateAlreadyExistsException$, TemplateAlreadyExistsException);
export var TemplateNotFoundException$ = [-3, n0, _TNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(TemplateNotFoundException$, TemplateNotFoundException);
export var TooManyTagsException$ = [-3, n0, _TMTE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(TooManyTagsException$, TooManyTagsException);
export var UnableToAccessSecretException$ = [-3, n0, _UTASE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UnableToAccessSecretException$, UnableToAccessSecretException);
export var UnableToDecryptSecretValueException$ = [-3, n0, _UTDSVE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UnableToDecryptSecretValueException$, UnableToDecryptSecretValueException);
export var UnableToGetUpstreamImageException$ = [-3, n0, _UTGUIE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UnableToGetUpstreamImageException$, UnableToGetUpstreamImageException);
export var UnableToGetUpstreamLayerException$ = [-3, n0, _UTGULE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UnableToGetUpstreamLayerException$, UnableToGetUpstreamLayerException);
export var UnableToListUpstreamImageReferrersException$ = [-3, n0, _UTLUIRE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UnableToListUpstreamImageReferrersException$, UnableToListUpstreamImageReferrersException);
export var UnsupportedImageTypeException$ = [-3, n0, _UITE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UnsupportedImageTypeException$, UnsupportedImageTypeException);
export var UnsupportedUpstreamRegistryException$ = [-3, n0, _UURE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UnsupportedUpstreamRegistryException$, UnsupportedUpstreamRegistryException);
export var UploadNotFoundException$ = [-3, n0, _UNFE,
    { [_e]: _c },
    [_m],
    [0]
];
n0_registry.registerError(UploadNotFoundException$, UploadNotFoundException);
export var ValidationException$ = [-3, n0, _VE,
    { [_e]: _c, [_hE]: 400 },
    [_m],
    [0]
];
n0_registry.registerError(ValidationException$, ValidationException);
export const errorTypeRegistries = [
    _s_registry,
    n0_registry,
];
export var Attribute$ = [3, n0, _A,
    0,
    [_k, _v],
    [0, 0], 1
];
export var AuthorizationData$ = [3, n0, _AD,
    0,
    [_aT, _eA, _pE],
    [0, 4, 0]
];
export var AwsEcrContainerImageDetails$ = [3, n0, _AECID,
    0,
    [_a, _au, _iH, _iT, _p, _pA, _lIUA, _iUC, _r, _rN],
    [0, 0, 0, 64 | 0, 0, 4, 4, 1, 0, 0]
];
export var BatchCheckLayerAvailabilityRequest$ = [3, n0, _BCLAR,
    0,
    [_rN, _lD, _rI],
    [0, 64 | 0, 0], 2
];
export var BatchCheckLayerAvailabilityResponse$ = [3, n0, _BCLARa,
    0,
    [_l, _f],
    [() => LayerList, () => LayerFailureList]
];
export var BatchDeleteImageRequest$ = [3, n0, _BDIR,
    0,
    [_rN, _iI, _rI],
    [0, () => ImageIdentifierList, 0], 2
];
export var BatchDeleteImageResponse$ = [3, n0, _BDIRa,
    0,
    [_iI, _f],
    [() => ImageIdentifierList, () => ImageFailureList]
];
export var BatchGetImageRequest$ = [3, n0, _BGIR,
    0,
    [_rN, _iI, _rI, _aMT],
    [0, () => ImageIdentifierList, 0, 64 | 0], 2
];
export var BatchGetImageResponse$ = [3, n0, _BGIRa,
    0,
    [_i, _f],
    [() => ImageList, () => ImageFailureList]
];
export var BatchGetRepositoryScanningConfigurationRequest$ = [3, n0, _BGRSCR,
    0,
    [_rNe],
    [64 | 0], 1
];
export var BatchGetRepositoryScanningConfigurationResponse$ = [3, n0, _BGRSCRa,
    0,
    [_sC, _f],
    [() => RepositoryScanningConfigurationList, () => RepositoryScanningConfigurationFailureList]
];
export var CompleteLayerUploadRequest$ = [3, n0, _CLUR,
    0,
    [_rN, _uI, _lD, _rI],
    [0, 0, 64 | 0, 0], 3
];
export var CompleteLayerUploadResponse$ = [3, n0, _CLURo,
    0,
    [_rI, _rN, _uI, _lDa],
    [0, 0, 0, 0]
];
export var CreatePullThroughCacheRuleRequest$ = [3, n0, _CPTCRR,
    0,
    [_eRP, _uRU, _rI, _uR, _cA, _cRA, _uRP],
    [0, 0, 0, 0, 0, 0, 0], 2
];
export var CreatePullThroughCacheRuleResponse$ = [3, n0, _CPTCRRr,
    0,
    [_eRP, _uRU, _cAr, _rI, _uR, _cA, _cRA, _uRP],
    [0, 0, 4, 0, 0, 0, 0, 0]
];
export var CreateRepositoryCreationTemplateRequest$ = [3, n0, _CRCTR,
    0,
    [_pr, _aF, _d, _eC, _rT, _iTM, _iTMEF, _rP, _lP, _cRA],
    [0, 64 | 0, 0, () => EncryptionConfigurationForRepositoryCreationTemplate$, () => TagList, 0, () => ImageTagMutabilityExclusionFilters, 0, 0, 0], 2
];
export var CreateRepositoryCreationTemplateResponse$ = [3, n0, _CRCTRr,
    0,
    [_rI, _rCT],
    [0, () => RepositoryCreationTemplate$]
];
export var CreateRepositoryRequest$ = [3, n0, _CRR,
    0,
    [_rN, _rI, _t, _iTM, _iTMEF, _iSC, _eC],
    [0, 0, () => TagList, 0, () => ImageTagMutabilityExclusionFilters, () => ImageScanningConfiguration$, () => EncryptionConfiguration$], 1
];
export var CreateRepositoryResponse$ = [3, n0, _CRRr,
    0,
    [_re],
    [() => Repository$]
];
export var CvssScore$ = [3, n0, _CS,
    0,
    [_bS, _sV, _so, _ve],
    [1, 0, 0, 0]
];
export var CvssScoreAdjustment$ = [3, n0, _CSA,
    0,
    [_me, _rea],
    [0, 0]
];
export var CvssScoreDetails$ = [3, n0, _CSD,
    0,
    [_ad, _sc, _sS, _sV, _ve],
    [() => CvssScoreAdjustmentList, 1, 0, 0, 0]
];
export var DeleteLifecyclePolicyRequest$ = [3, n0, _DLPR,
    0,
    [_rN, _rI],
    [0, 0], 1
];
export var DeleteLifecyclePolicyResponse$ = [3, n0, _DLPRe,
    0,
    [_rI, _rN, _lPT, _lEA],
    [0, 0, 0, 4]
];
export var DeletePullThroughCacheRuleRequest$ = [3, n0, _DPTCRR,
    0,
    [_eRP, _rI],
    [0, 0], 1
];
export var DeletePullThroughCacheRuleResponse$ = [3, n0, _DPTCRRe,
    0,
    [_eRP, _uRU, _cAr, _rI, _cA, _cRA, _uRP],
    [0, 0, 4, 0, 0, 0, 0]
];
export var DeleteRegistryPolicyRequest$ = [3, n0, _DRPR,
    0,
    [],
    []
];
export var DeleteRegistryPolicyResponse$ = [3, n0, _DRPRe,
    0,
    [_rI, _pT],
    [0, 0]
];
export var DeleteRepositoryCreationTemplateRequest$ = [3, n0, _DRCTR,
    0,
    [_pr],
    [0], 1
];
export var DeleteRepositoryCreationTemplateResponse$ = [3, n0, _DRCTRe,
    0,
    [_rI, _rCT],
    [0, () => RepositoryCreationTemplate$]
];
export var DeleteRepositoryPolicyRequest$ = [3, n0, _DRPRel,
    0,
    [_rN, _rI],
    [0, 0], 1
];
export var DeleteRepositoryPolicyResponse$ = [3, n0, _DRPRele,
    0,
    [_rI, _rN, _pT],
    [0, 0, 0]
];
export var DeleteRepositoryRequest$ = [3, n0, _DRR,
    0,
    [_rN, _rI, _fo],
    [0, 0, 2], 1
];
export var DeleteRepositoryResponse$ = [3, n0, _DRRe,
    0,
    [_re],
    [() => Repository$]
];
export var DeleteSigningConfigurationRequest$ = [3, n0, _DSCR,
    0,
    [],
    []
];
export var DeleteSigningConfigurationResponse$ = [3, n0, _DSCRe,
    0,
    [_rI, _sCi],
    [0, () => SigningConfiguration$]
];
export var DeregisterPullTimeUpdateExclusionRequest$ = [3, n0, _DPTUER,
    0,
    [_pAr],
    [0], 1
];
export var DeregisterPullTimeUpdateExclusionResponse$ = [3, n0, _DPTUERe,
    0,
    [_pAr],
    [0]
];
export var DescribeImageReplicationStatusRequest$ = [3, n0, _DIRSR,
    0,
    [_rN, _iIm, _rI],
    [0, () => ImageIdentifier$, 0], 2
];
export var DescribeImageReplicationStatusResponse$ = [3, n0, _DIRSRe,
    0,
    [_rN, _iIm, _rS],
    [0, () => ImageIdentifier$, () => ImageReplicationStatusList]
];
export var DescribeImageScanFindingsRequest$ = [3, n0, _DISFR,
    0,
    [_rN, _iIm, _rI, _nT, _mR],
    [0, () => ImageIdentifier$, 0, 0, 1], 2
];
export var DescribeImageScanFindingsResponse$ = [3, n0, _DISFRe,
    0,
    [_rI, _rN, _iIm, _iSS, _iSF, _nT],
    [0, 0, () => ImageIdentifier$, () => ImageScanStatus$, () => ImageScanFindings$, 0]
];
export var DescribeImagesFilter$ = [3, n0, _DIF,
    0,
    [_tS, _iS],
    [0, 0]
];
export var DescribeImageSigningStatusRequest$ = [3, n0, _DISSR,
    0,
    [_rN, _iIm, _rI],
    [0, () => ImageIdentifier$, 0], 2
];
export var DescribeImageSigningStatusResponse$ = [3, n0, _DISSRe,
    0,
    [_rN, _iIm, _rI, _sSi],
    [0, () => ImageIdentifier$, 0, () => ImageSigningStatusList]
];
export var DescribeImagesRequest$ = [3, n0, _DIR,
    0,
    [_rN, _rI, _iI, _nT, _mR, _fi],
    [0, 0, () => ImageIdentifierList, 0, 1, () => DescribeImagesFilter$], 1
];
export var DescribeImagesResponse$ = [3, n0, _DIRe,
    0,
    [_iD, _nT],
    [() => ImageDetailList, 0]
];
export var DescribePullThroughCacheRulesRequest$ = [3, n0, _DPTCRRes,
    0,
    [_rI, _eRPc, _nT, _mR],
    [0, 64 | 0, 0, 1]
];
export var DescribePullThroughCacheRulesResponse$ = [3, n0, _DPTCRResc,
    0,
    [_pTCR, _nT],
    [() => PullThroughCacheRuleList, 0]
];
export var DescribeRegistryRequest$ = [3, n0, _DRRes,
    0,
    [],
    []
];
export var DescribeRegistryResponse$ = [3, n0, _DRResc,
    0,
    [_rI, _rC],
    [0, () => ReplicationConfiguration$]
];
export var DescribeRepositoriesRequest$ = [3, n0, _DRRescr,
    0,
    [_rI, _rNe, _nT, _mR],
    [0, 64 | 0, 0, 1]
];
export var DescribeRepositoriesResponse$ = [3, n0, _DRRescri,
    0,
    [_rep, _nT],
    [() => RepositoryList, 0]
];
export var DescribeRepositoryCreationTemplatesRequest$ = [3, n0, _DRCTRes,
    0,
    [_pre, _nT, _mR],
    [64 | 0, 0, 1]
];
export var DescribeRepositoryCreationTemplatesResponse$ = [3, n0, _DRCTResc,
    0,
    [_rI, _rCTe, _nT],
    [0, () => RepositoryCreationTemplateList, 0]
];
export var EncryptionConfiguration$ = [3, n0, _EC,
    0,
    [_eT, _kK],
    [0, 0], 1
];
export var EncryptionConfigurationForRepositoryCreationTemplate$ = [3, n0, _ECFRCT,
    0,
    [_eT, _kK],
    [0, 0], 1
];
export var EnhancedImageScanFinding$ = [3, n0, _EISF,
    0,
    [_aAI, _d, _fA, _fOA, _lOA, _pVD, _rem, _res, _sc, _sD, _sev, _st, _ti, _ty, _uA, _fAi, _eAx],
    [0, 0, 0, 4, 4, () => PackageVulnerabilityDetails$, () => Remediation$, () => ResourceList, 1, () => ScoreDetails$, 0, 0, 0, 0, 4, 0, 0]
];
export var GetAccountSettingRequest$ = [3, n0, _GASR,
    0,
    [_n],
    [0], 1
];
export var GetAccountSettingResponse$ = [3, n0, _GASRe,
    0,
    [_n, _v],
    [0, 0]
];
export var GetAuthorizationTokenRequest$ = [3, n0, _GATR,
    0,
    [_rIe],
    [64 | 0]
];
export var GetAuthorizationTokenResponse$ = [3, n0, _GATRe,
    0,
    [_aD],
    [() => AuthorizationDataList]
];
export var GetDownloadUrlForLayerRequest$ = [3, n0, _GDUFLR,
    0,
    [_rN, _lDa, _rI],
    [0, 0, 0], 2
];
export var GetDownloadUrlForLayerResponse$ = [3, n0, _GDUFLRe,
    0,
    [_dU, _lDa],
    [0, 0]
];
export var GetLifecyclePolicyPreviewRequest$ = [3, n0, _GLPPR,
    0,
    [_rN, _rI, _iI, _nT, _mR, _fi],
    [0, 0, () => ImageIdentifierList, 0, 1, () => LifecyclePolicyPreviewFilter$], 1
];
export var GetLifecyclePolicyPreviewResponse$ = [3, n0, _GLPPRe,
    0,
    [_rI, _rN, _lPT, _st, _nT, _pR, _su],
    [0, 0, 0, 0, 0, () => LifecyclePolicyPreviewResultList, () => LifecyclePolicyPreviewSummary$]
];
export var GetLifecyclePolicyRequest$ = [3, n0, _GLPR,
    0,
    [_rN, _rI],
    [0, 0], 1
];
export var GetLifecyclePolicyResponse$ = [3, n0, _GLPRe,
    0,
    [_rI, _rN, _lPT, _lEA],
    [0, 0, 0, 4]
];
export var GetRegistryPolicyRequest$ = [3, n0, _GRPR,
    0,
    [],
    []
];
export var GetRegistryPolicyResponse$ = [3, n0, _GRPRe,
    0,
    [_rI, _pT],
    [0, 0]
];
export var GetRegistryScanningConfigurationRequest$ = [3, n0, _GRSCR,
    0,
    [],
    []
];
export var GetRegistryScanningConfigurationResponse$ = [3, n0, _GRSCRe,
    0,
    [_rI, _sCc],
    [0, () => RegistryScanningConfiguration$]
];
export var GetRepositoryPolicyRequest$ = [3, n0, _GRPRet,
    0,
    [_rN, _rI],
    [0, 0], 1
];
export var GetRepositoryPolicyResponse$ = [3, n0, _GRPRete,
    0,
    [_rI, _rN, _pT],
    [0, 0, 0]
];
export var GetSigningConfigurationRequest$ = [3, n0, _GSCR,
    0,
    [],
    []
];
export var GetSigningConfigurationResponse$ = [3, n0, _GSCRe,
    0,
    [_rI, _sCi],
    [0, () => SigningConfiguration$]
];
export var Image$ = [3, n0, _I,
    0,
    [_rI, _rN, _iIm, _iM, _iMMT],
    [0, 0, () => ImageIdentifier$, 0, 0]
];
export var ImageDetail$ = [3, n0, _ID,
    0,
    [_rI, _rN, _iDm, _iT, _iSIB, _iPA, _iSS, _iSFS, _iMMT, _aMTr, _lRPT, _sMD, _iS, _lAA, _lAAa],
    [0, 0, 0, 64 | 0, 1, 4, () => ImageScanStatus$, () => ImageScanFindingsSummary$, 0, 0, 4, 0, 0, 4, 4]
];
export var ImageFailure$ = [3, n0, _IF,
    0,
    [_iIm, _fC, _fR],
    [() => ImageIdentifier$, 0, 0]
];
export var ImageIdentifier$ = [3, n0, _II,
    0,
    [_iDm, _iTm],
    [0, 0]
];
export var ImageReferrer$ = [3, n0, _IR,
    0,
    [_di, _mT, _si, _aTr, _an, _aS],
    [0, 0, 1, 0, 128 | 0, 0], 3
];
export var ImageReplicationStatus$ = [3, n0, _IRS,
    0,
    [_reg, _rI, _st, _fC],
    [0, 0, 0, 0]
];
export var ImageScanFinding$ = [3, n0, _ISF,
    0,
    [_n, _d, _u, _sev, _at],
    [0, 0, 0, 0, () => AttributeList]
];
export var ImageScanFindings$ = [3, n0, _ISFm,
    0,
    [_iSCA, _vSUA, _fSC, _fin, _eF],
    [4, 4, 128 | 1, () => ImageScanFindingList, () => EnhancedImageScanFindingList]
];
export var ImageScanFindingsSummary$ = [3, n0, _ISFS,
    0,
    [_iSCA, _vSUA, _fSC],
    [4, 4, 128 | 1]
];
export var ImageScanningConfiguration$ = [3, n0, _ISC,
    0,
    [_sOP],
    [2]
];
export var ImageScanStatus$ = [3, n0, _ISS,
    0,
    [_st, _d],
    [0, 0]
];
export var ImageSigningStatus$ = [3, n0, _ISSm,
    0,
    [_sPA, _fC, _fR, _st],
    [0, 0, 0, 0]
];
export var ImageTagMutabilityExclusionFilter$ = [3, n0, _ITMEF,
    0,
    [_fT, _fi],
    [0, 0], 2
];
export var InitiateLayerUploadRequest$ = [3, n0, _ILUR,
    0,
    [_rN, _rI],
    [0, 0], 1
];
export var InitiateLayerUploadResponse$ = [3, n0, _ILURn,
    0,
    [_uI, _pS],
    [0, 1]
];
export var Layer$ = [3, n0, _L,
    0,
    [_lDa, _lA, _lS, _mT],
    [0, 0, 1, 0]
];
export var LayerFailure$ = [3, n0, _LF,
    0,
    [_lDa, _fC, _fR],
    [0, 0, 0]
];
export var LifecyclePolicyPreviewFilter$ = [3, n0, _LPPF,
    0,
    [_tS],
    [0]
];
export var LifecyclePolicyPreviewResult$ = [3, n0, _LPPR,
    0,
    [_iT, _iDm, _iPA, _ac, _aRP, _sCt],
    [64 | 0, 0, 4, () => LifecyclePolicyRuleAction$, 1, 0]
];
export var LifecyclePolicyPreviewSummary$ = [3, n0, _LPPS,
    0,
    [_eITC, _tITC],
    [1, () => TransitioningImageTotalCounts]
];
export var LifecyclePolicyRuleAction$ = [3, n0, _LPRA,
    0,
    [_ty, _tSC],
    [0, 0]
];
export var ListImageReferrersFilter$ = [3, n0, _LIRF,
    0,
    [_aTrt, _aS],
    [64 | 0, 0]
];
export var ListImageReferrersRequest$ = [3, n0, _LIRR,
    0,
    [_rN, _sI, _rI, _fi, _nT, _mR],
    [0, () => SubjectIdentifier$, 0, () => ListImageReferrersFilter$, 0, 1], 2
];
export var ListImageReferrersResponse$ = [3, n0, _LIRRi,
    0,
    [_ref, _nT],
    [() => ImageReferrerList, 0]
];
export var ListImagesFilter$ = [3, n0, _LIF,
    0,
    [_tS, _iS],
    [0, 0]
];
export var ListImagesRequest$ = [3, n0, _LIR,
    0,
    [_rN, _rI, _nT, _mR, _fi],
    [0, 0, 0, 1, () => ListImagesFilter$], 1
];
export var ListImagesResponse$ = [3, n0, _LIRi,
    0,
    [_iI, _nT],
    [() => ImageIdentifierList, 0]
];
export var ListPullTimeUpdateExclusionsRequest$ = [3, n0, _LPTUER,
    0,
    [_mR, _nT],
    [1, 0]
];
export var ListPullTimeUpdateExclusionsResponse$ = [3, n0, _LPTUERi,
    0,
    [_pTUE, _nT],
    [64 | 0, 0]
];
export var ListTagsForResourceRequest$ = [3, n0, _LTFRR,
    0,
    [_rA],
    [0], 1
];
export var ListTagsForResourceResponse$ = [3, n0, _LTFRRi,
    0,
    [_t],
    [() => TagList]
];
export var PackageVulnerabilityDetails$ = [3, n0, _PVD,
    0,
    [_cv, _rU, _rV, _so, _sU, _vCA, _vS, _vUA, _vI, _vP],
    [() => CvssScoreList, 64 | 0, 64 | 0, 0, 0, 4, 0, 4, 0, () => VulnerablePackagesList]
];
export var PullThroughCacheRule$ = [3, n0, _PTCR,
    0,
    [_eRP, _uRU, _cAr, _rI, _cA, _cRA, _uRP, _uR, _uA],
    [0, 0, 4, 0, 0, 0, 0, 0, 4]
];
export var PutAccountSettingRequest$ = [3, n0, _PASR,
    0,
    [_n, _v],
    [0, 0], 2
];
export var PutAccountSettingResponse$ = [3, n0, _PASRu,
    0,
    [_n, _v],
    [0, 0]
];
export var PutImageRequest$ = [3, n0, _PIR,
    0,
    [_rN, _iM, _rI, _iMMT, _iTm, _iDm],
    [0, 0, 0, 0, 0, 0], 2
];
export var PutImageResponse$ = [3, n0, _PIRu,
    0,
    [_im],
    [() => Image$]
];
export var PutImageScanningConfigurationRequest$ = [3, n0, _PISCR,
    0,
    [_rN, _iSC, _rI],
    [0, () => ImageScanningConfiguration$, 0], 2
];
export var PutImageScanningConfigurationResponse$ = [3, n0, _PISCRu,
    0,
    [_rI, _rN, _iSC],
    [0, 0, () => ImageScanningConfiguration$]
];
export var PutImageTagMutabilityRequest$ = [3, n0, _PITMR,
    0,
    [_rN, _iTM, _rI, _iTMEF],
    [0, 0, 0, () => ImageTagMutabilityExclusionFilters], 2
];
export var PutImageTagMutabilityResponse$ = [3, n0, _PITMRu,
    0,
    [_rI, _rN, _iTM, _iTMEF],
    [0, 0, 0, () => ImageTagMutabilityExclusionFilters]
];
export var PutLifecyclePolicyRequest$ = [3, n0, _PLPR,
    0,
    [_rN, _lPT, _rI],
    [0, 0, 0], 2
];
export var PutLifecyclePolicyResponse$ = [3, n0, _PLPRu,
    0,
    [_rI, _rN, _lPT],
    [0, 0, 0]
];
export var PutRegistryPolicyRequest$ = [3, n0, _PRPR,
    0,
    [_pT],
    [0], 1
];
export var PutRegistryPolicyResponse$ = [3, n0, _PRPRu,
    0,
    [_rI, _pT],
    [0, 0]
];
export var PutRegistryScanningConfigurationRequest$ = [3, n0, _PRSCR,
    0,
    [_sT, _ru],
    [0, () => RegistryScanningRuleList]
];
export var PutRegistryScanningConfigurationResponse$ = [3, n0, _PRSCRu,
    0,
    [_rSC],
    [() => RegistryScanningConfiguration$]
];
export var PutReplicationConfigurationRequest$ = [3, n0, _PRCR,
    0,
    [_rC],
    [() => ReplicationConfiguration$], 1
];
export var PutReplicationConfigurationResponse$ = [3, n0, _PRCRu,
    0,
    [_rC],
    [() => ReplicationConfiguration$]
];
export var PutSigningConfigurationRequest$ = [3, n0, _PSCR,
    0,
    [_sCi],
    [() => SigningConfiguration$], 1
];
export var PutSigningConfigurationResponse$ = [3, n0, _PSCRu,
    0,
    [_sCi],
    [() => SigningConfiguration$]
];
export var Recommendation$ = [3, n0, _R,
    0,
    [_ur, _te],
    [0, 0]
];
export var RegisterPullTimeUpdateExclusionRequest$ = [3, n0, _RPTUER,
    0,
    [_pAr],
    [0], 1
];
export var RegisterPullTimeUpdateExclusionResponse$ = [3, n0, _RPTUERe,
    0,
    [_pAr, _cAr],
    [0, 4]
];
export var RegistryScanningConfiguration$ = [3, n0, _RSC,
    0,
    [_sT, _ru],
    [0, () => RegistryScanningRuleList]
];
export var RegistryScanningRule$ = [3, n0, _RSR,
    0,
    [_sF, _rF],
    [0, () => ScanningRepositoryFilterList], 2
];
export var Remediation$ = [3, n0, _Re,
    0,
    [_rec],
    [() => Recommendation$]
];
export var ReplicationConfiguration$ = [3, n0, _RC,
    0,
    [_ru],
    [() => ReplicationRuleList], 1
];
export var ReplicationDestination$ = [3, n0, _RD,
    0,
    [_reg, _rI],
    [0, 0], 2
];
export var ReplicationRule$ = [3, n0, _RR,
    0,
    [_de, _rF],
    [() => ReplicationDestinationList, () => RepositoryFilterList], 1
];
export var Repository$ = [3, n0, _Rep,
    0,
    [_rAe, _rI, _rN, _rUe, _cAr, _iTM, _iTMEF, _iSC, _eC],
    [0, 0, 0, 0, 4, 0, () => ImageTagMutabilityExclusionFilters, () => ImageScanningConfiguration$, () => EncryptionConfiguration$]
];
export var RepositoryCreationTemplate$ = [3, n0, _RCT,
    0,
    [_pr, _d, _eC, _rT, _iTM, _iTMEF, _rP, _lP, _aF, _cRA, _cAr, _uA],
    [0, 0, () => EncryptionConfigurationForRepositoryCreationTemplate$, () => TagList, 0, () => ImageTagMutabilityExclusionFilters, 0, 0, 64 | 0, 0, 4, 4]
];
export var RepositoryFilter$ = [3, n0, _RF,
    0,
    [_fi, _fT],
    [0, 0], 2
];
export var RepositoryScanningConfiguration$ = [3, n0, _RSCe,
    0,
    [_rAe, _rN, _sOP, _sF, _aSF],
    [0, 0, 2, 0, () => ScanningRepositoryFilterList]
];
export var RepositoryScanningConfigurationFailure$ = [3, n0, _RSCF,
    0,
    [_rN, _fC, _fR],
    [0, 0, 0]
];
export var Resource$ = [3, n0, _Res,
    0,
    [_det, _id, _t, _ty],
    [() => ResourceDetails$, 0, 128 | 0, 0]
];
export var ResourceDetails$ = [3, n0, _RDe,
    0,
    [_aECI],
    [() => AwsEcrContainerImageDetails$]
];
export var ScanningRepositoryFilter$ = [3, n0, _SRF,
    0,
    [_fi, _fT],
    [0, 0], 2
];
export var ScoreDetails$ = [3, n0, _SD,
    0,
    [_cv],
    [() => CvssScoreDetails$]
];
export var SetRepositoryPolicyRequest$ = [3, n0, _SRPR,
    0,
    [_rN, _pT, _rI, _fo],
    [0, 0, 0, 2], 2
];
export var SetRepositoryPolicyResponse$ = [3, n0, _SRPRe,
    0,
    [_rI, _rN, _pT],
    [0, 0, 0]
];
export var SigningConfiguration$ = [3, n0, _SC,
    0,
    [_ru],
    [() => SigningRuleList], 1
];
export var SigningRepositoryFilter$ = [3, n0, _SRFi,
    0,
    [_fi, _fT],
    [0, 0], 2
];
export var SigningRule$ = [3, n0, _SR,
    0,
    [_sPA, _rF],
    [0, () => SigningRepositoryFilterList], 1
];
export var StartImageScanRequest$ = [3, n0, _SISR,
    0,
    [_rN, _iIm, _rI],
    [0, () => ImageIdentifier$, 0], 2
];
export var StartImageScanResponse$ = [3, n0, _SISRt,
    0,
    [_rI, _rN, _iIm, _iSS],
    [0, 0, () => ImageIdentifier$, () => ImageScanStatus$]
];
export var StartLifecyclePolicyPreviewRequest$ = [3, n0, _SLPPR,
    0,
    [_rN, _rI, _lPT],
    [0, 0, 0], 1
];
export var StartLifecyclePolicyPreviewResponse$ = [3, n0, _SLPPRt,
    0,
    [_rI, _rN, _lPT, _st],
    [0, 0, 0, 0]
];
export var SubjectIdentifier$ = [3, n0, _SI,
    0,
    [_iDm],
    [0], 1
];
export var Tag$ = [3, n0, _T,
    0,
    [_K, _V],
    [0, 0], 2
];
export var TagResourceRequest$ = [3, n0, _TRR,
    0,
    [_rA, _t],
    [0, () => TagList], 2
];
export var TagResourceResponse$ = [3, n0, _TRRa,
    0,
    [],
    []
];
export var TransitioningImageTotalCount$ = [3, n0, _TITC,
    0,
    [_tSC, _iTC],
    [0, 1]
];
export var UntagResourceRequest$ = [3, n0, _URR,
    0,
    [_rA, _tK],
    [0, 64 | 0], 2
];
export var UntagResourceResponse$ = [3, n0, _URRn,
    0,
    [],
    []
];
export var UpdateImageStorageClassRequest$ = [3, n0, _UISCR,
    0,
    [_rN, _iIm, _tSC, _rI],
    [0, () => ImageIdentifier$, 0, 0], 3
];
export var UpdateImageStorageClassResponse$ = [3, n0, _UISCRp,
    0,
    [_rI, _rN, _iIm, _iS],
    [0, 0, () => ImageIdentifier$, 0]
];
export var UpdatePullThroughCacheRuleRequest$ = [3, n0, _UPTCRR,
    0,
    [_eRP, _rI, _cA, _cRA],
    [0, 0, 0, 0], 1
];
export var UpdatePullThroughCacheRuleResponse$ = [3, n0, _UPTCRRp,
    0,
    [_eRP, _rI, _uA, _cA, _cRA, _uRP],
    [0, 0, 4, 0, 0, 0]
];
export var UpdateRepositoryCreationTemplateRequest$ = [3, n0, _URCTR,
    0,
    [_pr, _d, _eC, _rT, _iTM, _iTMEF, _rP, _lP, _aF, _cRA],
    [0, 0, () => EncryptionConfigurationForRepositoryCreationTemplate$, () => TagList, 0, () => ImageTagMutabilityExclusionFilters, 0, 0, 64 | 0, 0], 1
];
export var UpdateRepositoryCreationTemplateResponse$ = [3, n0, _URCTRp,
    0,
    [_rI, _rCT],
    [0, () => RepositoryCreationTemplate$]
];
export var UploadLayerPartRequest$ = [3, n0, _ULPR,
    0,
    [_rN, _uI, _pFB, _pLB, _lPB, _rI],
    [0, 0, 1, 1, 21, 0], 5
];
export var UploadLayerPartResponse$ = [3, n0, _ULPRp,
    0,
    [_rI, _rN, _uI, _lBR],
    [0, 0, 0, 1]
];
export var ValidatePullThroughCacheRuleRequest$ = [3, n0, _VPTCRR,
    0,
    [_eRP, _rI],
    [0, 0], 1
];
export var ValidatePullThroughCacheRuleResponse$ = [3, n0, _VPTCRRa,
    0,
    [_eRP, _rI, _uRU, _cA, _cRA, _uRP, _iV, _fa],
    [0, 0, 0, 0, 0, 0, 2, 0]
];
export var VulnerablePackage$ = [3, n0, _VP,
    0,
    [_ar, _ep, _fP, _n, _pM, _rel, _sLH, _ve, _fIV],
    [0, 1, 0, 0, 0, 0, 0, 0, 0]
];
var ArtifactTypeList = 64 | 0;
var AttributeList = [1, n0, _AL,
    0, () => Attribute$
];
var AuthorizationDataList = [1, n0, _ADL,
    0, () => AuthorizationData$
];
var BatchedOperationLayerDigestList = 64 | 0;
var CvssScoreAdjustmentList = [1, n0, _CSAL,
    0, () => CvssScoreAdjustment$
];
var CvssScoreList = [1, n0, _CSL,
    0, () => CvssScore$
];
var EnhancedImageScanFindingList = [1, n0, _EISFL,
    0, () => EnhancedImageScanFinding$
];
var GetAuthorizationTokenRegistryIdList = 64 | 0;
var ImageDetailList = [1, n0, _IDL,
    0, () => ImageDetail$
];
var ImageFailureList = [1, n0, _IFL,
    0, () => ImageFailure$
];
var ImageIdentifierList = [1, n0, _IIL,
    0, () => ImageIdentifier$
];
var ImageList = [1, n0, _IL,
    0, () => Image$
];
var ImageReferrerList = [1, n0, _IRL,
    0, () => ImageReferrer$
];
var ImageReplicationStatusList = [1, n0, _IRSL,
    0, () => ImageReplicationStatus$
];
var ImageScanFindingList = [1, n0, _ISFL,
    0, () => ImageScanFinding$
];
var ImageSigningStatusList = [1, n0, _ISSL,
    0, () => ImageSigningStatus$
];
var ImageTagList = 64 | 0;
var ImageTagMutabilityExclusionFilters = [1, n0, _ITMEFm,
    0, () => ImageTagMutabilityExclusionFilter$
];
var ImageTagsList = 64 | 0;
var LayerDigestList = 64 | 0;
var LayerFailureList = [1, n0, _LFL,
    0, () => LayerFailure$
];
var LayerList = [1, n0, _LL,
    0, () => Layer$
];
var LifecyclePolicyPreviewResultList = [1, n0, _LPPRL,
    0, () => LifecyclePolicyPreviewResult$
];
var MediaTypeList = 64 | 0;
var PrefixList = 64 | 0;
var PullThroughCacheRuleList = [1, n0, _PTCRL,
    0, () => PullThroughCacheRule$
];
var PullThroughCacheRuleRepositoryPrefixList = 64 | 0;
var PullTimeUpdateExclusionList = 64 | 0;
var RCTAppliedForList = 64 | 0;
var ReferenceUrlsList = 64 | 0;
var RegistryScanningRuleList = [1, n0, _RSRL,
    0, () => RegistryScanningRule$
];
var RelatedVulnerabilitiesList = 64 | 0;
var ReplicationDestinationList = [1, n0, _RDL,
    0, () => ReplicationDestination$
];
var ReplicationRuleList = [1, n0, _RRL,
    0, () => ReplicationRule$
];
var RepositoryCreationTemplateList = [1, n0, _RCTL,
    0, () => RepositoryCreationTemplate$
];
var RepositoryFilterList = [1, n0, _RFL,
    0, () => RepositoryFilter$
];
var RepositoryList = [1, n0, _RL,
    0, () => Repository$
];
var RepositoryNameList = 64 | 0;
var RepositoryScanningConfigurationFailureList = [1, n0, _RSCFL,
    0, () => RepositoryScanningConfigurationFailure$
];
var RepositoryScanningConfigurationList = [1, n0, _RSCL,
    0, () => RepositoryScanningConfiguration$
];
var ResourceList = [1, n0, _RLe,
    0, () => Resource$
];
var ScanningConfigurationRepositoryNameList = 64 | 0;
var ScanningRepositoryFilterList = [1, n0, _SRFL,
    0, () => ScanningRepositoryFilter$
];
var SigningRepositoryFilterList = [1, n0, _SRFLi,
    0, () => SigningRepositoryFilter$
];
var SigningRuleList = [1, n0, _SRL,
    0, () => SigningRule$
];
var TagKeyList = 64 | 0;
var TagList = [1, n0, _TL,
    0, () => Tag$
];
var TransitioningImageTotalCounts = [1, n0, _TITCr,
    0, () => TransitioningImageTotalCount$
];
var VulnerablePackagesList = [1, n0, _VPL,
    0, () => VulnerablePackage$
];
var Annotations = 128 | 0;
var FindingSeverityCounts = 128 | 1;
var Tags = 128 | 0;
export var BatchCheckLayerAvailability$ = [9, n0, _BCLA,
    0, () => BatchCheckLayerAvailabilityRequest$, () => BatchCheckLayerAvailabilityResponse$
];
export var BatchDeleteImage$ = [9, n0, _BDI,
    0, () => BatchDeleteImageRequest$, () => BatchDeleteImageResponse$
];
export var BatchGetImage$ = [9, n0, _BGI,
    0, () => BatchGetImageRequest$, () => BatchGetImageResponse$
];
export var BatchGetRepositoryScanningConfiguration$ = [9, n0, _BGRSC,
    0, () => BatchGetRepositoryScanningConfigurationRequest$, () => BatchGetRepositoryScanningConfigurationResponse$
];
export var CompleteLayerUpload$ = [9, n0, _CLU,
    0, () => CompleteLayerUploadRequest$, () => CompleteLayerUploadResponse$
];
export var CreatePullThroughCacheRule$ = [9, n0, _CPTCR,
    0, () => CreatePullThroughCacheRuleRequest$, () => CreatePullThroughCacheRuleResponse$
];
export var CreateRepository$ = [9, n0, _CR,
    0, () => CreateRepositoryRequest$, () => CreateRepositoryResponse$
];
export var CreateRepositoryCreationTemplate$ = [9, n0, _CRCT,
    0, () => CreateRepositoryCreationTemplateRequest$, () => CreateRepositoryCreationTemplateResponse$
];
export var DeleteLifecyclePolicy$ = [9, n0, _DLP,
    0, () => DeleteLifecyclePolicyRequest$, () => DeleteLifecyclePolicyResponse$
];
export var DeletePullThroughCacheRule$ = [9, n0, _DPTCR,
    0, () => DeletePullThroughCacheRuleRequest$, () => DeletePullThroughCacheRuleResponse$
];
export var DeleteRegistryPolicy$ = [9, n0, _DRP,
    0, () => DeleteRegistryPolicyRequest$, () => DeleteRegistryPolicyResponse$
];
export var DeleteRepository$ = [9, n0, _DR,
    0, () => DeleteRepositoryRequest$, () => DeleteRepositoryResponse$
];
export var DeleteRepositoryCreationTemplate$ = [9, n0, _DRCT,
    0, () => DeleteRepositoryCreationTemplateRequest$, () => DeleteRepositoryCreationTemplateResponse$
];
export var DeleteRepositoryPolicy$ = [9, n0, _DRPe,
    0, () => DeleteRepositoryPolicyRequest$, () => DeleteRepositoryPolicyResponse$
];
export var DeleteSigningConfiguration$ = [9, n0, _DSC,
    0, () => DeleteSigningConfigurationRequest$, () => DeleteSigningConfigurationResponse$
];
export var DeregisterPullTimeUpdateExclusion$ = [9, n0, _DPTUE,
    0, () => DeregisterPullTimeUpdateExclusionRequest$, () => DeregisterPullTimeUpdateExclusionResponse$
];
export var DescribeImageReplicationStatus$ = [9, n0, _DIRS,
    0, () => DescribeImageReplicationStatusRequest$, () => DescribeImageReplicationStatusResponse$
];
export var DescribeImages$ = [9, n0, _DI,
    0, () => DescribeImagesRequest$, () => DescribeImagesResponse$
];
export var DescribeImageScanFindings$ = [9, n0, _DISF,
    0, () => DescribeImageScanFindingsRequest$, () => DescribeImageScanFindingsResponse$
];
export var DescribeImageSigningStatus$ = [9, n0, _DISS,
    0, () => DescribeImageSigningStatusRequest$, () => DescribeImageSigningStatusResponse$
];
export var DescribePullThroughCacheRules$ = [9, n0, _DPTCRe,
    0, () => DescribePullThroughCacheRulesRequest$, () => DescribePullThroughCacheRulesResponse$
];
export var DescribeRegistry$ = [9, n0, _DRe,
    0, () => DescribeRegistryRequest$, () => DescribeRegistryResponse$
];
export var DescribeRepositories$ = [9, n0, _DRes,
    0, () => DescribeRepositoriesRequest$, () => DescribeRepositoriesResponse$
];
export var DescribeRepositoryCreationTemplates$ = [9, n0, _DRCTe,
    0, () => DescribeRepositoryCreationTemplatesRequest$, () => DescribeRepositoryCreationTemplatesResponse$
];
export var GetAccountSetting$ = [9, n0, _GAS,
    0, () => GetAccountSettingRequest$, () => GetAccountSettingResponse$
];
export var GetAuthorizationToken$ = [9, n0, _GAT,
    0, () => GetAuthorizationTokenRequest$, () => GetAuthorizationTokenResponse$
];
export var GetDownloadUrlForLayer$ = [9, n0, _GDUFL,
    0, () => GetDownloadUrlForLayerRequest$, () => GetDownloadUrlForLayerResponse$
];
export var GetLifecyclePolicy$ = [9, n0, _GLP,
    0, () => GetLifecyclePolicyRequest$, () => GetLifecyclePolicyResponse$
];
export var GetLifecyclePolicyPreview$ = [9, n0, _GLPP,
    0, () => GetLifecyclePolicyPreviewRequest$, () => GetLifecyclePolicyPreviewResponse$
];
export var GetRegistryPolicy$ = [9, n0, _GRP,
    0, () => GetRegistryPolicyRequest$, () => GetRegistryPolicyResponse$
];
export var GetRegistryScanningConfiguration$ = [9, n0, _GRSC,
    0, () => GetRegistryScanningConfigurationRequest$, () => GetRegistryScanningConfigurationResponse$
];
export var GetRepositoryPolicy$ = [9, n0, _GRPe,
    0, () => GetRepositoryPolicyRequest$, () => GetRepositoryPolicyResponse$
];
export var GetSigningConfiguration$ = [9, n0, _GSC,
    0, () => GetSigningConfigurationRequest$, () => GetSigningConfigurationResponse$
];
export var InitiateLayerUpload$ = [9, n0, _ILU,
    0, () => InitiateLayerUploadRequest$, () => InitiateLayerUploadResponse$
];
export var ListImageReferrers$ = [9, n0, _LIRis,
    0, () => ListImageReferrersRequest$, () => ListImageReferrersResponse$
];
export var ListImages$ = [9, n0, _LI,
    0, () => ListImagesRequest$, () => ListImagesResponse$
];
export var ListPullTimeUpdateExclusions$ = [9, n0, _LPTUE,
    0, () => ListPullTimeUpdateExclusionsRequest$, () => ListPullTimeUpdateExclusionsResponse$
];
export var ListTagsForResource$ = [9, n0, _LTFR,
    0, () => ListTagsForResourceRequest$, () => ListTagsForResourceResponse$
];
export var PutAccountSetting$ = [9, n0, _PAS,
    0, () => PutAccountSettingRequest$, () => PutAccountSettingResponse$
];
export var PutImage$ = [9, n0, _PI,
    0, () => PutImageRequest$, () => PutImageResponse$
];
export var PutImageScanningConfiguration$ = [9, n0, _PISC,
    0, () => PutImageScanningConfigurationRequest$, () => PutImageScanningConfigurationResponse$
];
export var PutImageTagMutability$ = [9, n0, _PITM,
    0, () => PutImageTagMutabilityRequest$, () => PutImageTagMutabilityResponse$
];
export var PutLifecyclePolicy$ = [9, n0, _PLP,
    0, () => PutLifecyclePolicyRequest$, () => PutLifecyclePolicyResponse$
];
export var PutRegistryPolicy$ = [9, n0, _PRP,
    0, () => PutRegistryPolicyRequest$, () => PutRegistryPolicyResponse$
];
export var PutRegistryScanningConfiguration$ = [9, n0, _PRSC,
    0, () => PutRegistryScanningConfigurationRequest$, () => PutRegistryScanningConfigurationResponse$
];
export var PutReplicationConfiguration$ = [9, n0, _PRC,
    0, () => PutReplicationConfigurationRequest$, () => PutReplicationConfigurationResponse$
];
export var PutSigningConfiguration$ = [9, n0, _PSC,
    0, () => PutSigningConfigurationRequest$, () => PutSigningConfigurationResponse$
];
export var RegisterPullTimeUpdateExclusion$ = [9, n0, _RPTUE,
    0, () => RegisterPullTimeUpdateExclusionRequest$, () => RegisterPullTimeUpdateExclusionResponse$
];
export var SetRepositoryPolicy$ = [9, n0, _SRP,
    0, () => SetRepositoryPolicyRequest$, () => SetRepositoryPolicyResponse$
];
export var StartImageScan$ = [9, n0, _SIS,
    0, () => StartImageScanRequest$, () => StartImageScanResponse$
];
export var StartLifecyclePolicyPreview$ = [9, n0, _SLPP,
    0, () => StartLifecyclePolicyPreviewRequest$, () => StartLifecyclePolicyPreviewResponse$
];
export var TagResource$ = [9, n0, _TR,
    0, () => TagResourceRequest$, () => TagResourceResponse$
];
export var UntagResource$ = [9, n0, _UR,
    0, () => UntagResourceRequest$, () => UntagResourceResponse$
];
export var UpdateImageStorageClass$ = [9, n0, _UISC,
    0, () => UpdateImageStorageClassRequest$, () => UpdateImageStorageClassResponse$
];
export var UpdatePullThroughCacheRule$ = [9, n0, _UPTCR,
    0, () => UpdatePullThroughCacheRuleRequest$, () => UpdatePullThroughCacheRuleResponse$
];
export var UpdateRepositoryCreationTemplate$ = [9, n0, _URCT,
    0, () => UpdateRepositoryCreationTemplateRequest$, () => UpdateRepositoryCreationTemplateResponse$
];
export var UploadLayerPart$ = [9, n0, _ULP,
    0, () => UploadLayerPartRequest$, () => UploadLayerPartResponse$
];
export var ValidatePullThroughCacheRule$ = [9, n0, _VPTCR,
    0, () => ValidatePullThroughCacheRuleRequest$, () => ValidatePullThroughCacheRuleResponse$
];
