// renovate/dist/constants/error-messages.js
//#region lib/constants/error-messages.ts
const SYSTEM_INSUFFICIENT_DISK_SPACE = "disk-space";
const SYSTEM_INSUFFICIENT_MEMORY = "out-of-memory";
const SystemErrors = [SYSTEM_INSUFFICIENT_DISK_SPACE, SYSTEM_INSUFFICIENT_MEMORY];
const PLATFORM_AUTHENTICATION_ERROR = "authentication-error";
const PLATFORM_BAD_CREDENTIALS = "bad-credentials";
const PLATFORM_GPG_FAILED = "gpg-failed";
const PLATFORM_INTEGRATION_UNAUTHORIZED = "integration-unauthorized";
const PLATFORM_NOT_FOUND = "platform-not-found";
const PLATFORM_RATE_LIMIT_EXCEEDED = "rate-limit-exceeded";
const PLATFORM_UNKNOWN_ERROR = "platform-unknown-error";
const PlatformErrors = [
	PLATFORM_AUTHENTICATION_ERROR,
	PLATFORM_BAD_CREDENTIALS,
	PLATFORM_GPG_FAILED,
	PLATFORM_INTEGRATION_UNAUTHORIZED,
	PLATFORM_NOT_FOUND,
	PLATFORM_RATE_LIMIT_EXCEEDED,
	PLATFORM_UNKNOWN_ERROR
];
const CONFIG_VALIDATION = "config-validation";
const CONFIG_PRESETS_INVALID = "config-presets-invalid";
const CONFIG_SECRETS_EXPOSED = "config-secrets-exposed";
const CONFIG_SECRETS_INVALID = "config-secrets-invalid";
const CONFIG_VARIABLES_INVALID = "config-variables-invalid";
const CONFIG_GIT_URL_UNAVAILABLE = "config-git-url-unavailable";
const CONFIG_INHERIT_NOT_FOUND = "config-inherit-not-found";
const CONFIG_INHERIT_PARSE_ERROR = "config-inherit-parse-error";
const ConfigErrors = [
	CONFIG_VALIDATION,
	CONFIG_PRESETS_INVALID,
	CONFIG_SECRETS_EXPOSED,
	CONFIG_SECRETS_INVALID,
	CONFIG_VARIABLES_INVALID,
	CONFIG_GIT_URL_UNAVAILABLE,
	CONFIG_INHERIT_NOT_FOUND,
	CONFIG_INHERIT_PARSE_ERROR
];
const REPOSITORY_ACCESS_FORBIDDEN = "forbidden";
const REPOSITORY_ARCHIVED = "archived";
const REPOSITORY_BLOCKED = "blocked";
const REPOSITORY_CANNOT_FORK = "cannot-fork";
const REPOSITORY_DISABLED = "disabled";
const REPOSITORY_CLOSED_ONBOARDING = "disabled-closed-onboarding";
const REPOSITORY_DISABLED_BY_CONFIG = "disabled-by-config";
const REPOSITORY_NO_CONFIG = "disabled-no-config";
const REPOSITORY_EMPTY = "empty";
const REPOSITORY_FORK_MISSING = "fork-missing";
const REPOSITORY_FORK_MODE_FORKED = "fork-mode-forked";
const REPOSITORY_FORKED = "fork";
const REPOSITORY_MIRRORED = "mirror";
const REPOSITORY_NOT_FOUND = "not-found";
const REPOSITORY_NO_PACKAGE_FILES = "no-package-files";
const REPOSITORY_RENAMED = "renamed";
const REPOSITORY_UNINITIATED = "uninitiated";
/** cause repo to be considered as disabled */
const RepositoryErrors = [
	REPOSITORY_ACCESS_FORBIDDEN,
	REPOSITORY_ARCHIVED,
	REPOSITORY_BLOCKED,
	REPOSITORY_CANNOT_FORK,
	REPOSITORY_DISABLED,
	REPOSITORY_CLOSED_ONBOARDING,
	REPOSITORY_DISABLED_BY_CONFIG,
	REPOSITORY_NO_CONFIG,
	REPOSITORY_EMPTY,
	REPOSITORY_FORK_MISSING,
	REPOSITORY_FORK_MODE_FORKED,
	REPOSITORY_FORKED,
	REPOSITORY_MIRRORED,
	REPOSITORY_NOT_FOUND,
	REPOSITORY_NO_PACKAGE_FILES,
	REPOSITORY_RENAMED,
	REPOSITORY_UNINITIATED
];
const REPOSITORY_CHANGED = "repository-changed";
const TEMPORARY_ERROR = "temporary-error";
const NO_VULNERABILITY_ALERTS = "no-vulnerability-alerts";
const TemporaryErrors = [
	REPOSITORY_CHANGED,
	TEMPORARY_ERROR,
	NO_VULNERABILITY_ALERTS
];
const MANAGER_LOCKFILE_ERROR = "lockfile-error";
const FILE_ACCESS_VIOLATION_ERROR = "file-access-violation-error";
const EXTERNAL_HOST_ERROR = "external-host-error";
const IGNORABLE_HOST_ERROR = "ignorable-host-error";
const HOST_DISABLED = "host-disabled";
const WORKER_FILE_UPDATE_FAILED = "update-failure";
const BUNDLER_INVALID_CREDENTIALS = "bundler-credentials";
const UNKNOWN_ERROR = "unknown-error";
const INVALID_PATH = "invalid-path";
const PAGE_NOT_FOUND_ERROR = "page-not-found";
const MISSING_API_CREDENTIALS = "missing-api-credentials";
/** Logger not initialized */
const LOGGER_NOT_INITIALIZED = "logger-not-initialized";
//#endregion
export { BUNDLER_INVALID_CREDENTIALS, CONFIG_GIT_URL_UNAVAILABLE, CONFIG_INHERIT_NOT_FOUND, CONFIG_INHERIT_PARSE_ERROR, CONFIG_PRESETS_INVALID, CONFIG_SECRETS_EXPOSED, CONFIG_SECRETS_INVALID, CONFIG_VALIDATION, CONFIG_VARIABLES_INVALID, ConfigErrors, EXTERNAL_HOST_ERROR, FILE_ACCESS_VIOLATION_ERROR, HOST_DISABLED, IGNORABLE_HOST_ERROR, INVALID_PATH, LOGGER_NOT_INITIALIZED, MANAGER_LOCKFILE_ERROR, MISSING_API_CREDENTIALS, NO_VULNERABILITY_ALERTS, PAGE_NOT_FOUND_ERROR, PLATFORM_AUTHENTICATION_ERROR, PLATFORM_BAD_CREDENTIALS, PLATFORM_GPG_FAILED, PLATFORM_INTEGRATION_UNAUTHORIZED, PLATFORM_NOT_FOUND, PLATFORM_RATE_LIMIT_EXCEEDED, PLATFORM_UNKNOWN_ERROR, PlatformErrors, REPOSITORY_ACCESS_FORBIDDEN, REPOSITORY_ARCHIVED, REPOSITORY_BLOCKED, REPOSITORY_CANNOT_FORK, REPOSITORY_CHANGED, REPOSITORY_CLOSED_ONBOARDING, REPOSITORY_DISABLED, REPOSITORY_DISABLED_BY_CONFIG, REPOSITORY_EMPTY, REPOSITORY_FORKED, REPOSITORY_FORK_MISSING, REPOSITORY_FORK_MODE_FORKED, REPOSITORY_MIRRORED, REPOSITORY_NOT_FOUND, REPOSITORY_NO_CONFIG, REPOSITORY_NO_PACKAGE_FILES, REPOSITORY_RENAMED, REPOSITORY_UNINITIATED, RepositoryErrors, SYSTEM_INSUFFICIENT_DISK_SPACE, SYSTEM_INSUFFICIENT_MEMORY, SystemErrors, TEMPORARY_ERROR, TemporaryErrors, UNKNOWN_ERROR, WORKER_FILE_UPDATE_FAILED };
