// @aws-sdk/client-ecr/dist-es/models/enums.js
export const LayerFailureCode = {
    InvalidLayerDigest: "InvalidLayerDigest",
    MissingLayerDigest: "MissingLayerDigest",
};
export const LayerAvailability = {
    ARCHIVED: "ARCHIVED",
    AVAILABLE: "AVAILABLE",
    UNAVAILABLE: "UNAVAILABLE",
};
export const ImageFailureCode = {
    ImageInaccessible: "ImageInaccessible",
    ImageNotFound: "ImageNotFound",
    ImageReferencedByManifestList: "ImageReferencedByManifestList",
    ImageTagDoesNotMatchDigest: "ImageTagDoesNotMatchDigest",
    InvalidImageDigest: "InvalidImageDigest",
    InvalidImageTag: "InvalidImageTag",
    KmsError: "KmsError",
    MissingDigestAndTag: "MissingDigestAndTag",
    UpstreamAccessDenied: "UpstreamAccessDenied",
    UpstreamTooManyRequests: "UpstreamTooManyRequests",
    UpstreamUnavailable: "UpstreamUnavailable",
};
export const ScanningConfigurationFailureCode = {
    REPOSITORY_NOT_FOUND: "REPOSITORY_NOT_FOUND",
};
export const ScanningRepositoryFilterType = {
    WILDCARD: "WILDCARD",
};
export const ScanFrequency = {
    CONTINUOUS_SCAN: "CONTINUOUS_SCAN",
    MANUAL: "MANUAL",
    SCAN_ON_PUSH: "SCAN_ON_PUSH",
};
export const UpstreamRegistry = {
    AzureContainerRegistry: "azure-container-registry",
    Chainguard: "chainguard",
    DockerHub: "docker-hub",
    Ecr: "ecr",
    EcrPublic: "ecr-public",
    GitHubContainerRegistry: "github-container-registry",
    GitLabContainerRegistry: "gitlab-container-registry",
    K8s: "k8s",
    Quay: "quay",
};
export const EncryptionType = {
    AES256: "AES256",
    KMS: "KMS",
    KMS_DSSE: "KMS_DSSE",
};
export const ImageTagMutability = {
    IMMUTABLE: "IMMUTABLE",
    IMMUTABLE_WITH_EXCLUSION: "IMMUTABLE_WITH_EXCLUSION",
    MUTABLE: "MUTABLE",
    MUTABLE_WITH_EXCLUSION: "MUTABLE_WITH_EXCLUSION",
};
export const ImageTagMutabilityExclusionFilterType = {
    WILDCARD: "WILDCARD",
};
export const RCTAppliedFor = {
    CREATE_ON_PUSH: "CREATE_ON_PUSH",
    PULL_THROUGH_CACHE: "PULL_THROUGH_CACHE",
    REPLICATION: "REPLICATION",
};
export const SigningRepositoryFilterType = {
    WILDCARD_MATCH: "WILDCARD_MATCH",
};
export const ReplicationStatus = {
    COMPLETE: "COMPLETE",
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
};
export const ImageStatusFilter = {
    ACTIVATING: "ACTIVATING",
    ACTIVE: "ACTIVE",
    ANY: "ANY",
    ARCHIVED: "ARCHIVED",
};
export const TagStatus = {
    ANY: "ANY",
    TAGGED: "TAGGED",
    UNTAGGED: "UNTAGGED",
};
export const FindingSeverity = {
    CRITICAL: "CRITICAL",
    HIGH: "HIGH",
    INFORMATIONAL: "INFORMATIONAL",
    LOW: "LOW",
    MEDIUM: "MEDIUM",
    UNDEFINED: "UNDEFINED",
};
export const ScanStatus = {
    ACTIVE: "ACTIVE",
    COMPLETE: "COMPLETE",
    FAILED: "FAILED",
    FINDINGS_UNAVAILABLE: "FINDINGS_UNAVAILABLE",
    IMAGE_ARCHIVED: "IMAGE_ARCHIVED",
    IN_PROGRESS: "IN_PROGRESS",
    LIMIT_EXCEEDED: "LIMIT_EXCEEDED",
    PENDING: "PENDING",
    SCAN_ELIGIBILITY_EXPIRED: "SCAN_ELIGIBILITY_EXPIRED",
    UNSUPPORTED_IMAGE: "UNSUPPORTED_IMAGE",
};
export const ImageStatus = {
    ACTIVATING: "ACTIVATING",
    ACTIVE: "ACTIVE",
    ARCHIVED: "ARCHIVED",
};
export const SigningStatus = {
    COMPLETE: "COMPLETE",
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
};
export const RepositoryFilterType = {
    PREFIX_MATCH: "PREFIX_MATCH",
};
export const LifecyclePolicyTargetStorageClass = {
    ARCHIVE: "ARCHIVE",
};
export const ImageActionType = {
    EXPIRE: "EXPIRE",
    TRANSITION: "TRANSITION",
};
export const LifecyclePolicyStorageClass = {
    ARCHIVE: "ARCHIVE",
    STANDARD: "STANDARD",
};
export const LifecyclePolicyPreviewStatus = {
    COMPLETE: "COMPLETE",
    EXPIRED: "EXPIRED",
    FAILED: "FAILED",
    IN_PROGRESS: "IN_PROGRESS",
};
export const ScanType = {
    BASIC: "BASIC",
    ENHANCED: "ENHANCED",
};
export const ArtifactStatusFilter = {
    ACTIVATING: "ACTIVATING",
    ACTIVE: "ACTIVE",
    ANY: "ANY",
    ARCHIVED: "ARCHIVED",
};
export const ArtifactStatus = {
    ACTIVATING: "ACTIVATING",
    ACTIVE: "ACTIVE",
    ARCHIVED: "ARCHIVED",
};
export const TargetStorageClass = {
    ARCHIVE: "ARCHIVE",
    STANDARD: "STANDARD",
};
