// @aws-sdk/client-eks/dist-cjs/schemas/schemas_0.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CapabilityConfigurationResponse$ = exports.CapabilityConfigurationRequest$ = exports.Capability$ = exports.BlockStorage$ = exports.AutoScalingGroup$ = exports.AssociateIdentityProviderConfigResponse$ = exports.AssociateIdentityProviderConfigRequest$ = exports.AssociateEncryptionConfigResponse$ = exports.AssociateEncryptionConfigRequest$ = exports.AssociatedAccessPolicy$ = exports.AssociateAccessPolicyResponse$ = exports.AssociateAccessPolicyRequest$ = exports.ArgoCdRoleMapping$ = exports.ArgoCdNetworkAccessConfigResponse$ = exports.ArgoCdNetworkAccessConfigRequest$ = exports.ArgoCdConfigResponse$ = exports.ArgoCdConfigRequest$ = exports.ArgoCdAwsIdcConfigResponse$ = exports.ArgoCdAwsIdcConfigRequest$ = exports.AddonVersionInfo$ = exports.AddonPodIdentityConfiguration$ = exports.AddonPodIdentityAssociations$ = exports.AddonNamespaceConfigResponse$ = exports.AddonNamespaceConfigRequest$ = exports.AddonIssue$ = exports.AddonInfo$ = exports.AddonHealth$ = exports.AddonCompatibilityDetail$ = exports.Addon$ = exports.AccessScope$ = exports.AccessPolicy$ = exports.AccessEntry$ = exports.AccessConfigResponse$ = exports.errorTypeRegistries = exports.UnsupportedAvailabilityZoneException$ = exports.ThrottlingException$ = exports.ServiceUnavailableException$ = exports.ServerException$ = exports.ResourcePropagationDelayException$ = exports.ResourceNotFoundException$ = exports.ResourceLimitExceededException$ = exports.ResourceInUseException$ = exports.NotFoundException$ = exports.InvalidStateException$ = exports.InvalidRequestException$ = exports.InvalidParameterException$ = exports.ClientException$ = exports.BadRequestException$ = exports.AccessDeniedException$ = exports.EKSServiceException$ = void 0;
exports.DeletePodIdentityAssociationResponse$ = exports.DeletePodIdentityAssociationRequest$ = exports.DeleteNodegroupResponse$ = exports.DeleteNodegroupRequest$ = exports.DeleteFargateProfileResponse$ = exports.DeleteFargateProfileRequest$ = exports.DeleteEksAnywhereSubscriptionResponse$ = exports.DeleteEksAnywhereSubscriptionRequest$ = exports.DeleteClusterResponse$ = exports.DeleteClusterRequest$ = exports.DeleteCapabilityResponse$ = exports.DeleteCapabilityRequest$ = exports.DeleteAddonResponse$ = exports.DeleteAddonRequest$ = exports.DeleteAccessEntryResponse$ = exports.DeleteAccessEntryRequest$ = exports.CreatePodIdentityAssociationResponse$ = exports.CreatePodIdentityAssociationRequest$ = exports.CreateNodegroupResponse$ = exports.CreateNodegroupRequest$ = exports.CreateFargateProfileResponse$ = exports.CreateFargateProfileRequest$ = exports.CreateEksAnywhereSubscriptionResponse$ = exports.CreateEksAnywhereSubscriptionRequest$ = exports.CreateClusterResponse$ = exports.CreateClusterRequest$ = exports.CreateCapabilityResponse$ = exports.CreateCapabilityRequest$ = exports.CreateAddonResponse$ = exports.CreateAddonRequest$ = exports.CreateAccessEntryResponse$ = exports.CreateAccessEntryRequest$ = exports.CreateAccessConfigRequest$ = exports.ControlPlaneScalingConfig$ = exports.ControlPlanePlacementResponse$ = exports.ControlPlanePlacementRequest$ = exports.ConnectorConfigResponse$ = exports.ConnectorConfigRequest$ = exports.ComputeConfigResponse$ = exports.ComputeConfigRequest$ = exports.Compatibility$ = exports.ClusterVersionInformation$ = exports.ClusterIssue$ = exports.ClusterHealth$ = exports.Cluster$ = exports.ClientStat$ = exports.Certificate$ = exports.CapabilitySummary$ = exports.CapabilityIssue$ = exports.CapabilityHealth$ = void 0;
exports.Insight$ = exports.IdentityProviderConfigResponse$ = exports.IdentityProviderConfig$ = exports.Identity$ = exports.FargateProfileSelector$ = exports.FargateProfileIssue$ = exports.FargateProfileHealth$ = exports.FargateProfile$ = exports.ErrorDetail$ = exports.EncryptionConfig$ = exports.ElasticLoadBalancing$ = exports.EksAnywhereSubscriptionTerm$ = exports.EksAnywhereSubscription$ = exports.DisassociateIdentityProviderConfigResponse$ = exports.DisassociateIdentityProviderConfigRequest$ = exports.DisassociateAccessPolicyResponse$ = exports.DisassociateAccessPolicyRequest$ = exports.DescribeUpdateResponse$ = exports.DescribeUpdateRequest$ = exports.DescribePodIdentityAssociationResponse$ = exports.DescribePodIdentityAssociationRequest$ = exports.DescribeNodegroupResponse$ = exports.DescribeNodegroupRequest$ = exports.DescribeInsightsRefreshResponse$ = exports.DescribeInsightsRefreshRequest$ = exports.DescribeInsightResponse$ = exports.DescribeInsightRequest$ = exports.DescribeIdentityProviderConfigResponse$ = exports.DescribeIdentityProviderConfigRequest$ = exports.DescribeFargateProfileResponse$ = exports.DescribeFargateProfileRequest$ = exports.DescribeEksAnywhereSubscriptionResponse$ = exports.DescribeEksAnywhereSubscriptionRequest$ = exports.DescribeClusterVersionsResponse$ = exports.DescribeClusterVersionsRequest$ = exports.DescribeClusterResponse$ = exports.DescribeClusterRequest$ = exports.DescribeCapabilityResponse$ = exports.DescribeCapabilityRequest$ = exports.DescribeAddonVersionsResponse$ = exports.DescribeAddonVersionsRequest$ = exports.DescribeAddonResponse$ = exports.DescribeAddonRequest$ = exports.DescribeAddonConfigurationResponse$ = exports.DescribeAddonConfigurationRequest$ = exports.DescribeAccessEntryResponse$ = exports.DescribeAccessEntryRequest$ = exports.DeregisterClusterResponse$ = exports.DeregisterClusterRequest$ = exports.DeprecationDetail$ = void 0;
exports.OidcIdentityProviderConfig$ = exports.OIDC$ = exports.NodeRepairConfigOverrides$ = exports.NodeRepairConfig$ = exports.NodegroupUpdateConfig$ = exports.NodegroupScalingConfig$ = exports.NodegroupResources$ = exports.NodegroupHealth$ = exports.Nodegroup$ = exports.MarketplaceInformation$ = exports.LogSetup$ = exports.Logging$ = exports.ListUpdatesResponse$ = exports.ListUpdatesRequest$ = exports.ListTagsForResourceResponse$ = exports.ListTagsForResourceRequest$ = exports.ListPodIdentityAssociationsResponse$ = exports.ListPodIdentityAssociationsRequest$ = exports.ListNodegroupsResponse$ = exports.ListNodegroupsRequest$ = exports.ListInsightsResponse$ = exports.ListInsightsRequest$ = exports.ListIdentityProviderConfigsResponse$ = exports.ListIdentityProviderConfigsRequest$ = exports.ListFargateProfilesResponse$ = exports.ListFargateProfilesRequest$ = exports.ListEksAnywhereSubscriptionsResponse$ = exports.ListEksAnywhereSubscriptionsRequest$ = exports.ListClustersResponse$ = exports.ListClustersRequest$ = exports.ListCapabilitiesResponse$ = exports.ListCapabilitiesRequest$ = exports.ListAssociatedAccessPoliciesResponse$ = exports.ListAssociatedAccessPoliciesRequest$ = exports.ListAddonsResponse$ = exports.ListAddonsRequest$ = exports.ListAccessPoliciesResponse$ = exports.ListAccessPoliciesRequest$ = exports.ListAccessEntriesResponse$ = exports.ListAccessEntriesRequest$ = exports.License$ = exports.LaunchTemplateSpecification$ = exports.KubernetesNetworkConfigResponse$ = exports.KubernetesNetworkConfigRequest$ = exports.Issue$ = exports.InsightSummary$ = exports.InsightStatus$ = exports.InsightsFilter$ = exports.InsightResourceDetail$ = exports.InsightCategorySpecificSummary$ = void 0;
exports.UpgradePolicyRequest$ = exports.UpdateTaintsPayload$ = exports.UpdateRoleMappings$ = exports.UpdatePodIdentityAssociationResponse$ = exports.UpdatePodIdentityAssociationRequest$ = exports.UpdateParam$ = exports.UpdateNodegroupVersionResponse$ = exports.UpdateNodegroupVersionRequest$ = exports.UpdateNodegroupConfigResponse$ = exports.UpdateNodegroupConfigRequest$ = exports.UpdateLabelsPayload$ = exports.UpdateEksAnywhereSubscriptionResponse$ = exports.UpdateEksAnywhereSubscriptionRequest$ = exports.UpdateClusterVersionResponse$ = exports.UpdateClusterVersionRequest$ = exports.UpdateClusterConfigResponse$ = exports.UpdateClusterConfigRequest$ = exports.UpdateCapabilityResponse$ = exports.UpdateCapabilityRequest$ = exports.UpdateCapabilityConfiguration$ = exports.UpdateArgoCdConfig$ = exports.UpdateAddonResponse$ = exports.UpdateAddonRequest$ = exports.UpdateAccessEntryResponse$ = exports.UpdateAccessEntryRequest$ = exports.UpdateAccessConfigRequest$ = exports.Update$ = exports.UntagResourceResponse$ = exports.UntagResourceRequest$ = exports.Taint$ = exports.TagResourceResponse$ = exports.TagResourceRequest$ = exports.StorageConfigResponse$ = exports.StorageConfigRequest$ = exports.StartInsightsRefreshResponse$ = exports.StartInsightsRefreshRequest$ = exports.SsoIdentity$ = exports.RemotePodNetwork$ = exports.RemoteNodeNetwork$ = exports.RemoteNetworkConfigResponse$ = exports.RemoteNetworkConfigRequest$ = exports.RemoteAccessConfig$ = exports.RegisterClusterResponse$ = exports.RegisterClusterRequest$ = exports.Provider$ = exports.PodIdentityAssociationSummary$ = exports.PodIdentityAssociation$ = exports.OutpostConfigResponse$ = exports.OutpostConfigRequest$ = exports.OidcIdentityProviderConfigRequest$ = void 0;
exports.ListEksAnywhereSubscriptions$ = exports.ListClusters$ = exports.ListCapabilities$ = exports.ListAssociatedAccessPolicies$ = exports.ListAddons$ = exports.ListAccessPolicies$ = exports.ListAccessEntries$ = exports.DisassociateIdentityProviderConfig$ = exports.DisassociateAccessPolicy$ = exports.DescribeUpdate$ = exports.DescribePodIdentityAssociation$ = exports.DescribeNodegroup$ = exports.DescribeInsightsRefresh$ = exports.DescribeInsight$ = exports.DescribeIdentityProviderConfig$ = exports.DescribeFargateProfile$ = exports.DescribeEksAnywhereSubscription$ = exports.DescribeClusterVersions$ = exports.DescribeCluster$ = exports.DescribeCapability$ = exports.DescribeAddonVersions$ = exports.DescribeAddonConfiguration$ = exports.DescribeAddon$ = exports.DescribeAccessEntry$ = exports.DeregisterCluster$ = exports.DeletePodIdentityAssociation$ = exports.DeleteNodegroup$ = exports.DeleteFargateProfile$ = exports.DeleteEksAnywhereSubscription$ = exports.DeleteCluster$ = exports.DeleteCapability$ = exports.DeleteAddon$ = exports.DeleteAccessEntry$ = exports.CreatePodIdentityAssociation$ = exports.CreateNodegroup$ = exports.CreateFargateProfile$ = exports.CreateEksAnywhereSubscription$ = exports.CreateCluster$ = exports.CreateCapability$ = exports.CreateAddon$ = exports.CreateAccessEntry$ = exports.AssociateIdentityProviderConfig$ = exports.AssociateEncryptionConfig$ = exports.AssociateAccessPolicy$ = exports.ZonalShiftConfigResponse$ = exports.ZonalShiftConfigRequest$ = exports.WarmPoolConfig$ = exports.VpcConfigResponse$ = exports.VpcConfigRequest$ = exports.UpgradePolicyResponse$ = void 0;
exports.UpdatePodIdentityAssociation$ = exports.UpdateNodegroupVersion$ = exports.UpdateNodegroupConfig$ = exports.UpdateEksAnywhereSubscription$ = exports.UpdateClusterVersion$ = exports.UpdateClusterConfig$ = exports.UpdateCapability$ = exports.UpdateAddon$ = exports.UpdateAccessEntry$ = exports.UntagResource$ = exports.TagResource$ = exports.StartInsightsRefresh$ = exports.RegisterCluster$ = exports.ListUpdates$ = exports.ListTagsForResource$ = exports.ListPodIdentityAssociations$ = exports.ListNodegroups$ = exports.ListInsights$ = exports.ListIdentityProviderConfigs$ = exports.ListFargateProfiles$ = void 0;
const _A = "Addon";
const _AAP = "AssociatedAccessPolicy";
const _AAPL = "AssociatedAccessPoliciesList";
const _AAPR = "AssociateAccessPolicyRequest";
const _AAPRs = "AssociateAccessPolicyResponse";
const _AAPs = "AssociateAccessPolicy";
const _ACAICR = "ArgoCdAwsIdcConfigRequest";
const _ACAICRr = "ArgoCdAwsIdcConfigResponse";
const _ACCR = "ArgoCdConfigRequest";
const _ACCRr = "ArgoCdConfigResponse";
const _ACD = "AddonCompatibilityDetail";
const _ACDd = "AddonCompatibilityDetails";
const _ACNACR = "ArgoCdNetworkAccessConfigRequest";
const _ACNACRr = "ArgoCdNetworkAccessConfigResponse";
const _ACR = "AccessConfigResponse";
const _ACRM = "ArgoCdRoleMapping";
const _ACRML = "ArgoCdRoleMappingList";
const _ADE = "AccessDeniedException";
const _AE = "AccessEntry";
const _AEC = "AssociateEncryptionConfig";
const _AECR = "AssociateEncryptionConfigRequest";
const _AECRs = "AssociateEncryptionConfigResponse";
const _AH = "AddonHealth";
const _AI = "AddonInfo";
const _AIL = "AddonIssueList";
const _AIPC = "AssociateIdentityProviderConfig";
const _AIPCR = "AssociateIdentityProviderConfigRequest";
const _AIPCRs = "AssociateIdentityProviderConfigResponse";
const _AId = "AddonIssue";
const _ANCR = "AddonNamespaceConfigRequest";
const _ANCRd = "AddonNamespaceConfigResponse";
const _AP = "AccessPolicy";
const _APIA = "AddonPodIdentityAssociations";
const _APIAL = "AddonPodIdentityAssociationsList";
const _APIC = "AddonPodIdentityConfiguration";
const _APICL = "AddonPodIdentityConfigurationList";
const _APL = "AccessPoliciesList";
const _AS = "AccessScope";
const _ASG = "AutoScalingGroup";
const _ASGL = "AutoScalingGroupList";
const _AVI = "AddonVersionInfo";
const _AVIL = "AddonVersionInfoList";
const _Ad = "Addons";
const _BRE = "BadRequestException";
const _BS = "BlockStorage";
const _C = "Capability";
const _CA = "CreateAddon";
const _CACR = "CreateAccessConfigRequest";
const _CAE = "CreateAccessEntry";
const _CAER = "CreateAccessEntryRequest";
const _CAERr = "CreateAccessEntryResponse";
const _CAR = "CreateAddonRequest";
const _CARr = "CreateAddonResponse";
const _CC = "CreateCapability";
const _CCR = "CapabilityConfigurationRequest";
const _CCRa = "CapabilityConfigurationResponse";
const _CCRo = "ComputeConfigRequest";
const _CCRom = "ComputeConfigResponse";
const _CCRon = "ConnectorConfigRequest";
const _CCRonn = "ConnectorConfigResponse";
const _CCRr = "CreateCapabilityRequest";
const _CCRre = "CreateCapabilityResponse";
const _CCRrea = "CreateClusterRequest";
const _CCRreat = "CreateClusterResponse";
const _CCr = "CreateCluster";
const _CE = "ClientException";
const _CEAS = "CreateEksAnywhereSubscription";
const _CEASR = "CreateEksAnywhereSubscriptionRequest";
const _CEASRr = "CreateEksAnywhereSubscriptionResponse";
const _CFP = "CreateFargateProfile";
const _CFPR = "CreateFargateProfileRequest";
const _CFPRr = "CreateFargateProfileResponse";
const _CH = "CapabilityHealth";
const _CHl = "ClusterHealth";
const _CI = "CapabilityIssue";
const _CIL = "CapabilityIssueList";
const _CILl = "ClusterIssueList";
const _CIl = "ClusterIssue";
const _CN = "CreateNodegroup";
const _CNR = "CreateNodegroupRequest";
const _CNRr = "CreateNodegroupResponse";
const _CPIA = "CreatePodIdentityAssociation";
const _CPIAR = "CreatePodIdentityAssociationRequest";
const _CPIARr = "CreatePodIdentityAssociationResponse";
const _CPPR = "ControlPlanePlacementRequest";
const _CPPRo = "ControlPlanePlacementResponse";
const _CPSC = "ControlPlaneScalingConfig";
const _CS = "CapabilitySummary";
const _CSL = "CapabilitySummaryList";
const _CSl = "ClientStat";
const _CSli = "ClientStats";
const _CVI = "ClusterVersionInformation";
const _CVL = "ClusterVersionList";
const _Ce = "Certificate";
const _Cl = "Cluster";
const _Co = "Compatibility";
const _Com = "Compatibilities";
const _DA = "DeleteAddon";
const _DAC = "DescribeAddonConfiguration";
const _DACR = "DescribeAddonConfigurationRequest";
const _DACRe = "DescribeAddonConfigurationResponse";
const _DAE = "DeleteAccessEntry";
const _DAER = "DeleteAccessEntryRequest";
const _DAERe = "DeleteAccessEntryResponse";
const _DAERes = "DescribeAccessEntryRequest";
const _DAEResc = "DescribeAccessEntryResponse";
const _DAEe = "DescribeAccessEntry";
const _DAP = "DisassociateAccessPolicy";
const _DAPR = "DisassociateAccessPolicyRequest";
const _DAPRi = "DisassociateAccessPolicyResponse";
const _DAR = "DeleteAddonRequest";
const _DARe = "DeleteAddonResponse";
const _DARes = "DescribeAddonRequest";
const _DAResc = "DescribeAddonResponse";
const _DAV = "DescribeAddonVersions";
const _DAVR = "DescribeAddonVersionsRequest";
const _DAVRe = "DescribeAddonVersionsResponse";
const _DAe = "DescribeAddon";
const _DC = "DeleteCapability";
const _DCR = "DeleteCapabilityRequest";
const _DCRe = "DeleteCapabilityResponse";
const _DCRel = "DeleteClusterRequest";
const _DCRele = "DeleteClusterResponse";
const _DCRer = "DeregisterClusterRequest";
const _DCRere = "DeregisterClusterResponse";
const _DCRes = "DescribeCapabilityRequest";
const _DCResc = "DescribeCapabilityResponse";
const _DCRescr = "DescribeClusterRequest";
const _DCRescri = "DescribeClusterResponse";
const _DCV = "DescribeClusterVersions";
const _DCVR = "DescribeClusterVersionsRequest";
const _DCVRe = "DescribeClusterVersionsResponse";
const _DCe = "DeleteCluster";
const _DCer = "DeregisterCluster";
const _DCes = "DescribeCapability";
const _DCesc = "DescribeCluster";
const _DD = "DeprecationDetail";
const _DDe = "DeprecationDetails";
const _DEAS = "DeleteEksAnywhereSubscription";
const _DEASR = "DeleteEksAnywhereSubscriptionRequest";
const _DEASRe = "DeleteEksAnywhereSubscriptionResponse";
const _DEASRes = "DescribeEksAnywhereSubscriptionRequest";
const _DEASResc = "DescribeEksAnywhereSubscriptionResponse";
const _DEASe = "DescribeEksAnywhereSubscription";
const _DFP = "DeleteFargateProfile";
const _DFPR = "DeleteFargateProfileRequest";
const _DFPRe = "DeleteFargateProfileResponse";
const _DFPRes = "DescribeFargateProfileRequest";
const _DFPResc = "DescribeFargateProfileResponse";
const _DFPe = "DescribeFargateProfile";
const _DI = "DescribeInsight";
const _DIPC = "DescribeIdentityProviderConfig";
const _DIPCR = "DescribeIdentityProviderConfigRequest";
const _DIPCRe = "DescribeIdentityProviderConfigResponse";
const _DIPCRi = "DisassociateIdentityProviderConfigRequest";
const _DIPCRis = "DisassociateIdentityProviderConfigResponse";
const _DIPCi = "DisassociateIdentityProviderConfig";
const _DIR = "DescribeInsightRequest";
const _DIRR = "DescribeInsightsRefreshRequest";
const _DIRRe = "DescribeInsightsRefreshResponse";
const _DIRe = "DescribeInsightResponse";
const _DIRes = "DescribeInsightsRefresh";
const _DN = "DeleteNodegroup";
const _DNR = "DeleteNodegroupRequest";
const _DNRe = "DeleteNodegroupResponse";
const _DNRes = "DescribeNodegroupRequest";
const _DNResc = "DescribeNodegroupResponse";
const _DNe = "DescribeNodegroup";
const _DPIA = "DeletePodIdentityAssociation";
const _DPIAR = "DeletePodIdentityAssociationRequest";
const _DPIARe = "DeletePodIdentityAssociationResponse";
const _DPIARes = "DescribePodIdentityAssociationRequest";
const _DPIAResc = "DescribePodIdentityAssociationResponse";
const _DPIAe = "DescribePodIdentityAssociation";
const _DU = "DescribeUpdate";
const _DUR = "DescribeUpdateRequest";
const _DURe = "DescribeUpdateResponse";
const _EAS = "EksAnywhereSubscription";
const _EASL = "EksAnywhereSubscriptionList";
const _EAST = "EksAnywhereSubscriptionTerm";
const _EC = "EncryptionConfig";
const _ECL = "EncryptionConfigList";
const _ED = "ErrorDetail";
const _EDr = "ErrorDetails";
const _ELB = "ElasticLoadBalancing";
const _FP = "FargateProfile";
const _FPH = "FargateProfileHealth";
const _FPI = "FargateProfileIssue";
const _FPIL = "FargateProfileIssueList";
const _FPS = "FargateProfileSelector";
const _FPSa = "FargateProfileSelectors";
const _I = "Identity";
const _ICSS = "InsightCategorySpecificSummary";
const _IF = "InsightsFilter";
const _IL = "IssueList";
const _IPC = "IdentityProviderConfig";
const _IPCR = "IdentityProviderConfigResponse";
const _IPCd = "IdentityProviderConfigs";
const _IPE = "InvalidParameterException";
const _IRD = "InsightResourceDetail";
const _IRDn = "InsightResourceDetails";
const _IRE = "InvalidRequestException";
const _IS = "InsightStatus";
const _ISE = "InvalidStateException";
const _ISn = "InsightSummary";
const _ISns = "InsightSummaries";
const _In = "Insight";
const _Is = "Issue";
const _KNCR = "KubernetesNetworkConfigRequest";
const _KNCRu = "KubernetesNetworkConfigResponse";
const _L = "License";
const _LA = "ListAddons";
const _LAAP = "ListAssociatedAccessPolicies";
const _LAAPR = "ListAssociatedAccessPoliciesRequest";
const _LAAPRi = "ListAssociatedAccessPoliciesResponse";
const _LAE = "ListAccessEntries";
const _LAER = "ListAccessEntriesRequest";
const _LAERi = "ListAccessEntriesResponse";
const _LAP = "ListAccessPolicies";
const _LAPR = "ListAccessPoliciesRequest";
const _LAPRi = "ListAccessPoliciesResponse";
const _LAR = "ListAddonsRequest";
const _LARi = "ListAddonsResponse";
const _LC = "ListCapabilities";
const _LCR = "ListCapabilitiesRequest";
const _LCRi = "ListCapabilitiesResponse";
const _LCRis = "ListClustersRequest";
const _LCRist = "ListClustersResponse";
const _LCi = "ListClusters";
const _LEAS = "ListEksAnywhereSubscriptions";
const _LEASR = "ListEksAnywhereSubscriptionsRequest";
const _LEASRi = "ListEksAnywhereSubscriptionsResponse";
const _LFP = "ListFargateProfiles";
const _LFPR = "ListFargateProfilesRequest";
const _LFPRi = "ListFargateProfilesResponse";
const _LI = "ListInsights";
const _LIPC = "ListIdentityProviderConfigs";
const _LIPCR = "ListIdentityProviderConfigsRequest";
const _LIPCRi = "ListIdentityProviderConfigsResponse";
const _LIR = "ListInsightsRequest";
const _LIRi = "ListInsightsResponse";
const _LL = "LicenseList";
const _LN = "ListNodegroups";
const _LNR = "ListNodegroupsRequest";
const _LNRi = "ListNodegroupsResponse";
const _LPIA = "ListPodIdentityAssociations";
const _LPIAR = "ListPodIdentityAssociationsRequest";
const _LPIARi = "ListPodIdentityAssociationsResponse";
const _LS = "LogSetup";
const _LSo = "LogSetups";
const _LTFR = "ListTagsForResource";
const _LTFRR = "ListTagsForResourceRequest";
const _LTFRRi = "ListTagsForResourceResponse";
const _LTS = "LaunchTemplateSpecification";
const _LU = "ListUpdates";
const _LUR = "ListUpdatesRequest";
const _LURi = "ListUpdatesResponse";
const _Lo = "Logging";
const _MI = "MarketplaceInformation";
const _N = "Nodegroup";
const _NFE = "NotFoundException";
const _NH = "NodegroupHealth";
const _NR = "NodegroupResources";
const _NRC = "NodeRepairConfig";
const _NRCO = "NodeRepairConfigOverrides";
const _NRCOL = "NodeRepairConfigOverridesList";
const _NSC = "NodegroupScalingConfig";
const _NUC = "NodegroupUpdateConfig";
const _OCR = "OutpostConfigRequest";
const _OCRu = "OutpostConfigResponse";
const _OIDC = "OIDC";
const _OIPC = "OidcIdentityProviderConfig";
const _OIPCR = "OidcIdentityProviderConfigRequest";
const _P = "Provider";
const _PIA = "PodIdentityAssociation";
const _PIAS = "PodIdentityAssociationSummary";
const _PIASo = "PodIdentityAssociationSummaries";
const _RAC = "RemoteAccessConfig";
const _RC = "RegisterCluster";
const _RCR = "RegisterClusterRequest";
const _RCRe = "RegisterClusterResponse";
const _RIUE = "ResourceInUseException";
const _RLEE = "ResourceLimitExceededException";
const _RNCR = "RemoteNetworkConfigRequest";
const _RNCRe = "RemoteNetworkConfigResponse";
const _RNFE = "ResourceNotFoundException";
const _RNN = "RemoteNodeNetwork";
const _RNNL = "RemoteNodeNetworkList";
const _RPDE = "ResourcePropagationDelayException";
const _RPN = "RemotePodNetwork";
const _RPNL = "RemotePodNetworkList";
const _SCR = "StorageConfigRequest";
const _SCRt = "StorageConfigResponse";
const _SE = "ServerException";
const _SI = "SsoIdentity";
const _SIL = "SsoIdentityList";
const _SIR = "StartInsightsRefresh";
const _SIRR = "StartInsightsRefreshRequest";
const _SIRRt = "StartInsightsRefreshResponse";
const _SUE = "ServiceUnavailableException";
const _T = "Taint";
const _TE = "ThrottlingException";
const _TR = "TagResource";
const _TRR = "TagResourceRequest";
const _TRRa = "TagResourceResponse";
const _U = "Update";
const _UA = "UpdateAddon";
const _UACC = "UpdateArgoCdConfig";
const _UACR = "UpdateAccessConfigRequest";
const _UAE = "UpdateAccessEntry";
const _UAER = "UpdateAccessEntryRequest";
const _UAERp = "UpdateAccessEntryResponse";
const _UAR = "UpdateAddonRequest";
const _UARp = "UpdateAddonResponse";
const _UAZE = "UnsupportedAvailabilityZoneException";
const _UC = "UpdateCapability";
const _UCC = "UpdateCapabilityConfiguration";
const _UCCR = "UpdateClusterConfigRequest";
const _UCCRp = "UpdateClusterConfigResponse";
const _UCCp = "UpdateClusterConfig";
const _UCR = "UpdateCapabilityRequest";
const _UCRp = "UpdateCapabilityResponse";
const _UCV = "UpdateClusterVersion";
const _UCVR = "UpdateClusterVersionRequest";
const _UCVRp = "UpdateClusterVersionResponse";
const _UEAS = "UpdateEksAnywhereSubscription";
const _UEASR = "UpdateEksAnywhereSubscriptionRequest";
const _UEASRp = "UpdateEksAnywhereSubscriptionResponse";
const _ULP = "UpdateLabelsPayload";
const _UNC = "UpdateNodegroupConfig";
const _UNCR = "UpdateNodegroupConfigRequest";
const _UNCRp = "UpdateNodegroupConfigResponse";
const _UNV = "UpdateNodegroupVersion";
const _UNVR = "UpdateNodegroupVersionRequest";
const _UNVRp = "UpdateNodegroupVersionResponse";
const _UP = "UpdateParam";
const _UPIA = "UpdatePodIdentityAssociation";
const _UPIAR = "UpdatePodIdentityAssociationRequest";
const _UPIARp = "UpdatePodIdentityAssociationResponse";
const _UPR = "UpgradePolicyRequest";
const _UPRp = "UpgradePolicyResponse";
const _UPp = "UpdateParams";
const _UR = "UntagResource";
const _URM = "UpdateRoleMappings";
const _URR = "UntagResourceRequest";
const _URRn = "UntagResourceResponse";
const _UTP = "UpdateTaintsPayload";
const _VCR = "VpcConfigRequest";
const _VCRp = "VpcConfigResponse";
const _WPC = "WarmPoolConfig";
const _ZSCR = "ZonalShiftConfigRequest";
const _ZSCRo = "ZonalShiftConfigResponse";
const _a = "arn";
const _aA = "addonArn";
const _aAP = "associatedAccessPolicy";
const _aAPs = "associatedAccessPolicies";
const _aAs = "associatedAt";
const _aAss = "associationArn";
const _aC = "argoCd";
const _aCD = "addonCompatibilityDetails";
const _aCc = "accessConfig";
const _aCct = "activationCode";
const _aE = "activationExpiry";
const _aEA = "accessEntryArn";
const _aEc = "accessEntry";
const _aEcc = "accessEntries";
const _aI = "awsIdc";
const _aIc = "activationId";
const _aId = "additionalInfo";
const _aIs = "associationId";
const _aM = "authenticationMode";
const _aN = "addonName";
const _aOUL = "addOrUpdateLabels";
const _aOURM = "addOrUpdateRoleMappings";
const _aOUT = "addOrUpdateTaints";
const _aP = "accessPolicies";
const _aPA = "associatedPolicyArn";
const _aR = "autoRenew";
const _aS = "accessScope";
const _aSG = "autoScalingGroups";
const _aT = "amiType";
const _aV = "addonVersion";
const _aVd = "addonVersions";
const _ad = "addon";
const _add = "addons";
const _ar = "architecture";
const _as = "association";
const _ass = "associations";
const _bCCAP = "bootstrapClusterCreatorAdminPermissions";
const _bS = "blockStorage";
const _bSMA = "bootstrapSelfManagedAddons";
const _c = "client";
const _cA = "createdAt";
const _cAe = "certificateAuthority";
const _cC = "connectorConfig";
const _cCo = "computeConfig";
const _cI = "clientId";
const _cL = "clusterLogging";
const _cN = "clusterName";
const _cNa = "capabilityName";
const _cPIT = "controlPlaneInstanceType";
const _cPP = "controlPlanePlacement";
const _cPSC = "controlPlaneScalingConfig";
const _cRT = "clientRequestToken";
const _cS = "clientStats";
const _cSGI = "clusterSecurityGroupId";
const _cSS = "categorySpecificSummary";
const _cSo = "configurationSchema";
const _cT = "computeTypes";
const _cTa = "capacityType";
const _cTl = "clusterType";
const _cV = "configurationValues";
const _cVl = "clusterVersion";
const _cVlu = "clusterVersions";
const _cVo = "compatibleVersions";
const _ca = "capability";
const _cap = "capabilities";
const _cat = "category";
const _cate = "categories";
const _ci = "cidrs";
const _cl = "cluster";
const _clu = "clusters";
const _co = "code";
const _com = "compatibilities";
const _con = "configuration";
const _d = "data";
const _dD = "deprecationDetails";
const _dN = "defaultNamespace";
const _dO = "defaultOnly";
const _dP = "deletionProtection";
const _dPP = "deletePropagationPolicy";
const _dPV = "defaultPlatformVersion";
const _dS = "diskSize";
const _dST = "disableSessionTags";
const _dSe = "desiredSize";
const _dV = "defaultVersion";
const _de = "description";
const _du = "duration";
const _e = "error";
const _eA = "endedAt";
const _eC = "encryptionConfig";
const _eCr = "errorCode";
const _eD = "effectiveDate";
const _eDx = "expirationDate";
const _eI = "externalId";
const _eLB = "elasticLoadBalancing";
const _eM = "errorMessage";
const _eOESD = "endOfExtendedSupportDate";
const _eOSSD = "endOfStandardSupportDate";
const _ePA = "endpointPublicAccess";
const _ePAn = "endpointPrivateAccess";
const _eSK = "ec2SshKey";
const _ef = "effect";
const _en = "enabled";
const _end = "endpoint";
const _er = "errors";
const _f = "filter";
const _fP = "fargateProfile";
const _fPA = "fargateProfileArn";
const _fPN = "fargateProfileName";
const _fPNa = "fargateProfileNames";
const _fo = "force";
const _gC = "groupsClaim";
const _gN = "groupName";
const _gP = "groupsPrefix";
const _h = "health";
const _hE = "httpError";
const _hQ = "httpQuery";
const _ht = "http";
const _i = "issues";
const _iA = "includeAll";
const _iF = "ipFamily";
const _iIA = "idcInstanceArn";
const _iMAA = "idcManagedApplicationArn";
const _iPC = "identityProviderConfig";
const _iPCA = "identityProviderConfigArn";
const _iPCN = "identityProviderConfigName";
const _iPCd = "identityProviderConfigs";
const _iR = "idcRegion";
const _iS = "insightStatus";
const _iSn = "includeStatus";
const _iT = "instanceTypes";
const _iU = "issuerUrl";
const _id = "identities";
const _id_ = "id";
const _ide = "identity";
const _in = "insight";
const _inc = "include";
const _ins = "insights";
const _is = "issuer";
const _k = "key";
const _kA = "keyArn";
const _kG = "kubernetesGroups";
const _kNC = "kubernetesNetworkConfig";
const _kPV = "kubernetesPatchVersion";
const _kRU = "kubernetesResourceUri";
const _kV = "kubernetesVersion";
const _kVu = "kubernetesVersions";
const _l = "logging";
const _lA = "licenseArns";
const _lQ = "licenseQuantity";
const _lRT = "lastRequestTime";
const _lRTa = "lastRefreshTime";
const _lT = "licenseType";
const _lTT = "lastTransitionTime";
const _lTa = "launchTemplate";
const _la = "labels";
const _li = "licenses";
const _m = "message";
const _mA = "modifiedAt";
const _mGPC = "maxGroupPreparedCapacity";
const _mI = "marketplaceInformation";
const _mPNRC = "maxParallelNodesRepairedCount";
const _mPNRP = "maxParallelNodesRepairedPercentage";
const _mR = "maxResults";
const _mRWTM = "minRepairWaitTimeMins";
const _mS = "minSize";
const _mSa = "maxSize";
const _mU = "maxUnavailable";
const _mUNTC = "maxUnhealthyNodeThresholdCount";
const _mUNTP = "maxUnhealthyNodeThresholdPercentage";
const _mUP = "maxUnavailablePercentage";
const _n = "name";
const _nA = "networkAccess";
const _nAo = "nodegroupArn";
const _nC = "namespaceConfig";
const _nMC = "nodeMonitoringCondition";
const _nN = "nodegroupName";
const _nORLD = "numberOfRequestsLast30Days";
const _nP = "nodePools";
const _nR = "nodeRole";
const _nRA = "nodeRoleArn";
const _nRC = "nodeRepairConfig";
const _nRCO = "nodeRepairConfigOverrides";
const _nT = "nextToken";
const _nUR = "nodeUnhealthyReason";
const _na = "namespaces";
const _nam = "namespace";
const _no = "nodegroup";
const _nod = "nodegroups";
const _o = "owner";
const _oA = "outpostArns";
const _oAw = "ownerArn";
const _oC = "outpostConfig";
const _oi = "oidc";
const _ow = "owners";
const _p = "publisher";
const _pA = "principalArn";
const _pAC = "publicAccessCidrs";
const _pAo = "policyArn";
const _pERA = "podExecutionRoleArn";
const _pI = "productId";
const _pIA = "podIdentityAssociations";
const _pIC = "podIdentityConfiguration";
const _pS = "poolState";
const _pU = "productUrl";
const _pV = "platformVersion";
const _pVl = "platformVersions";
const _pa = "params";
const _po = "policy";
const _pr = "provider";
const _pre = "preserve";
const _pu = "publishers";
const _r = "role";
const _rA = "roleArn";
const _rASG = "remoteAccessSecurityGroup";
const _rAe = "remoteAccess";
const _rAep = "repairAction";
const _rAes = "resourceArn";
const _rC = "requiresConfiguration";
const _rCe = "resolveConflicts";
const _rCeq = "requiredClaims";
const _rD = "releaseDate";
const _rI = "resourceIds";
const _rIP = "requiresIamPermissions";
const _rL = "removeLabels";
const _rMP = "recommendedManagedPolicies";
const _rNC = "remoteNetworkConfig";
const _rNN = "remoteNodeNetworks";
const _rOSI = "reuseOnScaleIn";
const _rPN = "remotePodNetworks";
const _rRM = "rbacRoleMappings";
const _rRMe = "removeRoleMappings";
const _rT = "removeTaints";
const _rV = "releaseVersion";
const _rVC = "resourcesVpcConfig";
const _rW = "replacedWith";
const _re = "resources";
const _rea = "reason";
const _rec = "recommendation";
const _s = "smithy.ts.sdk.synthetic.com.amazonaws.eks";
const _sA = "serviceAccount";
const _sARA = "serviceAccountRoleArn";
const _sAt = "startedAt";
const _sC = "storageConfig";
const _sCc = "scalingConfig";
const _sGI = "securityGroupIds";
const _sI = "subscriptionId";
const _sIC = "serviceIpv4Cidr";
const _sICe = "serviceIpv6Cidr";
const _sIu = "subnetIds";
const _sSG = "sourceSecurityGroups";
const _sSRV = "startServingReplacementVersion";
const _sSV = "stopServingVersion";
const _sT = "supportType";
const _sU = "serverUrl";
const _se = "server";
const _sel = "selectors";
const _st = "status";
const _sta = "statuses";
const _su = "subscription";
const _sub = "subnets";
const _subs = "subscriptions";
const _t = "tags";
const _tK = "tagKeys";
const _tL = "taintsList";
const _tRA = "targetRoleArn";
const _ta = "taints";
const _te = "term";
const _ti = "tier";
const _to = "token";
const _ty = "type";
const _typ = "types";
const _u = "username";
const _uA = "userAgent";
const _uC = "updateConfig";
const _uCs = "usernameClaim";
const _uI = "updateId";
const _uIp = "updateIds";
const _uP = "upgradePolicy";
const _uPs = "usernamePrefix";
const _uS = "updateStrategy";
const _un = "unit";
const _up = "update";
const _us = "usage";
const _v = "version";
const _vI = "vpceIds";
const _vIp = "vpcId";
const _vS = "versionStatus";
const _vZ = "validZones";
const _va = "value";
const _wPC = "warmPoolConfig";
const _zSC = "zonalShiftConfig";
const n0 = "com.amazonaws.eks";
const schema_1 = require("@smithy/core/schema");
const EKSServiceException_1 = require("../models/EKSServiceException");
const errors_1 = require("../models/errors");
const _s_registry = schema_1.TypeRegistry.for(_s);
exports.EKSServiceException$ = [-3, _s, "EKSServiceException", 0, [], []];
_s_registry.registerError(exports.EKSServiceException$, EKSServiceException_1.EKSServiceException);
const n0_registry = schema_1.TypeRegistry.for(n0);
exports.AccessDeniedException$ = [-3, n0, _ADE,
    { [_e]: _c, [_hE]: 403 },
    [_m],
    [0]
];
n0_registry.registerError(exports.AccessDeniedException$, errors_1.AccessDeniedException);
exports.BadRequestException$ = [-3, n0, _BRE,
    { [_e]: _c, [_hE]: 400 },
    [_m],
    [0]
];
n0_registry.registerError(exports.BadRequestException$, errors_1.BadRequestException);
exports.ClientException$ = [-3, n0, _CE,
    { [_e]: _c, [_hE]: 400 },
    [_cN, _nN, _aN, _sI, _m],
    [0, 0, 0, 0, 0]
];
n0_registry.registerError(exports.ClientException$, errors_1.ClientException);
exports.InvalidParameterException$ = [-3, n0, _IPE,
    { [_e]: _c, [_hE]: 400 },
    [_cN, _nN, _fPN, _aN, _sI, _m],
    [0, 0, 0, 0, 0, 0]
];
n0_registry.registerError(exports.InvalidParameterException$, errors_1.InvalidParameterException);
exports.InvalidRequestException$ = [-3, n0, _IRE,
    { [_e]: _c, [_hE]: 400 },
    [_cN, _nN, _aN, _sI, _m],
    [0, 0, 0, 0, 0]
];
n0_registry.registerError(exports.InvalidRequestException$, errors_1.InvalidRequestException);
exports.InvalidStateException$ = [-3, n0, _ISE,
    { [_e]: _c, [_hE]: 400 },
    [_cN, _m],
    [0, 0]
];
n0_registry.registerError(exports.InvalidStateException$, errors_1.InvalidStateException);
exports.NotFoundException$ = [-3, n0, _NFE,
    { [_e]: _c, [_hE]: 404 },
    [_m],
    [0]
];
n0_registry.registerError(exports.NotFoundException$, errors_1.NotFoundException);
exports.ResourceInUseException$ = [-3, n0, _RIUE,
    { [_e]: _c, [_hE]: 409 },
    [_cN, _nN, _aN, _m],
    [0, 0, 0, 0]
];
n0_registry.registerError(exports.ResourceInUseException$, errors_1.ResourceInUseException);
exports.ResourceLimitExceededException$ = [-3, n0, _RLEE,
    { [_e]: _c, [_hE]: 400 },
    [_cN, _nN, _sI, _m],
    [0, 0, 0, 0]
];
n0_registry.registerError(exports.ResourceLimitExceededException$, errors_1.ResourceLimitExceededException);
exports.ResourceNotFoundException$ = [-3, n0, _RNFE,
    { [_e]: _c, [_hE]: 404 },
    [_cN, _nN, _fPN, _aN, _sI, _m],
    [0, 0, 0, 0, 0, 0]
];
n0_registry.registerError(exports.ResourceNotFoundException$, errors_1.ResourceNotFoundException);
exports.ResourcePropagationDelayException$ = [-3, n0, _RPDE,
    { [_e]: _c, [_hE]: 428 },
    [_m],
    [0]
];
n0_registry.registerError(exports.ResourcePropagationDelayException$, errors_1.ResourcePropagationDelayException);
exports.ServerException$ = [-3, n0, _SE,
    { [_e]: _se, [_hE]: 500 },
    [_cN, _nN, _aN, _sI, _m],
    [0, 0, 0, 0, 0]
];
n0_registry.registerError(exports.ServerException$, errors_1.ServerException);
exports.ServiceUnavailableException$ = [-3, n0, _SUE,
    { [_e]: _se, [_hE]: 503 },
    [_m],
    [0]
];
n0_registry.registerError(exports.ServiceUnavailableException$, errors_1.ServiceUnavailableException);
exports.ThrottlingException$ = [-3, n0, _TE,
    { [_e]: _c, [_hE]: 429 },
    [_cN, _m],
    [0, 0]
];
n0_registry.registerError(exports.ThrottlingException$, errors_1.ThrottlingException);
exports.UnsupportedAvailabilityZoneException$ = [-3, n0, _UAZE,
    { [_e]: _c, [_hE]: 400 },
    [_m, _cN, _nN, _vZ],
    [0, 0, 0, 64 | 0]
];
n0_registry.registerError(exports.UnsupportedAvailabilityZoneException$, errors_1.UnsupportedAvailabilityZoneException);
exports.errorTypeRegistries = [
    _s_registry,
    n0_registry,
];
exports.AccessConfigResponse$ = [3, n0, _ACR,
    0,
    [_bCCAP, _aM],
    [2, 0]
];
exports.AccessEntry$ = [3, n0, _AE,
    0,
    [_cN, _pA, _kG, _aEA, _cA, _mA, _t, _u, _ty],
    [0, 0, 64 | 0, 0, 4, 4, 128 | 0, 0, 0]
];
exports.AccessPolicy$ = [3, n0, _AP,
    0,
    [_n, _a],
    [0, 0]
];
exports.AccessScope$ = [3, n0, _AS,
    0,
    [_ty, _na],
    [0, 64 | 0]
];
exports.Addon$ = [3, n0, _A,
    0,
    [_aN, _cN, _st, _aV, _h, _aA, _cA, _mA, _sARA, _t, _p, _o, _mI, _cV, _pIA, _nC],
    [0, 0, 0, 0, () => exports.AddonHealth$, 0, 4, 4, 0, 128 | 0, 0, 0, () => exports.MarketplaceInformation$, 0, 64 | 0, () => exports.AddonNamespaceConfigResponse$]
];
exports.AddonCompatibilityDetail$ = [3, n0, _ACD,
    0,
    [_n, _cVo],
    [0, 64 | 0]
];
exports.AddonHealth$ = [3, n0, _AH,
    0,
    [_i],
    [() => AddonIssueList]
];
exports.AddonInfo$ = [3, n0, _AI,
    0,
    [_aN, _ty, _aVd, _p, _o, _mI, _dN],
    [0, 0, () => AddonVersionInfoList, 0, 0, () => exports.MarketplaceInformation$, 0]
];
exports.AddonIssue$ = [3, n0, _AId,
    0,
    [_co, _m, _rI],
    [0, 0, 64 | 0]
];
exports.AddonNamespaceConfigRequest$ = [3, n0, _ANCR,
    0,
    [_nam],
    [0]
];
exports.AddonNamespaceConfigResponse$ = [3, n0, _ANCRd,
    0,
    [_nam],
    [0]
];
exports.AddonPodIdentityAssociations$ = [3, n0, _APIA,
    0,
    [_sA, _rA],
    [0, 0], 2
];
exports.AddonPodIdentityConfiguration$ = [3, n0, _APIC,
    0,
    [_sA, _rMP],
    [0, 64 | 0]
];
exports.AddonVersionInfo$ = [3, n0, _AVI,
    0,
    [_aV, _ar, _cT, _com, _rC, _rIP],
    [0, 64 | 0, 64 | 0, () => Compatibilities, 2, 2]
];
exports.ArgoCdAwsIdcConfigRequest$ = [3, n0, _ACAICR,
    0,
    [_iIA, _iR],
    [0, 0], 1
];
exports.ArgoCdAwsIdcConfigResponse$ = [3, n0, _ACAICRr,
    0,
    [_iIA, _iR, _iMAA],
    [0, 0, 0]
];
exports.ArgoCdConfigRequest$ = [3, n0, _ACCR,
    0,
    [_aI, _nam, _rRM, _nA],
    [() => exports.ArgoCdAwsIdcConfigRequest$, 0, () => ArgoCdRoleMappingList, () => exports.ArgoCdNetworkAccessConfigRequest$], 1
];
exports.ArgoCdConfigResponse$ = [3, n0, _ACCRr,
    0,
    [_nam, _aI, _rRM, _nA, _sU],
    [0, () => exports.ArgoCdAwsIdcConfigResponse$, () => ArgoCdRoleMappingList, () => exports.ArgoCdNetworkAccessConfigResponse$, 0]
];
exports.ArgoCdNetworkAccessConfigRequest$ = [3, n0, _ACNACR,
    0,
    [_vI],
    [64 | 0]
];
exports.ArgoCdNetworkAccessConfigResponse$ = [3, n0, _ACNACRr,
    0,
    [_vI],
    [64 | 0]
];
exports.ArgoCdRoleMapping$ = [3, n0, _ACRM,
    0,
    [_r, _id],
    [0, () => SsoIdentityList], 2
];
exports.AssociateAccessPolicyRequest$ = [3, n0, _AAPR,
    0,
    [_cN, _pA, _pAo, _aS],
    [[0, 1], [0, 1], 0, () => exports.AccessScope$], 4
];
exports.AssociateAccessPolicyResponse$ = [3, n0, _AAPRs,
    0,
    [_cN, _pA, _aAP],
    [0, 0, () => exports.AssociatedAccessPolicy$]
];
exports.AssociatedAccessPolicy$ = [3, n0, _AAP,
    0,
    [_pAo, _aS, _aAs, _mA],
    [0, () => exports.AccessScope$, 4, 4]
];
exports.AssociateEncryptionConfigRequest$ = [3, n0, _AECR,
    0,
    [_cN, _eC, _cRT],
    [[0, 1], () => EncryptionConfigList, [0, 4]], 2
];
exports.AssociateEncryptionConfigResponse$ = [3, n0, _AECRs,
    0,
    [_up],
    [() => exports.Update$]
];
exports.AssociateIdentityProviderConfigRequest$ = [3, n0, _AIPCR,
    0,
    [_cN, _oi, _t, _cRT],
    [[0, 1], () => exports.OidcIdentityProviderConfigRequest$, 128 | 0, [0, 4]], 2
];
exports.AssociateIdentityProviderConfigResponse$ = [3, n0, _AIPCRs,
    0,
    [_up, _t],
    [() => exports.Update$, 128 | 0]
];
exports.AutoScalingGroup$ = [3, n0, _ASG,
    0,
    [_n],
    [0]
];
exports.BlockStorage$ = [3, n0, _BS,
    0,
    [_en],
    [2]
];
exports.Capability$ = [3, n0, _C,
    0,
    [_cNa, _a, _cN, _ty, _rA, _st, _v, _con, _t, _h, _cA, _mA, _dPP],
    [0, 0, 0, 0, 0, 0, 0, () => exports.CapabilityConfigurationResponse$, 128 | 0, () => exports.CapabilityHealth$, 4, 4, 0]
];
exports.CapabilityConfigurationRequest$ = [3, n0, _CCR,
    0,
    [_aC],
    [() => exports.ArgoCdConfigRequest$]
];
exports.CapabilityConfigurationResponse$ = [3, n0, _CCRa,
    0,
    [_aC],
    [() => exports.ArgoCdConfigResponse$]
];
exports.CapabilityHealth$ = [3, n0, _CH,
    0,
    [_i],
    [() => CapabilityIssueList]
];
exports.CapabilityIssue$ = [3, n0, _CI,
    0,
    [_co, _m],
    [0, 0]
];
exports.CapabilitySummary$ = [3, n0, _CS,
    0,
    [_cNa, _a, _ty, _st, _v, _cA, _mA],
    [0, 0, 0, 0, 0, 4, 4]
];
exports.Certificate$ = [3, n0, _Ce,
    0,
    [_d],
    [0]
];
exports.ClientStat$ = [3, n0, _CSl,
    0,
    [_uA, _nORLD, _lRT],
    [0, 1, 4]
];
exports.Cluster$ = [3, n0, _Cl,
    0,
    [_n, _a, _cA, _v, _end, _rA, _rVC, _kNC, _l, _ide, _st, _cAe, _cRT, _pV, _t, _eC, _cC, _id_, _h, _oC, _aCc, _uP, _zSC, _rNC, _cCo, _sC, _dP, _cPSC],
    [0, 0, 4, 0, 0, 0, () => exports.VpcConfigResponse$, () => exports.KubernetesNetworkConfigResponse$, () => exports.Logging$, () => exports.Identity$, 0, () => exports.Certificate$, 0, 0, 128 | 0, () => EncryptionConfigList, () => exports.ConnectorConfigResponse$, 0, () => exports.ClusterHealth$, () => exports.OutpostConfigResponse$, () => exports.AccessConfigResponse$, () => exports.UpgradePolicyResponse$, () => exports.ZonalShiftConfigResponse$, () => exports.RemoteNetworkConfigResponse$, () => exports.ComputeConfigResponse$, () => exports.StorageConfigResponse$, 2, () => exports.ControlPlaneScalingConfig$]
];
exports.ClusterHealth$ = [3, n0, _CHl,
    0,
    [_i],
    [() => ClusterIssueList]
];
exports.ClusterIssue$ = [3, n0, _CIl,
    0,
    [_co, _m, _rI],
    [0, 0, 64 | 0]
];
exports.ClusterVersionInformation$ = [3, n0, _CVI,
    0,
    [_cVl, _cTl, _dPV, _dV, _rD, _eOSSD, _eOESD, _st, _vS, _kPV],
    [0, 0, 0, 2, 4, 4, 4, 0, 0, 0]
];
exports.Compatibility$ = [3, n0, _Co,
    0,
    [_cVl, _pVl, _dV],
    [0, 64 | 0, 2]
];
exports.ComputeConfigRequest$ = [3, n0, _CCRo,
    0,
    [_en, _nP, _nRA],
    [2, 64 | 0, 0]
];
exports.ComputeConfigResponse$ = [3, n0, _CCRom,
    0,
    [_en, _nP, _nRA],
    [2, 64 | 0, 0]
];
exports.ConnectorConfigRequest$ = [3, n0, _CCRon,
    0,
    [_rA, _pr],
    [0, 0], 2
];
exports.ConnectorConfigResponse$ = [3, n0, _CCRonn,
    0,
    [_aIc, _aCct, _aE, _pr, _rA],
    [0, 0, 4, 0, 0]
];
exports.ControlPlanePlacementRequest$ = [3, n0, _CPPR,
    0,
    [_gN],
    [0]
];
exports.ControlPlanePlacementResponse$ = [3, n0, _CPPRo,
    0,
    [_gN],
    [0]
];
exports.ControlPlaneScalingConfig$ = [3, n0, _CPSC,
    0,
    [_ti],
    [0]
];
exports.CreateAccessConfigRequest$ = [3, n0, _CACR,
    0,
    [_bCCAP, _aM],
    [2, 0]
];
exports.CreateAccessEntryRequest$ = [3, n0, _CAER,
    0,
    [_cN, _pA, _kG, _t, _cRT, _u, _ty],
    [[0, 1], 0, 64 | 0, 128 | 0, [0, 4], 0, 0], 2
];
exports.CreateAccessEntryResponse$ = [3, n0, _CAERr,
    0,
    [_aEc],
    [() => exports.AccessEntry$]
];
exports.CreateAddonRequest$ = [3, n0, _CAR,
    0,
    [_cN, _aN, _aV, _sARA, _rCe, _cRT, _t, _cV, _pIA, _nC],
    [[0, 1], 0, 0, 0, 0, [0, 4], 128 | 0, 0, () => AddonPodIdentityAssociationsList, () => exports.AddonNamespaceConfigRequest$], 2
];
exports.CreateAddonResponse$ = [3, n0, _CARr,
    0,
    [_ad],
    [() => exports.Addon$]
];
exports.CreateCapabilityRequest$ = [3, n0, _CCRr,
    0,
    [_cNa, _cN, _ty, _rA, _dPP, _cRT, _con, _t],
    [0, [0, 1], 0, 0, 0, [0, 4], () => exports.CapabilityConfigurationRequest$, 128 | 0], 5
];
exports.CreateCapabilityResponse$ = [3, n0, _CCRre,
    0,
    [_ca],
    [() => exports.Capability$]
];
exports.CreateClusterRequest$ = [3, n0, _CCRrea,
    0,
    [_n, _rA, _rVC, _v, _kNC, _l, _cRT, _t, _eC, _oC, _aCc, _bSMA, _uP, _zSC, _rNC, _cCo, _sC, _dP, _cPSC],
    [0, 0, () => exports.VpcConfigRequest$, 0, () => exports.KubernetesNetworkConfigRequest$, () => exports.Logging$, [0, 4], 128 | 0, () => EncryptionConfigList, () => exports.OutpostConfigRequest$, () => exports.CreateAccessConfigRequest$, 2, () => exports.UpgradePolicyRequest$, () => exports.ZonalShiftConfigRequest$, () => exports.RemoteNetworkConfigRequest$, () => exports.ComputeConfigRequest$, () => exports.StorageConfigRequest$, 2, () => exports.ControlPlaneScalingConfig$], 3
];
exports.CreateClusterResponse$ = [3, n0, _CCRreat,
    0,
    [_cl],
    [() => exports.Cluster$]
];
exports.CreateEksAnywhereSubscriptionRequest$ = [3, n0, _CEASR,
    0,
    [_n, _te, _lQ, _lT, _aR, _cRT, _t],
    [0, () => exports.EksAnywhereSubscriptionTerm$, 1, 0, 2, [0, 4], 128 | 0], 2
];
exports.CreateEksAnywhereSubscriptionResponse$ = [3, n0, _CEASRr,
    0,
    [_su],
    [() => exports.EksAnywhereSubscription$]
];
exports.CreateFargateProfileRequest$ = [3, n0, _CFPR,
    0,
    [_fPN, _cN, _pERA, _sub, _sel, _cRT, _t],
    [0, [0, 1], 0, 64 | 0, () => FargateProfileSelectors, [0, 4], 128 | 0], 3
];
exports.CreateFargateProfileResponse$ = [3, n0, _CFPRr,
    0,
    [_fP],
    [() => exports.FargateProfile$]
];
exports.CreateNodegroupRequest$ = [3, n0, _CNR,
    0,
    [_cN, _nN, _sub, _nR, _sCc, _dS, _iT, _aT, _rAe, _la, _ta, _t, _cRT, _lTa, _uC, _nRC, _cTa, _v, _rV, _wPC],
    [[0, 1], 0, 64 | 0, 0, () => exports.NodegroupScalingConfig$, 1, 64 | 0, 0, () => exports.RemoteAccessConfig$, 128 | 0, () => taintsList, 128 | 0, [0, 4], () => exports.LaunchTemplateSpecification$, () => exports.NodegroupUpdateConfig$, () => exports.NodeRepairConfig$, 0, 0, 0, () => exports.WarmPoolConfig$], 4
];
exports.CreateNodegroupResponse$ = [3, n0, _CNRr,
    0,
    [_no],
    [() => exports.Nodegroup$]
];
exports.CreatePodIdentityAssociationRequest$ = [3, n0, _CPIAR,
    0,
    [_cN, _nam, _sA, _rA, _cRT, _t, _dST, _tRA, _po],
    [[0, 1], 0, 0, 0, [0, 4], 128 | 0, 2, 0, 0], 4
];
exports.CreatePodIdentityAssociationResponse$ = [3, n0, _CPIARr,
    0,
    [_as],
    [() => exports.PodIdentityAssociation$]
];
exports.DeleteAccessEntryRequest$ = [3, n0, _DAER,
    0,
    [_cN, _pA],
    [[0, 1], [0, 1]], 2
];
exports.DeleteAccessEntryResponse$ = [3, n0, _DAERe,
    0,
    [],
    []
];
exports.DeleteAddonRequest$ = [3, n0, _DAR,
    0,
    [_cN, _aN, _pre],
    [[0, 1], [0, 1], [2, { [_hQ]: _pre }]], 2
];
exports.DeleteAddonResponse$ = [3, n0, _DARe,
    0,
    [_ad],
    [() => exports.Addon$]
];
exports.DeleteCapabilityRequest$ = [3, n0, _DCR,
    0,
    [_cN, _cNa],
    [[0, 1], [0, 1]], 2
];
exports.DeleteCapabilityResponse$ = [3, n0, _DCRe,
    0,
    [_ca],
    [() => exports.Capability$]
];
exports.DeleteClusterRequest$ = [3, n0, _DCRel,
    0,
    [_n],
    [[0, 1]], 1
];
exports.DeleteClusterResponse$ = [3, n0, _DCRele,
    0,
    [_cl],
    [() => exports.Cluster$]
];
exports.DeleteEksAnywhereSubscriptionRequest$ = [3, n0, _DEASR,
    0,
    [_id_],
    [[0, 1]], 1
];
exports.DeleteEksAnywhereSubscriptionResponse$ = [3, n0, _DEASRe,
    0,
    [_su],
    [() => exports.EksAnywhereSubscription$]
];
exports.DeleteFargateProfileRequest$ = [3, n0, _DFPR,
    0,
    [_cN, _fPN],
    [[0, 1], [0, 1]], 2
];
exports.DeleteFargateProfileResponse$ = [3, n0, _DFPRe,
    0,
    [_fP],
    [() => exports.FargateProfile$]
];
exports.DeleteNodegroupRequest$ = [3, n0, _DNR,
    0,
    [_cN, _nN],
    [[0, 1], [0, 1]], 2
];
exports.DeleteNodegroupResponse$ = [3, n0, _DNRe,
    0,
    [_no],
    [() => exports.Nodegroup$]
];
exports.DeletePodIdentityAssociationRequest$ = [3, n0, _DPIAR,
    0,
    [_cN, _aIs],
    [[0, 1], [0, 1]], 2
];
exports.DeletePodIdentityAssociationResponse$ = [3, n0, _DPIARe,
    0,
    [_as],
    [() => exports.PodIdentityAssociation$]
];
exports.DeprecationDetail$ = [3, n0, _DD,
    0,
    [_us, _rW, _sSV, _sSRV, _cS],
    [0, 0, 0, 0, () => ClientStats]
];
exports.DeregisterClusterRequest$ = [3, n0, _DCRer,
    0,
    [_n],
    [[0, 1]], 1
];
exports.DeregisterClusterResponse$ = [3, n0, _DCRere,
    0,
    [_cl],
    [() => exports.Cluster$]
];
exports.DescribeAccessEntryRequest$ = [3, n0, _DAERes,
    0,
    [_cN, _pA],
    [[0, 1], [0, 1]], 2
];
exports.DescribeAccessEntryResponse$ = [3, n0, _DAEResc,
    0,
    [_aEc],
    [() => exports.AccessEntry$]
];
exports.DescribeAddonConfigurationRequest$ = [3, n0, _DACR,
    0,
    [_aN, _aV],
    [[0, { [_hQ]: _aN }], [0, { [_hQ]: _aV }]], 2
];
exports.DescribeAddonConfigurationResponse$ = [3, n0, _DACRe,
    0,
    [_aN, _aV, _cSo, _pIC],
    [0, 0, 0, () => AddonPodIdentityConfigurationList]
];
exports.DescribeAddonRequest$ = [3, n0, _DARes,
    0,
    [_cN, _aN],
    [[0, 1], [0, 1]], 2
];
exports.DescribeAddonResponse$ = [3, n0, _DAResc,
    0,
    [_ad],
    [() => exports.Addon$]
];
exports.DescribeAddonVersionsRequest$ = [3, n0, _DAVR,
    0,
    [_kV, _mR, _nT, _aN, _typ, _pu, _ow],
    [[0, { [_hQ]: _kV }], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }], [0, { [_hQ]: _aN }], [64 | 0, { [_hQ]: _typ }], [64 | 0, { [_hQ]: _pu }], [64 | 0, { [_hQ]: _ow }]]
];
exports.DescribeAddonVersionsResponse$ = [3, n0, _DAVRe,
    0,
    [_add, _nT],
    [() => Addons, 0]
];
exports.DescribeCapabilityRequest$ = [3, n0, _DCRes,
    0,
    [_cN, _cNa],
    [[0, 1], [0, 1]], 2
];
exports.DescribeCapabilityResponse$ = [3, n0, _DCResc,
    0,
    [_ca],
    [() => exports.Capability$]
];
exports.DescribeClusterRequest$ = [3, n0, _DCRescr,
    0,
    [_n],
    [[0, 1]], 1
];
exports.DescribeClusterResponse$ = [3, n0, _DCRescri,
    0,
    [_cl],
    [() => exports.Cluster$]
];
exports.DescribeClusterVersionsRequest$ = [3, n0, _DCVR,
    0,
    [_cTl, _mR, _nT, _dO, _iA, _cVlu, _st, _vS],
    [[0, { [_hQ]: _cTl }], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }], [2, { [_hQ]: _dO }], [2, { [_hQ]: _iA }], [64 | 0, { [_hQ]: _cVlu }], [0, { [_hQ]: _st }], [0, { [_hQ]: _vS }]]
];
exports.DescribeClusterVersionsResponse$ = [3, n0, _DCVRe,
    0,
    [_nT, _cVlu],
    [0, () => ClusterVersionList]
];
exports.DescribeEksAnywhereSubscriptionRequest$ = [3, n0, _DEASRes,
    0,
    [_id_],
    [[0, 1]], 1
];
exports.DescribeEksAnywhereSubscriptionResponse$ = [3, n0, _DEASResc,
    0,
    [_su],
    [() => exports.EksAnywhereSubscription$]
];
exports.DescribeFargateProfileRequest$ = [3, n0, _DFPRes,
    0,
    [_cN, _fPN],
    [[0, 1], [0, 1]], 2
];
exports.DescribeFargateProfileResponse$ = [3, n0, _DFPResc,
    0,
    [_fP],
    [() => exports.FargateProfile$]
];
exports.DescribeIdentityProviderConfigRequest$ = [3, n0, _DIPCR,
    0,
    [_cN, _iPC],
    [[0, 1], () => exports.IdentityProviderConfig$], 2
];
exports.DescribeIdentityProviderConfigResponse$ = [3, n0, _DIPCRe,
    0,
    [_iPC],
    [() => exports.IdentityProviderConfigResponse$]
];
exports.DescribeInsightRequest$ = [3, n0, _DIR,
    0,
    [_cN, _id_],
    [[0, 1], [0, 1]], 2
];
exports.DescribeInsightResponse$ = [3, n0, _DIRe,
    0,
    [_in],
    [() => exports.Insight$]
];
exports.DescribeInsightsRefreshRequest$ = [3, n0, _DIRR,
    0,
    [_cN],
    [[0, 1]], 1
];
exports.DescribeInsightsRefreshResponse$ = [3, n0, _DIRRe,
    0,
    [_m, _st, _sAt, _eA],
    [0, 0, 4, 4]
];
exports.DescribeNodegroupRequest$ = [3, n0, _DNRes,
    0,
    [_cN, _nN],
    [[0, 1], [0, 1]], 2
];
exports.DescribeNodegroupResponse$ = [3, n0, _DNResc,
    0,
    [_no],
    [() => exports.Nodegroup$]
];
exports.DescribePodIdentityAssociationRequest$ = [3, n0, _DPIARes,
    0,
    [_cN, _aIs],
    [[0, 1], [0, 1]], 2
];
exports.DescribePodIdentityAssociationResponse$ = [3, n0, _DPIAResc,
    0,
    [_as],
    [() => exports.PodIdentityAssociation$]
];
exports.DescribeUpdateRequest$ = [3, n0, _DUR,
    0,
    [_n, _uI, _nN, _aN, _cNa],
    [[0, 1], [0, 1], [0, { [_hQ]: _nN }], [0, { [_hQ]: _aN }], [0, { [_hQ]: _cNa }]], 2
];
exports.DescribeUpdateResponse$ = [3, n0, _DURe,
    0,
    [_up],
    [() => exports.Update$]
];
exports.DisassociateAccessPolicyRequest$ = [3, n0, _DAPR,
    0,
    [_cN, _pA, _pAo],
    [[0, 1], [0, 1], [0, 1]], 3
];
exports.DisassociateAccessPolicyResponse$ = [3, n0, _DAPRi,
    0,
    [],
    []
];
exports.DisassociateIdentityProviderConfigRequest$ = [3, n0, _DIPCRi,
    0,
    [_cN, _iPC, _cRT],
    [[0, 1], () => exports.IdentityProviderConfig$, [0, 4]], 2
];
exports.DisassociateIdentityProviderConfigResponse$ = [3, n0, _DIPCRis,
    0,
    [_up],
    [() => exports.Update$]
];
exports.EksAnywhereSubscription$ = [3, n0, _EAS,
    0,
    [_id_, _a, _cA, _eD, _eDx, _lQ, _lT, _te, _st, _aR, _lA, _li, _t],
    [0, 0, 4, 4, 4, 1, 0, () => exports.EksAnywhereSubscriptionTerm$, 0, 2, 64 | 0, () => LicenseList, 128 | 0]
];
exports.EksAnywhereSubscriptionTerm$ = [3, n0, _EAST,
    0,
    [_du, _un],
    [1, 0]
];
exports.ElasticLoadBalancing$ = [3, n0, _ELB,
    0,
    [_en],
    [2]
];
exports.EncryptionConfig$ = [3, n0, _EC,
    0,
    [_re, _pr],
    [64 | 0, () => exports.Provider$]
];
exports.ErrorDetail$ = [3, n0, _ED,
    0,
    [_eCr, _eM, _rI],
    [0, 0, 64 | 0]
];
exports.FargateProfile$ = [3, n0, _FP,
    0,
    [_fPN, _fPA, _cN, _cA, _pERA, _sub, _sel, _st, _t, _h],
    [0, 0, 0, 4, 0, 64 | 0, () => FargateProfileSelectors, 0, 128 | 0, () => exports.FargateProfileHealth$]
];
exports.FargateProfileHealth$ = [3, n0, _FPH,
    0,
    [_i],
    [() => FargateProfileIssueList]
];
exports.FargateProfileIssue$ = [3, n0, _FPI,
    0,
    [_co, _m, _rI],
    [0, 0, 64 | 0]
];
exports.FargateProfileSelector$ = [3, n0, _FPS,
    0,
    [_nam, _la],
    [0, 128 | 0]
];
exports.Identity$ = [3, n0, _I,
    0,
    [_oi],
    [() => exports.OIDC$]
];
exports.IdentityProviderConfig$ = [3, n0, _IPC,
    0,
    [_ty, _n],
    [0, 0], 2
];
exports.IdentityProviderConfigResponse$ = [3, n0, _IPCR,
    0,
    [_oi],
    [() => exports.OidcIdentityProviderConfig$]
];
exports.Insight$ = [3, n0, _In,
    0,
    [_id_, _n, _cat, _kV, _lRTa, _lTT, _de, _iS, _rec, _aId, _re, _cSS],
    [0, 0, 0, 0, 4, 4, 0, () => exports.InsightStatus$, 0, 128 | 0, () => InsightResourceDetails, () => exports.InsightCategorySpecificSummary$]
];
exports.InsightCategorySpecificSummary$ = [3, n0, _ICSS,
    0,
    [_dD, _aCD],
    [() => DeprecationDetails, () => AddonCompatibilityDetails]
];
exports.InsightResourceDetail$ = [3, n0, _IRD,
    0,
    [_iS, _kRU, _a],
    [() => exports.InsightStatus$, 0, 0]
];
exports.InsightsFilter$ = [3, n0, _IF,
    0,
    [_cate, _kVu, _sta],
    [64 | 0, 64 | 0, 64 | 0]
];
exports.InsightStatus$ = [3, n0, _IS,
    0,
    [_st, _rea],
    [0, 0]
];
exports.InsightSummary$ = [3, n0, _ISn,
    0,
    [_id_, _n, _cat, _kV, _lRTa, _lTT, _de, _iS],
    [0, 0, 0, 0, 4, 4, 0, () => exports.InsightStatus$]
];
exports.Issue$ = [3, n0, _Is,
    0,
    [_co, _m, _rI],
    [0, 0, 64 | 0]
];
exports.KubernetesNetworkConfigRequest$ = [3, n0, _KNCR,
    0,
    [_sIC, _iF, _eLB],
    [0, 0, () => exports.ElasticLoadBalancing$]
];
exports.KubernetesNetworkConfigResponse$ = [3, n0, _KNCRu,
    0,
    [_sIC, _sICe, _iF, _eLB],
    [0, 0, 0, () => exports.ElasticLoadBalancing$]
];
exports.LaunchTemplateSpecification$ = [3, n0, _LTS,
    0,
    [_n, _v, _id_],
    [0, 0, 0]
];
exports.License$ = [3, n0, _L,
    0,
    [_id_, _to],
    [0, 0]
];
exports.ListAccessEntriesRequest$ = [3, n0, _LAER,
    0,
    [_cN, _aPA, _mR, _nT],
    [[0, 1], [0, { [_hQ]: _aPA }], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]], 1
];
exports.ListAccessEntriesResponse$ = [3, n0, _LAERi,
    0,
    [_aEcc, _nT],
    [64 | 0, 0]
];
exports.ListAccessPoliciesRequest$ = [3, n0, _LAPR,
    0,
    [_mR, _nT],
    [[1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]]
];
exports.ListAccessPoliciesResponse$ = [3, n0, _LAPRi,
    0,
    [_aP, _nT],
    [() => AccessPoliciesList, 0]
];
exports.ListAddonsRequest$ = [3, n0, _LAR,
    0,
    [_cN, _mR, _nT],
    [[0, 1], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]], 1
];
exports.ListAddonsResponse$ = [3, n0, _LARi,
    0,
    [_add, _nT],
    [64 | 0, 0]
];
exports.ListAssociatedAccessPoliciesRequest$ = [3, n0, _LAAPR,
    0,
    [_cN, _pA, _mR, _nT],
    [[0, 1], [0, 1], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]], 2
];
exports.ListAssociatedAccessPoliciesResponse$ = [3, n0, _LAAPRi,
    0,
    [_cN, _pA, _nT, _aAPs],
    [0, 0, 0, () => AssociatedAccessPoliciesList]
];
exports.ListCapabilitiesRequest$ = [3, n0, _LCR,
    0,
    [_cN, _nT, _mR],
    [[0, 1], [0, { [_hQ]: _nT }], [1, { [_hQ]: _mR }]], 1
];
exports.ListCapabilitiesResponse$ = [3, n0, _LCRi,
    0,
    [_cap, _nT],
    [() => CapabilitySummaryList, 0]
];
exports.ListClustersRequest$ = [3, n0, _LCRis,
    0,
    [_mR, _nT, _inc],
    [[1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }], [64 | 0, { [_hQ]: _inc }]]
];
exports.ListClustersResponse$ = [3, n0, _LCRist,
    0,
    [_clu, _nT],
    [64 | 0, 0]
];
exports.ListEksAnywhereSubscriptionsRequest$ = [3, n0, _LEASR,
    0,
    [_mR, _nT, _iSn],
    [[1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }], [64 | 0, { [_hQ]: _iSn }]]
];
exports.ListEksAnywhereSubscriptionsResponse$ = [3, n0, _LEASRi,
    0,
    [_subs, _nT],
    [() => EksAnywhereSubscriptionList, 0]
];
exports.ListFargateProfilesRequest$ = [3, n0, _LFPR,
    0,
    [_cN, _mR, _nT],
    [[0, 1], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]], 1
];
exports.ListFargateProfilesResponse$ = [3, n0, _LFPRi,
    0,
    [_fPNa, _nT],
    [64 | 0, 0]
];
exports.ListIdentityProviderConfigsRequest$ = [3, n0, _LIPCR,
    0,
    [_cN, _mR, _nT],
    [[0, 1], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]], 1
];
exports.ListIdentityProviderConfigsResponse$ = [3, n0, _LIPCRi,
    0,
    [_iPCd, _nT],
    [() => IdentityProviderConfigs, 0]
];
exports.ListInsightsRequest$ = [3, n0, _LIR,
    0,
    [_cN, _f, _mR, _nT],
    [[0, 1], () => exports.InsightsFilter$, 1, 0], 1
];
exports.ListInsightsResponse$ = [3, n0, _LIRi,
    0,
    [_ins, _nT],
    [() => InsightSummaries, 0]
];
exports.ListNodegroupsRequest$ = [3, n0, _LNR,
    0,
    [_cN, _mR, _nT],
    [[0, 1], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]], 1
];
exports.ListNodegroupsResponse$ = [3, n0, _LNRi,
    0,
    [_nod, _nT],
    [64 | 0, 0]
];
exports.ListPodIdentityAssociationsRequest$ = [3, n0, _LPIAR,
    0,
    [_cN, _nam, _sA, _mR, _nT],
    [[0, 1], [0, { [_hQ]: _nam }], [0, { [_hQ]: _sA }], [1, { [_hQ]: _mR }], [0, { [_hQ]: _nT }]], 1
];
exports.ListPodIdentityAssociationsResponse$ = [3, n0, _LPIARi,
    0,
    [_ass, _nT],
    [() => PodIdentityAssociationSummaries, 0]
];
exports.ListTagsForResourceRequest$ = [3, n0, _LTFRR,
    0,
    [_rAes],
    [[0, 1]], 1
];
exports.ListTagsForResourceResponse$ = [3, n0, _LTFRRi,
    0,
    [_t],
    [128 | 0]
];
exports.ListUpdatesRequest$ = [3, n0, _LUR,
    0,
    [_n, _nN, _aN, _cNa, _nT, _mR],
    [[0, 1], [0, { [_hQ]: _nN }], [0, { [_hQ]: _aN }], [0, { [_hQ]: _cNa }], [0, { [_hQ]: _nT }], [1, { [_hQ]: _mR }]], 1
];
exports.ListUpdatesResponse$ = [3, n0, _LURi,
    0,
    [_uIp, _nT],
    [64 | 0, 0]
];
exports.Logging$ = [3, n0, _Lo,
    0,
    [_cL],
    [() => LogSetups]
];
exports.LogSetup$ = [3, n0, _LS,
    0,
    [_typ, _en],
    [64 | 0, 2]
];
exports.MarketplaceInformation$ = [3, n0, _MI,
    0,
    [_pI, _pU],
    [0, 0]
];
exports.Nodegroup$ = [3, n0, _N,
    0,
    [_nN, _nAo, _cN, _v, _rV, _cA, _mA, _st, _cTa, _sCc, _iT, _sub, _rAe, _aT, _nR, _la, _ta, _re, _dS, _h, _uC, _nRC, _lTa, _t, _wPC],
    [0, 0, 0, 0, 0, 4, 4, 0, 0, () => exports.NodegroupScalingConfig$, 64 | 0, 64 | 0, () => exports.RemoteAccessConfig$, 0, 0, 128 | 0, () => taintsList, () => exports.NodegroupResources$, 1, () => exports.NodegroupHealth$, () => exports.NodegroupUpdateConfig$, () => exports.NodeRepairConfig$, () => exports.LaunchTemplateSpecification$, 128 | 0, () => exports.WarmPoolConfig$]
];
exports.NodegroupHealth$ = [3, n0, _NH,
    0,
    [_i],
    [() => IssueList]
];
exports.NodegroupResources$ = [3, n0, _NR,
    0,
    [_aSG, _rASG],
    [() => AutoScalingGroupList, 0]
];
exports.NodegroupScalingConfig$ = [3, n0, _NSC,
    0,
    [_mS, _mSa, _dSe],
    [1, 1, 1]
];
exports.NodegroupUpdateConfig$ = [3, n0, _NUC,
    0,
    [_mU, _mUP, _uS],
    [1, 1, 0]
];
exports.NodeRepairConfig$ = [3, n0, _NRC,
    0,
    [_en, _mUNTC, _mUNTP, _mPNRC, _mPNRP, _nRCO],
    [2, 1, 1, 1, 1, () => NodeRepairConfigOverridesList]
];
exports.NodeRepairConfigOverrides$ = [3, n0, _NRCO,
    0,
    [_nMC, _nUR, _mRWTM, _rAep],
    [0, 0, 1, 0]
];
exports.OIDC$ = [3, n0, _OIDC,
    0,
    [_is],
    [0]
];
exports.OidcIdentityProviderConfig$ = [3, n0, _OIPC,
    0,
    [_iPCN, _iPCA, _cN, _iU, _cI, _uCs, _uPs, _gC, _gP, _rCeq, _t, _st],
    [0, 0, 0, 0, 0, 0, 0, 0, 0, 128 | 0, 128 | 0, 0]
];
exports.OidcIdentityProviderConfigRequest$ = [3, n0, _OIPCR,
    0,
    [_iPCN, _iU, _cI, _uCs, _uPs, _gC, _gP, _rCeq],
    [0, 0, 0, 0, 0, 0, 0, 128 | 0], 3
];
exports.OutpostConfigRequest$ = [3, n0, _OCR,
    0,
    [_oA, _cPIT, _cPP],
    [64 | 0, 0, () => exports.ControlPlanePlacementRequest$], 2
];
exports.OutpostConfigResponse$ = [3, n0, _OCRu,
    0,
    [_oA, _cPIT, _cPP],
    [64 | 0, 0, () => exports.ControlPlanePlacementResponse$], 2
];
exports.PodIdentityAssociation$ = [3, n0, _PIA,
    0,
    [_cN, _nam, _sA, _rA, _aAss, _aIs, _t, _cA, _mA, _oAw, _dST, _tRA, _eI, _po],
    [0, 0, 0, 0, 0, 0, 128 | 0, 4, 4, 0, 2, 0, 0, 0]
];
exports.PodIdentityAssociationSummary$ = [3, n0, _PIAS,
    0,
    [_cN, _nam, _sA, _aAss, _aIs, _oAw],
    [0, 0, 0, 0, 0, 0]
];
exports.Provider$ = [3, n0, _P,
    0,
    [_kA],
    [0]
];
exports.RegisterClusterRequest$ = [3, n0, _RCR,
    0,
    [_n, _cC, _cRT, _t],
    [0, () => exports.ConnectorConfigRequest$, [0, 4], 128 | 0], 2
];
exports.RegisterClusterResponse$ = [3, n0, _RCRe,
    0,
    [_cl],
    [() => exports.Cluster$]
];
exports.RemoteAccessConfig$ = [3, n0, _RAC,
    0,
    [_eSK, _sSG],
    [0, 64 | 0]
];
exports.RemoteNetworkConfigRequest$ = [3, n0, _RNCR,
    0,
    [_rNN, _rPN],
    [() => RemoteNodeNetworkList, () => RemotePodNetworkList]
];
exports.RemoteNetworkConfigResponse$ = [3, n0, _RNCRe,
    0,
    [_rNN, _rPN],
    [() => RemoteNodeNetworkList, () => RemotePodNetworkList]
];
exports.RemoteNodeNetwork$ = [3, n0, _RNN,
    0,
    [_ci],
    [64 | 0]
];
exports.RemotePodNetwork$ = [3, n0, _RPN,
    0,
    [_ci],
    [64 | 0]
];
exports.SsoIdentity$ = [3, n0, _SI,
    0,
    [_id_, _ty],
    [0, 0], 2
];
exports.StartInsightsRefreshRequest$ = [3, n0, _SIRR,
    0,
    [_cN],
    [[0, 1]], 1
];
exports.StartInsightsRefreshResponse$ = [3, n0, _SIRRt,
    0,
    [_m, _st],
    [0, 0]
];
exports.StorageConfigRequest$ = [3, n0, _SCR,
    0,
    [_bS],
    [() => exports.BlockStorage$]
];
exports.StorageConfigResponse$ = [3, n0, _SCRt,
    0,
    [_bS],
    [() => exports.BlockStorage$]
];
exports.TagResourceRequest$ = [3, n0, _TRR,
    0,
    [_rAes, _t],
    [[0, 1], 128 | 0], 2
];
exports.TagResourceResponse$ = [3, n0, _TRRa,
    0,
    [],
    []
];
exports.Taint$ = [3, n0, _T,
    0,
    [_k, _va, _ef],
    [0, 0, 0]
];
exports.UntagResourceRequest$ = [3, n0, _URR,
    0,
    [_rAes, _tK],
    [[0, 1], [64 | 0, { [_hQ]: _tK }]], 2
];
exports.UntagResourceResponse$ = [3, n0, _URRn,
    0,
    [],
    []
];
exports.Update$ = [3, n0, _U,
    0,
    [_id_, _st, _ty, _pa, _cA, _er],
    [0, 0, 0, () => UpdateParams, 4, () => ErrorDetails]
];
exports.UpdateAccessConfigRequest$ = [3, n0, _UACR,
    0,
    [_aM],
    [0]
];
exports.UpdateAccessEntryRequest$ = [3, n0, _UAER,
    0,
    [_cN, _pA, _kG, _cRT, _u],
    [[0, 1], [0, 1], 64 | 0, [0, 4], 0], 2
];
exports.UpdateAccessEntryResponse$ = [3, n0, _UAERp,
    0,
    [_aEc],
    [() => exports.AccessEntry$]
];
exports.UpdateAddonRequest$ = [3, n0, _UAR,
    0,
    [_cN, _aN, _aV, _sARA, _rCe, _cRT, _cV, _pIA],
    [[0, 1], [0, 1], 0, 0, 0, [0, 4], 0, () => AddonPodIdentityAssociationsList], 2
];
exports.UpdateAddonResponse$ = [3, n0, _UARp,
    0,
    [_up],
    [() => exports.Update$]
];
exports.UpdateArgoCdConfig$ = [3, n0, _UACC,
    0,
    [_rRM, _nA],
    [() => exports.UpdateRoleMappings$, () => exports.ArgoCdNetworkAccessConfigRequest$]
];
exports.UpdateCapabilityConfiguration$ = [3, n0, _UCC,
    0,
    [_aC],
    [() => exports.UpdateArgoCdConfig$]
];
exports.UpdateCapabilityRequest$ = [3, n0, _UCR,
    0,
    [_cN, _cNa, _rA, _con, _cRT, _dPP],
    [[0, 1], [0, 1], 0, () => exports.UpdateCapabilityConfiguration$, [0, 4], 0], 2
];
exports.UpdateCapabilityResponse$ = [3, n0, _UCRp,
    0,
    [_up],
    [() => exports.Update$]
];
exports.UpdateClusterConfigRequest$ = [3, n0, _UCCR,
    0,
    [_n, _rVC, _l, _cRT, _aCc, _uP, _zSC, _cCo, _kNC, _sC, _rNC, _dP, _cPSC],
    [[0, 1], () => exports.VpcConfigRequest$, () => exports.Logging$, [0, 4], () => exports.UpdateAccessConfigRequest$, () => exports.UpgradePolicyRequest$, () => exports.ZonalShiftConfigRequest$, () => exports.ComputeConfigRequest$, () => exports.KubernetesNetworkConfigRequest$, () => exports.StorageConfigRequest$, () => exports.RemoteNetworkConfigRequest$, 2, () => exports.ControlPlaneScalingConfig$], 1
];
exports.UpdateClusterConfigResponse$ = [3, n0, _UCCRp,
    0,
    [_up],
    [() => exports.Update$]
];
exports.UpdateClusterVersionRequest$ = [3, n0, _UCVR,
    0,
    [_n, _v, _cRT, _fo],
    [[0, 1], 0, [0, 4], 2], 2
];
exports.UpdateClusterVersionResponse$ = [3, n0, _UCVRp,
    0,
    [_up],
    [() => exports.Update$]
];
exports.UpdateEksAnywhereSubscriptionRequest$ = [3, n0, _UEASR,
    0,
    [_id_, _aR, _cRT],
    [[0, 1], 2, [0, 4]], 2
];
exports.UpdateEksAnywhereSubscriptionResponse$ = [3, n0, _UEASRp,
    0,
    [_su],
    [() => exports.EksAnywhereSubscription$]
];
exports.UpdateLabelsPayload$ = [3, n0, _ULP,
    0,
    [_aOUL, _rL],
    [128 | 0, 64 | 0]
];
exports.UpdateNodegroupConfigRequest$ = [3, n0, _UNCR,
    0,
    [_cN, _nN, _la, _ta, _sCc, _uC, _nRC, _wPC, _cRT],
    [[0, 1], [0, 1], () => exports.UpdateLabelsPayload$, () => exports.UpdateTaintsPayload$, () => exports.NodegroupScalingConfig$, () => exports.NodegroupUpdateConfig$, () => exports.NodeRepairConfig$, () => exports.WarmPoolConfig$, [0, 4]], 2
];
exports.UpdateNodegroupConfigResponse$ = [3, n0, _UNCRp,
    0,
    [_up],
    [() => exports.Update$]
];
exports.UpdateNodegroupVersionRequest$ = [3, n0, _UNVR,
    0,
    [_cN, _nN, _v, _rV, _lTa, _fo, _cRT],
    [[0, 1], [0, 1], 0, 0, () => exports.LaunchTemplateSpecification$, 2, [0, 4]], 2
];
exports.UpdateNodegroupVersionResponse$ = [3, n0, _UNVRp,
    0,
    [_up],
    [() => exports.Update$]
];
exports.UpdateParam$ = [3, n0, _UP,
    0,
    [_ty, _va],
    [0, 0]
];
exports.UpdatePodIdentityAssociationRequest$ = [3, n0, _UPIAR,
    0,
    [_cN, _aIs, _rA, _cRT, _dST, _tRA, _po],
    [[0, 1], [0, 1], 0, [0, 4], 2, 0, 0], 2
];
exports.UpdatePodIdentityAssociationResponse$ = [3, n0, _UPIARp,
    0,
    [_as],
    [() => exports.PodIdentityAssociation$]
];
exports.UpdateRoleMappings$ = [3, n0, _URM,
    0,
    [_aOURM, _rRMe],
    [() => ArgoCdRoleMappingList, () => ArgoCdRoleMappingList]
];
exports.UpdateTaintsPayload$ = [3, n0, _UTP,
    0,
    [_aOUT, _rT],
    [() => taintsList, () => taintsList]
];
exports.UpgradePolicyRequest$ = [3, n0, _UPR,
    0,
    [_sT],
    [0]
];
exports.UpgradePolicyResponse$ = [3, n0, _UPRp,
    0,
    [_sT],
    [0]
];
exports.VpcConfigRequest$ = [3, n0, _VCR,
    0,
    [_sIu, _sGI, _ePA, _ePAn, _pAC],
    [64 | 0, 64 | 0, 2, 2, 64 | 0]
];
exports.VpcConfigResponse$ = [3, n0, _VCRp,
    0,
    [_sIu, _sGI, _cSGI, _vIp, _ePA, _ePAn, _pAC],
    [64 | 0, 64 | 0, 0, 0, 2, 2, 64 | 0]
];
exports.WarmPoolConfig$ = [3, n0, _WPC,
    0,
    [_en, _mS, _mGPC, _pS, _rOSI],
    [2, 1, 1, 0, 2]
];
exports.ZonalShiftConfigRequest$ = [3, n0, _ZSCR,
    0,
    [_en],
    [2]
];
exports.ZonalShiftConfigResponse$ = [3, n0, _ZSCRo,
    0,
    [_en],
    [2]
];
var AccessPoliciesList = [1, n0, _APL,
    0, () => exports.AccessPolicy$
];
var AddonCompatibilityDetails = [1, n0, _ACDd,
    0, () => exports.AddonCompatibilityDetail$
];
var AddonIssueList = [1, n0, _AIL,
    0, () => exports.AddonIssue$
];
var AddonPodIdentityAssociationsList = [1, n0, _APIAL,
    0, () => exports.AddonPodIdentityAssociations$
];
var AddonPodIdentityConfigurationList = [1, n0, _APICL,
    0, () => exports.AddonPodIdentityConfiguration$
];
var Addons = [1, n0, _Ad,
    0, () => exports.AddonInfo$
];
var AddonVersionInfoList = [1, n0, _AVIL,
    0, () => exports.AddonVersionInfo$
];
var ArgoCdRoleMappingList = [1, n0, _ACRML,
    0, () => exports.ArgoCdRoleMapping$
];
var AssociatedAccessPoliciesList = [1, n0, _AAPL,
    0, () => exports.AssociatedAccessPolicy$
];
var AutoScalingGroupList = [1, n0, _ASGL,
    0, () => exports.AutoScalingGroup$
];
var CapabilityIssueList = [1, n0, _CIL,
    0, () => exports.CapabilityIssue$
];
var CapabilitySummaryList = [1, n0, _CSL,
    0, () => exports.CapabilitySummary$
];
var CategoryList = 64 | 0;
var ClientStats = [1, n0, _CSli,
    0, () => exports.ClientStat$
];
var ClusterIssueList = [1, n0, _CILl,
    0, () => exports.ClusterIssue$
];
var ClusterVersionList = [1, n0, _CVL,
    0, () => exports.ClusterVersionInformation$
];
var Compatibilities = [1, n0, _Com,
    0, () => exports.Compatibility$
];
var DeprecationDetails = [1, n0, _DDe,
    0, () => exports.DeprecationDetail$
];
var EksAnywhereSubscriptionList = [1, n0, _EASL,
    0, () => exports.EksAnywhereSubscription$
];
var EksAnywhereSubscriptionStatusValues = 64 | 0;
var EncryptionConfigList = [1, n0, _ECL,
    0, () => exports.EncryptionConfig$
];
var ErrorDetails = [1, n0, _EDr,
    0, () => exports.ErrorDetail$
];
var FargateProfileIssueList = [1, n0, _FPIL,
    0, () => exports.FargateProfileIssue$
];
var FargateProfileSelectors = [1, n0, _FPSa,
    0, () => exports.FargateProfileSelector$
];
var IdentityProviderConfigs = [1, n0, _IPCd,
    0, () => exports.IdentityProviderConfig$
];
var IncludeClustersList = 64 | 0;
var InsightResourceDetails = [1, n0, _IRDn,
    0, () => exports.InsightResourceDetail$
];
var InsightStatusValueList = 64 | 0;
var InsightSummaries = [1, n0, _ISns,
    0, () => exports.InsightSummary$
];
var IssueList = [1, n0, _IL,
    0, () => exports.Issue$
];
var labelsKeyList = 64 | 0;
var LicenseList = [1, n0, _LL,
    0, () => exports.License$
];
var LogSetups = [1, n0, _LSo,
    0, () => exports.LogSetup$
];
var LogTypes = 64 | 0;
var NodeRepairConfigOverridesList = [1, n0, _NRCOL,
    0, () => exports.NodeRepairConfigOverrides$
];
var PodIdentityAssociationSummaries = [1, n0, _PIASo,
    0, () => exports.PodIdentityAssociationSummary$
];
var RemoteNodeNetworkList = [1, n0, _RNNL,
    0, () => exports.RemoteNodeNetwork$
];
var RemotePodNetworkList = [1, n0, _RPNL,
    0, () => exports.RemotePodNetwork$
];
var SsoIdentityList = [1, n0, _SIL,
    0, () => exports.SsoIdentity$
];
var StringList = 64 | 0;
var TagKeyList = 64 | 0;
var taintsList = [1, n0, _tL,
    0, () => exports.Taint$
];
var UpdateParams = [1, n0, _UPp,
    0, () => exports.UpdateParam$
];
var AdditionalInfoMap = 128 | 0;
var FargateProfileLabel = 128 | 0;
var labelsMap = 128 | 0;
var requiredClaimsMap = 128 | 0;
var TagMap = 128 | 0;
exports.AssociateAccessPolicy$ = [9, n0, _AAPs,
    { [_ht]: ["POST", "/clusters/{clusterName}/access-entries/{principalArn}/access-policies", 200] }, () => exports.AssociateAccessPolicyRequest$, () => exports.AssociateAccessPolicyResponse$
];
exports.AssociateEncryptionConfig$ = [9, n0, _AEC,
    { [_ht]: ["POST", "/clusters/{clusterName}/encryption-config/associate", 200] }, () => exports.AssociateEncryptionConfigRequest$, () => exports.AssociateEncryptionConfigResponse$
];
exports.AssociateIdentityProviderConfig$ = [9, n0, _AIPC,
    { [_ht]: ["POST", "/clusters/{clusterName}/identity-provider-configs/associate", 200] }, () => exports.AssociateIdentityProviderConfigRequest$, () => exports.AssociateIdentityProviderConfigResponse$
];
exports.CreateAccessEntry$ = [9, n0, _CAE,
    { [_ht]: ["POST", "/clusters/{clusterName}/access-entries", 200] }, () => exports.CreateAccessEntryRequest$, () => exports.CreateAccessEntryResponse$
];
exports.CreateAddon$ = [9, n0, _CA,
    { [_ht]: ["POST", "/clusters/{clusterName}/addons", 200] }, () => exports.CreateAddonRequest$, () => exports.CreateAddonResponse$
];
exports.CreateCapability$ = [9, n0, _CC,
    { [_ht]: ["POST", "/clusters/{clusterName}/capabilities", 200] }, () => exports.CreateCapabilityRequest$, () => exports.CreateCapabilityResponse$
];
exports.CreateCluster$ = [9, n0, _CCr,
    { [_ht]: ["POST", "/clusters", 200] }, () => exports.CreateClusterRequest$, () => exports.CreateClusterResponse$
];
exports.CreateEksAnywhereSubscription$ = [9, n0, _CEAS,
    { [_ht]: ["POST", "/eks-anywhere-subscriptions", 200] }, () => exports.CreateEksAnywhereSubscriptionRequest$, () => exports.CreateEksAnywhereSubscriptionResponse$
];
exports.CreateFargateProfile$ = [9, n0, _CFP,
    { [_ht]: ["POST", "/clusters/{clusterName}/fargate-profiles", 200] }, () => exports.CreateFargateProfileRequest$, () => exports.CreateFargateProfileResponse$
];
exports.CreateNodegroup$ = [9, n0, _CN,
    { [_ht]: ["POST", "/clusters/{clusterName}/node-groups", 200] }, () => exports.CreateNodegroupRequest$, () => exports.CreateNodegroupResponse$
];
exports.CreatePodIdentityAssociation$ = [9, n0, _CPIA,
    { [_ht]: ["POST", "/clusters/{clusterName}/pod-identity-associations", 200] }, () => exports.CreatePodIdentityAssociationRequest$, () => exports.CreatePodIdentityAssociationResponse$
];
exports.DeleteAccessEntry$ = [9, n0, _DAE,
    { [_ht]: ["DELETE", "/clusters/{clusterName}/access-entries/{principalArn}", 200] }, () => exports.DeleteAccessEntryRequest$, () => exports.DeleteAccessEntryResponse$
];
exports.DeleteAddon$ = [9, n0, _DA,
    { [_ht]: ["DELETE", "/clusters/{clusterName}/addons/{addonName}", 200] }, () => exports.DeleteAddonRequest$, () => exports.DeleteAddonResponse$
];
exports.DeleteCapability$ = [9, n0, _DC,
    { [_ht]: ["DELETE", "/clusters/{clusterName}/capabilities/{capabilityName}", 200] }, () => exports.DeleteCapabilityRequest$, () => exports.DeleteCapabilityResponse$
];
exports.DeleteCluster$ = [9, n0, _DCe,
    { [_ht]: ["DELETE", "/clusters/{name}", 200] }, () => exports.DeleteClusterRequest$, () => exports.DeleteClusterResponse$
];
exports.DeleteEksAnywhereSubscription$ = [9, n0, _DEAS,
    { [_ht]: ["DELETE", "/eks-anywhere-subscriptions/{id}", 200] }, () => exports.DeleteEksAnywhereSubscriptionRequest$, () => exports.DeleteEksAnywhereSubscriptionResponse$
];
exports.DeleteFargateProfile$ = [9, n0, _DFP,
    { [_ht]: ["DELETE", "/clusters/{clusterName}/fargate-profiles/{fargateProfileName}", 200] }, () => exports.DeleteFargateProfileRequest$, () => exports.DeleteFargateProfileResponse$
];
exports.DeleteNodegroup$ = [9, n0, _DN,
    { [_ht]: ["DELETE", "/clusters/{clusterName}/node-groups/{nodegroupName}", 200] }, () => exports.DeleteNodegroupRequest$, () => exports.DeleteNodegroupResponse$
];
exports.DeletePodIdentityAssociation$ = [9, n0, _DPIA,
    { [_ht]: ["DELETE", "/clusters/{clusterName}/pod-identity-associations/{associationId}", 200] }, () => exports.DeletePodIdentityAssociationRequest$, () => exports.DeletePodIdentityAssociationResponse$
];
exports.DeregisterCluster$ = [9, n0, _DCer,
    { [_ht]: ["DELETE", "/cluster-registrations/{name}", 200] }, () => exports.DeregisterClusterRequest$, () => exports.DeregisterClusterResponse$
];
exports.DescribeAccessEntry$ = [9, n0, _DAEe,
    { [_ht]: ["GET", "/clusters/{clusterName}/access-entries/{principalArn}", 200] }, () => exports.DescribeAccessEntryRequest$, () => exports.DescribeAccessEntryResponse$
];
exports.DescribeAddon$ = [9, n0, _DAe,
    { [_ht]: ["GET", "/clusters/{clusterName}/addons/{addonName}", 200] }, () => exports.DescribeAddonRequest$, () => exports.DescribeAddonResponse$
];
exports.DescribeAddonConfiguration$ = [9, n0, _DAC,
    { [_ht]: ["GET", "/addons/configuration-schemas", 200] }, () => exports.DescribeAddonConfigurationRequest$, () => exports.DescribeAddonConfigurationResponse$
];
exports.DescribeAddonVersions$ = [9, n0, _DAV,
    { [_ht]: ["GET", "/addons/supported-versions", 200] }, () => exports.DescribeAddonVersionsRequest$, () => exports.DescribeAddonVersionsResponse$
];
exports.DescribeCapability$ = [9, n0, _DCes,
    { [_ht]: ["GET", "/clusters/{clusterName}/capabilities/{capabilityName}", 200] }, () => exports.DescribeCapabilityRequest$, () => exports.DescribeCapabilityResponse$
];
exports.DescribeCluster$ = [9, n0, _DCesc,
    { [_ht]: ["GET", "/clusters/{name}", 200] }, () => exports.DescribeClusterRequest$, () => exports.DescribeClusterResponse$
];
exports.DescribeClusterVersions$ = [9, n0, _DCV,
    { [_ht]: ["GET", "/cluster-versions", 200] }, () => exports.DescribeClusterVersionsRequest$, () => exports.DescribeClusterVersionsResponse$
];
exports.DescribeEksAnywhereSubscription$ = [9, n0, _DEASe,
    { [_ht]: ["GET", "/eks-anywhere-subscriptions/{id}", 200] }, () => exports.DescribeEksAnywhereSubscriptionRequest$, () => exports.DescribeEksAnywhereSubscriptionResponse$
];
exports.DescribeFargateProfile$ = [9, n0, _DFPe,
    { [_ht]: ["GET", "/clusters/{clusterName}/fargate-profiles/{fargateProfileName}", 200] }, () => exports.DescribeFargateProfileRequest$, () => exports.DescribeFargateProfileResponse$
];
exports.DescribeIdentityProviderConfig$ = [9, n0, _DIPC,
    { [_ht]: ["POST", "/clusters/{clusterName}/identity-provider-configs/describe", 200] }, () => exports.DescribeIdentityProviderConfigRequest$, () => exports.DescribeIdentityProviderConfigResponse$
];
exports.DescribeInsight$ = [9, n0, _DI,
    { [_ht]: ["GET", "/clusters/{clusterName}/insights/{id}", 200] }, () => exports.DescribeInsightRequest$, () => exports.DescribeInsightResponse$
];
exports.DescribeInsightsRefresh$ = [9, n0, _DIRes,
    { [_ht]: ["GET", "/clusters/{clusterName}/insights-refresh", 200] }, () => exports.DescribeInsightsRefreshRequest$, () => exports.DescribeInsightsRefreshResponse$
];
exports.DescribeNodegroup$ = [9, n0, _DNe,
    { [_ht]: ["GET", "/clusters/{clusterName}/node-groups/{nodegroupName}", 200] }, () => exports.DescribeNodegroupRequest$, () => exports.DescribeNodegroupResponse$
];
exports.DescribePodIdentityAssociation$ = [9, n0, _DPIAe,
    { [_ht]: ["GET", "/clusters/{clusterName}/pod-identity-associations/{associationId}", 200] }, () => exports.DescribePodIdentityAssociationRequest$, () => exports.DescribePodIdentityAssociationResponse$
];
exports.DescribeUpdate$ = [9, n0, _DU,
    { [_ht]: ["GET", "/clusters/{name}/updates/{updateId}", 200] }, () => exports.DescribeUpdateRequest$, () => exports.DescribeUpdateResponse$
];
exports.DisassociateAccessPolicy$ = [9, n0, _DAP,
    { [_ht]: ["DELETE", "/clusters/{clusterName}/access-entries/{principalArn}/access-policies/{policyArn}", 200] }, () => exports.DisassociateAccessPolicyRequest$, () => exports.DisassociateAccessPolicyResponse$
];
exports.DisassociateIdentityProviderConfig$ = [9, n0, _DIPCi,
    { [_ht]: ["POST", "/clusters/{clusterName}/identity-provider-configs/disassociate", 200] }, () => exports.DisassociateIdentityProviderConfigRequest$, () => exports.DisassociateIdentityProviderConfigResponse$
];
exports.ListAccessEntries$ = [9, n0, _LAE,
    { [_ht]: ["GET", "/clusters/{clusterName}/access-entries", 200] }, () => exports.ListAccessEntriesRequest$, () => exports.ListAccessEntriesResponse$
];
exports.ListAccessPolicies$ = [9, n0, _LAP,
    { [_ht]: ["GET", "/access-policies", 200] }, () => exports.ListAccessPoliciesRequest$, () => exports.ListAccessPoliciesResponse$
];
exports.ListAddons$ = [9, n0, _LA,
    { [_ht]: ["GET", "/clusters/{clusterName}/addons", 200] }, () => exports.ListAddonsRequest$, () => exports.ListAddonsResponse$
];
exports.ListAssociatedAccessPolicies$ = [9, n0, _LAAP,
    { [_ht]: ["GET", "/clusters/{clusterName}/access-entries/{principalArn}/access-policies", 200] }, () => exports.ListAssociatedAccessPoliciesRequest$, () => exports.ListAssociatedAccessPoliciesResponse$
];
exports.ListCapabilities$ = [9, n0, _LC,
    { [_ht]: ["GET", "/clusters/{clusterName}/capabilities", 200] }, () => exports.ListCapabilitiesRequest$, () => exports.ListCapabilitiesResponse$
];
exports.ListClusters$ = [9, n0, _LCi,
    { [_ht]: ["GET", "/clusters", 200] }, () => exports.ListClustersRequest$, () => exports.ListClustersResponse$
];
exports.ListEksAnywhereSubscriptions$ = [9, n0, _LEAS,
    { [_ht]: ["GET", "/eks-anywhere-subscriptions", 200] }, () => exports.ListEksAnywhereSubscriptionsRequest$, () => exports.ListEksAnywhereSubscriptionsResponse$
];
exports.ListFargateProfiles$ = [9, n0, _LFP,
    { [_ht]: ["GET", "/clusters/{clusterName}/fargate-profiles", 200] }, () => exports.ListFargateProfilesRequest$, () => exports.ListFargateProfilesResponse$
];
exports.ListIdentityProviderConfigs$ = [9, n0, _LIPC,
    { [_ht]: ["GET", "/clusters/{clusterName}/identity-provider-configs", 200] }, () => exports.ListIdentityProviderConfigsRequest$, () => exports.ListIdentityProviderConfigsResponse$
];
exports.ListInsights$ = [9, n0, _LI,
    { [_ht]: ["POST", "/clusters/{clusterName}/insights", 200] }, () => exports.ListInsightsRequest$, () => exports.ListInsightsResponse$
];
exports.ListNodegroups$ = [9, n0, _LN,
    { [_ht]: ["GET", "/clusters/{clusterName}/node-groups", 200] }, () => exports.ListNodegroupsRequest$, () => exports.ListNodegroupsResponse$
];
exports.ListPodIdentityAssociations$ = [9, n0, _LPIA,
    { [_ht]: ["GET", "/clusters/{clusterName}/pod-identity-associations", 200] }, () => exports.ListPodIdentityAssociationsRequest$, () => exports.ListPodIdentityAssociationsResponse$
];
exports.ListTagsForResource$ = [9, n0, _LTFR,
    { [_ht]: ["GET", "/tags/{resourceArn}", 200] }, () => exports.ListTagsForResourceRequest$, () => exports.ListTagsForResourceResponse$
];
exports.ListUpdates$ = [9, n0, _LU,
    { [_ht]: ["GET", "/clusters/{name}/updates", 200] }, () => exports.ListUpdatesRequest$, () => exports.ListUpdatesResponse$
];
exports.RegisterCluster$ = [9, n0, _RC,
    { [_ht]: ["POST", "/cluster-registrations", 200] }, () => exports.RegisterClusterRequest$, () => exports.RegisterClusterResponse$
];
exports.StartInsightsRefresh$ = [9, n0, _SIR,
    { [_ht]: ["POST", "/clusters/{clusterName}/insights-refresh", 200] }, () => exports.StartInsightsRefreshRequest$, () => exports.StartInsightsRefreshResponse$
];
exports.TagResource$ = [9, n0, _TR,
    { [_ht]: ["POST", "/tags/{resourceArn}", 200] }, () => exports.TagResourceRequest$, () => exports.TagResourceResponse$
];
exports.UntagResource$ = [9, n0, _UR,
    { [_ht]: ["DELETE", "/tags/{resourceArn}", 200] }, () => exports.UntagResourceRequest$, () => exports.UntagResourceResponse$
];
exports.UpdateAccessEntry$ = [9, n0, _UAE,
    { [_ht]: ["POST", "/clusters/{clusterName}/access-entries/{principalArn}", 200] }, () => exports.UpdateAccessEntryRequest$, () => exports.UpdateAccessEntryResponse$
];
exports.UpdateAddon$ = [9, n0, _UA,
    { [_ht]: ["POST", "/clusters/{clusterName}/addons/{addonName}/update", 200] }, () => exports.UpdateAddonRequest$, () => exports.UpdateAddonResponse$
];
exports.UpdateCapability$ = [9, n0, _UC,
    { [_ht]: ["POST", "/clusters/{clusterName}/capabilities/{capabilityName}", 200] }, () => exports.UpdateCapabilityRequest$, () => exports.UpdateCapabilityResponse$
];
exports.UpdateClusterConfig$ = [9, n0, _UCCp,
    { [_ht]: ["POST", "/clusters/{name}/update-config", 200] }, () => exports.UpdateClusterConfigRequest$, () => exports.UpdateClusterConfigResponse$
];
exports.UpdateClusterVersion$ = [9, n0, _UCV,
    { [_ht]: ["POST", "/clusters/{name}/updates", 200] }, () => exports.UpdateClusterVersionRequest$, () => exports.UpdateClusterVersionResponse$
];
exports.UpdateEksAnywhereSubscription$ = [9, n0, _UEAS,
    { [_ht]: ["POST", "/eks-anywhere-subscriptions/{id}", 200] }, () => exports.UpdateEksAnywhereSubscriptionRequest$, () => exports.UpdateEksAnywhereSubscriptionResponse$
];
exports.UpdateNodegroupConfig$ = [9, n0, _UNC,
    { [_ht]: ["POST", "/clusters/{clusterName}/node-groups/{nodegroupName}/update-config", 200] }, () => exports.UpdateNodegroupConfigRequest$, () => exports.UpdateNodegroupConfigResponse$
];
exports.UpdateNodegroupVersion$ = [9, n0, _UNV,
    { [_ht]: ["POST", "/clusters/{clusterName}/node-groups/{nodegroupName}/update-version", 200] }, () => exports.UpdateNodegroupVersionRequest$, () => exports.UpdateNodegroupVersionResponse$
];
exports.UpdatePodIdentityAssociation$ = [9, n0, _UPIA,
    { [_ht]: ["POST", "/clusters/{clusterName}/pod-identity-associations/{associationId}", 200] }, () => exports.UpdatePodIdentityAssociationRequest$, () => exports.UpdatePodIdentityAssociationResponse$
];
